---
title: "Table 1 sample summary"
author: "Kimberly Olney"
date: "2023-07-28"
output: html_document
---
This script will examine the variables in the metadata to determine if ther are differences by disease type and/or sex differences for clinical data such as brain weight. 
# Setup
```{r setup, message=FALSE, warning=FALSE, tidy=TRUE}
knitr::opts_knit$set(root.dir = ".")
```
# User defined variables
sourcing file_paths_and_colours will load the libraries and user defined variables for colors of the graphs 
```{r set_variables, message=FALSE, warning=FALSE, tidy=TRUE}
source(here::here("scripts/R", "file_paths_and_colours.R"))
remove(metadata, metadata_ATS, df)
```
# Read data
```{r read_data, message=FALSE, warning=FALSE, tidy=TRUE}
# read in metadata
metadata <-
  read.delim(paste0(pathToRawData, "metadata_expanded.tsv"),
             header = TRUE,
             sep = "\t")

# remove technical replicates (i.e lane 1 and lane 2)
metadata <- metadata[!duplicated(metadata[, c('NPID')]), ]
# remove samples that have missing A-T-S scores or that don't fit into an A-T-S category 
excluded_IDs <-  c("NA18-272", "5510", "NA01-135", "NA20-102", "NA20-104", "NA06-159")
metadata <- metadata[ ! metadata$NPID %in% excluded_IDs, ]

# read in APOE genotype information 
APOE <- read.delim(paste0(pathToRawData, "APOE.tsv"),
             header = TRUE,
             sep = "\t")
metadata <- merge(metadata, APOE, by = "NPID", all.x = TRUE, all.y= TRUE)
colnames(metadata)[colnames(metadata) == "APOE.y"] <- "APOE"

# remove samples that don't have RNA data
metadata <- metadata %>% drop_na(flowcell_and_lane)
metadata <- metadata %>%
  mutate(across('AD.subtype', str_replace, 'typical', 'Typical'))
# rename columns for easier plotting 
metadata <- metadata %>%
  mutate(across('TYPE', str_replace, 'CONTROL - AD', 'AD'))
metadata <- metadata %>%
  mutate(across('TYPE', str_replace, 'CONTROL - PA', 'PA'))
metadata$TYPE <- factor(metadata$TYPE, levels = c("CONTROL", "PA", "AD", "LBD"))
# add sex chromosome information 
metadata <- metadata %>% mutate(sex_chr = case_when(
  startsWith(sex_inferred, "female") ~ "XX", 
  startsWith(sex_inferred, "male") ~ "XY"))
# remove rows with no inferred sex.
# No counts data for this samples because there is no RNAseq data for that individual
metadata <- metadata[!is.na(metadata$sex_inferred),]
```

## RNA sample count
```{r bar_plot, message=FALSE, warning=FALSE, tidy=TRUE}
# Type count 
as.data.frame(table(metadata$TYPE))
# LBD.Type count 
as.data.frame(table(metadata$LBD.type))

# percent male by type
metadata %>% 
  group_by(sex_inferred,TYPE) %>%
  dplyr::count() %>% 
  ungroup(sex_inferred) %>% 
  mutate(perc = `n` / sum(`n`)) %>% 
  arrange(perc) %>%
  mutate(labels = scales::percent(perc))

# percent male by LBD.type
metadata %>% 
  group_by(sex_inferred,LBD.type) %>%
  dplyr::count() %>% 
  ungroup(sex_inferred) %>% 
  mutate(perc = `n` / sum(`n`)) %>% 
  arrange(perc) %>%
  mutate(labels = scales::percent(perc))
```
# Age and brain weight 
```{r}
# Age by type
tapply(metadata$Age, metadata$TYPE, t.test) 
# Age by LBD type
tapply(metadata$Age, metadata$LBD.type, t.test) 


# Brain weight by type
tapply(metadata$Brain.wt, metadata$TYPE, t.test) 
#  Brain weight by type
tapply(metadata$Brain.wt, metadata$LBD.type, t.test) 
```
# Braak stage count 
```{r}
Braak_TYPE <- metadata %>% 
  group_by(Braak.NFT,TYPE) %>%
  dplyr::count() %>% 
  ungroup(Braak.NFT) %>% 
  mutate(perc = `n` / sum(`n`)) %>% 
  arrange(perc) %>%
  mutate(labels = scales::percent(perc))

Braak_LBD.type <- metadata %>% 
  group_by(Braak.NFT,LBD.type) %>%
  dplyr::count() %>% 
  ungroup(Braak.NFT) %>% 
  mutate(perc = `n` / sum(`n`)) %>% 
  arrange(perc) %>%
  mutate(labels = scales::percent(perc))
```

# APOE E4+ %
```{r}
# Yes to E4+
metadata$APOE_E4 <-
  ifelse(
    metadata$APOE == "E2E4",
    "yes",
    ifelse(
      metadata$APOE == "E3E4",
      "yes",
      ifelse(
        metadata$APOE == "E4E4",
        "yes", "no")))

APOE_E4_TYPE <- metadata %>% 
  group_by(APOE_E4,TYPE) %>%
  dplyr::count() %>% 
  ungroup(APOE_E4) %>% 
  mutate(perc = `n` / sum(`n`)) %>% 
  arrange(perc) %>%
  mutate(labels = scales::percent(perc))

# LBD.type
APOE_E4_LBD.type <- metadata %>% 
  group_by(APOE_E4,LBD.type) %>%
  dplyr::count() %>% 
  ungroup(APOE_E4) %>% 
  mutate(perc = `n` / sum(`n`)) %>% 
  arrange(perc) %>%
  mutate(labels = scales::percent(perc))
```
