---
title: "Supplemental Figure ROSMAP upset and correlation within each sex"
author: "Kimberly Olney"
date: "07/10/2023"
output: html_document
---
# Setup
```{r setup}
knitr::opts_knit$set(root.dir = ".")
```
# User defined variables
```{r set_variables}
source(here::here("scripts/R", "file_paths_and_colours.R"))
library("readxl")
#source(here::here("scripts/R", "gtf_path.R"))
condition <- c("TYPE")
tool <- "star"
dge.filtered.norm <- readRDS(paste0("../../../rObjects/dge.filtered.norm.rds"))
```
# Binary list shared gene function
```{r}
fromList <- function (input) {
  # Same as original fromList()...
  elements <- unique(unlist(input))
  data <- unlist(lapply(input, function(x) {
      x <- as.vector(match(elements, x))
      }))
  data[is.na(data)] <- as.integer(0)
  data[data != 0] <- as.integer(1)
  data <- data.frame(matrix(data, ncol = length(input), byrow = F))
  data <- data[which(rowSums(data) != 0), ]
  names(data) <- names(input)
  # ... Except now it conserves your original value names!
  row.names(data) <- elements
  return(data)
  }
```


# ROSMAP correlation
```{r}
# read in ROSMAP data 
ROSMAP_AD <- read.delim("/research/labs/neurology/fryer/m239830/synapse_RNAseq_harmonization/ROSMAP_Differential_Expression_diagnosis_PCC.tsv", sep = "\t", header = TRUE)

ROSMAP_AD$gene_name <- ROSMAP_AD$hgnc_symbol

i <- "ADvsControl"
ADvsControl <- read.delim(paste0("../../../results/", tool, "/DEGs/", condition, "_", i, "_gene_DEGs_FDRq1.00.txt"))

data <-
  merge(ADvsControl,
        ROSMAP_AD,
        by = "gene_name",
        all = T)
data <- data[c(1,2,6,8,11,14,18,21,24)] #data[c(1,2,6,9,10,13,22,23,26)]
colnames(data) = c("gene_name", "chr", "gene_id", "AD_CWOW_logFC", "AD_CWOW_AveExpr", "AD_CWOW_adj.P.Val", "AD_ROSMAP_logFC", "AD_ROSMAP_AveExpr", "AD_ROSMAP_adj.P.Val")

#data[is.na(data)] <- 0
data <- na.omit(data)

dataV2 <- subset(data, (AD_CWOW_adj.P.Val <= 0.05 & AD_CWOW_logFC > 0.25) | 
                      (AD_CWOW_adj.P.Val <= 0.05 & AD_CWOW_logFC < -0.25) |
                    (AD_ROSMAP_adj.P.Val <= 0.05 & AD_ROSMAP_logFC > 0.25) |
                    (AD_ROSMAP_adj.P.Val <= 0.05 & AD_ROSMAP_logFC < -0.25))

sig_CWOW <- subset(dataV2, (AD_CWOW_adj.P.Val <= 0.05 & AD_CWOW_logFC > 0.25) | (AD_CWOW_adj.P.Val <= 0.05 & AD_CWOW_logFC < -0.25))
sig_ROSMAP <- subset(dataV2, (AD_ROSMAP_adj.P.Val < 0.05 & AD_ROSMAP_logFC > 0.25) | (AD_ROSMAP_adj.P.Val < 0.05 & AD_ROSMAP_logFC < -0.25))

sig_both <- subset(dataV2, (AD_CWOW_adj.P.Val < 0.05 & AD_ROSMAP_adj.P.Val < 0.05 & AD_ROSMAP_logFC < -0.25 & AD_CWOW_logFC < -0.25) |
                     (AD_CWOW_adj.P.Val < 0.05 & AD_ROSMAP_adj.P.Val < 0.05 & AD_ROSMAP_logFC  > 0.25 & AD_CWOW_logFC > 0.25) |
                     (AD_CWOW_adj.P.Val < 0.05 & AD_ROSMAP_adj.P.Val < 0.05 & AD_ROSMAP_logFC < -0.25 & AD_CWOW_logFC > 0.25) |(AD_CWOW_adj.P.Val < 0.05 & AD_ROSMAP_adj.P.Val < 0.05 & AD_ROSMAP_logFC  > 0.25 & AD_CWOW_logFC < -0.25))

sig_CWOW_new <- subset(sig_CWOW, !(sig_CWOW$gene_name %in% sig_both$gene_name) | !(sig_CWOW$gene_name %in% sig_ROSMAP$gene_name))
sig_ROSMAP_new <- subset(sig_ROSMAP, !(sig_ROSMAP$gene_name %in% sig_both$gene_name) | !(sig_ROSMAP$gene_name %in% sig_CWOW$gene_name))

#data$difference <- data$logFC - data$gene.log2FC
AD_ROSAMP_sig <- ggplot(data = dataV2, 
  aes(x = AD_ROSMAP_logFC, y = AD_CWOW_logFC, text = paste(gene_name))) +
  annotate(
    "rect",
    xmin = 0,
    xmax = 1.75,
    ymin = 1.75,
    ymax = 0,
    fill = "lightpink3",
    alpha = .1
  ) +  annotate(
    "rect",
    xmin = 0,
    xmax = -1.5,
    ymin = 0,
    ymax = -1.5,
    fill = "cadetblue3",
    alpha = .1) +
  geom_abline(color = "gray80", linetype = 2) +
  geom_point(data = sig_CWOW_new, color=AD_color, shape=AD_shape) + 
  geom_point(data = sig_ROSMAP_new, color="sienna4", shape=PA_shape) + 
  geom_point(data = sig_both, color="black", shape=25) + 
  geom_text_repel(
    data = sig_both, 
    aes(
    x = AD_ROSMAP_logFC, 
    y = AD_CWOW_logFC,
    label = gene_name
    ),
    color = "black",
    fontface = "italic",
    size = 2, 
    max.overlaps = getOption("ggrepel.max.overlaps", default = 1)
    ) +
  theme_bw() +
    theme(strip.text = element_text(size = 8), 
          axis.text.y = element_text(size = 8),
          axis.title.y = element_text(size = 8),
          axis.text.x = element_text(size = 8), 
          axis.title.x = element_text(size = 8), 
          plot.margin = margin(0.1, 0.3, 0.2, 0.4, "cm"), #t,r,b,l
          plot.title = element_text(size = 8, margin = margin(.1,.1,.1,0), hjust = 0)) +
  labs(
    title = "Differentially expressed genes, FDRq < 0.05 & absolute log2FC > 0.25
Shared and unique between AD and ROSMAP PCC AD",
    x = expression(paste("ROSMAP ", log[2](AD/Control))),
    y = expression(paste("CWOW ", log[2](AD/Control))))+
  scale_y_continuous(breaks = seq(-1.25, 1.5, by = .5), 
                     limits = c(-1.5, 1.75), expand = c(0,0)) +
  scale_x_continuous(breaks = seq(-1.25, 1.5, by = .5), 
                     limits = c(-1.5, 1.75), expand = c(0,0))
AD_ROSAMP_sig


# Scatter plot with correlation coefficient
sp <- ggscatter(dataV2, x = "AD_ROSMAP_logFC", y = "AD_CWOW_logFC",
                add = "reg.line",  # Add regressin line
                add.params = list(color = "blue", fill = "lightgray"), 
                conf.int = TRUE)
# Add correlation coefficient
sp + stat_cor(method = "pearson", label.x = -.5, label.y = 1)

path <- paste0("../../../results/star/correlation/TYPE_AD_CWOW_vs_ROSMAP_shared_and_unique_pval")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 4)

youknow <- ggplot(data = dataV2, 
  aes(x = AD_ROSMAP_logFC, y = AD_CWOW_logFC, text = paste(gene_name)))  +
  annotate(
    "rect",
    xmin = 0,
    xmax = 1.75,
    ymin = 1.75,
    ymax = 0,
    fill = "lightpink3",
    alpha = .1
  ) +  annotate(
    "rect",
    xmin = 0,
    xmax = -1.5,
    ymin = 0,
    ymax = -1.5,
    fill = "cadetblue3",
    alpha = .1) +
  geom_abline(color = "gray80", linetype = 2) +
  geom_point(data = sig_CWOW_new, color=AD_color, shape=AD_shape) + 
  geom_point(data = sig_ROSMAP_new, color="sienna4", shape=PA_shape) + 
  geom_point(data = sig_both, color="black", shape=25) 

#ggplotly(youknow)
```

```{r}
# DEGs by FDRq <= 0.05 & log2FC 
ROSMAPvsControl_up <- subset(ROSMAP_AD, logFC > 0.25 & adj.P.Val <= 0.05)
ROSMAPvsControl_down <- subset(ROSMAP_AD, logFC < -0.25 & adj.P.Val <= 0.05)
ADvsControl_up <- subset(ADvsControl, logFC > 0.25 & adj.P.Val <= 0.05)
ADvsControl_down <- subset(ADvsControl, logFC < -0.25 & adj.P.Val <= 0.05)
```
# Upset
```{r}
list_input <- list("CWOW AD down-regulated" = ADvsControl_down$gene_name,
                   "CWOW AD up-regulated" = ADvsControl_up$gene_name,
                   "ROSMAP AD down-regulated" = ROSMAPvsControl_down$gene_name,
                   "ROSMAP AD up-regulated" = ROSMAPvsControl_up$gene_name)
data <- fromList(list_input)
dataX <- subset(data, rownames(data) %in% dataV2$gene_name)
detach("package:UpSetR", unload = TRUE)
# detach UpSetR
upset_gene <- upset(dataX, name='', set_sizes = FALSE,
      c('CWOW AD down-regulated',
        'ROSMAP AD down-regulated',
        'CWOW AD up-regulated',
        'ROSMAP AD up-regulated'),
      themes=upset_modify_themes(
        list('intersections_matrix'=theme(text=element_text(size=10),
              plot.margin = margin(0, 0, 0, 0, "cm")),
          'overall_sizes' = theme(axis.text.x = element_text(size =10), 
              plot.margin = margin(0, 0, 0, 0, "cm")))),
  queries=list(
    upset_query(set='CWOW AD down-regulated', fill='blue'),
    upset_query(set='ROSMAP AD down-regulated', fill='blue'),
    upset_query(set='CWOW AD up-regulated', fill='red'),
    upset_query(set='ROSMAP AD up-regulated', fill='red')
  ),
  intersections = list(c('CWOW AD down-regulated', 'ROSMAP AD down-regulated'), 
                       'CWOW AD down-regulated',
                       'ROSMAP AD down-regulated', 
                       c('CWOW AD up-regulated', 'ROSMAP AD up-regulated'), 
                       'CWOW AD up-regulated',
                       'ROSMAP AD up-regulated', 
                       c('CWOW AD down-regulated', 'ROSMAP AD up-regulated')
  ),
  base_annotations=list(
    'Intersection size'=(
        intersection_size(
        size=2,
        text = list(size = 2.5),
        text_mapping = aes(),
        bar_number_threshold=3,  # show all numbers on top of bars
        width=0.3,   # reduce width of the bars
        mapping=aes(fill='bars_color')
      )
      + scale_fill_manual(values=c('grey'), guide='none')
      + scale_y_continuous(expand=expansion(mult=c(0, 0.5)))
      + theme(axis.text.y = element_text(size = 8),
              axis.title.y = element_text(size = 8),
          panel.grid.major=element_blank(),
          panel.grid.minor=element_blank(),
          plot.margin = margin(0, 0, 0, 0, "cm"), 
          axis.line=element_line(colour='black')))),
  matrix=intersection_matrix(
      geom=geom_point(
        shape='circle filled',
        size=2,
        stroke=0.45)) +
        theme(axis.text.y = element_text(size = 8),
              axis.title.y = element_text(size = 8),
              axis.text.x = element_blank(),
              axis.title.x = element_blank(), 
              plot.margin = margin(0, 0, 0, 0, "cm")),
    sort_sets=FALSE,
  sort_intersections=FALSE)
upset_gene
```


## Combine volcano and enrichment plot 
```{r}
row1 <- ggarrange(
  volcano,
  up_and_down_enrich_plot,
  ncol = 2,
  labels = c("a)", "b)"),
  widths = c(2.5, 4.25), 
  font.label = list(size = 8)
  )
row1

row2 <- ggarrange(
  upset_gene,
  AD_ROSAMP_sig,
  ncol = 2,
  labels = c("c)", "d)"),
  widths = c(.9, 1.2), 
  font.label = list(size = 8)
  )
row2

combined <- ggarrange(
  row1, 
  row2,
  nrow = 2, 
  heights = c(1.2,.8)
)
combined
path <- paste0("../../../results/manuscript_figures/Figure_2_AD_vs_Control_with_ROSMAP")
saveToPDF(paste0(path, ".pdf"), width = 8.5, height = 8)
```
```{r}
```
