---
title: "Supplemental Figure. Sex check"
output: html_document
date: "2023-09-01"
---

# Setup
```{r setup}
# Also do Session > Set Working Directory > Choose Directory
knitr::opts_knit$set(root.dir = ".")
```
# Read in data
```{r}
source(here::here("scripts/R", "file_paths_and_colours.R"))

# default alignment 
dge <- readRDS("../../../../rObjects/all_genes_default_alignment_dge.raw.rds")
lcpm <- edgeR::cpm(dge$counts, log = TRUE)

# alignment to sex chromosome complement 
dge.scc <- readRDS("../../../../rObjects/dge.raw.rds")
lcpm.scc <- edgeR::cpm(dge.scc$counts, log = TRUE)
```
# Default alignment - Sex check
View the expression of sex-linked genes to determine if samples are correctly annotated as XX or XY. 
Genetic sex is not gender. 
```{r}
genes_and_counts <- cbind(dge$genes$gene_name, lcpm)
genes_and_counts <- as.data.frame(genes_and_counts)
names(genes_and_counts)[names(genes_and_counts) == "V1"] <- "Geneid"
rownames(genes_and_counts) <- NULL

sex_genes <- c("XIST", "EIF1AY", "KDM5D", "UTY", "DDX3Y", "RPS4Y1")
sex_genes_counts <- subset(genes_and_counts, Geneid %in% sex_genes)
rownames(sex_genes_counts) <- sex_genes_counts$Geneid
sex_genes_counts$Geneid <- NULL
sex_gene_df <- as.data.frame(t(sex_genes_counts))
sex_gene_df$NPID <- rownames(sex_gene_df)
rownames(sex_gene_df) <- NULL

meta_short <- metadata[,c("NPID","Sex")] 
sex_gene_meta <- merge(sex_gene_df, meta_short, by = "NPID")
cols.num <- c("XIST","EIF1AY", "KDM5D", "UTY", "DDX3Y", "RPS4Y1")
sex_gene_meta[cols.num] <- sapply(sex_gene_meta[cols.num], as.integer)
sex_gene_meta <- sex_gene_meta[,c(2,3,4,5,6,7,8,1)]
# replace M with male and F with female 
sex_gene_meta <- sex_gene_meta %>% 
  mutate(
    Sex = ifelse(Sex %in% c("M"), "male", "female") 
  )
```

### Format sex check for plotting
```{r}
genes_counts <- reshape2::melt(genes_and_counts, id=c("Geneid"))
names(genes_counts)[names(genes_counts) == "variable"] <- "NPID"

df <- cbind(metadata$NPID, metadata$Sex)
df <- as.data.frame(df)
names(df)[names(df) == "V1"] <- "NPID"
names(df)[names(df) == "V2"] <- "Sex"

data <- merge(genes_counts, df, by = "NPID")

sexGenes <- c("DDX3X, DDX3Y")
SelectGenes_counts <-
  subset(
    data,
    Geneid %in% c(
      "DDX3X",
      "DDX3Y",
      "ZFX",
      "ZFY",
      "USP9X",
      "USP9Y",
      "KDM6A",
      "UTY",
      "PCDH11X",
      "PCDH11Y",
      "XIST",
      "SRY"
    )
  )
SelectGenes_counts[, "geneComb"] <- NA
SelectGenes_counts[, "group"] <- NA


SelectGenes_counts$geneComb <-
  ifelse(
    SelectGenes_counts$Geneid == "DDX3X",
    "DDX3X:DDX3Y",
    ifelse(
      SelectGenes_counts$Geneid == "DDX3Y",
      "DDX3X:DDX3Y",
      ifelse(
        SelectGenes_counts$Geneid == "ZFX",
        "ZFX:ZFY",
        ifelse(
          SelectGenes_counts$Geneid == "ZFY",
          "ZFX:ZFY",
          ifelse(
            SelectGenes_counts$Geneid == "USP9X",
            "USP9X:USP9Y",
            ifelse(
              SelectGenes_counts$Geneid == "USP9Y",
              "USP9X:USP9Y",
              ifelse(
                SelectGenes_counts$Geneid == "KDM6A",
                "UTX:UTY",
                ifelse(
                  SelectGenes_counts$Geneid == "UTY",
                  "UTX:UTY",
                  ifelse(
                    SelectGenes_counts$Geneid == "PCDH11X",
                    "PCDH11X:PCDH11Y",
                    ifelse(
                      SelectGenes_counts$Geneid == "PCDH11Y",
                      "PCDH11X:PCDH11Y",
                      ifelse(
                        SelectGenes_counts$Geneid == "XIST",
                        "XIST",
                        ifelse(SelectGenes_counts$Geneid == "SRY", "SRY", "NA")
                      )
                    )
                  )
                )
              )
            )
          )
        )
      )
    )
  )

SelectGenes_counts$group <-
  ifelse(
    SelectGenes_counts$geneComb == "DDX3X:DDX3Y",
    1,
    ifelse(
      SelectGenes_counts$geneComb == "ZFX:ZFY",
      4,
      ifelse(
        SelectGenes_counts$geneComb == "USP9X:USP9Y",
        3,
        ifelse(
          SelectGenes_counts$geneComb == "UTX:UTY",
          5,
          ifelse(
            SelectGenes_counts$geneComb == "PCDH11X:PCDH11Y",
            2,
            ifelse(
              SelectGenes_counts$geneComb == "XIST",
              6,
              ifelse(SelectGenes_counts$geneComb == "SRY", 7, "NA")
            )
          )
        )
      )
    )
  )
data <- SelectGenes_counts
data$value <- as.numeric(data$value)
```

```{r}
datav2 <- merge(data, metadata, by = "NPID")
```
### Plot sex check
```{r}
leg_lab <- "reported sex"
cbPaletteJITTER = SexColors

default_sex_plot <- ggplot(datav2, aes(x = Geneid, y = value)) +
  geom_jitter(aes(color = sex_inferred, shape = sex_inferred),
              width = 0.25,
              size = 1) +
  scale_color_manual(leg_lab, values = cbPaletteJITTER) + 
  scale_shape_manual(leg_lab, values = c(19, 15)) +
  labs(x = "", y = "lcpm", title = "") +
  facet_grid(
    . ~ group.x + geneComb,
    switch = "x",
    scales = "free_x",
    space = "free_x"
  ) +
  theme_bw() +
  theme(
    strip.background = element_blank(),
    strip.text.x = element_blank(),
    axis.ticks = element_blank(),
    axis.text.x = element_text(face = "italic", size = 7),
    strip.text = element_text(size = 8), 
    axis.text.y = element_text(size = 8),
    axis.title.y = element_text(size = 8),
    legend.text = element_text(size = 8),
    legend.title = element_text(size = 8),
    legend.margin=margin(0,0.5,0,0),
    legend.box.margin=margin(-10,-2,-10,-7.5), 
    plot.margin = margin(0.1, 0.1, 0.3, 0.1, "cm"), #t,r,b,l
    panel.border = element_rect(
      colour = "black",
      fill = NA,
      size = 1
    )
  ) +
  scale_y_continuous(breaks = seq(-5, 10, by = 1), limits = c(-5, 10))
default_sex_plot
```
clean up 
```{r}
remove(
  data,
  datav2,
  df,
  dge,
  genes_and_counts,
  genes_counts,
  lcpm,
  meta_short,
  SelectGenes_counts,
  sex_gene_df,
  sex_gene_meta,
  sex_genes_counts
)
```
# Sex chromosome complement alignment 
View the expression of sex-linked genes to determine if samples are correctly annotated as XX or XY. 
Genetic sex is not gender. 
```{r}
genes_and_counts <- cbind(dge.scc$genes$gene_name, lcpm.scc)
genes_and_counts <- as.data.frame(genes_and_counts)
names(genes_and_counts)[names(genes_and_counts) == "V1"] <- "Geneid"
rownames(genes_and_counts) <- NULL

sex_genes <- c("XIST", "EIF1AY", "KDM5D", "UTY", "DDX3Y", "RPS4Y1")
sex_genes_counts <- subset(genes_and_counts, Geneid %in% sex_genes)
rownames(sex_genes_counts) <- sex_genes_counts$Geneid
sex_genes_counts$Geneid <- NULL
sex_gene_df <- as.data.frame(t(sex_genes_counts))
sex_gene_df$NPID <- rownames(sex_gene_df)
rownames(sex_gene_df) <- NULL

meta_short <- metadata[,c("NPID","Sex")] 
sex_gene_meta <- merge(sex_gene_df, meta_short, by = "NPID")
cols.num <- c("XIST","EIF1AY", "KDM5D", "UTY", "DDX3Y", "RPS4Y1")
sex_gene_meta[cols.num] <- sapply(sex_gene_meta[cols.num], as.integer)
sex_gene_meta <- sex_gene_meta[,c(2,3,4,5,6,7,8,1)]
# replace M with male and F with female 
sex_gene_meta <- sex_gene_meta %>% 
  mutate(
    Sex = ifelse(Sex %in% c("M"), "male", "female") 
  )
```

### Format sex check for plotting
```{r}
genes_counts <- reshape2::melt(genes_and_counts, id=c("Geneid"))
names(genes_counts)[names(genes_counts) == "variable"] <- "NPID"

df <- cbind(metadata$NPID, metadata$Sex)
df <- as.data.frame(df)
names(df)[names(df) == "V1"] <- "NPID"
names(df)[names(df) == "V2"] <- "Sex"

data <- merge(genes_counts, df, by = "NPID")

sexGenes <- c("DDX3X, DDX3Y")
SelectGenes_counts <-
  subset(
    data,
    Geneid %in% c(
      "DDX3X",
      "DDX3Y",
      "ZFX",
      "ZFY",
      "USP9X",
      "USP9Y",
      "KDM6A",
      "UTY",
      "PCDH11X",
      "PCDH11Y",
      "XIST",
      "SRY"
    )
  )
SelectGenes_counts[, "geneComb"] <- NA
SelectGenes_counts[, "group"] <- NA


SelectGenes_counts$geneComb <-
  ifelse(
    SelectGenes_counts$Geneid == "DDX3X",
    "DDX3X:DDX3Y",
    ifelse(
      SelectGenes_counts$Geneid == "DDX3Y",
      "DDX3X:DDX3Y",
      ifelse(
        SelectGenes_counts$Geneid == "ZFX",
        "ZFX:ZFY",
        ifelse(
          SelectGenes_counts$Geneid == "ZFY",
          "ZFX:ZFY",
          ifelse(
            SelectGenes_counts$Geneid == "USP9X",
            "USP9X:USP9Y",
            ifelse(
              SelectGenes_counts$Geneid == "USP9Y",
              "USP9X:USP9Y",
              ifelse(
                SelectGenes_counts$Geneid == "KDM6A",
                "UTX:UTY",
                ifelse(
                  SelectGenes_counts$Geneid == "UTY",
                  "UTX:UTY",
                  ifelse(
                    SelectGenes_counts$Geneid == "PCDH11X",
                    "PCDH11X:PCDH11Y",
                    ifelse(
                      SelectGenes_counts$Geneid == "PCDH11Y",
                      "PCDH11X:PCDH11Y",
                      ifelse(
                        SelectGenes_counts$Geneid == "XIST",
                        "XIST",
                        ifelse(SelectGenes_counts$Geneid == "SRY", "SRY", "NA")
                      )
                    )
                  )
                )
              )
            )
          )
        )
      )
    )
  )

SelectGenes_counts$group <-
  ifelse(
    SelectGenes_counts$geneComb == "DDX3X:DDX3Y",
    1,
    ifelse(
      SelectGenes_counts$geneComb == "ZFX:ZFY",
      4,
      ifelse(
        SelectGenes_counts$geneComb == "USP9X:USP9Y",
        3,
        ifelse(
          SelectGenes_counts$geneComb == "UTX:UTY",
          5,
          ifelse(
            SelectGenes_counts$geneComb == "PCDH11X:PCDH11Y",
            2,
            ifelse(
              SelectGenes_counts$geneComb == "XIST",
              6,
              ifelse(SelectGenes_counts$geneComb == "SRY", 7, "NA")
            )
          )
        )
      )
    )
  )
data <- SelectGenes_counts
data$value <- as.numeric(data$value)
```

```{r}
datav2 <- merge(data, metadata, by = "NPID")
```
### Plot sex check
```{r}
leg_lab <- "inferred sex"
cbPaletteJITTER = SexColors

SCC_sex_plot <- ggplot(datav2, aes(x = Geneid, y = value)) +
  geom_jitter(aes(color = sex_inferred, shape = sex_inferred),
              width = 0.25,
              size = 1) +
  scale_color_manual(leg_lab, values = cbPaletteJITTER) + 
  scale_shape_manual(leg_lab, values = c(19, 15)) +
  labs(x = "", y = "lcpm", title = "") +
  facet_grid(
    . ~ group.x + geneComb,
    switch = "x",
    scales = "free_x",
    space = "free_x"
  ) +
  theme_bw() +
  theme(
    strip.background = element_blank(),
    strip.text.x = element_blank(),
    axis.ticks = element_blank(),
    axis.text.x = element_text(face = "italic", size = 7),
    strip.text = element_text(size = 8), 
    axis.text.y = element_text(size = 8),
    axis.title.y = element_text(size = 8),
    legend.text = element_text(size = 8),
    legend.title = element_text(size = 8),
    legend.margin=margin(0,0.5,0,0),
    legend.box.margin=margin(-10,-2,-10,-7.5), 
    plot.margin = margin(0.1, 0.1, 0.3, 0.1, "cm"), #t,r,b,l
    panel.border = element_rect(
      colour = "black",
      fill = NA,
      size = 1
    )
  ) +
  scale_y_continuous(breaks = seq(-5, 10, by = 1), limits = c(-5, 10))
SCC_sex_plot
```
clean up 
```{r}
remove(
  data,
  datav2,
  df,
  dge.scc,
  genes_and_counts,
  genes_counts,
  lcpm.scc,
  meta_short,
  SelectGenes_counts,
  sex_gene_df,
  sex_gene_meta,
  sex_genes_counts
)
```
# Combined plot
```{r}
combind <-
  ggarrange(
    default_sex_plot,
    SCC_sex_plot,
    nrow = 2,
    labels = c("a)", "b)"), 
    font.label = list(size = 8)
    )
combind

path <- paste0("../../../../results/manuscript_figures/Supplemental_Sex_check_default_and_SCC_alignment")
saveToPDF(paste0(path, ".pdf"), width = 8.5, height = 11)
```
```{r}
```


