---
title: "Differential expression in LBD samples"
author: "Kimberly Olney, Ph.D."
date: "01/05/2023"
output:
  html_document:
    df_print: paged
  pdf_document: default
params:
  args: myarg
---

# Setup
```{r setup}
# Also do Session > Set Working Directory > Choose Directory
knitr::opts_knit$set(root.dir = ".")
```

# User defined variables
```{r set_variables}
source(here::here("scripts/R", "file_paths_and_colours.R"))
condition <- c("all_genes_default_alignment")
myContrasts <- c("LBDvsControl = LBD - CONTROL",
  "LBDvsAD = LBD - AD",
  "LBDvsPA = LBD - PA",
  "ADvsControl = AD - CONTROL", 
  "PAvsControl = PA - CONTROL",
  "ADvsPA = AD - PA")
tool = c("star")
typeOfCount <-  c("_STAR.bamReadsPerGene.out.tab")
```

# Read data
Samples are aligned to a default human reference genome GRCh38. See Snakemake pipeline for details. 
Read in the STAR aligned counts files for each sample. 
```{r read_data}
# read in metadata
metadata <- vroom(paste0(pathToRawData, "metadata_cleaned.tsv"))
metadata$TYPE <- gsub(" - ", "_", metadata$TYPE) 
# RNA integrity number (RIN)
RIN <- vroom(paste0(pathToRawData, "RIN.tsv"))
RIN$NPID <- RIN$Sample
# Add RIN information to metadata
metadata <- merge(metadata, RIN, by = "NPID", all = TRUE)

# read in snakemake sample information
sample_read_group_info <-
  read.delim("../snakemake/sample_read_group_info.txt", header = FALSE)
sample_read_group_info_split <-
  as.data.frame(str_split_fixed(sample_read_group_info$V1, "_R1", 2))
counts_ids <-
  as.data.frame(gsub("\\.", "_", sample_read_group_info_split$V1))
names(counts_ids)[1] <- "V1"
counts_ids <- as.data.frame(str_split_fixed(counts_ids$V1, "_", 2))
names(counts_ids)[names(counts_ids) == 'V1'] <- 'NPID'
names(counts_ids)[names(counts_ids) == 'V2'] <- 'flow_lane'
# merge sample_count_id and metadata files
counts_metadata <- merge(metadata, counts_ids, by = "NPID")

# path to counts files
count_files <-
  file.path(paste0(
    "../../starAligned/",
    counts_metadata$NPID,
    "_",
    counts_metadata$flow_lane,
    typeOfCount
  ))
# add sample name to counts files
names(count_files) <- paste0(counts_metadata$NPID)
# add sample count path information to metadata
counts_metadata$path <- count_files

# create a counts matrix
counts <- data.frame(fread(count_files[1]))[c(1, 2)]
# Loop and read the 2nd column remaining files
for (i in 2:length(count_files)) {
  counts <- cbind(counts, data.frame(fread(count_files[i]))[2])
}
```
# Star mapping metrics
```{r}
star_metrics <- data.frame(counts[counts$V1 %like% "N_", ])
star_metrics$V1 <- as.vector(star_metrics$V1)
melt_star_metrics <- reshape::melt(star_metrics, id = c("V1"))
ggplot(melt_star_metrics, aes(x = value, fill = V1)) + 
  geom_histogram(position = "identity", bins = 100) + 
  facet_grid(V1 ~ .)  + theme_bw()
```
### Clean up
```{r}
# clean up
remove(melt_star_metrics, 
       metadata, 
       counts_ids, 
       params, 
       RIN, 
       sample_read_group_info, 
       sample_read_group_info_split, 
       star_metrics)
```
### Format counts table
```{r}
# remove star metric information
counts <- counts[!grepl("N_", counts$V1),]
# set gene_id (i.e. V1) as the row name
row.names(counts) <- counts$V1
counts$V1 <- NULL
# set column names to NPID sample ID
colnames(counts) <- names(count_files)
```
# Get gene information
```{r}
# read in annotation file
gtf.file <- paste0(pathToRef, "gencode.v38.annotation.gtf")
genes.gtf <- rtracklayer::import(gtf.file)
genes.gtf <- as.data.frame(genes.gtf)
genes.gtf <- genes.gtf[genes.gtf$type == "gene",]
gene_type_table <- table(genes.gtf$gene_type)
write.table(gene_type_table, "gene_type_table.tsv", row.names = F, quote = F, sep = "\t")
```
# Check inputs
```{r}
all.equal(rownames(counts), genes.gtf$gene_id)
all.equal(colnames(counts), (counts_metadata$NPID))
```
# Create DGE object
```{r DGE_object}
# create object
dge <- DGEList(counts = counts,
               samples = counts_metadata,
               genes = genes.gtf)

table(dge$samples$TYPE)
```
# Save R object
```{r, eval=FALSE}
saveRDS(dge, file = paste0("../../rObjects/", condition, "_dge.raw.rds"))
```

# Remove mitochondrial genes
```{r MT_genes}
dim(dge)
removeMT <- dge$genes$seqnames != "chrM"  # true when NOT MT
dge <- dge[removeMT,,keep.lib.sizes = FALSE]
dim(dge)
```

# Raw MDS (multidimensional scaling)
convert counts to cpm and lcpm
set colors and get data
```{r cpm, warning=FALSE}
lcpm <- edgeR::cpm(dge$counts, log = TRUE)
cpm <- edgeR::cpm(dge$counts, log = FALSE)
# sex colors 
dge$samples$Sex <- as.factor(dge$samples$Sex)
sex_colors <- c(SexColors)[dge$samples$Sex]
```

```{r}
par(bg = 'white')

# plot MDS
MDS_1 <- plotMDS(
  lcpm,
  top = 100, 
  labels = dge$samples$Sex,
  cex = .8, 
  dim.plot = c(1,2), 
  plot = TRUE, 
  col = sex_colors,
  gene.selection = "common"
)
title(expression('Top 100 Genes (Log'[2]~'CPM)'))

path <- paste0("../../results/", tool, "/MDS/", 
               condition,"_gene_MDS_techreps_label_sex_dim1&2")
saveToPDF(paste0(path, ".pdf"), width = 5, height = 5)

# plot MDS
MDS_2 <- plotMDS(
  lcpm,
  top = 100, 
  labels = dge$samples$Sex,
  cex = .8, 
  dim.plot = c(2,3), 
  plot = TRUE, 
  col = sex_colors,
  gene.selection = "common"
)
title(expression('Top 100 Genes (Log'[2]~'CPM)'))
path <- paste0("../../results/", tool, "/MDS/", 
               condition,"_gene_MDS_techreps_label_sex_dim2&3")
saveToPDF(paste0(path, ".pdf"), width = 5, height = 5)
```
# Sex check
View the expression of sex-linked genes to determine if samples are correctly annoated as XX or XY. 
Genetic sex is not gender. 
```{r}
genes_and_counts <- cbind(dge$genes$gene_name, dge$counts)
genes_and_counts <- as.data.frame(genes_and_counts)
names(genes_and_counts)[names(genes_and_counts) == "V1"] <- "Geneid"
rownames(genes_and_counts) <- NULL

sex_genes <- c("XIST", "EIF1AY", "KDM5D", "UTY", "DDX3Y", "RPS4Y1")
sex_genes_counts <- subset(genes_and_counts, Geneid %in% sex_genes)
rownames(sex_genes_counts) <- sex_genes_counts$Geneid
sex_genes_counts$Geneid <- NULL
sex_gene_df <- as.data.frame(t(sex_genes_counts))
sex_gene_df$NPID <- rownames(sex_gene_df)
rownames(sex_gene_df) <- NULL

meta_short <- counts_metadata[,c("NPID","Sex")] 
sex_gene_meta <- merge(sex_gene_df, meta_short, by = "NPID")
cols.num <- c("XIST","EIF1AY", "KDM5D", "UTY", "DDX3Y", "RPS4Y1")
sex_gene_meta[cols.num] <- sapply(sex_gene_meta[cols.num], as.integer)
sex_gene_meta <- sex_gene_meta[,c(2,3,4,5,6,7,8,1)]
# replace M with male and F with female 
sex_gene_meta <- sex_gene_meta %>% 
  mutate(
    Sex = ifelse(Sex %in% c("M"), "male", "female") 
  )
```
# Sex inference model
Infer the genetic sex of the sample using GTEx data as a reference or typical sex-linked gene expression. 
```{r}
# for_rna_sex_check.tsv is GTEx expression data 
training_model_data <- "for_rna_sex_check.tsv"

# Build the model using the GTEx data
# Load the data and remove NAs
train_data <- read.csv(training_model_data, sep="\t")

# Split the data into training and test set
set.seed(123)
training.samples <- train_data$sex %>% 
  createDataPartition(p = 0.8, list = FALSE)
train.data  <- train_data[training.samples, ]
test.data <- train_data[-training.samples, ]

# Dummy code categorical predictor variables
x <- model.matrix(sex~., train.data)[,-1]
# Convert the outcome (class) to a numerical variable
y <- ifelse(train.data$sex == "female", 1, 0)

cv.lasso <- cv.glmnet(x, y, alpha = 1, family = "binomial")
plot(cv.lasso)
cv.lasso$lambda.min

coef(cv.lasso, cv.lasso$lambda.min)
coef(cv.lasso, cv.lasso$lambda.1se)

# Final model with lambda.min
# Glmnet is a package that fits generalized linear and similar models via penalized maximum likelihood. 
lasso.model <- glmnet(x, y, alpha = 1, family = "binomial",
                      lambda = cv.lasso$lambda.min)
# Make prediction on test data
x.test <- model.matrix(sex ~., test.data)[,-1]
probabilities <- lasso.model %>% predict(newx = x.test)
predicted.classes <- ifelse(probabilities > 0.5, "female", "male")
# Model accuracy
observed.classes <- test.data$sex
mean(predicted.classes == observed.classes)

# ----------------------
# Run on experiment data
# ----------------------
NPID <- sex_gene_meta$NPID 
sex_gene_meta$NPID <- NULL
typeof(sex_gene_meta$XIST)
test.experiment <- model.matrix(Sex ~., sex_gene_meta)[,-1]

probabilities <- lasso.model %>% predict(newx = test.experiment)
predicted.classes <- ifelse(probabilities > 4, "female", "male")
# Model accuracy
observed.classes <- sex_gene_meta$Sex
mean(predicted.classes == observed.classes)
sex_check <- cbind(as.data.frame(predicted.classes)$s0, sex_gene_meta$Sex)
sex_check <- as.data.frame(sex_check)
# Rename column where names is "Sepal.Length"
names(sex_check)[names(sex_check) == "V1"] <- "reported"
names(sex_check)[names(sex_check) == "V2"] <- "observed"

# add back in sample IDs
sex_check$NPID <- NPID

# what doesn't match between reported and observed? 
sex_check$sex_check <- ifelse(sex_check$reported == sex_check$observed,"Yes","No")
# output results 
write.table(
  sex_check,
  paste0("../../results/",
         tool,
         "/sex_check/sex_check_RNA.txt"),
  sep = "\t",
  row.names = FALSE,
  quote = FALSE
)

check_samples <- subset(sex_check, sex_check == "No")
samples_to_investigate <- subset(sex_gene_df, NPID %in% check_samples$NPID)
sex_gene_df[cols.num] <- sapply(sex_gene_df[cols.num], as.numeric)
sapply(sex_gene_df, class)
low_coverage_female_samples <- subset(sex_gene_meta, XIST < 5 & Sex == "female")
low_coverage_male_samples <- subset(sex_gene_meta, DDX3Y < 5 & Sex == "male" | 
                                      RPS4Y1 < 5 & Sex == "male" | 
                                      UTY < 5 & Sex == "male" |
                                      KDM5D < 5 & Sex == "male" |
                                      EIF1AY < 5 & Sex == "male")
atest <- cbind(sex_gene_df, sex_gene_meta)
low_coverage_samples_to_investigate <- subset(sex_gene_df, 
                                    NPID %in% low_coverage_female_samples$NPID)

```
# Clean up
```{r}
remove(genes_and_counts, 
       cv.lasso, 
       lasso.model, 
       probabilities, 
       predicted.classes,
       test.data, 
       test.experiment, 
       train_data, 
       train.data, 
       training.samples, 
       x, 
       x.test, 
       cols.num, 
       observed.classes, 
       sex_genes, 
       y,
       meta_short)
```
# Format sex check for plotting
```{r}
genes_and_counts <- cbind(dge$genes$gene_name, lcpm)
genes_and_counts <- as.data.frame(genes_and_counts)
names(genes_and_counts)[names(genes_and_counts) == "V1"] <- "Geneid"
rownames(genes_and_counts)<-NULL
genes_counts <- reshape2::melt(genes_and_counts, id=c("Geneid"))
names(genes_counts)[names(genes_counts) == "variable"] <- "NPID"

df <- cbind(counts_metadata$NPID, counts_metadata$Sex)
df <- as.data.frame(df)
names(df)[names(df) == "V1"] <- "NPID"
names(df)[names(df) == "V2"] <- "Sex"

data <- merge(genes_counts, df, by = "NPID")

sexGenes <- c("DDX3X, DDX3Y")
SelectGenes_counts <-
  subset(
    data,
    Geneid %in% c(
      "DDX3X",
      "DDX3Y",
      "ZFX",
      "ZFY",
      "USP9X",
      "USP9Y",
      "KDM6A",
      "UTY",
      "PCDH11X",
      "PCDH11Y",
      "XIST",
      "SRY"
    )
  )
SelectGenes_counts[, "geneComb"] <- NA
SelectGenes_counts[, "group"] <- NA


SelectGenes_counts$geneComb <-
  ifelse(
    SelectGenes_counts$Geneid == "DDX3X",
    "DDX3X:DDX3Y",
    ifelse(
      SelectGenes_counts$Geneid == "DDX3Y",
      "DDX3X:DDX3Y",
      ifelse(
        SelectGenes_counts$Geneid == "ZFX",
        "ZFX:ZFY",
        ifelse(
          SelectGenes_counts$Geneid == "ZFY",
          "ZFX:ZFY",
          ifelse(
            SelectGenes_counts$Geneid == "USP9X",
            "USP9X:USP9Y",
            ifelse(
              SelectGenes_counts$Geneid == "USP9Y",
              "USP9X:USP9Y",
              ifelse(
                SelectGenes_counts$Geneid == "KDM6A",
                "UTX:UTY",
                ifelse(
                  SelectGenes_counts$Geneid == "UTY",
                  "UTX:UTY",
                  ifelse(
                    SelectGenes_counts$Geneid == "PCDH11X",
                    "PCDH11X:PCDH11Y",
                    ifelse(
                      SelectGenes_counts$Geneid == "PCDH11Y",
                      "PCDH11X:PCDH11Y",
                      ifelse(
                        SelectGenes_counts$Geneid == "XIST",
                        "XIST",
                        ifelse(SelectGenes_counts$Geneid == "SRY", "SRY", "NA")
                      )
                    )
                  )
                )
              )
            )
          )
        )
      )
    )
  )

SelectGenes_counts$group <-
  ifelse(
    SelectGenes_counts$geneComb == "DDX3X:DDX3Y",
    1,
    ifelse(
      SelectGenes_counts$geneComb == "ZFX:ZFY",
      4,
      ifelse(
        SelectGenes_counts$geneComb == "USP9X:USP9Y",
        3,
        ifelse(
          SelectGenes_counts$geneComb == "UTX:UTY",
          5,
          ifelse(
            SelectGenes_counts$geneComb == "PCDH11X:PCDH11Y",
            2,
            ifelse(
              SelectGenes_counts$geneComb == "XIST",
              6,
              ifelse(SelectGenes_counts$geneComb == "SRY", 7, "NA")
            )
          )
        )
      )
    )
  )
data <- SelectGenes_counts
data$value <- as.numeric(data$value)
```
# Plot sex check
```{r}
wrong_sex_samples <-
  data[data$NPID %in% samples_to_investigate$NPID, ]
wrong_sex_samples <- unique(wrong_sex_samples)
low_coverage_samples <-
  data[data$NPID %in% low_coverage_samples_to_investigate$NPID, ]
wrong_sex_samples <- unique(wrong_sex_samples)
leg_lab <- "reported sex"
cbPaletteJITTER = SexColors
write.table(data, paste0("../../results/", tool, 
                         "/sex_check/sex_check_data.txt"), sep = "\t")
write.table(wrong_sex_samples, paste0("../../results/", tool, 
                                      "/sex_check/wrong_sex_samples.txt"), sep = "\t")

geneticSEXgenes_plot <- ggplot(data, aes(x = Geneid, y = value)) +
  geom_jitter(aes(color = Sex, shape = Sex),
              width = 0.25,
              size = 1) +
  scale_color_manual(leg_lab, values = cbPaletteJITTER) + 
  scale_shape_manual(leg_lab, values = c(19, 15)) +
  labs(x = "", y = "lcpm", title = "") +
  facet_grid(
    . ~ group + geneComb,
    switch = "x",
    scales = "free_x",
    space = "free_x"
  ) +
  theme_bw() +
  theme(
    strip.background = element_blank(),
    strip.text.x = element_blank(),
    axis.ticks = element_blank(),
    axis.text.x = element_text(face = "italic"),
    panel.border = element_rect(
      colour = "black",
      fill = NA,
      size = 1
    )
  ) +
  scale_y_continuous(breaks = seq(-4, 9, by = 1), limits = c(-4, 9)) +
  geom_text_repel(
    data = wrong_sex_samples,
    aes(x = Geneid,
        y = value,
        label = NPID),
    color = "black",
    size = 3,
    nudge_x = .5,
    min.segment.length = unit(0, 'lines'),
    max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
  )

geneticSEXgenes_plot
path <-
  paste0("../../results/", tool, "/sex_check/sex_check_samples_labeled")
saveToPDF(paste0(path, ".pdf"), width = 11, height = 6)
```
### Clean up 
```{r}
remove(df, 
       geneticSEXgenes_plot,
       SelectGenes_counts, 
       sex_check, 
       sex_gene_df,
       sex_gene_meta,
       wrong_sex_samples)
```
# RIN check with replicates
```{r}
# first filter by expression and normalize the data
keep.expr <- filterByExpr(dge, group = dge$samples$TYPE)
dim(dge)
dge.filtered.reps <- dge[keep.expr, , keep.lib.sizes = FALSE]

dim(dge.filtered.reps)
table(dge.filtered.reps$genes$gene_type)

# Now, normalization by the method of trimmed mean of M-values (TMM)
dge.filtered.reps.norm <- calcNormFactors(dge.filtered.reps, method = "TMM")

# norm factor summary
summary(dge.filtered.reps.norm$samples$norm.factors)
log2cpm.norm.reps <- edgeR::cpm(dge.filtered.reps.norm, log = T)
nsamples <- ncol(dge.filtered.reps.norm)
boxplot(log2cpm.norm.reps, 
        main="Filtered normalized lcpm data with replicates", 
        xlab="RIN", 
        ylab=expression('Counts per gene (Log'[2]~'CPM)'),
        axes=FALSE)
axis(2)
axis(1,at=1:nsamples,labels=(dge.filtered.reps.norm$samples$RIN),las=2,cex.axis=0.8)

path <- paste0("../../results/", tool, "/library/", condition, "_gene_boxplot_RIN")
saveToPDF(paste0(path, ".pdf"), width = 35, height = 6)
```

# Sum technical replicates
Each sample was sequenced on two different lanes. 
Samples have already been combined together during the alignment process. 
See snakemake pipeline for details. 
```{r techReps}
dim(dge)
dge.tech <- sumTechReps(dge, dge$samples$NPID)
dim(dge.tech$counts)
colnames(dge.tech$counts) <- dge.tech$samples$NPID

dge.tech$samples$TYPE <- as.factor(dge.tech$samples$TYPE)
group_colors <- c(TypeColors)[dge.tech$samples$TYPE]

lcpm <- edgeR::cpm(dge.tech$counts, log = TRUE)
```

# Raw MDS
```{r}
par(bg = 'white')

# plot MDS
plotMDS(
  lcpm,
  top = 500, 
  labels = dge.tech$samples$Sex,
  cex = .8, 
  dim.plot = c(1,2), 
  plot = TRUE, 
  col = group_colors,
  gene.selection = "common"
)
title(expression('Top 500 Genes (Log'[2]~'CPM)'))

path <- paste0("../../results/", tool, "/MDS/", 
               condition,"_label_sex_dim1&2")
saveToPDF(paste0(path, ".pdf"), width = 5.2, height = 5.2)

# plot MDS
plotMDS(
  lcpm,
  top = 500, 
  labels = dge.tech$samples$Sex,
  cex = .8, 
  dim.plot = c(2,3), 
  plot = TRUE, 
  col = group_colors,
  gene.selection = "common"
)
title(expression('Top 500 Genes (Log'[2]~'CPM)'))
path <- paste0("../../results/", tool, "/MDS/", 
               condition,"_label_sex_dim2&3")
saveToPDF(paste0(path, ".pdf"), width = 5.2, height = 5.2)
```

# Filter lowly expressed genes
The filterByExpr() function in the edgeR package determines which genes have a 
great enough count value to keep.  We will filter by group.  This means at least 
x samples (x is the smallest group sample size) must express a minimum count of 
10 (in cpm, default value). 

```{r filter}
keep.expr <- filterByExpr(dge.tech, group = dge.tech$samples$TYPE)
dim(dge.tech)
dge.filtered <- dge.tech[keep.expr, , keep.lib.sizes = FALSE]
help(filterByExpr)
dim(dge.filtered)
table(dge.filtered$genes$gene_type)
```

# TMM normalization
For estimating relative RNA production levels from RNA-seq data. 
The TMM method estimates scale factors between samples that can be incorporated 
into currently used statistical methods for DE analysis.
```{r TMM_normalize}
# Now, normalization by the method of trimmed mean of M-values (TMM)
dge.filtered.norm <- calcNormFactors(dge.filtered, method = "TMM")

# norm factor summary
summary(dge.filtered.norm$samples$norm.factors)
```

# Density plot
Density plots of log-intensity distribution of each library can be superposed 
on a single graph for a better comparison between libraries and for 
identification of libraries with weird distribution. 
```{r density_plots}
# set graphical parameter
par(mfrow = c(1,3))

# Normalize data for library size and expression intensity
log2cpm.tech <- edgeR::cpm(dge.tech, log = TRUE)
log2cpm.filtered <- edgeR::cpm(dge.filtered, log = TRUE)
log2cpm.norm <- edgeR::cpm(dge.filtered.norm, log = TRUE)

# set colors
colors <- c("red","orange","green","yellow","blue","purple", 
            "lightgray","brown","pink","cyan")
nsamples <- ncol(dge.tech)

# First, plot the first column of the log2cpm.tech density
plot(density(log2cpm.tech[,1]), col = colors[1], lwd = 2, ylim = c(0,0.45), 
     las = 2, main = "A. Raw", xlab = expression('Log'[2]~CPM))

# For each sample plot the lcpm density
for (i in 2:nsamples){
  den <- density(log2cpm.tech[,i]) #subset each column
  lines(den$x, den$y, col = colors[i], lwd = 2) 
}

# Second, plot log2cpm.filtered
plot(density(log2cpm.filtered[,1]), col = colors[1], lwd = 2, ylim = c(0,0.45), 
     las = 2, main = "B. Filtered", xlab = expression('Log'[2]~CPM))
abline(v = edgeR::cpm(3, log = TRUE), lty = 3)
for (i in 2:nsamples) {
  den <- density(log2cpm.filtered[,i])
  lines(den$x, den$y, col = colors[i], lwd = 2)
}

# Third, plot log2cpm.norm
plot(density(log2cpm.norm[,1]), col = colors[1], lwd = 2, ylim = c(0,0.45), 
     las = 2, main = "C. Normalized", xlab = expression('Log'[2]~CPM))
abline(v = edgeR::cpm(3, log = TRUE), lty = 3)
for (i in 2:nsamples) {
  den <- density(log2cpm.norm[,i])
  lines(den$x, den$y, col = colors[i], lwd = 2)
}

# save
path <- paste0("../../results/", tool, "/density/", condition, "_gene_density")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 4)
```

# RIN check 
check if there is correlation between RIN and library size
```{r}
box <- dge.filtered.norm$samples
cor(box$RIN, box$lib.size, method = c("pearson", "kendall", "spearman"))
cor.test(box$RIN, box$lib.size, method=c("pearson", "kendall", "spearman"))

# is the data normally distributed?
ggqqplot(box$lib.size, ylab = "library size")
ggqqplot(box$RIN, ylab = "RIN")
res <- cor.test(box$lib.size, box$RIN, 
                method = "pearson")
res

ggscatter(box, x = "RIN", y = "lib.size", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          ylab = "library size", xlab = "RIN value") 

path <- paste0("../../results/", tool, "/library/", condition, "_corr_RIN_lib_size")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 6)
```

```{r}
remove(box,
       res)
```

# Variance partition
CCA Heatmap
```{r CCA_heatmap}
form <- ~ TYPE +
PathDx +
AD.subtype +
LBD.type +
CDLB +
Braak.NFT +
Thal.amyloid +
MF.SP +
MF.NFT +
MF.LB +
Cing.LB +
MF.Amyloid +
MF.Tau +
Cing.Synuclein +
CWOW.Category +
VaD  +
TDP.type +
Brain.wt +
ClinicalDx +
FHx +
Duration +
Sex +
Age +
Race +
PMI +
APOE +
MAPT +
GRN +
TMEM106b +
RIN +
Total.RNA.ng 
# Compute Canonical Correlation Analysis (CCA)
# between all pairs of variables
# returns absolute correlation value
info <- as.data.frame(dge.filtered.norm$samples)
C = canCorPairs( form, info)

# replace NA with zero
C[is.na(C)] = 0
cor.mtest <- function(C, ...) {
    C <- as.matrix(C)
    n <- ncol(C)
    p.mat<- matrix(NA, n, n)
    diag(p.mat) <- 0
    for (i in 1:(n - 1)) {
        for (j in (i + 1):n) {
            tmp <- cor.test(C[, i], C[, j], ...)
            p.mat[i, j] <- p.mat[j, i] <- tmp$p.value
        }
    }
  colnames(p.mat) <- rownames(p.mat) <- colnames(C)
  p.mat
}
# matrix of the p-value of the correlation
p.mat <- cor.mtest(C)
col <- colorRampPalette(c("#4477AA", "#77AADD", "#FFFFFF", "#EE9988", "#BB4444"))
  corrplot(C, method="color", col=col(200),  
         type="upper", order="hclust", 
         addCoef.col = "black", # Add coefficient of correlation
         tl.col="black", tl.srt=45, #Text label color and rotation
         diag=FALSE 
         )

path <- paste0("../../results/", tool ,"/varpart/", condition, "_gene_CCA")
saveToPDF(paste0(path, ".pdf"), width = 25, height = 25)
```
### Clean up 
```{r}
remove(C, 
       info,
       p.mat,
       dge, 
       dge.tech,
       dge.filtered,
       dge.filtered.reps,
       dge.filtered.reps.norm,
       lcpm,
       log2cpm.filtered,
       log2cpm.norm,
       log2cpm.norm.reps,
       log2cpm.tech)
```
# Save R object
```{r}
saveRDS(dge.filtered.norm, file = paste0("../../rObjects/", condition, "_dge.filtered.norm.rds"))
dge.filtered.norm <- readRDS(paste0("../../rObjects/", condition, "_dge.filtered.norm.rds"))
```
# Design matrix
```{r design_matrix}
# first make sure that numerical values are numerical and factors are set as factors
age <- as.numeric(dge.filtered.norm$samples$Age)
RIN <- as.numeric(dge.filtered.norm$samples$RIN)
sex <- as.factor(dge.filtered.norm$samples$Sex)
race <- as.factor(dge.filtered.norm$samples$Race)
group <- as.factor(dge.filtered.norm$samples$TYPE)

design <- model.matrix(~ 0 + group + sex + (1|RIN))
colnames(design) <- c(CONTROL, AD, PA, LBD, "sex", "RIN") # rename columns 
```

# Voom
The voom method estimates the mean-variance relationship of the log-counts.
Generates a precision weight for each observation and enters these into the limma empirical Bayes analysis pipeline.
```{r voom}
# no weights needed with this large sample size
v <- voom(dge.filtered.norm, design, plot=TRUE)
path <- paste0("../../results/", tool, "/voom/", condition, "_gene_mean_voom")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 4)
```
# PCA on all/the most variable genes
Select most variable genes based on the biological 
coefficient of variance
```{r pca, cache=TRUE}
# Voom transformed counts
voomCounts <- v$E
voomCountsMatrix <- data.matrix(voomCounts, rownames.force = NA)
# Setting the N of genes to use
ntop = length(dge.filtered.norm$genes$gene_id)
# Sorting by the coefficient of variance
means <- rowMeans(voomCountsMatrix)
Pvars <- rowVars(voomCountsMatrix)
cv2 <- Pvars / means ^ 2
select <-
  order(cv2, decreasing = TRUE)[seq_len(min(ntop, length(cv2)))]
highly_variable_exp <- ((voomCountsMatrix)[select,])
# Running PCA
pca_exp <- prcomp(t(highly_variable_exp), scale = F, center = T)
# scale a logical value indicating whether the variables should be scaled.
# A logical value indicating whether the variables should be shifted to be zero centered.

# Dataframe with the first 10 PCs
dim1_10 <- data.frame(pca_exp$x[, 1:10])
# Adding metadata
pcaWithMetadata <- merge(dim1_10, dge.filtered.norm$samples, by = 0, all = TRUE)
pcaWithMetadata$TYPE <- gsub("CONTROL_PA", "PA", pcaWithMetadata$TYPE)
pcaWithMetadata$TYPE <- gsub("CONTROL_AD", "AD", pcaWithMetadata$TYPE)
pcaWithMetadata$group <- factor(pcaWithMetadata$TYPE,
                                 levels = c(CONTROL, AD, PA, LBD, NA))
# Plotting
ggplot(data = pcaWithMetadata, aes(
  x = PC1,
  y = PC2,
  shape = group,
  color = group
)) +
  geom_point(size = 2) +
  theme_bw() +
  scale_color_manual(values = c(control_color, AD_color,
                                PA_color, LBD_color,
                                "azure3")) +
  scale_shape_manual(values = c(control_shape, AD_shape,
                                PA_shape, LBD_shape)) +
  theme(
    plot.title = element_text(size = 10, face = "bold"),
    legend.title = element_text(size = 10),
    legend.text = element_text(size = 10),
    axis.text.x = element_text(size = 10),
    axis.text.y = element_text(size = 10),
    axis.title.x = element_text(size = 10),
    axis.title.y = element_text(size = 10)
  ) +
  theme(legend.title = element_blank()) +
  xlab("PC1 ") +
  ylab("PC2 ")
```
# Linear fit
fits linear model for each gene given a series of arrays
```{r}
fit <- lmFit(v, design)
coef.fit <- fit$coefficients
head(coef.fit)
# contrast design for differential expression
contrasts <- makeContrasts(
  LBDvsControl = LBD - CONTROL,
  LBDvsAD = LBD - AD,
  LBDvsPA = LBD - PA,
  ADvsControl = AD - CONTROL, 
  PAvsControl = PA - CONTROL,
  ADvsPA = AD - PA, 
  levels = colnames(design))
head(contrasts)
# save contrast names
allComparisons <- colnames(contrasts)
allComparisons # check

# run contrast analysis
vfit <- contrasts.fit(fit, contrasts = contrasts)

# Compute differential expression based on the empirical Bayes moderation of the
# standard errors towards a common value.
veBayesFit <- eBayes(vfit)
plotSA(veBayesFit, main = "Final Model: Mean-variance Trend")

# save
path <- paste0("../../results/", tool, "/voom/", condition, "_gene_final_mean_var")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 4)
```

# Save voom rObjects
```{r save_EList}
saveRDS(v, file = paste0("../../rObjects/", condition, "_gene_voom.rds"))
v <- readRDS(paste0("../../rObjects/", condition, "_gene_voom.rds"))
```

# Number of DEGs
Identify number of differential expressed genes.
```{r decide_tests}
pval <- 0.05
sumTable <- 
  summary(decideTests(
    vfit,  # object
    adjust.method = "BH", # by default the method = "separate"
    p.value = pval,
    lfc = 0  # numeric, minimum absolute log2-fold change required
  ))
print(paste0(" FDRq < ", pval))
sumTable
topTable(veBayesFit, coef = "LBDvsControl")
```

# Output DEG tables
```{r output_DEG_tables}
coef <- 1

for (i in allComparisons) {
  # p < 1, log2fc > 0 
  vTopTableAll <-
    topTable(
      veBayesFit, 
      coef = coef,  
      n = Inf, 
      p.value = 1,
      lfc = 0 
    )
vTopTableAll <-
  subset(
    vTopTableAll,
    select = -c(
      score,
      phase,
      transcript_id,
      transcript_type,
      transcript_name,
      transcript_support_level,
      exon_number,
      exon_id,
      ont,
      protein_id,
      ccdsid
    )
  )
    saveRDS(vTopTableAll, file = 
            paste0("../../rObjects/",condition, "_", 
                   i,"_gene_table.rds"))
  path <- paste0("../../results/", tool, "/DEGs/",condition, "_",
  i, "_gene_DEGs_FDRq1.00.txt", sep = "") 
  write.table(
    vTopTableAll,
    path,
    sep = "\t",
    row.names = FALSE,
    quote = FALSE
  )
  # p < 0.05, log2fc > 0
  vTopTable1 <-
    topTable( 
      veBayesFit,  
      coef = coef,  
      n = Inf, 
      p.value = 0.05,
      lfc = 0
    )
  path <- paste0("../../results/", tool, "/DEGs/",condition, "_",
  i, "_gene_DEGs_FDRq0.05.txt", sep = "") 
  write.table(
    vTopTable1,
    path,
    sep = "\t",
    row.names = FALSE,
    quote = FALSE
  )
  # increment 
  coef <- coef + 1
}
remove(coef)
```
# Read in DEGs
Read table with all genes (FDRq = 1).
```{r read_DEG_table}
coef <- 1

for (i in allComparisons) {
  filepath <- paste0(
    "../../results/",
    tool,
    "/DEGs/",condition, "_",
    i,
    "_gene_DEGs_FDRq1.00.txt"
  )
  assign(i,
         read.delim(
           filepath,
           header = TRUE,
           sep = "\t",
           stringsAsFactors = FALSE
         ))
  # increment
  coef <- coef + 1
}
```

# Assign colors
Assign colors  values based on FDRq cutoff of 0.05.
```{r assign_colors}
color_values <- vector()
max <- nrow(LBDvsControl)

for(i in 1:max){
  if (LBDvsControl$adj.P.Val[i] < 0.05){
    if (LBDvsControl$logFC[i] > 0){
      color_values <- c(color_values, 1) # 1 when logFC > 0 and FDRq < 0.05
    }
    else if (LBDvsControl$logFC[i] < 0){
      color_values <- c(color_values, 2) # 2 when logFC < 0 and FDRq < 0.05
    }
  }
  else{
    color_values <- c(color_values, 3) # 3 when FDRq >= 0.05
  }
}

LBDvsControl$color_p0.05 <- factor(color_values)
```

# Subset genes to label
Subset the top 10 up and down-regulated genes
```{r}
up <- LBDvsControl[LBDvsControl$color_p0.05 == 1,]
up10 <- up[1:10,]

down <- LBDvsControl[LBDvsControl$color_p0.05 == 2,]
down <- subset(down, down$logFC < -1.5)
down10 <- down[1:7,]
```

# Volcano plot
```{r volcano}
hadjpval <- (-log10(max(
  LBDvsControl$P.Value[LBDvsControl$adj.P.Val < 0.05], 
  na.rm=TRUE)))

p_vol <-
  ggplot(data = LBDvsControl, 
         aes(x = logFC,  # x-axis is logFC
             y = -log10(P.Value),  # y-axis will be -log10 of P.Value
             color = color_p0.05)) +  # color is based on factored color column
  geom_point(alpha = 1.5, size = 1.7) +  # create scatterplot, alpha makes points transparent
  theme_bw() +  # set color theme
  theme(legend.position = "none") +  # no legend
  scale_color_manual(values = c("red", "blue","grey")) +  # set factor colors
  labs(
    title = "", # no main title
    x = expression(log[2](FC)), # x-axis title
    y = expression(-log[10] ~ "(" ~ italic("p") ~ "-value)") # y-axis title
  ) +
  theme(axis.title.x = element_text(size = 10),
        axis.text.x = element_text(size = 10)) +
  theme(axis.title.y = element_text(size = 10),
        axis.text.y = element_text(size = 10)) +
  geom_hline(yintercept = hadjpval,  #  horizontal line
                     colour = "#000000",
                     linetype = "dashed") +
  ggtitle("LBD vs Control\nFDRq < 0.05") +
  theme(plot.title = element_text(size = 10)) +
  geom_text_repel(data = up10,
                  aes(x = logFC, y= -log10(P.Value), label = gene_name), 
                  color = "maroon", 
                  fontface="italic",
                  size = 4, 
                  max.overlaps = getOption("ggrepel.max.overlaps", default = 30)
                  ) +
  geom_text_repel(data = down10,
                  aes(x = logFC, y= -log10(P.Value), label = gene_name), 
                  color = "navyblue", 
                  fontface="italic",
                  size = 4, 
                  max.overlaps = getOption("ggrepel.max.overlaps", default = 15)
                  ) 
p_vol
# save
path <- paste0("../../results/", tool, "/volcano/",condition,"_gene_volcano_FDRq0.05")
saveToPDF(paste0(path, ".pdf"), width = 5.2, height = 5.2)
```
# Interactive volcanos via glimmaVolcano 
```{r, eval = FALSE}
glimmaVolcano(veBayesFit, coef = 1,  dge = dge.filtered.norm, groups = dge.filtered.norm$samples$TYPE, display.columns=c("gene_name", "AdjPValue", "seqnames"))
# glimmaVolcano()
```

