---
title: "Sex check when samples were aligned to the default reference genome"
author: "Kimberly Olney, Ph.D."
date: "01/05/2023"
output:
  html_document:
    df_print: paged
  pdf_document: default
params:
  args: myarg
---
# Sex check
Samples have already been sex checked. All samples match the reported with inferred sex. The metadata.tsv in the scripts fold contains both reported and referred sex. This script can be skipped if you wish to proceed with the inferred sex in metadata.tsv file. 
# Setup
```{r setup}
# Also do Session > Set Working Directory > Choose Directory
knitr::opts_knit$set(root.dir = ".")
```

# User defined variables
```{r set_variables}
source(here::here("scripts/R", "file_paths_and_colours.R"))
condition <- c("all_genes_default_alignment")
myContrasts <- c("LBDvsControl = LBD - CONTROL",
  "LBDvsAD = LBD - AD",
  "LBDvsPA = LBD - PA",
  "ADvsControl = AD - CONTROL", 
  "PAvsControl = PA - CONTROL",
  "ADvsPA = AD - PA")
tool = c("star")
typeOfCount <-  c("_STAR.bamReadsPerGene.out.tab")
```

# Read data
Samples are aligned to a default human reference genome GRCh38. See Snakemake pipeline for details. 
Read in the STAR aligned counts files for each sample. 
```{r read_data}
# read in metadata
metadata <- vroom("../metadata.tsv"))

# read in snakemake sample information
sample_read_group_info <-
  read.delim("../snakemake/sample_read_group_info.txt", header = FALSE)
sample_read_group_info_split <-
  as.data.frame(str_split_fixed(sample_read_group_info$V1, "_R1", 2))
counts_ids <-
  as.data.frame(gsub("\\.", "_", sample_read_group_info_split$V1))
names(counts_ids)[1] <- "V1"
counts_ids <- as.data.frame(str_split_fixed(counts_ids$V1, "_", 2))
names(counts_ids)[names(counts_ids) == 'V1'] <- 'NPID'
names(counts_ids)[names(counts_ids) == 'V2'] <- 'flow_lane'
# merge sample_count_id and metadata files
counts_metadata <- merge(metadata, counts_ids, by = "NPID")

# path to counts files
count_files <-
  file.path(paste0(
    "../../starAligned/",
    counts_metadata$NPID,
    "_",
    counts_metadata$flow_lane,
    typeOfCount
  ))
# add sample name to counts files
names(count_files) <- paste0(counts_metadata$NPID)
# add sample count path information to metadata
counts_metadata$path <- count_files

# create a counts matrix
counts <- data.frame(fread(count_files[1]))[c(1, 4)]
for (i in 2:length(count_files)) {
  counts <- cbind(counts, data.frame(fread(count_files[i]))[4])
}
```
# Star mapping metrics
```{r plot_mapping_metrics}
star_metrics <- data.frame(counts[counts$V1 %like% "N_", ])
star_metrics$V1 <- as.vector(star_metrics$V1)
melt_star_metrics <- reshape::melt(star_metrics, id = c("V1"))
ggplot(melt_star_metrics, aes(x = value, fill = V1)) + 
  geom_histogram(position = "identity", bins = 100) + 
  facet_grid(V1 ~ .)  + theme_bw()
```
### Clean up
```{r}
# clean up
remove(melt_star_metrics, 
       metadata, 
       counts_ids, 
       params, 
       RIN, 
       sample_read_group_info, 
       sample_read_group_info_split, 
       star_metrics)
```
### Format counts table
```{r}
# remove star metric information
counts <- counts[!grepl("N_", counts$V1),]
# set gene_id (i.e. V1) as the row name
row.names(counts) <- counts$V1
counts$V1 <- NULL
# set column names to NPID sample ID
colnames(counts) <- names(count_files)
```
# Get gene information
```{r}
# read in annotation file
gtf.file <- paste0(pathToRef, "gencode.v38.annotation.gtf")
genes.gtf <- rtracklayer::import(gtf.file)
genes.gtf <- as.data.frame(genes.gtf)
genes.gtf <- genes.gtf[genes.gtf$type == "gene",]
gene_type_table <- table(genes.gtf$gene_type)
write.table(gene_type_table, "gene_type_table.tsv", row.names = F, quote = F, sep = "\t")
```
# Check inputs
```{r}
all.equal(rownames(counts), genes.gtf$gene_id)
all.equal(colnames(counts), (counts_metadata$NPID))
```
# Create DGE object
```{r DGE_object}
# create object
dge <- DGEList(counts = counts,
               samples = counts_metadata,
               genes = genes.gtf)
table(dge$samples$TYPE) # examine 
```
# Save R object
```{r, eval=FALSE}
saveRDS(dge, file = paste0("../../rObjects/", condition, "_dge.raw.rds"))
```

# Remove mitochondrial genes
```{r MT_genes}
dim(dge)
removeMT <- dge$genes$seqnames != "chrM"  # true when NOT MT
dge <- dge[removeMT,,keep.lib.sizes = FALSE]
dim(dge)
```

# Raw MDS (multidimensional scaling)
convert counts to cpm and lcpm
set colors and get data
```{r cpm, warning=FALSE}
lcpm <- edgeR::cpm(dge$counts, log = TRUE)
cpm <- edgeR::cpm(dge$counts, log = FALSE)
# sex colors 
dge$samples$Sex <- as.factor(dge$samples$Sex)
sex_colors <- c(SexColors)[dge$samples$Sex]
```

```{r MDS}
par(bg = 'white')

# plot MDS
MDS_1 <- plotMDS(
  lcpm,
  top = 100, 
  labels = dge$samples$Sex,
  cex = .8, 
  dim.plot = c(1,2), 
  plot = TRUE, 
  col = sex_colors,
  gene.selection = "common"
)
title(expression('Top 100 Genes (Log'[2]~'CPM)'))

path <- paste0("../../results/", tool, "/MDS/", 
               condition,"_gene_MDS_label_sex_dim1&2")
saveToPDF(paste0(path, ".pdf"), width = 5, height = 5)

# plot MDS
MDS_2 <- plotMDS(
  lcpm,
  top = 100, 
  labels = dge$samples$Sex,
  cex = .8, 
  dim.plot = c(2,3), 
  plot = TRUE, 
  col = sex_colors,
  gene.selection = "common"
)
title(expression('Top 100 Genes (Log'[2]~'CPM)'))
path <- paste0("../../results/", tool, "/MDS/", 
               condition,"_gene_MDS_label_sex_dim2&3")
saveToPDF(paste0(path, ".pdf"), width = 5, height = 5)
```
# Sex check
View the expression of sex-linked genes to determine if samples are correctly annotated as XX or XY. 
```{r sex_genes}
genes_and_counts <- cbind(dge$genes$gene_name, dge$counts)
genes_and_counts <- as.data.frame(genes_and_counts)
names(genes_and_counts)[names(genes_and_counts) == "V1"] <- "Geneid"
rownames(genes_and_counts) <- NULL

sex_genes <- c("XIST", "EIF1AY", "KDM5D", "UTY", "DDX3Y", "RPS4Y1")
sex_genes_counts <- subset(genes_and_counts, Geneid %in% sex_genes)
rownames(sex_genes_counts) <- sex_genes_counts$Geneid
sex_genes_counts$Geneid <- NULL
sex_gene_df <- as.data.frame(t(sex_genes_counts))
sex_gene_df$NPID <- rownames(sex_gene_df)
rownames(sex_gene_df) <- NULL

meta_short <- counts_metadata[,c("NPID","Sex")] 
sex_gene_meta <- merge(sex_gene_df, meta_short, by = "NPID")
cols.num <- c("XIST","EIF1AY", "KDM5D", "UTY", "DDX3Y", "RPS4Y1")
sex_gene_meta[cols.num] <- sapply(sex_gene_meta[cols.num], as.integer)
sex_gene_meta <- sex_gene_meta[,c(2,3,4,5,6,7,8,1)]
# replace M with male and F with female 
sex_gene_meta <- sex_gene_meta %>% 
  mutate(
    Sex = ifelse(Sex %in% c("M"), "male", "female") 
  )
```
# Sex inference model
Infer the genetic sex of the sample using GTEx data as a reference for typical sex-linked gene expression. 
```{r}
# for_rna_sex_check.tsv is GTEx expression data 
training_model_data <- "for_rna_sex_check.tsv"

# Build the model using the GTEx data
# Load the data and remove NAs
train_data <- read.csv(training_model_data, sep="\t")

# Split the data into training and test set
set.seed(123)
training.samples <- train_data$sex %>% 
  createDataPartition(p = 0.8, list = FALSE)
train.data  <- train_data[training.samples, ]
test.data <- train_data[-training.samples, ]

# Dummy code for categorical predictor variables
x <- model.matrix(sex~., train.data)[,-1]
# Convert the outcome (class) to a numerical variable
y <- ifelse(train.data$sex == "female", 1, 0)

cv.lasso <- cv.glmnet(x, y, alpha = 1, family = "binomial")
plot(cv.lasso)
cv.lasso$lambda.min

coef(cv.lasso, cv.lasso$lambda.min)
coef(cv.lasso, cv.lasso$lambda.1se)

# Final model with lambda.min
# Glmnet is a package that fits generalized linear and similar models via penalized maximum likelihood. 
lasso.model <- glmnet(x, y, alpha = 1, family = "binomial",
                      lambda = cv.lasso$lambda.min)
# Make predictions on test data
x.test <- model.matrix(sex ~., test.data)[,-1]
probabilities <- lasso.model %>% predict(newx = x.test)
predicted.classes <- ifelse(probabilities > 0.5, "female", "male")
# Model accuracy
observed.classes <- test.data$sex
mean(predicted.classes == observed.classes)

# ----------------------
# Run on experiment data
# ----------------------
NPID <- sex_gene_meta$NPID 
sex_gene_meta$NPID <- NULL
typeof(sex_gene_meta$XIST)
test.experiment <- model.matrix(Sex ~., sex_gene_meta)[,-1]

probabilities <- lasso.model %>% predict(newx = test.experiment)
predicted.classes <- ifelse(probabilities > 4, "female", "male")
# Model accuracy
observed.classes <- sex_gene_meta$Sex
mean(predicted.classes == observed.classes)
sex_check <- cbind(as.data.frame(predicted.classes)$s0, sex_gene_meta$Sex)
sex_check <- as.data.frame(sex_check)
# Rename column where names is "Sepal.Length"
names(sex_check)[names(sex_check) == "V1"] <- "reported"
names(sex_check)[names(sex_check) == "V2"] <- "observed"

# add back in sample IDs
sex_check$NPID <- NPID

# what doesn't match between reported and observed? 
sex_check$sex_check <- ifelse(sex_check$reported == sex_check$observed,"Yes","No")
# output results 
write.table(
  sex_check,
  paste0("../../results/",
         tool,
         "/sex_check/sex_check_RNA.txt"),
  sep = "\t",
  row.names = FALSE,
  quote = FALSE
)

check_samples <- subset(sex_check, sex_check == "No")
samples_to_investigate <- subset(sex_gene_df, NPID %in% check_samples$NPID)
sex_gene_df[cols.num] <- sapply(sex_gene_df[cols.num], as.numeric)
sapply(sex_gene_df, class)
low_coverage_female_samples <- subset(sex_gene_meta, XIST < 5 & Sex == "female")
low_coverage_male_samples <- subset(sex_gene_meta, DDX3Y < 5 & Sex == "male" | 
                                      RPS4Y1 < 5 & Sex == "male" | 
                                      UTY < 5 & Sex == "male" |
                                      KDM5D < 5 & Sex == "male" |
                                      EIF1AY < 5 & Sex == "male")
atest <- cbind(sex_gene_df, sex_gene_meta)
low_coverage_samples_to_investigate <- subset(sex_gene_df, 
                                    NPID %in% low_coverage_female_samples$NPID)

```
# Clean up
```{r}
remove(genes_and_counts, 
       cv.lasso, 
       lasso.model, 
       probabilities, 
       predicted.classes,
       test.data, 
       test.experiment, 
       train_data, 
       train.data, 
       training.samples, 
       x, 
       x.test, 
       cols.num, 
       observed.classes, 
       sex_genes, 
       y,
       meta_short)
```
# Format sex check for plotting
```{r sex_check_plot}
genes_and_counts <- cbind(dge$genes$gene_name, lcpm)
genes_and_counts <- as.data.frame(genes_and_counts)
names(genes_and_counts)[names(genes_and_counts) == "V1"] <- "Geneid"
rownames(genes_and_counts)<-NULL
genes_counts <- reshape2::melt(genes_and_counts, id=c("Geneid"))
names(genes_counts)[names(genes_counts) == "variable"] <- "NPID"

df <- cbind(counts_metadata$NPID, counts_metadata$Sex)
df <- as.data.frame(df)
names(df)[names(df) == "V1"] <- "NPID"
names(df)[names(df) == "V2"] <- "Sex"

data <- merge(genes_counts, df, by = "NPID")

sexGenes <- c("DDX3X, DDX3Y")
SelectGenes_counts <-
  subset(
    data,
    Geneid %in% c(
      "DDX3X",
      "DDX3Y",
      "ZFX",
      "ZFY",
      "USP9X",
      "USP9Y",
      "KDM6A",
      "UTY",
      "PCDH11X",
      "PCDH11Y",
      "XIST",
      "SRY"
    )
  )
SelectGenes_counts[, "geneComb"] <- NA
SelectGenes_counts[, "group"] <- NA


SelectGenes_counts$geneComb <-
  ifelse(
    SelectGenes_counts$Geneid == "DDX3X",
    "DDX3X:DDX3Y",
    ifelse(
      SelectGenes_counts$Geneid == "DDX3Y",
      "DDX3X:DDX3Y",
      ifelse(
        SelectGenes_counts$Geneid == "ZFX",
        "ZFX:ZFY",
        ifelse(
          SelectGenes_counts$Geneid == "ZFY",
          "ZFX:ZFY",
          ifelse(
            SelectGenes_counts$Geneid == "USP9X",
            "USP9X:USP9Y",
            ifelse(
              SelectGenes_counts$Geneid == "USP9Y",
              "USP9X:USP9Y",
              ifelse(
                SelectGenes_counts$Geneid == "KDM6A",
                "UTX:UTY",
                ifelse(
                  SelectGenes_counts$Geneid == "UTY",
                  "UTX:UTY",
                  ifelse(
                    SelectGenes_counts$Geneid == "PCDH11X",
                    "PCDH11X:PCDH11Y",
                    ifelse(
                      SelectGenes_counts$Geneid == "PCDH11Y",
                      "PCDH11X:PCDH11Y",
                      ifelse(
                        SelectGenes_counts$Geneid == "XIST",
                        "XIST",
                        ifelse(SelectGenes_counts$Geneid == "SRY", "SRY", "NA")
                      )
                    )
                  )
                )
              )
            )
          )
        )
      )
    )
  )

SelectGenes_counts$group <-
  ifelse(
    SelectGenes_counts$geneComb == "DDX3X:DDX3Y",
    1,
    ifelse(
      SelectGenes_counts$geneComb == "ZFX:ZFY",
      4,
      ifelse(
        SelectGenes_counts$geneComb == "USP9X:USP9Y",
        3,
        ifelse(
          SelectGenes_counts$geneComb == "UTX:UTY",
          5,
          ifelse(
            SelectGenes_counts$geneComb == "PCDH11X:PCDH11Y",
            2,
            ifelse(
              SelectGenes_counts$geneComb == "XIST",
              6,
              ifelse(SelectGenes_counts$geneComb == "SRY", 7, "NA")
            )
          )
        )
      )
    )
  )
data <- SelectGenes_counts
data$value <- as.numeric(data$value)
```
# Plot sex check
```{r sex_check_plot2}
wrong_sex_samples <-
  data[data$NPID %in% samples_to_investigate$NPID, ]
wrong_sex_samples <- unique(wrong_sex_samples)
low_coverage_samples <-
  data[data$NPID %in% low_coverage_samples_to_investigate$NPID, ]
wrong_sex_samples <- unique(wrong_sex_samples)
leg_lab <- "reported sex"
cbPaletteJITTER = SexColors
write.table(data, paste0("../../results/", tool, 
                         "/sex_check/sex_check_data.txt"), sep = "\t")
write.table(wrong_sex_samples, paste0("../../results/", tool, 
                                      "/sex_check/wrong_sex_samples.txt"), sep = "\t")

geneticSEXgenes_plot <- ggplot(data, aes(x = Geneid, y = value)) +
  geom_jitter(aes(color = Sex, shape = Sex),
              width = 0.25,
              size = 1) +
  scale_color_manual(leg_lab, values = cbPaletteJITTER) + 
  scale_shape_manual(leg_lab, values = c(19, 15)) +
  labs(x = "", y = "lcpm", title = "") +
  facet_grid(
    . ~ group + geneComb,
    switch = "x",
    scales = "free_x",
    space = "free_x"
  ) +
  theme_bw() +
  theme(
    strip.background = element_blank(),
    strip.text.x = element_blank(),
    axis.ticks = element_blank(),
    axis.text.x = element_text(face = "italic"),
    panel.border = element_rect(
      colour = "black",
      fill = NA,
      size = 1
    )
  ) +
  scale_y_continuous(breaks = seq(-4, 9, by = 1), limits = c(-4, 9)) +
  geom_text_repel(
    data = wrong_sex_samples,
    aes(x = Geneid,
        y = value,
        label = NPID),
    color = "black",
    size = 3,
    nudge_x = .5,
    min.segment.length = unit(0, 'lines'),
    max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
  )

geneticSEXgenes_plot
path <-
  paste0("../../results/", tool, "/sex_check/sex_check_samples_labeled")
saveToPDF(paste0(path, ".pdf"), width = 11, height = 6)
```
### Clean up 
```{r}
remove(df, 
       geneticSEXgenes_plot,
       SelectGenes_counts, 
       sex_check, 
       sex_gene_df,
       sex_gene_meta,
       wrong_sex_samples)
```
