---
title: "Examine clinical stats"
author: "Kimberly Olney, Ph.D"
date: "10/07/2022"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---
This script will examine the variables in the metadata to determine if there are differences by disease type and/or sex differences for clinical data such as brain weight. 
# Setup
```{r setup, message=FALSE, warning=FALSE, tidy=TRUE}
knitr::opts_knit$set(root.dir = ".")
```
# User defined variables
sourcing file_paths_and_colours will load the libraries and user defined variables for colors of the graphs 
```{r set_variables, message=FALSE, warning=FALSE, tidy=TRUE}
source(here::here("scripts/R", "file_paths_and_colours.R"))
```

# Read in RNAseq and Alignment metrics text files
### RNAseq metrics
```{r RNA_metrics, message=FALSE, warning=FALSE, tidy=TRUE}
# read in RNA_metrics header.
RNA_metrics_header <- read.delim("../../starAligned_SCC/RNA_metrics_header.txt")
RNA_metrics_files <-
  file.path(paste0("../../starAligned_SCC/",
    metadata$NPID,
    "_STAR_metrics_only_",
    metadata$sex_chr,
    ".txt"
  ))
# add sample name to counts files
names(RNA_metrics_files) <- paste0(metadata$NPID)
# add sample count path information to metadata
metadata$RNA_metrics_path <- RNA_metrics_files
# make sure there are only unique paths

RNA_metrics_df <- data.frame()
# Loop and read the 2nd column remaining files
for (i in 1:length(RNA_metrics_files)) {
  RNA_metrics_df <- rbind(RNA_metrics_df, data.frame(fread(RNA_metrics_files[i])))
}
colnames(RNA_metrics_df) <- colnames(RNA_metrics_header)
RNA_metrics_df$NPID <- names(RNA_metrics_files)
RNA_metrics_metadata <- merge(metadata, RNA_metrics_df, by = "NPID")
# remove the last three columns as they don't contain any information
RNA_metrics_metadata <- RNA_metrics_metadata[1:(length(RNA_metrics_metadata)-3)]
```
clean up 
```{r message=FALSE, warning=FALSE, tidy=TRUE}
remove(RNA_metrics_path, 
       RNA_metrics_df, 
       RNA_metrics_files, 
       RNA_metrics_header)
```
### Alignment metrics
```{r alignment_metrics, message=FALSE, warning=FALSE, tidy=TRUE}
# read in Alignment_metrics header
Alignment_metrics_header <- read.delim("../../starAligned_SCC/Alignment_metrics_header.txt")
Alignment_metrics_files <-
  file.path(paste0("../../starAligned_SCC/",
    metadata$NPID,
    "_Alignment_metrics_only_",
    metadata$sex_chr,
    ".txt"
  ))
# add sample name to counts files
names(Alignment_metrics_files) <- paste0(metadata$NPID)
# add sample count path information to metadata
metadata$Alignment_metrics_path <- Alignment_metrics_files
# make sure there is only unique paths

Alignment_metrics_df <- data.frame()
# Loop and read the 2nd column remaining files
for (i in 1:length(Alignment_metrics_files)) {
  Alignment_metrics_df <- rbind(Alignment_metrics_df, data.frame(fread(Alignment_metrics_files[i])))
}
colnames(Alignment_metrics_df) <- colnames(Alignment_metrics_header)
Alignment_metrics_df$NPID <- names(Alignment_metrics_files)
Alignment_metrics_metadata <- merge(RNA_metrics_metadata, Alignment_metrics_df, by = "NPID")
# remove the last four columns as they don't contain any information
Alignment_metrics_metadata <- Alignment_metrics_metadata[1:(length(Alignment_metrics_metadata)-4)]
metadata <- Alignment_metrics_metadata
```
clean up 
```{r message=FALSE, warning=FALSE, tidy=TRUE}
remove(Alignment_metrics_path, 
       Alignment_metrics_df, 
       Alignment_metrics_files, 
       Alignment_metrics_header, 
       Alignment_metrics_metadata)
```
# APOE E4 allele count
```{r APOE_E4_allele_count, message=FALSE, warning=FALSE, tidy=TRUE}
metadata$APOE_E4_allele_count <-
  ifelse(
    metadata$APOE == "E2E3",
    0,
    ifelse(
      metadata$APOE == "E2E4",
      1,
      ifelse(
        metadata$APOE == "E3E3",
        0,
        ifelse(
          metadata$APOE == "E3E4",
          1,
          ifelse(
            metadata$APOE == "E4E4", 2, "NA")))))
metadata$APOE_E4_allele_count <- as.numeric(metadata$APOE_E4_allele_count)
metadata$sex_numeric <- ifelse(metadata$sex_inferred == "male", 0, 1)
metadata$Race_numeric <- ifelse(metadata$Race == "Caucasian", 0, 1)
```
# Library size
```{r library_size, message=FALSE, warning=FALSE, tidy=TRUE}
plot(metadata$TYPE, metadata$TOTAL_READS)
hist(metadata$TOTAL_READS, breaks = 100)
summary(metadata$TOTAL_READS)
```

## RNA sample count
```{r bar_plot, message=FALSE, warning=FALSE, tidy=TRUE}
# create bar plot showing number of samples for each group
bar <- ggplot(metadata, aes(TYPE, after_stat(count), fill = TYPE)) +
  geom_bar() +
  theme_bw() +
  xlab("Disease group") +
  ggtitle("Bulk RNAseq sample count by disease group") +
  geom_text(
    stat = 'count',
    aes(label = after_stat(count)),
    vjust = 1.6,
    color = "white",
    size = 3.5
  ) +
    theme_bw() + 
  scale_fill_manual(values = TypeColors)
bar

# pie chart
data <- metadata %>%
  group_by(TYPE) %>%
  dplyr::count() %>%
  ungroup() %>%
  mutate(perc = `n` / sum(`n`)) %>%
  arrange(perc) %>%
  mutate(labels = scales::percent(perc))

pie <- ggplot(data, aes(x = "", y = perc, fill = TYPE)) +
  geom_col(color = "black") +
  geom_label(
    aes(label = labels),
    color = c("white"),
    position = position_stack(vjust = 0.5),
    show.legend = FALSE
  ) +
  guides(fill = guide_legend(title = "Disease group")) +
  scale_fill_manual(values = TypeColors) +
  coord_polar(theta = "y") +
  theme_void() +
  theme(legend.position = "none")
pie
```
save plot
```{r message=FALSE, warning=FALSE, tidy=TRUE}
ggarrange(bar,
          pie,
          ncol = 2,
          widths = c(2.5, 1))

path <- paste0("../../results/clinical/TYPE.RNA.count")
saveToPDF(paste0(path, ".pdf"), width = 7.25, height = 3.5)

# clean up
remove(bar, pie, data)
```

## Split by sex 
```{r message=FALSE, warning=FALSE, tidy=TRUE}
bar_sex <- ggplot(metadata, aes(x = TYPE, fill = sex_inferred)) + 
  geom_bar(aes(y = after_stat(count)), position = "dodge") + 
  theme_bw() +
  xlab("Disease group") + 
  ggtitle("Bulk RNAseq sample count by inferred sex") +
  geom_text(stat='count', aes(x = TYPE, label=after_stat(count)), 
            position = position_dodge(width = 1), 
              vjust=-.25, 
            color="black", size=3.5) + 
  scale_fill_manual(values = SexColors) +
  guides(fill = guide_legend(title = "Sex inferred"))+
  scale_y_continuous(breaks = seq(0, 350, by = 100), limits = c(0, 350))
bar_sex

data <- metadata %>% 
  group_by(sex_inferred,TYPE) %>%
  dplyr::count() %>% 
  ungroup(sex_inferred) %>% 
  mutate(perc = `n` / sum(`n`)) %>% 
  arrange(perc) %>%
  mutate(labels = scales::percent(perc))

pie_sex <- ggplot(data, aes(x = "", y = perc, fill = sex_inferred)) +
  geom_col(color = "black") +
  geom_label(aes(label = labels), color = c("white"),
            position = position_stack(vjust = 0.5),
            show.legend = FALSE) +
    facet_wrap(~TYPE, ncol = 4) +
  guides(fill = guide_legend(title = "Sex inferred")) +
  scale_fill_manual(values=SexColors) +
  coord_polar(theta = "y") + 
  theme_void() +
  theme(legend.position = "none") +
    theme(strip.text = element_blank())
pie_sex
```
```{r message=FALSE, warning=FALSE, tidy=TRUE}
ggarrange(bar_sex,
          pie_sex,
          nrow = 2,
          widths = c(2, 1), heights = c(2.8,1))

path <- paste0("../../results/clinical/TYPE.sex.RNA.count")
saveToPDF(paste0(path, ".pdf"), width = 6.88, height = 4.33)

# clean up
remove(bar_sex, pie_sex, data)
```

# Save metadata
```{r message=FALSE, warning=FALSE, tidy=TRUE}
saveRDS(metadata, paste0("../../rObjects/metadata.rds"))
```

# Categorial variables
### TYPE
```{r message=FALSE, warning=FALSE, tidy=TRUE}
metadata_categorical <-
  data.frame(
    metadata$LBD.type,
    metadata$AD.subtype,
    metadata$CDLB,
    metadata$TDP.type,
    metadata$APOE,
    metadata$Braak.NFT,
    metadata$Thal.amyloid,
    metadata$TDP.43,
    metadata$Cing.LB,
    metadata$Cing.Synuclein,
    metadata$Race
  )
# To do it for all names
column_variables <-
  c(
    "LBD.type",
    "AD.subtype",
    "CDLB",
    "TDP.type",
    "APOE",
    "Braak.NFT",
    "Thal.amyloid",
    "TDP.43",
    "Cing.LB",
    "Cing.Synuclein",
    "Race"
  )

bar_plot_fun <- function(i, j) {
  ggplot(metadata, aes(TYPE, after_stat(count), fill = factor(i))) +
    geom_bar(colour = "black") +
    theme_bw() +
    ggtitle(j) + 
    guides(fill = guide_legend(title = j)) +
    xlab("Disease group") +
    ylab("Count") +
    scale_fill_manual(values=colorbindColors) +
    geom_text(
      stat = 'count',
      aes(label = after_stat(count)),
      position = position_stack(vjust = .5),
      color = "black",
      size = 3.5
    )

}
bar_plots <-
  Map(bar_plot_fun, i = metadata_categorical, j = column_variables)
bar_plots

pie_plot_fun <- function(j) {
  data <- metadata %>%
    group_by(metadata[j], TYPE) %>%
    dplyr::count() %>%
    ungroup(all_of(j)) %>%
    mutate(perc = `n` / sum(`n`)) %>%
    arrange(perc) %>%
    mutate(labels = scales::percent(perc))
  k <- unlist(data[, c(1)])
  
  pie <- ggplot(data, aes(x = "", y = perc, fill = factor(k))) +
    geom_col(color = "black") +
    geom_label(
      aes(label = labels),
      color = c("black"),
      position = position_stack(vjust = 0.5),
      show.legend = FALSE
    ) +
    facet_wrap( ~ TYPE, ncol = 4) +
    guides(fill = guide_legend(title = j)) +
    coord_polar(theta = "y") +
    theme_void() +
    theme(legend.position = "none")+ 
    scale_fill_manual(values=colorbindColors)
}
pie_plots <- Map(pie_plot_fun, j = column_variables)

p <- list()
for (i in 1:length(column_variables)) {
  p[[i]] <- grid.arrange(grobs = c(bar_plots[i], pie_plots[i]), heights = c(2,1))
}

plotnames = imap(pie_plots, ~ paste0("../../results/clinical/TYPE.", .y, ".pdf")) %>%
  flatten()
plotnames
walk2(plotnames,
      p,
      ~ ggsave(
        filename = .x,
        plot = .y,
        height = 7,
        width = 6.5
      ))
```

```{r message=FALSE, warning=FALSE, tidy=TRUE}
# clean up
remove(bar_plots, pie_plots, plotnames, p) # do not remove metadata_categorical
```
## TYPE.sex
```{r message=FALSE, warning=FALSE, tidy=TRUE}
metadata_categorical <-
  data.frame(
    metadata$LBD.type,
    metadata$AD.subtype,
    metadata$CDLB,
    metadata$TDP.type,
    metadata$APOE,
    metadata$Braak.NFT,
    metadata$Thal.amyloid,
    metadata$TDP.43,
    metadata$Cing.LB,
    metadata$Cing.Synuclein,
    metadata$Race
  )
# To do it for all names
column_variables <-
  c(
    "LBD.type",
    "AD.subtype",
    "CDLB",
    "TDP.type",
    "APOE",
    "Braak.NFT",
    "Thal.amyloid",
    "TDP.43",
    "Cing.LB",
    "Cing.Synuclein",
    "Race"
  )

bar_plot_fun <- function(i, j) {
  ggplot(metadata, aes(TYPE, after_stat(count), fill = factor(i))) +
    geom_bar(colour = "black") +
    theme_bw() +
    ggtitle(j) + 
    guides(fill = guide_legend(title = j)) +
    xlab("Disease group") +
    ylab("Count") +
    scale_fill_manual(values=colorbindColors) +
    geom_text(
      stat = 'count',
      aes(label = after_stat(count)),
      position = position_stack(vjust = .5),
      color = "black",
      size = 3.5
    ) + facet_grid(. ~ sex_inferred)

}
bar_plots <-
  Map(bar_plot_fun, i = metadata_categorical, j = column_variables)
bar_plots

pie_plot_fun <- function(j) {
  data <- metadata %>%
    group_by(metadata[j], TYPE) %>%
    dplyr::count() %>%
    ungroup(all_of(j)) %>%
    mutate(perc = `n` / sum(`n`)) %>%
    arrange(perc) %>%
    mutate(labels = scales::percent(perc))
  k <- unlist(data[, c(1)])
  
  pie <- ggplot(data, aes(x = "", y = perc, fill = factor(k))) +
    geom_col(color = "black") +
    geom_label(
      aes(label = labels),
      color = c("black"),
      position = position_stack(vjust = 0.5),
      show.legend = FALSE
    ) +
    facet_wrap( ~ TYPE, ncol = 4) +
    guides(fill = guide_legend(title = j)) +
    coord_polar(theta = "y") +
    theme_void() +
    theme(legend.position = "none")+ 
    scale_fill_manual(values=colorbindColors)+ 
    facet_grid(. ~ sex_inferred)
}
pie_plots <- Map(pie_plot_fun, j = column_variables)

p <- list()
for (i in 1:length(column_variables)) {
  p[[i]] <- grid.arrange(grobs = c(bar_plots[i], pie_plots[i]), heights = c(2,1))
}

plotnames = imap(pie_plots, ~ paste0("../../results/clinical/TYPE.sex", .y, ".pdf")) %>%
  flatten()
plotnames
walk2(plotnames,
      p,
      ~ ggsave(
        filename = .x,
        plot = .y,
        height = 7,
        width = 8.5
      ))
```

```{r message=FALSE, warning=FALSE, tidy=TRUE}
# clean up
remove(bar_plots, pie_plots, plotnames, p) # do not remove metadata_categorical
```

# Continuous variables 
### TYPE
```{r message=FALSE, warning=FALSE, tidy=TRUE}

#tapply(metadata$Age, metadata$TYPE, t.test) 
#result <- t.test(data)
#result$conf.int

metadata_continuous <-
  data.frame(
    metadata$Brain.wt,
    metadata$Age,
    metadata$PMI,
    metadata$Concentration.ng.ul,
    metadata$Volume.ul,
    metadata$Total.RNA.ng,
    metadata$RIN
  )
column_variables <-
  c(
    "Brain.wt",
    "Age",
    "PMI",
    "Concentration.ng.ul",
    "Volume.ul",
    "Total.RNA.ng",
    "RIN"
  )
TYPE_comparison <-
  list(
    c("CONTROL", "PA"),
    c("CONTROL", "AD"),
    c("CONTROL", "LBD"),
    c("PA", "AD"),
    c("PA", "LBD"),
    c("AD", "LBD")
  )

violin_plot_fun <- function(i, j) {
  ggplot(metadata, aes(TYPE, i, fill = TYPE)) +
    geom_violin() +
    geom_boxplot(width = 0.1, outlier.shape = NA) + 
    geom_jitter(shape=16, position=position_jitter(0.2)) +
    theme_bw() + 
    ggtitle(j) +
    xlab("Disease group") +
    ylab(j) +
    stat_compare_means(comparisons = TYPE_comparison) +
    scale_fill_manual(values=TypeColors) +
    theme(legend.position = "none")
}
violin_plots <-
  Map(violin_plot_fun, i = metadata_continuous, j = column_variables)
violin_plots

plotnames = imap(violin_plots, ~paste0("../../results/clinical/TYPE.", .y, ".pdf")) %>%
  flatten()
plotnames <- gsub("metadata.", "", plotnames)
walk2(plotnames, violin_plots, ~ggsave(filename = .x, plot = .y, 
                                    height = 4, width = 5.5))

remove(violin_plots) 
```

## TYPE.sex
```{r message=FALSE, warning=FALSE, tidy=TRUE}
violin_plot_fun <- function(i, j) {
  ggplot(metadata, aes(TYPE, i, fill = TYPE)) +
    geom_violin() +
    geom_boxplot(width = 0.1, outlier.shape = NA) + 
    geom_jitter(shape=16, position=position_jitter(0.2)) +
    theme_bw() + 
    ggtitle(j) +
    xlab("Disease group") +
    ylab(j) +
    stat_compare_means(comparisons = TYPE_comparison) +
    scale_fill_manual(values=TypeColors) +
    theme(legend.position = "none") +
    facet_grid(. ~ sex_inferred)
}
violin_plots <-
  Map(violin_plot_fun, i = metadata_continuous, j = column_variables)
violin_plots

plotnames = imap(violin_plots, ~paste0("../../results/clinical/TYPE.Sex.", .y, ".pdf")) %>%
  flatten()
plotnames <- gsub("metadata.", "", plotnames)
walk2(plotnames, violin_plots, ~ggsave(filename = .x, plot = .y, 
                                    height = 4, width = 7.5))
remove(violin_plots) 
```

# Alignment metrics
### TYPE
```{r message=FALSE, warning=FALSE, tidy=TRUE}
metrics_metadata <-
  data.frame(
    metadata$PCT_CODING_BASES,
    metadata$PCT_INTERGENIC_BASES,
    metadata$PCT_INTRONIC_BASES,
    metadata$TOTAL_READS
  )
column_variables <-
  c(
    "PCT_CODING_BASES",
    "PCT_INTERGENIC_BASES",
    "PCT_INTRONIC_BASES",
    "TOTAL_READS"
  )

violin_plot_fun <- function(i, j) {
  ggplot(metadata, aes(TYPE, i, fill = TYPE)) +
    geom_violin() +
    geom_boxplot(width = 0.1, outlier.shape = NA) + 
    geom_jitter(shape=16, position=position_jitter(0.2)) +
    theme_bw() + 
    ggtitle(j) +
    xlab("Disease group") +
    ylab(j) +
    stat_compare_means(comparisons = TYPE_comparison) +
    scale_fill_manual(values=TypeColors) +
    theme(legend.position = "none")
}
violin_plots <-
  Map(violin_plot_fun, i = metrics_metadata, j = column_variables)
violin_plots

plotnames = imap(violin_plots, ~paste0("../../results/clinical/TYPE.", .y, ".pdf")) %>%
  flatten()
plotnames <- gsub("metadata.", "", plotnames)
walk2(plotnames, violin_plots, ~ggsave(filename = .x, plot = .y, 
                                    height = 4, width = 5.5))
remove(violin_plots) 
```
