---
title: "Examine clinical stats"
author: "Kimberly Olney, Ph.D"
date: "10/07/2022"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---
This script will examine the variables in the metadata to determine if ther are differences by disease type and/or sex differences for clinical data such as brain weight. 
# Setup
```{r setup, message=FALSE, warning=FALSE, tidy=TRUE}
knitr::opts_knit$set(root.dir = ".")
```
# User defined variables
sourcing file_paths_and_colours will load the libraries and user defined variables for colors of the graphs 
```{r set_variables, message=FALSE, warning=FALSE, tidy=TRUE}
source(here::here("scripts/R", "file_paths_and_colours.R"))
remove(metadata, metadata_ATS, df)
```
# Read data
```{r read_data, message=FALSE, warning=FALSE, tidy=TRUE}
# read in metadata
metadata <-
  read.delim(paste0(pathToRawData, "metadata_expanded.tsv"),
             header = TRUE,
             sep = "\t")

# remove technical replicates (i.e lane 1 and lane 2)
metadata <- metadata[!duplicated(metadata[, c('NPID')]), ]
# remove samples that have missing A-T-S scores or that don't fit into an A-T-S category 
excluded_IDs <-  c("NA18-272", "5510", "NA01-135", "NA20-102", "NA20-104", "NA06-159")
metadata <- metadata[ ! metadata$NPID %in% excluded_IDs, ]

# read in APOE genotype information 
APOE <- read.delim(paste0(pathToRawData, "APOE.tsv"),
             header = TRUE,
             sep = "\t")
metadata <- merge(metadata, APOE, by = "NPID", all.x = TRUE, all.y= TRUE)
colnames(metadata)[colnames(metadata) == "APOE.y"] <- "APOE"

# remove samples that don't have RNA data
metadata <- metadata %>% drop_na(flowcell_and_lane)
metadata <- metadata %>%
  mutate(across('AD.subtype', str_replace, 'typical', 'Typical'))
# rename columns for easier plotting 
metadata <- metadata %>%
  mutate(across('TYPE', str_replace, 'CONTROL - AD', 'AD'))
metadata <- metadata %>%
  mutate(across('TYPE', str_replace, 'CONTROL - PA', 'PA'))
metadata$TYPE <- factor(metadata$TYPE, levels = c("CONTROL", "PA", "AD", "LBD"))
# add sex chromosome information 
metadata <- metadata %>% mutate(sex_chr = case_when(
  startsWith(sex_inferred, "female") ~ "XX", 
  startsWith(sex_inferred, "male") ~ "XY"))
# remove rows with no inferred sex.
# No counts data for this samples because there is no RNAseq data for that individual
metadata <- metadata[!is.na(metadata$sex_inferred),]
```

# Read in RNAseq and Alignment metrics text files
### RNAseq metrics
```{r RNA_metrics, message=FALSE, warning=FALSE, tidy=TRUE}
# read in RNA_metrics header.
# note that is ONLY the header. 
RNA_metrics_header <- read.delim("../../starAligned_SCC/RNA_metrics_header.txt")
RNA_metrics_files <-
  file.path(paste0("../../starAligned_SCC/",
    metadata$NPID,
    "_STAR_metrics_only_",
    metadata$sex_chr,
    ".txt"
  ))
# add sample name to counts files
names(RNA_metrics_files) <- paste0(metadata$NPID)
# add sample count path information to metadata
metadata$RNA_metrics_path <- RNA_metrics_files
# make sure there is only unique paths

RNA_metrics_df <- data.frame()
# Loop and read the 2nd column remaining files
for (i in 1:length(RNA_metrics_files)) {
  RNA_metrics_df <- rbind(RNA_metrics_df, data.frame(fread(RNA_metrics_files[i])))
}
colnames(RNA_metrics_df) <- colnames(RNA_metrics_header)
RNA_metrics_df$NPID <- names(RNA_metrics_files)
RNA_metrics_metadata <- merge(metadata, RNA_metrics_df, by = "NPID")
# remove the last three columns as they don't contain any information
RNA_metrics_metadata <- RNA_metrics_metadata[1:(length(RNA_metrics_metadata)-3)]
```
clean up 
```{r message=FALSE, warning=FALSE, tidy=TRUE}
remove(RNA_metrics_path, 
       RNA_metrics_df, 
       RNA_metrics_files, 
       RNA_metrics_header)
```
### Alignment metrics
```{r alignment_metrics, message=FALSE, warning=FALSE, tidy=TRUE}
# read in Alignment_metrics header
Alignment_metrics_header <- read.delim("../../starAligned_SCC/Alignment_metrics_header.txt")
Alignment_metrics_files <-
  file.path(paste0("../../starAligned_SCC/",
    metadata$NPID,
    "_Alignment_metrics_only_",
    metadata$sex_chr,
    ".txt"
  ))
# add sample name to counts files
names(Alignment_metrics_files) <- paste0(metadata$NPID)
# add sample count path information to metadata
metadata$Alignment_metrics_path <- Alignment_metrics_files
# make sure there is only unique paths

Alignment_metrics_df <- data.frame()
# Loop and read the 2nd column remaining files
for (i in 1:length(Alignment_metrics_files)) {
  Alignment_metrics_df <- rbind(Alignment_metrics_df, data.frame(fread(Alignment_metrics_files[i])))
}
colnames(Alignment_metrics_df) <- colnames(Alignment_metrics_header)
Alignment_metrics_df$NPID <- names(Alignment_metrics_files)
Alignment_metrics_metadata <- merge(RNA_metrics_metadata, Alignment_metrics_df, by = "NPID")
# remove the last four columns as they don't contain any information
Alignment_metrics_metadata <- Alignment_metrics_metadata[1:(length(Alignment_metrics_metadata)-4)]
metadata <- Alignment_metrics_metadata
```
clean up 
```{r message=FALSE, warning=FALSE, tidy=TRUE}
remove(Alignment_metrics_path, 
       Alignment_metrics_df, 
       Alignment_metrics_files, 
       Alignment_metrics_header, 
       Alignment_metrics_metadata)
```
# APOE E4 allele count
```{r APOE_E4_allele_count, message=FALSE, warning=FALSE, tidy=TRUE}
metadata$APOE_E4_allele_count <-
  ifelse(
    metadata$APOE == "E2E3",
    0,
    ifelse(
      metadata$APOE == "E2E4",
      1,
      ifelse(
        metadata$APOE == "E3E3",
        0,
        ifelse(
          metadata$APOE == "E3E4",
          1,
          ifelse(
            metadata$APOE == "E4E4", 2, "NA")))))
metadata$APOE_E4_allele_count <- as.numeric(metadata$APOE_E4_allele_count)
metadata$sex_numeric <- ifelse(metadata$sex_inferred == "male", 0, 1)
metadata$Race_numeric <- ifelse(metadata$Race == "Caucasian", 0, 1)
```
# Remove sample with very large library size 
```{r library_size, message=FALSE, warning=FALSE, tidy=TRUE}
plot(metadata$TYPE, metadata$TOTAL_READS)
hist(metadata$TOTAL_READS, breaks = 100)
metadata <- subset(metadata, NPID != "NA18-272")
#metadata <- subset(metadata, TOTAL_READS < 100557772)
plot(metadata$TYPE, metadata$TOTAL_READS)
hist(metadata$TOTAL_READS, breaks = 100)
summary(metadata$TOTAL_READS)
```

# Output updated RNA metadata table
```{r metadata_output, message=FALSE, warning=FALSE, tidy=TRUE}
write.table(
  metadata,
  paste0(pathToRawData, "RNA_metadata.tsv"),
  sep = "\t",
  quote = FALSE,
  row.names = FALSE
)
# clean up
remove(APOE, metadata_ATS, df, RNA_metrics_metadata)
```
## RNA sample count
```{r bar_plot, message=FALSE, warning=FALSE, tidy=TRUE}
# create bar plot showing number of samples for each group
bar <- ggplot(metadata, aes(TYPE, after_stat(count), fill = TYPE)) +
  geom_bar() +
  theme_bw() +
  xlab("Disease group") +
  ggtitle("Bulk RNAseq sample count by disease group") +
  geom_text(
    stat = 'count',
    aes(label = after_stat(count)),
    vjust = 1.6,
    color = "white",
    size = 3.5
  ) +
    theme_bw() + 
  scale_fill_manual(values = TypeColors)
bar

# pie chart
data <- metadata %>%
  group_by(TYPE) %>%
  dplyr::count() %>%
  ungroup() %>%
  mutate(perc = `n` / sum(`n`)) %>%
  arrange(perc) %>%
  mutate(labels = scales::percent(perc))

pie <- ggplot(data, aes(x = "", y = perc, fill = TYPE)) +
  geom_col(color = "black") +
  geom_label(
    aes(label = labels),
    color = c("white"),
    position = position_stack(vjust = 0.5),
    show.legend = FALSE
  ) +
  guides(fill = guide_legend(title = "Disease group")) +
  scale_fill_manual(values = TypeColors) +
  coord_polar(theta = "y") +
  theme_void() +
  theme(legend.position = "none")
pie
```
save plot
```{r message=FALSE, warning=FALSE, tidy=TRUE}
ggarrange(bar,
          pie,
          ncol = 2,
          widths = c(2.5, 1))

path <- paste0("../../results/clinical/TYPE.RNA.count")
saveToPDF(paste0(path, ".pdf"), width = 7.25, height = 3.5)

# clean up
remove(bar, pie, data)
```

## Split by sex 
```{r message=FALSE, warning=FALSE, tidy=TRUE}
#SexColors <- c("orange", "olivedrab")
bar_sex <- ggplot(metadata, aes(x = TYPE, fill = sex_inferred)) + 
  geom_bar(aes(y = after_stat(count)), position = "dodge") + 
  theme_bw() +
  xlab("Disease group") + 
  ggtitle("Bulk RNAseq sample count by inferred sex") +
  geom_text(stat='count', aes(x = TYPE, label=after_stat(count)), 
            position = position_dodge(width = 1), 
              vjust=-.25, 
            color="black", size=3.5) + 
  scale_fill_manual(values = SexColors) +
  guides(fill = guide_legend(title = "Sex inferred"))+
  scale_y_continuous(breaks = seq(0, 350, by = 100), limits = c(0, 350))
bar_sex

data <- metadata %>% 
  group_by(sex_inferred,TYPE) %>%
  dplyr::count() %>% 
  ungroup(sex_inferred) %>% 
  mutate(perc = `n` / sum(`n`)) %>% 
  arrange(perc) %>%
  mutate(labels = scales::percent(perc))

pie_sex <- ggplot(data, aes(x = "", y = perc, fill = sex_inferred)) +
  geom_col(color = "black") +
  geom_label(aes(label = labels), color = c("white"),
            position = position_stack(vjust = 0.5),
            show.legend = FALSE) +
    facet_wrap(~TYPE, ncol = 4) +
  guides(fill = guide_legend(title = "Sex inferred")) +
  scale_fill_manual(values=SexColors) +
  coord_polar(theta = "y") + 
  theme_void() +
  theme(legend.position = "none") +
    theme(strip.text = element_blank())
pie_sex
```
```{r message=FALSE, warning=FALSE, tidy=TRUE}
ggarrange(bar_sex,
          pie_sex,
          nrow = 2,
          widths = c(2, 1), heights = c(2.8,1))

path <- paste0("../../results/clinical/TYPE.sex.RNA.count")
saveToPDF(paste0(path, ".pdf"), width = 6.88, height = 4.33)

# clean up
remove(bar_sex, pie_sex, data)
```

# RNA sample count by A-T-S scores
Group 1: A-T-S amyloid + synuclein + tau (A >= 1, T >= 1, S >= 1 )
Group 2: Ahigh-S high amyloid + synuclein (A >= 2, T = 0, S >= 1 )
Group 3: Alow-S low amyloid + synuclein  (A < 2 , T = 0, S >= 1 )
Group 4: S "pure" synuclein (A = 0, T = 0, S >= 1)
Group 5: A-T amyloid + tau (A >= 1, T >= 1, S = 0)
Group 6: A amyloid (A >= 1, T = 0, S = 0)
Group 7: Min minimal (i.e normal controls) (A = 0, T = 0, S = 0) 
```{r message=FALSE, warning=FALSE, tidy=TRUE}
# Split CWOW column into A - T - S scores
metadata[c('A', 'T', 'S')] <- str_split_fixed(metadata$CWOW.Category, ', ', 3)
metadata$A <- gsub('A', '', metadata$A)
metadata$T <- gsub('T', '', metadata$T)
metadata$S <- gsub('S', '', metadata$S)

# set as numeric in order to assign to groups 
metadata$A <- as.numeric(metadata$A)
metadata$T <- as.numeric(metadata$T)
metadata$S <- as.numeric(metadata$S)
group1 <- subset(metadata, A >= 1 &  T >= 1 & S >= 1 )
group2 <- subset(metadata, A >= 2 & T == 0 & S >= 1)
group3 <- subset(metadata, A < 2 & T == 0 & S >= 1)
group4 <- subset(metadata, A == 0 & T == 0 & S >= 1)
group5 <- subset(metadata, A >= 1 & T >= 1 & S == 0)
group6 <- subset(metadata, A >= 1 & T == 0 & S == 0)
group7 <- subset(metadata, A == 0 & T == 0 & S == 0)

list_input <- list("group1" = group1$NPID,
                   "group2" = group2$NPID,
                   "group3" = group3$NPID,
                   "group4" = group4$NPID,
                   "group5" = group5$NPID,
                   "group6" = group6$NPID,
                   "group7" = group7$NPID)
group_data <- fromList(list_input)
upset(group_data)

metadata$ATS <-
  ifelse(
    (metadata$A >= 1 & metadata$T >= 1 & metadata$S >= 1),
    1,
    ifelse(
      (metadata$A >= 2 & metadata$T == 0 & metadata$S >= 1),
      2,
      ifelse(
        (metadata$A < 2 & metadata$A > 0 & metadata$T == 0 & metadata$S >= 1),
        3,
        ifelse(
          (metadata$A == 0 & metadata$T == 0 & metadata$S >= 1),
          4,
          ifelse(
            (metadata$A >= 1 & metadata$T >= 1 & metadata$S == 0),
            5,
            ifelse(
              (metadata$A >= 1 & metadata$T == 0 & metadata$S == 0),
              6,
              ifelse((metadata$A == 0 & metadata$T == 0 & metadata$S == 0), 7, "NA")
            )
          )
        )
      )
    )
  )

table(metadata$ATS)
remove(group1,group2,group3,group4,group5,group6,group7)
```
plot
```{r message=FALSE, warning=FALSE, tidy=TRUE}
# Drop samples with NA ATS score
df <- metadata %>% drop_na(ATS)
df <- subset(df, ATS != "NA")

# create bar plot showing number of samples for each group 
bar <- ggplot(df, aes(ATS, after_stat(count), fill = ATS)) +
  geom_bar() +
  theme_bw() +
  xlab("Disease group") +
  ggtitle("Bulk RNAseq sample count by A-T-S scores") +
  geom_text(
    stat = 'count',
    aes(label = after_stat(count)),
    vjust = 1.6,
    color = "black",
    size = 3.5
  ) +
  theme_bw() + 
  theme(axis.title.x = element_blank(),
        axis.text.x = element_text(size = 10, angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_manual(values = colorbindColors)
bar
path <- paste0("../../results/clinical/ATS.RNA.count")
saveToPDF(paste0(path, ".pdf"), width = 7, height = 5)
```
# A-T-S score names
```{r message=FALSE, warning=FALSE, tidy=TRUE}
metadata$ATS_names <-
  ifelse(
    metadata$A >= 1 & metadata$T >= 1 & metadata$S >= 1,
    "amyloid + synuclein + tau",
    ifelse(
      metadata$A >= 2 & metadata$T == 0 & metadata$S >= 1,
      "high amyloid + synuclein",
      ifelse(
        metadata$A < 2 & metadata$A > 0 & metadata$T == 0 & metadata$S >= 1,
        "low amyloid + synuclein",
        ifelse(
          metadata$A == 0 & metadata$T == 0 & metadata$S >= 1,
          "pure synuclein",
          ifelse(
            metadata$A >= 1 & metadata$T >= 1 & metadata$S == 0,
            "amyloid + tau",
            ifelse(
              metadata$A >= 1 & metadata$T == 0 & metadata$S == 0,
              "amyloid",
              ifelse(metadata$A == 0 & metadata$T == 0 & metadata$S == 0, "no pathology", "NA")
            )
          )
        )
      )
    )
  )
```

plot by A-T-S score
```{r message=FALSE, warning=FALSE, tidy=TRUE}
# Samples with missing A-T-S scores
subset(metadata, ATS_names == "NA")
subset(metadata, is.na(metadata$ATS_names))
# Drop samples with NA ATS score
df <- metadata %>% drop_na(ATS_names)
df <- subset(df, ATS_names != "NA")
df$ATS_names <-
  factor(
    df$ATS_names,
    levels = c(
      "no pathology",
      "amyloid",
      "amyloid + tau",
      "pure synuclein",
      "low amyloid + synuclein",
      "high amyloid + synuclein",
      "amyloid + synuclein + tau"
    )
  )
metadata$ATS_names <-
  factor(
    metadata$ATS_names,
    levels = c(
      "no pathology",
      "amyloid",
      "amyloid + tau",
      "pure synuclein",
      "low amyloid + synuclein",
      "high amyloid + synuclein",
      "amyloid + synuclein + tau"
    )
  )
```

```{r message=FALSE, warning=FALSE, tidy=TRUE}
write.table(
  df,
  paste0("../../rObjects/metadata_A-T-S_scores.txt"),
  sep = "\t",
  quote = FALSE,
  row.names = FALSE
)
```

```{r message=FALSE, warning=FALSE, tidy=TRUE}
# create bar plot showing number of samples for each group
bar <- ggplot(df, aes(TYPE, after_stat(count), fill = ATS_names)) +
  geom_bar() +
  theme_bw() +
  xlab("Disease group") +
  ggtitle("Bulk RNAseq sample count by A-T-S scores") +
  theme_bw() + 
  theme(axis.title.x = element_blank(),
        axis.text.x = element_text(size = 10, angle = 90, vjust = 0.5, hjust=1),
        axis.title.y = element_text(size = 10),
        axis.text.y = element_text(size = 10),
        plot.title = element_text(size = 10),
        legend.text=element_text(size=10),
        legend.title=element_text(size=10)) +
  scale_fill_manual(values = colorbindColors)
bar
path <- paste0("../../results/clinical/ATS.RNA.count.ByTYPE")
saveToPDF(paste0(path, ".pdf"), width = 7, height = 5)

# create bar plot showing number of samples for each end point criteria
bar <- ggplot(df, aes(ATS_names, after_stat(count), fill = ATS_names)) +
  geom_bar() +
  theme_bw() +
  xlab("Disease group") +
  ggtitle("Bulk RNAseq sample count by A-T-S scores") +
  theme_bw() + 
  geom_text(
    stat = 'count',
    aes(label = after_stat(count)),
    vjust = 1.6,
    color = "black",
    size = 3.5
  ) +
  theme(axis.title.x = element_blank(),
        axis.text.x = element_text(size = 10, angle = 75, vjust = 1, hjust=1),
        axis.title.y = element_text(size = 10),
        axis.text.y = element_text(size = 10),
        plot.title = element_text(size = 10),
        legend.text=element_text(size=10),
        legend.title=element_text(size=10)) +
  scale_fill_manual(values = colorbindColors)
bar
path <- paste0("../../results/clinical/ATS.RNA.count")
saveToPDF(paste0(path, ".pdf"), width = 7, height = 5)
```

```{r message=FALSE, warning=FALSE, tidy=TRUE}
# pie chart
data <- df %>%
  group_by(ATS_names) %>%
  dplyr::count() %>%
  ungroup() %>%
  mutate(perc = `n` / sum(`n`)) %>%
  arrange(perc) %>%
  mutate(labels = scales::percent(perc))

pie <- ggplot(data, aes(x = "", y = perc, fill = ATS_names)) +
  geom_col(color = "black") +
  geom_label(
    aes(label = labels),
    color = c("black"),
    position = position_stack(vjust = 0.5),
    show.legend = FALSE
  ) +
  guides(fill = guide_legend(title = "Disease group")) +
  scale_fill_manual(values = colorbindColors) +
  coord_polar(theta = "y") +
  theme_void() +
  theme(legend.position = "none")
pie
```
save plot
```{r message=FALSE, warning=FALSE, tidy=TRUE}
ggarrange(bar,
          pie,
          ncol = 2,
          widths = c(2.5, 1))

path <- paste0("../../results/clinical/ATS.RNA.count.pie")
saveToPDF(paste0(path, ".pdf"), width = 9, height = 6)

# clean up
remove(bar, data, pie) # do not remove df, used for split by sex 
```
## A-T-S split by sex 
```{r message=FALSE, warning=FALSE, tidy=TRUE}
bar_sex <- ggplot(df, aes(x = ATS_names, fill = sex_inferred)) + 
  geom_bar(aes(y = after_stat(count)), position = "dodge") + 
  theme_bw() + 
  theme(axis.title.x = element_blank(),
        axis.text.x = element_text(size = 10, angle = 90, vjust = 0.5, hjust=1),
        axis.title.y = element_text(size = 10),
        axis.text.y = element_text(size = 10),
        plot.title = element_text(size = 10),
        legend.text=element_text(size=10),
        legend.title=element_text(size=10)) +
  xlab("Disease group") + 
  ggtitle("Bulk RNAseq sample count by inferred sex") +
  geom_text(stat='count', aes(x = ATS_names, label=after_stat(count)), 
            position = position_dodge(width = 1), 
              vjust=-.25, 
            color="black", size=3.5) + 
  scale_fill_manual(values = SexColors) +
  guides(fill = guide_legend(title = "Sex inferred"))+
  scale_y_continuous(breaks = seq(0, 225, by = 25), limits = c(0, 225))
bar_sex

data <- df %>% 
  group_by(sex_inferred,ATS_names) %>%
  dplyr::count() %>% 
  ungroup(sex_inferred) %>% 
  mutate(perc = `n` / sum(`n`)) %>% 
  arrange(perc) %>%
  mutate(labels = scales::percent(perc))

pie_sex <- ggplot(data, aes(x = "", y = perc, fill = sex_inferred)) +
  geom_col(color = "black") +
  geom_label(aes(label = labels), color = c("white"),
            position = position_stack(vjust = 0.5),
            show.legend = FALSE) +
    facet_wrap(~ATS_names, ncol = 7) +
  guides(fill = guide_legend(title = "Sex inferred")) +
  scale_fill_manual(values=SexColors) +
  coord_polar(theta = "y") + 
  theme_void() +
  theme(legend.position = "none") +
    theme(strip.text = element_blank())
pie_sex
```
```{r message=FALSE, warning=FALSE, tidy=TRUE}
ggarrange(bar_sex,
          pie_sex,
          nrow = 2, 
          heights = c(2,1))

path <- paste0("../../results/clinical/ATS.sex.RNA.count")
saveToPDF(paste0(path, ".pdf"), width = 9, height = 7)

remove(bar_sex, pie_sex, data, df)
```
# Save metadata
```{r message=FALSE, warning=FALSE, tidy=TRUE}
saveRDS(metadata, paste0("../../rObjects/metadata.rds"))
```

# Categorial variables
### TYPE
```{r message=FALSE, warning=FALSE, tidy=TRUE}
metadata_categorical <-
  data.frame(
    metadata$LBD.type,
    metadata$AD.subtype,
    metadata$CDLB,
    metadata$TDP.type,
    metadata$APOE,
    metadata$MAPT,
    metadata$GRN,
    metadata$TMEM106b,
    metadata$Braak.NFT,
    metadata$Thal.amyloid,
    metadata$MF.Tau,
    metadata$MF.Amyloid,
    metadata$VaD,
    metadata$TDP.43,
    metadata$FHx,
    metadata$MF.SP,
    metadata$MF.NFT,
    metadata$MF.LB,
    metadata$Cing.LB,
    metadata$Cing.Synuclein,
    metadata$Race
  )
# To do it for all names
column_variables <-
  c(
    "LBD.type",
    "AD.subtype",
    "CDLB",
    "TDP.type",
    "APOE",
    "MAPT",
    "GRN",
    "TMEM106b",
    "Braak.NFT",
    "Thal.amyloid",
    "MF.Tau",
    "MF.Amyloid",
    "VaD",
    "TDP.43",
    "FHx",
    "MF.SP",
    "MF.NFT",
    "MF.LB",
    "Cing.LB",
    "Cing.Synuclein",
    "Race"
  )

bar_plot_fun <- function(i, j) {
  ggplot(metadata, aes(TYPE, after_stat(count), fill = factor(i))) +
    geom_bar(colour = "black") +
    theme_bw() +
    ggtitle(j) + 
    guides(fill = guide_legend(title = j)) +
    xlab("Disease group") +
    ylab("Count") +
    scale_fill_manual(values=colorbindColors) +
    geom_text(
      stat = 'count',
      aes(label = after_stat(count)),
      position = position_stack(vjust = .5),
      color = "black",
      size = 3.5
    )
}
bar_plots <-
  Map(bar_plot_fun, i = metadata_categorical, j = column_variables)
bar_plots

pie_plot_fun <- function(j) {
  data <- metadata %>%
    group_by(metadata[j], TYPE) %>%
    dplyr::count() %>%
    ungroup(all_of(j)) %>%
    mutate(perc = `n` / sum(`n`)) %>%
    arrange(perc) %>%
    mutate(labels = scales::percent(perc))
  k <- unlist(data[, c(1)])
  
  pie <- ggplot(data, aes(x = "", y = perc, fill = factor(k))) +
    geom_col(color = "black") +
    geom_label(
      aes(label = labels),
      color = c("black"),
      position = position_stack(vjust = 0.5),
      show.legend = FALSE
    ) +
    facet_wrap( ~ TYPE, ncol = 4) +
    guides(fill = guide_legend(title = j)) +
    coord_polar(theta = "y") +
    theme_void() +
    theme(legend.position = "none")+ 
    scale_fill_manual(values=colorbindColors)
}
pie_plots <- Map(pie_plot_fun, j = column_variables)

p <- list()
for (i in 1:length(column_variables)) {
  p[[i]] <- grid.arrange(grobs = c(bar_plots[i], pie_plots[i]), heights = c(2,1))
}

plotnames = imap(pie_plots, ~ paste0("../../results/clinical/TYPE.", .y, ".pdf")) %>%
  flatten()
plotnames
walk2(plotnames,
      p,
      ~ ggsave(
        filename = .x,
        plot = .y,
        height = 7,
        width = 6.5
      ))
```

```{r message=FALSE, warning=FALSE, tidy=TRUE}
# clean up
remove(bar_plots, pie_plots, plotnames, p) # do not remove metadata_categorical
```
### ATS
```{r message=FALSE, warning=FALSE, tidy=TRUE}
bar_plot_fun <- function(i, j) {
  ggplot(metadata, aes(ATS_names, after_stat(count), fill = factor(i))) +
    geom_bar(colour = "black") +
    theme_bw() +
    ggtitle(j) + 
    guides(fill = guide_legend(title = j)) +
    xlab("Disease group") +
    ylab("Count") +
    scale_fill_manual(values=colorbindColors) +
    geom_text(
      stat = 'count',
      aes(label = after_stat(count)),
      position = position_stack(vjust = .5),
      color = "black",
      size = 3.5
    ) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
}
bar_plots <-
  Map(bar_plot_fun, i = metadata_categorical, j = column_variables)
bar_plots

pie_plot_fun <- function(j) {
  data <- metadata %>%
    group_by(metadata[j], ATS_names) %>%
    dplyr::count() %>%
    ungroup(all_of(j)) %>%
    mutate(perc = `n` / sum(`n`)) %>%
    arrange(perc) %>%
    mutate(labels = scales::percent(perc))
  k <- unlist(data[, c(1)])
  
  pie <- ggplot(data, aes(x = "", y = perc, fill = factor(k))) +
    geom_col(color = "black") +
    geom_label(
      aes(label = labels),
      color = c("black"),
      position = position_stack(vjust = 0.5),
      show.legend = FALSE
    ) +
    facet_wrap( ~ ATS_names, ncol = 7) +
    guides(fill = guide_legend(title = j)) +
    coord_polar(theta = "y") +
    theme_void() +
    theme(legend.position = "none")+ 
    scale_fill_manual(values=colorbindColors)
}
pie_plots <- Map(pie_plot_fun, j = column_variables)

p <- list()
for (i in 1:length(column_variables)) {
  p[[i]] <- grid.arrange(grobs = c(bar_plots[i], pie_plots[i]), heights = c(2,1))
}

plotnames = imap(pie_plots, ~ paste0("../../results/clinical/ATS.", .y, ".pdf")) %>%
  flatten()
walk2(plotnames,
      p,
      ~ ggsave(
        filename = .x,
        plot = .y,
        height = 8.5,
        width = 9.5
      ))
```
```{r message=FALSE, warning=FALSE, tidy=TRUE}
# clean up
remove(bar_plots, pie_plots, plotnames, p) # do not remove metadata_categorical
```

# Continuous variables 
### TYPE
```{r message=FALSE, warning=FALSE, tidy=TRUE}

#tapply(metadata$Age, metadata$TYPE, t.test) 
#result <- t.test(data)
#result$conf.int

metadata_continuous <-
  data.frame(
    metadata$Brain.wt,
    metadata$Duration,
    metadata$Age,
    metadata$PMI,
    metadata$Concentration.ng.ul,
    metadata$Volume.ul,
    metadata$Total.RNA.ng,
    metadata$RIN
  )
column_variables <-
  c(
    "Brain.wt",
    "Duration",
    "Age",
    "PMI",
    "Concentration.ng.ul",
    "Volume.ul",
    "Total.RNA.ng",
    "RIN"
  )
TYPE_comparison <-
  list(
    c("CONTROL", "PA"),
    c("CONTROL", "AD"),
    c("CONTROL", "LBD"),
    c("PA", "AD"),
    c("PA", "LBD"),
    c("AD", "LBD")
  )

violin_plot_fun <- function(i, j) {
  ggplot(metadata, aes(TYPE, i, fill = TYPE)) +
    geom_violin() +
    geom_boxplot(width = 0.1, outlier.shape = NA) + 
    geom_jitter(shape=16, position=position_jitter(0.2)) +
    theme_bw() + 
    ggtitle(j) +
    xlab("Disease group") +
    ylab(j) +
    stat_compare_means(comparisons = TYPE_comparison) +
    scale_fill_manual(values=TypeColors) +
    theme(legend.position = "none")
}
violin_plots <-
  Map(violin_plot_fun, i = metadata_continuous, j = column_variables)
violin_plots

plotnames = imap(violin_plots, ~paste0("../../results/clinical/TYPE.", .y, ".pdf")) %>%
  flatten()
plotnames <- gsub("metadata.", "", plotnames)
walk2(plotnames, violin_plots, ~ggsave(filename = .x, plot = .y, 
                                    height = 4, width = 5.5))

remove(violin_plots) 
```

## TYPE.sex
```{r message=FALSE, warning=FALSE, tidy=TRUE}
violin_plot_fun <- function(i, j) {
  ggplot(metadata, aes(TYPE, i, fill = TYPE)) +
    geom_violin() +
    geom_boxplot(width = 0.1, outlier.shape = NA) + 
    geom_jitter(shape=16, position=position_jitter(0.2)) +
    theme_bw() + 
    ggtitle(j) +
    xlab("Disease group") +
    ylab(j) +
    stat_compare_means(comparisons = TYPE_comparison) +
    scale_fill_manual(values=TypeColors) +
    theme(legend.position = "none") +
    facet_grid(. ~ sex_inferred)
}
violin_plots <-
  Map(violin_plot_fun, i = metadata_continuous, j = column_variables)
violin_plots

plotnames = imap(violin_plots, ~paste0("../../results/clinical/TYPE.Sex.", .y, ".pdf")) %>%
  flatten()
plotnames <- gsub("metadata.", "", plotnames)
walk2(plotnames, violin_plots, ~ggsave(filename = .x, plot = .y, 
                                    height = 4, width = 7.5))
remove(violin_plots) 
```


### ATS
```{r message=FALSE, warning=FALSE, tidy=TRUE}
metadata$ATS_names
ATS_comparison <-
  list(
    c("no pathology", "amyloid"),
    c("no pathology", "amyloid + tau"),
    c("no pathology", "pure synuclein"),
    c("no pathology", "low amyloid + synuclein"),
    c("no pathology", "high amyloid + synuclein"),
    c("no pathology", "amyloid + synuclein + tau"),
    c("amyloid", "amyloid + tau"),
    c("amyloid", "pure synuclein"),
    c("amyloid", "low amyloid + synuclein"),
    c("amyloid", "high amyloid + synuclein"),
    c("amyloid", "amyloid + synuclein + tau"),
    c("amyloid + tau", "pure synuclein"),
    c("amyloid + tau", "low amyloid + synuclein"),
    c("amyloid + tau", "high amyloid + synuclein"),
    c("amyloid + tau", "amyloid + synuclein + tau"),
    c("pure synuclein", "low amyloid + synuclein"),
    c("pure synuclein", "high amyloid + synuclein"),
    c("pure synuclein", "amyloid + synuclein + tau"),
    c("low amyloid + synuclein", "high amyloid + synuclein"),  
    c("low amyloid + synuclein", "amyloid + synuclein + tau"),   
    c("high amyloid + synuclein", "amyloid + synuclein + tau")
  )
violin_plot_fun <- function(i, j) {
  ggplot(metadata, aes(ATS_names, i, fill = ATS_names)) +
    geom_violin() +
    geom_boxplot(width = 0.1, outlier.shape = NA) + 
    geom_jitter(shape=16, position=position_jitter(0.2)) +
    theme_bw() + 
    ggtitle(j) +
    xlab("Disease group") +
    ylab(j) +
    stat_compare_means(comparisons = ATS_comparison) +
    scale_fill_manual(values=colorbindColors) +
    theme(legend.position = "none",
          axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
}
violin_plots <-
  Map(violin_plot_fun, i = metadata_continuous, j = column_variables)
violin_plots

plotnames = imap(violin_plots, ~paste0("../../results/clinical/ATS.", .y, ".pdf")) %>%
  flatten()
plotnames <- gsub("metadata.", "", plotnames)
walk2(plotnames, violin_plots, ~ggsave(filename = .x, plot = .y, 
                                    height = 7, width = 7))
remove(violin_plots) 
```

## ATS.sex
```{r message=FALSE, warning=FALSE, tidy=TRUE}
violin_plot_fun <- function(i, j) {
  ggplot(metadata, aes(ATS_names, i, fill = ATS_names)) +
    geom_violin() +
    geom_boxplot(width = 0.1, outlier.shape = NA) + 
    geom_jitter(shape=16, position=position_jitter(0.2)) +
    theme_bw() + 
    ggtitle(j) +
    xlab("Disease group") +
    ylab(j) +
    stat_compare_means(comparisons = ATS_comparison) +
    scale_fill_manual(values=colorbindColors) +
    theme(legend.position = "none",
          axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    facet_grid(. ~ sex_inferred)
}
violin_plots <-
  Map(violin_plot_fun, i = metadata_continuous, j = column_variables)
violin_plots

plotnames = imap(violin_plots, ~paste0("../../results/clinical/ATS.Sex.", .y, ".pdf")) %>%
  flatten()
plotnames <- gsub("metadata.", "", plotnames)
walk2(plotnames, violin_plots, ~ggsave(filename = .x, plot = .y, 
                                    height = 7, width = 9.5))
remove(violin_plots) 
```

### APOE
```{r message=FALSE, warning=FALSE, tidy=TRUE}
metadata$APOE <- as.factor(metadata$APOE)
metadata_continuous <-
  data.frame(
    metadata$Brain.wt,
    metadata$Duration,
    metadata$Age,
    metadata$Cing.LB,
    metadata$MF.LB,
    metadata$Braak.NFT,
    metadata$Thal.amyloid,
    metadata$MF.SP,
    metadata$MF.NFT,
    metadata$MF.Amyloid,
    metadata$Cing.Synuclein
  )
column_variables <-
  c(
    "Brain.wt",
    "Duration",
    "Age",
    "Cing.LB",
    "MF.LB",
    "Braak.NFT",
    "Thal.amyloid",
    "MF.SP",
    "MF.NFT",
    "MF.Amyloid",
    "Cing.Synuclein"
  )
APOE_comparison <-
  list(
    c("E2E3", "E2E4"),
    c("E2E3", "E3E3"),
    c("E2E3", "E3E4"),
    c("E2E3", "E4E4"),
    c("E2E4", "E3E3"),
    c("E2E4", "E3E4"),
    c("E2E4", "E4E4"),
    c("E3E3", "E3E4"),
    c("E3E3", "E4E4"),
    c("E3E4", "E4E4")
  )

violin_plot_fun <- function(i, j) {
  ggplot(metadata, aes(APOE, i, fill = APOE)) +
    geom_violin() +
    geom_boxplot(width = 0.1, outlier.shape = NA) + 
    geom_jitter(shape=16, position=position_jitter(0.2)) +
    theme_bw() + 
    ggtitle(j) +
    xlab("APOE") +
    ylab(j) +
    stat_compare_means(comparisons = APOE_comparison) +
    scale_fill_manual(values=colorbindColors) +
    theme(legend.position = "none")
}
violin_plots <-
  Map(violin_plot_fun, i = metadata_continuous, j = column_variables)
violin_plots

plotnames = imap(violin_plots, ~paste0("../../results/clinical/TYPE.APOE.", .y, ".pdf")) %>%
  flatten()
plotnames <- gsub("metadata.", "", plotnames)
walk2(plotnames, violin_plots, ~ggsave(filename = .x, plot = .y, 
                                    height = 4, width = 5.5))
remove(violin_plots) 
```

# Alignment metrics
### TYPE
```{r message=FALSE, warning=FALSE, tidy=TRUE}
metrics_metadata <-
  data.frame(
    metadata$PCT_PF_READS,
    metadata$PCT_CODING_BASES,
    metadata$PCT_INTERGENIC_BASES,
    metadata$PCT_INTRONIC_BASES,
    metadata$TOTAL_READS
  )
column_variables <-
  c(
    "PCT_PF_READS",
    "PCT_CODING_BASES",
    "PCT_INTERGENIC_BASES",
    "PCT_INTRONIC_BASES",
    "TOTAL_READS"
  )

violin_plot_fun <- function(i, j) {
  ggplot(metadata, aes(TYPE, i, fill = TYPE)) +
    geom_violin() +
    geom_boxplot(width = 0.1, outlier.shape = NA) + 
    geom_jitter(shape=16, position=position_jitter(0.2)) +
    theme_bw() + 
    ggtitle(j) +
    xlab("Disease group") +
    ylab(j) +
    stat_compare_means(comparisons = TYPE_comparison) +
    scale_fill_manual(values=TypeColors) +
    theme(legend.position = "none")
}
violin_plots <-
  Map(violin_plot_fun, i = metrics_metadata, j = column_variables)
violin_plots

plotnames = imap(violin_plots, ~paste0("../../results/clinical/TYPE.", .y, ".pdf")) %>%
  flatten()
plotnames <- gsub("metadata.", "", plotnames)
walk2(plotnames, violin_plots, ~ggsave(filename = .x, plot = .y, 
                                    height = 4, width = 5.5))

remove(violin_plots) 
```

### ATS
```{r message=FALSE, warning=FALSE, tidy=TRUE}
violin_plot_fun <- function(i, j) {
  ggplot(metadata, aes(ATS_names, i, fill = ATS_names)) +
    geom_violin() +
    geom_boxplot(width = 0.1, outlier.shape = NA) + 
    geom_jitter(shape=16, position=position_jitter(0.2)) +
    theme_bw() + 
    ggtitle(j) +
    xlab("Disease group") +
    ylab(j) +
    stat_compare_means(comparisons = ATS_comparison) +
    scale_fill_manual(values=colorbindColors) +
    theme(legend.position = "none",
          axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
}
violin_plots <-
  Map(violin_plot_fun, i = metrics_metadata, j = column_variables)
violin_plots

plotnames = imap(violin_plots, ~paste0("../../results/clinical/ATS.", .y, ".pdf")) %>%
  flatten()
plotnames <- gsub("metadata.", "", plotnames)
walk2(plotnames, violin_plots, ~ggsave(filename = .x, plot = .y, 
                                    height = 5.5, width = 6.5))
remove(violin_plots) 
```
# Correlation heatmap
scale() function is used to ensure that variables measured in different scales (e.g. age of death vs RIN) are comparable.
same results with and without the use of scale
```{r correlation, message=FALSE, warning=FALSE, tidy=TRUE}
# scale continous data 
metadata <-
  metadata[, c(
    "MF.NFT",
    "MF.Amyloid",
    "MF.SP",
    "MF.Tau",
    "Brain.wt",
   # "Duration",
    "Age",
   # "PMI",
    "RIN",
    "Braak.NFT",
    "Thal.amyloid",
    "Cing.LB",
    "Cing.Synuclein"
  )] %>% scale()
info <- as.data.frame(metadata)
help(rcorr)
# correlation
cor_mat <- rcorr(as.matrix(info))
corrplot(
  cor_mat$r,
  method = "color",
  col = correlationColors(200),
  type = "upper",
  order = "hclust",
  p.mat = cor_mat$P,
  addCoef.col = "black",
  # Add coefficient of correlation
  tl.col = "black",
  tl.srt = 45,
  #Text label color and rotation
  diag = FALSE,
  col.lim = c(-1, 1)
)
path <- paste0("../../results/clinical/correlation_continuous_variables_pathology_scores")
saveToPDF(paste0(path, ".pdf"), width = 10, height = 10)
```
```{r}
```

