---
title: "cell type enrichment"
author: "Kimberly Olney, Ph.D."
date: "03/09/2023"
output:
  html_document:
    df_print: paged
  pdf_document: default
params:
  args: myarg
---
# Setup
```{r setup}
knitr::opts_knit$set(root.dir = ".")
```
# User defined variables
```{r set_variables, warning=FALSE}
source(here::here("scripts/R", "file_paths_and_colours.R"))
source(here::here("scripts/R", "gtf_path.R"))
condition <- c("")
tool <- "star"
```
# Read in DGE object & metadata
```{r dge}
dge.filtered.norm <-
  readRDS(
    paste0(
      "../../rObjects/",
      condition, "dge.filtered.norm.rds"
    )
  )
info <- as.data.frame(dge.filtered.norm$samples)
# drop columns with cell type information if it already exists 
info <- info[,!names(info) %in%
               c(
                 "Astrocyte.Zscore",
                 "Endothelial.Zscore",
                 "Microglia.Zscore",
                 "Mural.Zscore",
                 "Neuron_All.Zscore",
                 "Neuron_Interneuron.Zscore",
                 "Neuron_Projection.Zscore",
                 "Oligodendrocyte.Zscore",
                 "Oligodendrocyte_Immature.Zscore",
                 "RBC.Zscore",
                 "Choroid_Plexus.Zscore"
               )]
biomart_gene_lengths <- read.delim("gene_id_gene_length_QC.txt")
# see biomart_gene_length_and_GC_content.R for more information on how the text file was created
# gene lenght is used to compute RPKM values for xCell 
```
# Voom transform 
The voom method estimates the mean-variance relationship of the log-counts.
Generates a precision weight for each observation and enters these into the limma empirical Bayes analysis pipeline.
voomWithDreamWeights() is the same as voom(), except that it allows random effects in the formula. limma::voomWithQualityWeights() includes observational and sample specific quality weights whereas voomWithDreamWeights only includes observational quality weights. 
```{r voom}
formula <- (~ 0 + ATS_names)
voom <-
  variancePartition::voomWithDreamWeights(
    counts = dge.filtered.norm$counts,
    formula = formula,
    data = dge.filtered.norm$samples,
    BPPARAM = BiocParallel::SnowParam(cores),
    plot = TRUE
  )
path <- paste0("../../results/", tool, "/voom/", condition,
               "voom")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 4)
voomCounts <- voom$E
```
# Design simple
```{r design, warning=FALSE}
design <-
  model.matrix(~ 0 + 
      dge.filtered.norm$samples$ATS_names)

colnames(voom$design) <-
  c("no_pathology",
    "amyloid",
    "amyloid_tau",
    "pure_synuclein",
    "low_amyloid_synuclein",
    "high_amyloid_synuclein",
    "amyloid_synuclein_tau")
```
# RPKM values
```{r RPKM}
# remove version information from gene_id so files can be merged
dge.filtered.norm$genes$gene_id <-
  gsub("\\..*", "", dge.filtered.norm$genes$gene_id)
gene_info <-
  merge(dge.filtered.norm$genes,
        biomart_gene_lengths,
        by = "gene_id",
        all.x = T) # all by dge TRUE

# reorder gene_info so it matches the order of the dge object
gene_info <-
  gene_info[match(dge.filtered.norm$genes$gene_id, gene_info$gene_id), ]
# compute RPKM values
rpkm_protein_coding <-
  rpkm(dge.filtered.norm, gene.length = gene_info$length)

# There is a gene that is duplicated. Remove or else run into errors.
rownames(rpkm_protein_coding) <- gene_info$gene_name
RPKM_no_dups <-
  rpkm_protein_coding[!duplicated(rownames(rpkm_protein_coding)),]

# output RPKM table
write.table(
  RPKM_no_dups,
  paste0(
    "../../results/",
    tool,
    "/counts/",
    condition,
    "RPKM.txt"
  ),
  sep = "\t",
  quote = FALSE,
  row.names = FALSE
)
```
# xCell 
xCell performs cell type enrichment analysis from gene expression data for 64 immune and stroma cell types. xCell is a gene signatures-based method learned from thousands of pure cell types from various sources. xCell applies a novel technique for reducing associations between closely related cell types. xCell signatures were validated using extensive in-silico simulations and also cytometry immunophenotyping, and were shown to outperform previous methods. xCell allows researchers to reliably portray the cellular heterogeneity landscape of tissue expression profiles. For more informations please refer to the xCell manuscript.
https://link.springer.com/epdf/10.1186/s13059-017-1349-1?author_access_token=DVtns3PR3raQv61Z6RD4Ym_BpE1tBhCbnbw3BuzI2RO7SD-w75iAhrQ7gjSGzw_zJO6jHEpDqZzy8CsttNZVysUprXQi0WGX-FRCguKfv2d96DcweRBG2ni-01x6T6bpU3-cfZ5nkfzCQeNeg-mMUQ== 

The expression matrix should be a matrix with genes in rows and samples in columns. The rownames should be gene symbols. If the data contains non-unique gene symbols, rows with same gene symbols will be averaged. xCell uses the expression levels ranking and not the actual values, thus normalization does not have an effect, however normalizing to gene length (RPKM/FPKM/TPM/RSEM) is required. Hence why we obtained RPKM values first. 

Importantly, xCell performs best with heterogenous dataset. Thus it is recommended to use all data combined in one run, and not break down to pieces (especially not cases and control in different runs).
```{r xCell, warning=FALSE}
# run xCellAnalysis
xCell_scores <- xCellAnalysis(RPKM_no_dups)
# save as dataframe for reformatting and plotting 
df <- as.data.frame(xCell_scores) 
xCell_dt <- df %>%
  rownames_to_column() %>%
  gather(colname, value, -rowname)
# rowname is cell type
# colname is sample ID
# value is the cell type enrichment score. 

# plot 
ggplot(xCell_dt, aes(x = colname, y = rowname, fill = value)) +
  geom_tile()
```
re-plot the xCell enrichment scores by clustering samples using the euclidean method
```{r replot_xCell}
# Obtain euclidean distance 
hc.cols <- hclust(dist(t(df)))
ord <- hclust(dist(df, method = "euclidean"), method = "ward.D")$order

# change column names to disease type so we can see if it groups by pathology 
colnames(xCell_scores) <- dge.filtered.norm$samples$TYPE
heatmap(
   as.matrix(xCell_scores), Rowv=NA,
   Colv=as.dendrogram(hclust(dist(t(as.matrix(xCell_scores)))))
 )
path <- paste0("../../results/", tool ,"/heatmap/", condition, "xCell")
saveToPDF(paste0(path, ".pdf"), width = 30, height = 7)

# group by cell type enrichment & pathology 
heatmap(xCell_scores, scale = "none")
path <- paste0("../../results/", tool ,"/cell_type_enrichment/xCell/", condition, "xCell.heatmap")
saveToPDF(paste0(path, ".pdf"), width = 12, height = 12)
```
correlation of xCell enrichment scores among pathology types
```{r corr_xCell}
dge_xCell <- xCell_scores
cormat <- round(cor(dge_xCell),2)
melted_cormat <- reshape2::melt(cormat)
ggplot(data = melted_cormat, aes(x=Var1, y=Var2, fill=value)) + 
  geom_tile()
```
save xCell cell type enrichment information
```{r save_xCell}
# output RPKM table
write.table(
  xCell_scores,
  paste0(
    "../../results/",
    tool,
    "/cell_type_enrichment/xCell/",
    condition,
    "xCell_scores.txt"
  ),
  sep = "\t",
  quote = FALSE,
  row.names = FALSE
)
```

```{r xCell_clean_up}
# clean up
remove(
  xCell_scores,
  xCell_dt,
  melted_cormat,
  hc.cols,
  dge_xCell,
  cormat,
  df,
  biomart_gene_lengths,
  voom,
  design,
  RPKM_no_dups,
  genes,
  genes.gtf
)
```

# Brain in a blender
Input: A dataframe containing the gene expression data for the samples (RNAseq or microarray data that has already been variance stabilized and received appropriate quality control to remove outlier samples and large-scale technical artifacts), including one column of gene symbols.
You must also have single character string indicating the species from which the data were derived, currently allows the values "mouse", "Mouse", "human", or "Human".

- ouput will be a list of 2 data frames \
- PublicationSpecific_CellTypeIndex - provides estimates based on the cell type specific gene lists provided by particular publications ("cell type indices") \ 
- AveragePrimary_CellTypeIndex - averages each of the publication-specific cell type indices to create an average cell type index for each primary cell type \
- A z-score measures how many standard deviations a data point is from the mean in a distribution \
```{r BinB, warning=FALSE}
# save as a dataframe (required for Sir_UnMixALot function below)
voomCounts <- as.data.frame(voomCounts)
# add gene_id column to voomCounts
voomCounts$gene_id <- rownames(voomCounts)

# re-arrange data by disease type
ByType <- dge.filtered.norm$samples %>% arrange(TYPE)
ByType_colors <- setcolorder(voomCounts, ByType$NPID)
ByType_colors$gene_id <- rownames(ByType_colors)

# remove gene id version information
ByType_colors$gene_id <- gsub("\\..*", "", ByType_colors$gene_id)
# merge voom counts with gene_info
voomCounts_gene_name <-
  merge(ByType_colors, gene_info, by = "gene_id", all.x = T)
# change directory or else the results will output to the current working directory which is the scripts folder.
setwd("/research/labs/neurology/fryer/m239830/LBD_CWOW/results/star/cell_type_enrichment/BrainInABlender")
# run brain in a blender
BinB_voom <-
  Sir_UnMixALot(
    userInput = voomCounts_gene_name,
    dataColumns = c(2:620), # samples are columns 2-620 (i.e 619 sample IDs)
    geneColumn = 631, # column with gene name 
    species = "human"
  )
saveRDS(BinB_voom, file = paste0("../../rObjects/", condition, "BinB.rds"))
```
reformat
```{r format_BinB}
zscores <- as.data.frame(t(BinB_voom[[2]]))
colnames(zscores) <- paste0(colnames(zscores),".Zscore")
zscores$group <- dge.filtered.norm$samples$ATS_names
dge.filtered.norm$samples <- cbind(dge.filtered.norm$samples, zscores)
```
### read BinB outputs
This file includes the average cell type indices for each primary cell type for each sample.
the average of the publication- specific cell type indices.
```{r read_BinB}
Zscore_Expression_CellType_NoPrimaryOverlap_Mean <-
  read_csv(
    "/research/labs/neurology/fryer/m239830/LBD_CWOW/results/star/cell_type_enrichment/BrainInABlender/Zscore_Expression_CellType_NoPrimaryOverlap_Mean.csv"
  )
# transform the data 
Zscore_CellType_melt <- melt(Zscore_Expression_CellType_NoPrimaryOverlap_Mean)
```

### merge Zscore information with metadata to compare between disease groups
```{r zscore_byType}
# get cell types
Zscore_CellType_melt$...1 <- as.factor(Zscore_CellType_melt$...1)
# save levels to loop through to make a dataframe for each
CellTypes <- levels(Zscore_CellType_melt$...1)

for (i in CellTypes) {
  CellType_df <-
    Zscore_CellType_melt[Zscore_CellType_melt$...1 == i,]
  CellType_df <- CellType_df %>%
    rename("CellType" = "...1",
           "NPID" = "variable",
           "Zscore" = "value")
  colnames(CellType_df)[3] = paste0(i, ".Zscore")
  CellType_df$CellType <- NULL
  CellType_df$NPID <- gsub(".", "-", CellType_df$NPID, fixed = TRUE)
  CellType_df$NPID <- sub("X", "", CellType_df$NPID, fixed = TRUE)
  path <-
    paste0(
      "../../results/",
      tool,
      "/cell_type_enrichment/BrainInABlender/",
      condition,
      i,
      ".txt"
    )
  write.table(
    CellType_df,
    path,
    sep = "\t",
    row.names = FALSE,
    quote = FALSE
  )
}

# read in each CellType_df
for (i in CellTypes) {
  path <-
    paste0(
      "../../results/",
      tool,
      "/cell_type_enrichment/BrainInABlender/",
      condition,
      i,
      ".txt"
    )
  assign(paste0(i),
         read.delim(
           path,
           header = TRUE,
           sep = "\t",
           stringsAsFactors = FALSE
         ))
}
# put all cell type data frames into list
df_list <-
  list(
    Astrocyte,
    Endothelial,
    Microglia,
    Mural,
    Neuron_All,
    Neuron_Interneuron,
    Neuron_Projection,
    Oligodendrocyte,
    Oligodendrocyte_Immature,
    RBC,
    Choroid_Plexus
  )
# merge all data frames in list
all_CellTypes <- df_list %>% reduce(full_join, by = 'NPID')
# merge with metadata data info
cellType <- merge(info, all_CellTypes, by = "NPID")
```
### plot cell type zscore between disease groups
```{r BinB_violins}
metadata_continuous <-
  data.frame(
    cellType$Astrocyte.Zscore,
    cellType$Endothelial.Zscore,
    cellType$Microglia.Zscore,
    cellType$Mural.Zscore,
    cellType$Neuron_All.Zscore,
    cellType$Neuron_Interneuron.Zscore,
    cellType$Neuron_Projection.Zscore,
    cellType$Oligodendrocyte.Zscore,
    cellType$Oligodendrocyte_Immature.Zscore,
    cellType$RBC.Zscore,
    cellType$Choroid_Plexus.Zscore
  )

column_variables <-
  c(
    "Astrocyte.Zscore",
    "Endothelial.Zscore",
    "Microglia.Zscore",
    "Mural.Zscore",
    "Neuron_All.Zscore",
    "Neuron_Interneuron.Zscore",
    "Neuron_Projection.Zscore",
    "Oligodendrocyte.Zscore",
    "Oligodendrocyte_Immature.Zscore",
    "RBC.Zscore",
    "Choroid_Plexus.Zscore"
  )
my_comparisons <-
  list(
    c("CONTROL", "PA"),
    c("CONTROL", "AD"),
    c("CONTROL", "LBD"),
    c("PA", "AD"),
    c("PA", "LBD"),
    c("AD", "LBD")
  )

violin_plot_fun <- function(i, j) {
  ggplot(cellType, aes(TYPE, i, fill = TYPE)) +
    geom_violin() +
    geom_boxplot(width = 0.1, outlier.shape = NA) + 
    geom_jitter(shape=16, position=position_jitter(0.2)) +
    theme_bw() + 
    ggtitle(j) +
    xlab("Disease group") +
    ylab(j) +
    stat_compare_means(comparisons = my_comparisons, method="wilcox.test") +
    scale_fill_manual(values=TypeColors) +
    theme(legend.position = "none")
}
violin_plots <-
  Map(violin_plot_fun, i = metadata_continuous, j = column_variables)
violin_plots

plotnames = imap(violin_plots, ~paste0("../../results/star/cell_type_enrichment/BrainInABlender/TYPE.", .y, ".pdf")) %>%
  flatten()
plotnames
walk2(plotnames, violin_plots, ~ggsave(filename = .x, plot = .y, 
                                    height = 4, width = 5.5))
```
### plot by disease, split by sex 
```{r BinB_TYPE_sex}
my_comparisons <-
  list(
    c("CONTROL", "PA"),
    c("CONTROL", "AD"),
    c("CONTROL", "LBD"),
    c("PA", "AD"),
    c("PA", "LBD"),
    c("AD", "LBD")
  )
violin_plot_fun <- function(i, j) {
  ggplot(cellType, aes(TYPE, i,  fill = sex_inferred)) +
    geom_violin(position=position_dodge(.8)) +
    geom_boxplot(width = 0.1, outlier.shape = NA, aes(fill=sex_inferred), position=position_dodge(.8), colour = "white") + 
    theme_bw() + 
    ggtitle(j) +
    xlab("Disease group") +
    ylab(j) +
    stat_compare_means(comparisons = my_comparisons, method="wilcox.test") +
    scale_fill_manual(values=SexColors) +
    theme(legend.position = "none")
}
violin_plots <-
  Map(violin_plot_fun, i = metadata_continuous, j = column_variables)
violin_plots

plotnames = imap(violin_plots, ~paste0("../../results/star/cell_type_enrichment/BrainInABlender/TYPE.EachSex.", .y, ".pdf")) %>%
  flatten()
walk2(plotnames, violin_plots, ~ggsave(filename = .x, plot = .y, 
                                    height = 4, width = 5.5))

violin_plot_fun <- function(i, j) {
  ggplot(cellType, aes(TYPE, i, fill = TYPE)) +
    geom_violin() +
    geom_boxplot(width = 0.1, outlier.shape = NA) + 
    geom_jitter(shape=16, position=position_jitter(0.2)) +
    theme_bw() + 
    ggtitle(j) +
    xlab("Disease group") +
    ylab(j) +
    stat_compare_means(comparisons = my_comparisons, method="wilcox.test") +
    scale_fill_manual(values=TypeColors) +
    theme(legend.position = "none") +
    facet_grid(. ~ sex_inferred)
}
violin_plots <-
  Map(violin_plot_fun, i = metadata_continuous, j = column_variables)
violin_plots

plotnames = imap(violin_plots, ~paste0("../../results/star/cell_type_enrichment/BrainInABlender/TYPE.FacetSex.", .y, ".pdf")) %>%
  flatten()
walk2(plotnames, violin_plots, ~ggsave(filename = .x, plot = .y, 
                                    height = 4, width = 8.5))
```
### plot cell type zscore between A-T-S groups 
```{r BinB_ATS}
cellType$ATS_names <-
  factor(
    cellType$ATS_names,
    levels = c(
      "no pathology",
      "amyloid",
      "amyloid + tau",
      "pure synuclein",
      "low amyloid + synuclein",
      "high amyloid + synuclein",
      "amyloid + synuclein + tau"
    )
  )

my_comparisons <-
  list(
    c("no pathology", "amyloid"),
    c("no pathology", "amyloid + tau"),
    c("no pathology", "low amyloid + synuclein"),
    c("no pathology", "high amyloid + synuclein"),
    c("no pathology", "amyloid + synuclein + tau")
  )

violin_plot_fun <- function(i, j) {
  ggplot(cellType, aes(ATS_names, i, fill = ATS_names)) +
    geom_violin() +
    geom_boxplot(width = 0.1, outlier.shape = NA) + 
    geom_jitter(shape=16, position=position_jitter(0.2)) +
    theme_bw() + 
    ggtitle(j) +
    xlab("A-T-S groups") +
    ylab(j) +
    stat_compare_means(comparisons = my_comparisons, method="wilcox.test") +
    scale_fill_manual(values=colorbindColors) +
    theme(legend.position = "none",
          axis.text.x = element_text(angle = 75, vjust = 1, hjust=1))
}
violin_plots <-
  Map(violin_plot_fun, i = metadata_continuous, j = column_variables)
violin_plots

plotnames = imap(violin_plots, ~paste0("../../results/star/cell_type_enrichment/BrainInABlender/ATS.", .y, ".pdf")) %>%
  flatten()
plotnames
walk2(plotnames, violin_plots, ~ggsave(filename = .x, plot = .y, 
                                    height = 6, width = 6.5))

violin_plot_fun <- function(i, j) {
  ggplot(cellType, aes(ATS_names, i, fill = ATS_names)) +
    geom_violin() +
    geom_boxplot(width = 0.1, outlier.shape = NA) + 
    geom_jitter(shape=16, position=position_jitter(0.2)) +
    theme_bw() + 
    ggtitle(j) +
    xlab("A-T-S groups") +
    ylab(j) +
    stat_compare_means(comparisons = my_comparisons, method="wilcox.test") +
    scale_fill_manual(values=colorbindColors) +
    theme(legend.position = "none",
          axis.text.x = element_text(angle = 75, vjust = 1, hjust=1))+
        facet_grid(. ~ sex_inferred)

}
violin_plots <-
  Map(violin_plot_fun, i = metadata_continuous, j = column_variables)
violin_plots

plotnames = imap(violin_plots, ~paste0("../../results/star/cell_type_enrichment/BrainInABlender/ATS.FacetSex.", .y, ".pdf")) %>%
  flatten()
plotnames
walk2(plotnames, violin_plots, ~ggsave(filename = .x, plot = .y, 
                                    height = 6, width = 11))
```

### formal test of difference among groups 
## Pairwise wilcoxon test
Calculate pairwise comparisons between group levels with corrections for multiple testing.
```{r wilcox}
# Pairwise wilcoxon test for cell type zscores and group
df <- Matrix(0, 11, 6)
rownames(df) <- colnames(all_CellTypes)[2:12]
colnames(df) <- c("CONTROL_vs_PA","CONTROL_vs_AD","CONTROL_vs_LBD",
                  "PA_vs_AD","PA_vs_LBD","AD_vs_LBD")
all_CellTypes$TYPE <- cellType$TYPE
for (i in colnames(all_CellTypes)[2:12]) {
  pwt <- pairwise.wilcox.test(all_CellTypes[[i]], 
                              all_CellTypes$TYPE,
                              p.adjust.method = "BH")
  pwt <- pwt$p.value
  df[i,"CONTROL_vs_PA"] <- pwt["PA","CONTROL"]
  df[i,"CONTROL_vs_AD"] <- pwt["AD","CONTROL"]
  df[i,"CONTROL_vs_LBD"] <- pwt["LBD","CONTROL"]
  df[i,"PA_vs_AD"] <- pwt["AD","PA"]
  df[i,"PA_vs_LBD"] <- pwt["LBD","PA"]
  df[i,"AD_vs_LBD"] <- pwt["LBD","AD"]
}
pwt <- df
remove(df)
pwt <- as.data.frame(as.matrix(pwt))
write.table(pwt, 
  paste0(
    "../../results/",
    tool,
    "/cell_type_enrichment/BrainInABlender/",
    condition,
    "TYPE.pairwise_Wilcox_zscore_BH.tsv"
  ),
  sep = "\t",
  quote = FALSE,
  row.names = FALSE
)
```

## kruskal.test
The Kruskal-Wallis test is a nonparametric (distribution free) test, and is used 
when the assumptions of one-way ANOVA are not met. Both the Kruskal-Wallis test 
and one-way ANOVA assess for significant differences on a continuous dependent 
variable by a categorical independent variable (with two or more groups).
```{r kruskal}
# by Type 
df <- data.frame(matrix(ncol = 1, nrow = 0))
colnames(df) <- c("kruskal.test.p.val")

for (i in colnames(all_CellTypes)[2:12]) {
  kt.pval <- kruskal.test(all_CellTypes[[i]] ~ TYPE, data = cellType)[[3]]
  print(kt.pval)
  df[i, "kruskal.test.p.val"] <- kt.pval
}
write.table(df, 
  paste0(
    "../../results/",
    tool,
    "/cell_type_enrichment/BrainInABlender/",
    condition,
    "TYPE.cellType.kruskal.test.p.val.tsv"
  ),
  sep = "\t",
  quote = FALSE,
  row.names = TRUE
)

# by A-T-S
df <- data.frame(matrix(ncol = 1, nrow = 0))
colnames(df) <- c("kruskal.test.p.val")

for (i in colnames(all_CellTypes)[2:12]) {
  kt.pval <- kruskal.test(all_CellTypes[[i]] ~ ATS_names, data = cellType)[[3]]
  print(kt.pval)
  df[i, "kruskal.test.p.val"] <- kt.pval
}

write.table(df, 
  paste0(
    "../../results/",
    tool,
    "/cell_type_enrichment/BrainInABlender/",
    condition,
    "ATS.cellType.kruskal.test.p.val.tsv"
  ),
  sep = "\t",
  quote = FALSE,
  row.names = TRUE
)
```
### heatmap 
```{r BinB_heatmap}
# save results as dataframe
BinB_ave_celltype <- as.data.frame(BinB_voom$AveragePrimary_CellTypeIndex)
# heatmap of results 
heatmap(as.matrix(BinB_ave_celltype))
path <- paste0("../../results/", tool ,"/cell_type_enrichment/BrainInABlender/heatmap_AveragePrimary_CellTypeIndex")
saveToPDF(paste0(path, ".pdf"), width = 5, height = 8)
```

### save metadata with cell type information
```{r save_BinB}
write.table(
  cellType,
  paste0("/research/labs/neurology/fryer/m239830/LBD_CWOW/rObjects/metadata_BinB_cellType_Zscore.txt"),
  sep = "\t",
  quote = FALSE,
  row.names = FALSE
)
```

