---
title: "Variance in expression data"
author: "Kimberly Olney, Ph.D."
date: "04/05/2023"
output:
  html_document:
    df_print: paged
  pdf_document: default
params:
  args: myarg
---
# Setup
```{r setup}
knitr::opts_knit$set(root.dir = ".")
```

# User defined variables
```{r set_variables, warning=FALSE}
source(here::here("scripts/R", "file_paths_and_colours.R"))
source(here::here("scripts/R", "gtf_path.R"))
condition <- c("TYPE.SEX") #  TYPE
tool = c("star")
```
# Read in DGE object & metadata
```{r dge}
dge.filtered.norm <- readRDS(paste0("../../rObjects/dge.filtered.norm.rds"))
# some samples are missing RIN values. 
# Replace NA with median RIN. 
# This is necessary to be able include RIN as a covariate in voom
# fill missing values with median
dge.filtered.norm$samples$RIN <- impute(dge.filtered.norm$samples$RIN, median)
# one sample is missing VaD information
dge.filtered.norm$samples$VaD <- impute(dge.filtered.norm$samples$VaD, median)
dge.filtered.norm$samples$flowcell_and_lane <- factor(dge.filtered.norm$samples$flowcell_and_lane)
dge.filtered.norm$samples$APOE <- factor(dge.filtered.norm$samples$APOE)

info <- as.data.frame(dge.filtered.norm$samples)
genes <- dge.filtered.norm$genes

# add a new column with TYPE + sex inferred
dge.filtered.norm$samples$TYPE_sex_inferred <-
  paste0(dge.filtered.norm$samples$TYPE,
         "_",
         dge.filtered.norm$samples$sex_inferred)
dge.filtered.norm$samples$TYPE_sex_inferred <-
  factor(
    dge.filtered.norm$samples$TYPE_sex_inferred,
    levels = c(
      "CONTROL_male",
      "CONTROL_female",
      "AD_male",
      "AD_female",
      "PA_male",
      "PA_female",
      "LBD_male",
      "LBD_female"
    )
  )
info$TYPE_sex_inferred <- dge.filtered.norm$samples$TYPE_sex_inferred
info <- as.data.frame(dge.filtered.norm$samples)
genes <- dge.filtered.norm$genes
```
# Add biomarker expression to dge metadata 
```{r genes_lcpm}
genes <- dge.filtered.norm$genes # gene information 
lcpm <- edgeR::cpm(dge.filtered.norm$counts, log = TRUE) # obtain log CPM counts
```

```{r biomarker_expression}
biomarkers <- c("ENO2", "GFAP", "OLIG2", "CD34", "P2RY12")

for (i in biomarkers) {
  biomarker <- subset(genes, gene_name == i) # gene of interest 
  biomarker_counts <- subset(lcpm, rownames(lcpm) %in% biomarker)
  biomarker_melt <- reshape2::melt(biomarker_counts) # reshape data 
  # rename columns to merge with metadata 
  names(biomarker_melt)[names(biomarker_melt) == "value"] <- i 
  # rename columns to merge with metadata 
  names(biomarker_melt)[names(biomarker_melt) == "Var2"] <- "NPID" 
  names(biomarker_melt)[names(biomarker_melt) == "Var1"] <- "gene_id"
  biomarker_melt$gene_id <- NULL
  assign(paste0(i),biomarker_melt)
}

# add gene expression values into one dataframe 
# put all data frames into list
df_list <- list(ENO2, OLIG2, CD34, P2RY12, GFAP)

# merge all data frames in list
cell_biomarker_lcpm <- df_list %>% reduce(full_join, by='NPID')
```
# Scale data
rescaling a predictor in a regression has absolutely no effect on the magnitude of the relation being studiedâ€”the slope itself will not change its steepness, nor will the p-values or variance explained be changed. Rescaling is merely a means of communicating the nature of the regression line in different, hopefully more intuitive language.
```{r scale}
df <- merge(cell_biomarker_lcpm, info, by = "NPID")
scaled.info <-
  df[c(
    "Race_numeric",
    "RIN",
    "Age",
    "PCT_CODING_BASES",
    "PCT_INTERGENIC_BASES",
    "PCT_INTRONIC_BASES",
    "APOE_E4_allele_count", 
    "ENO2", 
    "GFAP", 
    "OLIG2", 
    "CD34", 
    "P2RY12",
    "Brain.wt"
  )] %>% scale()
scaled.info.df <- as.data.frame(scaled.info)
# remove columns with unscaled data 
df <- (subset(df, select = -c(Race_numeric, RIN, Age, PCT_CODING_BASES, PCT_INTRONIC_BASES, APOE_E4_allele_count, ENO2, GFAP, OLIG2, CD34, P2RY12, Brain.wt)))
# Add scaled information to the metadata called "info"
info_with_scale <- cbind(df, scaled.info.df)
```

```{r}
all.equal(dge.filtered.norm$samples$NPID, as.character(info_with_scale$NPID))
# replace sample information with the updated info that includes biomakrer expression
dge.filtered.norm$samples <- info_with_scale
```
Voom transform counts to use for BIC 
```{r voom, warning=FALSE}
formula <- (~ 0 + TYPE_sex_inferred)
voom_with_weights <-
  variancePartition::voomWithDreamWeights(
    counts = dge.filtered.norm$counts,
    formula = formula,
    data = dge.filtered.norm$samples,
    BPPARAM = BiocParallel::SnowParam(cores),
    plot = TRUE
  )
path <- paste0("../../results/", tool, "/voom/", condition, ".raw.voom")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 4)
voomCounts <- voom_with_weights$E
```
# Fit variance 
variancePartition quantifies and interprets multiple sources of biological and technical variation in gene expression experiments. The package a linear mixed model to quantify variation in gene expression attributable to individual, tissue, time point, or technical variables.
```{r varpart}
form_varPart <- ~ (1|TYPE_sex_inferred) + 
  (1|flowcell_and_lane) +
  PCT_CODING_BASES +
  PCT_INTERGENIC_BASES + 
  PCT_INTRONIC_BASES +
  RIN +
  ENO2 

# fit model and extract variance percents
varPart <- fitExtractVarPartModel(voomCounts, form_varPart, info_with_scale,
                                  showWarnings=FALSE)
```

```{r}
setnames(varPart, old = c('flowcell_and_lane'), 
         new = c('Batch'))
plotVarPart(sortCols(varPart), label.angle = 80)
path <-
  paste0(
    "../../results/",
    tool,
    "/varpart/",
    condition,
    ".finalmodel.varpart"
  )
saveToPDF(paste0(path, ".pdf"), width = 8, height = 7)


# sort variables (i.e. columns) by median fraction # of variance explained
vp <- sortCols( varPart )
# Bar plot of variance fractions for the first 10 genes plotPercentBars( vp[1:10,] )
plotPercentBars( vp[1:10,] )

varPart$gene_id <- rownames(varPart)
# merge with gene information to get gene names for gene_id
variance_explained <- merge(varPart, genes, by = "gene_id")
# Reorder columns
variance_explained_reorder <- variance_explained[, c(1,21,11,2:10)]

# remove unnecessary columns 
write.table(
  variance_explained_reorder,
  paste0(
    "../../results/",
    tool ,
    "/varpart/",
    condition,
    ".finalmodel.variance.explained.tsv"
  ),
  sep = "\t",
  quote = FALSE
)
```
# CCA 
```{r}
form <- ~ TYPE_sex_inferred + 
  flowcell_and_lane +
  PCT_CODING_BASES +
  PCT_INTERGENIC_BASES + 
  PCT_INTRONIC_BASES +
  RIN +
  ENO2 

# Compute Canonical Correlation Analysis (CCA) # between all pairs of variables
# returns absolute correlation value
C = canCorPairs( form, info_with_scale)
# Plot correlation matrix
plotCorrMatrix( C )

path <-
  paste0(
    "../../results/",
    tool,
    "/varpart/",
    condition,
    ".finalmodel.CCA"
  )
saveToPDF(paste0(path, ".pdf"), width = 8, height = 8)
```

# BIC with forward stepwise regression
First, we will scale some of the continuous variables. 
In regression, it is often recommended to scale the features so that the predictors have a mean of 0. This makes it easier to interpret the intercept term as the expected value of Y when the predictor values are set to their means.
scale is generic function whose default method centers and/or scales the columns of a numeric matrix.

Secondly, obtian voom$E counts to use in the BIC. 
Finally, perform forwards stepwise regression Bayesian information criterion (BIC) to determine the best model. 
see: https://rdrr.io/github/GabrielHoffman/mvIC/man/mvIC_fit.html 
```{r BIC, eval = FALSE}
baseFormula <- ~ (1 | TYPE_sex_inferred)
# Combine responses on *rows*
# Add brain weight, brain weight and sex interaction, Age and sex interaction 
Y = with(
  info,
  rbind(
    flowcell_and_lane,
    scaled.info.df$RIN,
    scaled.info.df$PCT_CODING_BASES,
    scaled.info.df$PCT_INTERGENIC_BASES,
    scaled.info.df$PCT_INTRONIC_BASES, 
    scaled.info.df$ENO2
  )
)

rownames(Y) <-
  c(
    "flowcell_and_lane",
    "RIN",
    "PCT_CODING_BASES",
    "PCT_INTERGENIC_BASES",
    "PCT_INTRONIC_BASES",
    "ENO2"
  )
# variables to consider in the model
# categorical variables must be modeled using (1|)
variables = c(
  "(1|flowcell_and_lane)",
  "RIN",
  "PCT_CODING_BASES",
  "PCT_INTERGENIC_BASES",
  "PCT_INTRONIC_BASES",
  "ENO2"
)

# fit forward stepwise regression starting
bestModel_voomcounts = mvForwardStepwise(voomCounts,
                                         baseFormula,
                                         data = info_with_scale,
                                         variables = variables)
bestModel_voomcounts
```
# Design matrix
```{r design}
design <-
  model.matrix(~ 0 + 
      TYPE_sex_inferred + 
      flowcell_and_lane + 
      RIN +
      PCT_CODING_BASES +
      PCT_INTERGENIC_BASES + 
      PCT_INTRONIC_BASES +
      ENO2,
    dge.filtered.norm$samples
  )

colnames(design) <-
  c(
    "CONTROL_male", 
    "CONTROL_female", 
    "AD_male", 
    "AD_female", 
    "PA_male",
    "PA_female", 
    "LBD_male",
    "LBD_female",
    "Batch1",
    "Batch2",
    "Batch3",
    "Batch4",
    "Batch5",
    "Batch6",
    "Batch7",
    "RIN",
    "PCT_CODING_BASES", 
    "PCT_INTERGENIC_BASES", 
    "PCT_INTRONIC_BASES",
    "ENO2"
  )
```
# Voom
When the library sizes are quite variable between samples, then the voom approach is theoretically more powerful than limma-trend. 
The voom method estimates the mean-variance relationship of the log-counts.
Generates a precision weight for each observation and enters these into the limma empirical Bayes analysis pipeline.
```{r voom_BIC}
form <- (
  ~ 0 +
      TYPE_sex_inferred + 
      flowcell_and_lane + 
      RIN +
      PCT_CODING_BASES +
      PCT_INTERGENIC_BASES + 
      PCT_INTRONIC_BASES +
      ENO2
)

voom_cov <-
  variancePartition::voomWithDreamWeights(
    counts = dge.filtered.norm$counts,
    formula = form,
    data = dge.filtered.norm$samples,
    BPPARAM = BiocParallel::SnowParam(cores),
    plot = TRUE
  )
path <-
  paste0("../../results/",
         tool,
         "/voom/",
         condition,
         ".voom.model")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 4)
voomCounts <- voom_cov$E
```
# Visualize contrast matrix 

Consider that, if there were no interaction, then the combined effect of treatment and mutation would be purely additive, i.e., equal to the sum of the separate effects of treatment and mutation. 
```{r}
simple_form <- ~ 0 + TYPE_sex_inferred
myDreamContrasts <- c(
  # pathology differences
  ADvsControl = "(TYPE_sex_inferredAD_male + TYPE_sex_inferredAD_female)/2 - (TYPE_sex_inferredCONTROL_male + TYPE_sex_inferredCONTROL_female)/2",
  PAvsControl = "(TYPE_sex_inferredPA_male + TYPE_sex_inferredPA_female)/2 - (TYPE_sex_inferredCONTROL_male + TYPE_sex_inferredCONTROL_female)/2",
  LBDvsControl = "(TYPE_sex_inferredLBD_male + TYPE_sex_inferredLBD_female)/2 - (TYPE_sex_inferredCONTROL_male + TYPE_sex_inferredCONTROL_female)/2",
  # sex differences - effect of sex 
  sex_diff_Control = "TYPE_sex_inferredCONTROL_female -TYPE_sex_inferredCONTROL_male", 
  sex_diff_AD = "TYPE_sex_inferredAD_female - TYPE_sex_inferredAD_male",
  sex_diff_PA = "TYPE_sex_inferredPA_female - TYPE_sex_inferredPA_male",
  sex_diff_LBD = "TYPE_sex_inferredLBD_female -TYPE_sex_inferredLBD_male",
  # effect of disease within males 
  male_ADvsControl = "TYPE_sex_inferredAD_male - TYPE_sex_inferredCONTROL_male",
  male_PAvsControl = "TYPE_sex_inferredPA_male - TYPE_sex_inferredCONTROL_male",
  male_LBDvsControl = "TYPE_sex_inferredLBD_male - TYPE_sex_inferredCONTROL_male",
  male_LBDvsAD = "TYPE_sex_inferredLBD_male - TYPE_sex_inferredAD_male",
  # effect of disease within females 
  female_ADvsControl = "TYPE_sex_inferredAD_female - TYPE_sex_inferredCONTROL_female",
  female_PAvsControl = "TYPE_sex_inferredPA_female - TYPE_sex_inferredCONTROL_female",
  female_LBDvsControl = "TYPE_sex_inferredLBD_female - TYPE_sex_inferredCONTROL_female",
  female_LBDvsAD = "TYPE_sex_inferredLBD_female - TYPE_sex_inferredAD_female",
  #  want to know whether the effect of sex is the same in both disease type and control. 
  sex_interaction_ADvsControl = "(TYPE_sex_inferredAD_female - TYPE_sex_inferredAD_male) - (TYPE_sex_inferredCONTROL_male - TYPE_sex_inferredCONTROL_female)",
  sex_interaction_PAvsControl = "(TYPE_sex_inferredPA_female - TYPE_sex_inferredPA_male) - (TYPE_sex_inferredCONTROL_male - TYPE_sex_inferredCONTROL_female)",
  sex_interaction_LBDvsControl = "(TYPE_sex_inferredLBD_female - TYPE_sex_inferredLBD_male) - (TYPE_sex_inferredCONTROL_male - TYPE_sex_inferredCONTROL_female)", 
  
  disease_interaction_ADvsControl = "(TYPE_sex_inferredAD_female - TYPE_sex_inferredCONTROL_female) - (TYPE_sex_inferredAD_male  - TYPE_sex_inferredCONTROL_male)",
  disease_interaction_PAvsControl = "(TYPE_sex_inferredPA_female - TYPE_sex_inferredCONTROL_female) - (TYPE_sex_inferredPA_male - TYPE_sex_inferredCONTROL_male)",
  disease_interaction_LBDvsControl = "(TYPE_sex_inferredLBD_female - TYPE_sex_inferredCONTROL_female) - (TYPE_sex_inferredLBD_male - TYPE_sex_inferredCONTROL_male)", 
)

myDreamContrasts <- rev(myDreamContrasts)

L = makeContrastsDream(simple_form, dge.filtered.norm$samples, contrasts = myDreamContrasts)
colnames(L) <- gsub("TYPE_sex_inferred", "", colnames(L))
rownames(L) <- gsub("TYPE_sex_inferred", "", rownames(L))

# Visualize contrast matrix
plotContrasts(L)
```
# Contrast plot
### pairwise TYPE.SEX 
```{r contrasts}
# fits linear model for each gene given a series of arrays
fit <- lmFit(voom_cov, design)
coef.fit <- fit$coefficients

contrasts <- makeContrasts(
  LBDvsControl_no_sex_correction = (LBD_male + LBD_female)/2 - (CONTROL_male + CONTROL_female)/2,
  LBDvsAD_no_sex_correction  = (LBD_male + LBD_female)/2 - (AD_male + AD_female)/2,
  LBDvsPA_no_sex_correction  = (LBD_male + LBD_female)/2 - (PA_male + PA_female)/2,
  ADvsControl_no_sex_correction  = (AD_male + AD_female)/2 - (CONTROL_male + CONTROL_female)/2,
  PAvsControl_no_sex_correction  = (PA_male + PA_female)/2 - (CONTROL_male + CONTROL_female)/2,
  ADvsPA_no_sex_correction  = (AD_male + AD_female)/2 - (PA_male + PA_female)/2,
  male_XY_LBDvsControl = LBD_male - CONTROL_male,
  male_XY_LBDvsAD = LBD_male  - AD_male,
  male_XY_LBDvsPA = LBD_male - PA_male,
  male_XY_ADvsControl = AD_male - CONTROL_male,
  male_XY_PAvsControl = PA_male - CONTROL_male,
  male_XY_ADvsPA = AD_male  - PA_male,
  female_XX_LBDvsControl = LBD_female - CONTROL_female,
  female_XX_LBDvsAD = LBD_female  - AD_female,
  female_XX_LBDvsPA = LBD_female - PA_female,
  female_XX_ADvsControl = AD_female - CONTROL_female,
  female_XX_PAvsControl = PA_female - CONTROL_female,
  female_XX_ADvsPA = AD_female  - PA_female,
  female_LBD_vs_male_LBD = LBD_female - LBD_male, 
  female_AD_vs_male_AD = AD_female - AD_male, 
  female_PA_vs_male_PA = PA_female - PA_male, 
  female_Control_vs_male_Control = CONTROL_female - CONTROL_male, 
  ADvsControl_sex_interaction = ((AD_female -  CONTROL_female) - (AD_male - CONTROL_male)),
  PAvsControl_sex_interaction = ((PA_female - CONTROL_female) - (PA_male - CONTROL_male)),
  LBDvsControl_sex_interaction = ((LBD_female - CONTROL_female) - (LBD_male - CONTROL_male)),
  LBDvsAD_sex_interaction = ((LBD_female - AD_female ) - (LBD_male - AD_male)),
  LBDvsPA_sex_interaction = ((LBD_female - PA_female ) - (LBD_male - PA_male)),
  ADvsPA_sex_interaction = ((AD_female - PA_female) - (AD_male - PA_male)),
  levels = colnames(design))
contrasts

# save contrast names
allComparisons <- colnames(contrasts)
allComparisons # check

# run contrast analysis
vfit <- contrasts.fit(fit, contrasts = contrasts)

# Compute differential expression based on the empirical Bayes moderation of the
# standard errors towards a common value.
# The logCPM values can then be used in any standard limma pipeline, using the trend=TRUE
# argument when running eBayes or treat. For example:
veBayesFit <- eBayes(vfit, trend = TRUE, robust=TRUE)
plotSA(veBayesFit, main = "Final Model: Mean-variance Trend")
path <-
  paste0("../../results/",
         tool,
         "/voom/",
         condition,
         ".voom.eBayesFinalModel")
saveToPDF(paste0(path, ".pdf"), width = 9, height = 5)
# Rather than worry about the normalization too much, better to explore the data. E.g. try a BCV plot to look for dispersion outliers, or try robust=TRUE with eBayes() to downweight dispersion outliers. 

#disp <- estimateDisp(dge.filtered.norm, design, robust=TRUE)
#plotBCV(disp)
```

# DEGs summary
```{r DGE_summary}
pval <- 0.05
lfc.cutoff <- 0.25

sumTable <- 
  summary(decideTests(
    veBayesFit,  # object
    adjust.method = "BH", # by default the method = "separate"
    p.value = pval,
    lfc = lfc.cutoff  # numeric, minimum absolute log2-fold change required
  ))

print(paste0(" FDRq < ", pval,
             " & absolute log2-fold change > ", lfc.cutoff))
sumTable
write.table(sumTable, 
            paste0("../../results/", tool, "/DEGs/", condition, ".DEGs.summary.txt"), 
            quote = FALSE, sep = "\t")
```
# Add gene information to DEGs
reformat genes table to only include relevant information
```{r}
genes_relevant <- dplyr::select(genes, 1:4,10:12)
```
Check 
```{r DGE_check, eval=FALSE}
test <- topTable(
  veBayesFit, 
  coef = "LBDvsControl_sex_interaction",  
  n = Inf, 
  p.value = 1,
  lfc = 0, 
  sort.by = "P", 
  genelist = genes_relevant, 
  confint = TRUE # column of confidence interval 
    )
#head(test, 20)
#subset(test, gene_name == "SNCB") 
```
# Save objects
```{r save_voom}
saveRDS(veBayesFit, file = paste0("../../rObjects/", condition, ".veBayesFit.rds"))
saveRDS(voomCounts, file = paste0("../../rObjects/", condition, ".voomCountsMatrix.rds"))
```
# cool map
```{r}
LBDvsControl <- topTable(veBayesFit, coef = 'female_XX_LBDvsControl', 
                         p.value = 0.05, adjust.method = 'fdr',
                  number = Inf, genelist = genes_relevant)
LBDvsControl$gene_id
up <- LBDvsControl$gene_id[LBDvsControl$logFC > .25][1:15]
down <- LBDvsControl$gene_id[LBDvsControl$logFC < -.25][1:15]
select <- c(up, down)
coolmap(voom_cov[select,])
```

# Output DEG tables
```{r DGE_output}
coef <- 1

for (i in allComparisons) {
  vTopTableAll <- topTable(
    veBayesFit, 
    coef = coef,  
    n = Inf, 
    p.value = 1,
    lfc = 0, 
    sort.by = "P", 
    genelist = genes_relevant, 
    confint = TRUE # column of confidence interval 
    )
    saveRDS(vTopTableAll, file = 
            paste0("../../rObjects/gene_tables/", condition, "_", 
                   i,"_gene_table.rds"))
  path <- paste0("../../results/", tool, "/DEGs/", condition, "_", 
  i, "_gene_DEGs_FDRq1.00.txt", sep = "") 
  write.table(
    vTopTableAll,
    path,
    sep = "\t",
    row.names = FALSE,
    quote = FALSE
  )
  # p < 0.05, log2fc > 0
  vTopTable1 <-
    topTable( 
      veBayesFit,  
      coef = coef,  
      n = Inf, 
      p.value = pval,
      lfc = lfc.cutoff,
      genelist = genes_relevant, 
      confint = TRUE # column of confidence interval 
    )
  path <- paste0("../../results/", tool, "/DEGs/", condition, "_", 
  i, "_gene_DEGs_FDRq0.05_logFC_0.25.txt", sep = "") 
  write.table(
    vTopTable1,
    path,
    sep = "\t",
    row.names = FALSE,
    quote = FALSE
  )
  # increment 
  coef <- coef + 1
}
remove(coef)
```

# PCA
Principal component analysis, or PCA, is a dimensionality reduction method that is often used to reduce the dimensionality of large data sets, by transforming a large set of variables into a smaller one that still contains most of the information in the large set.
```{r PCA}
# Setting the N of genes to use
ntop = length(dge.filtered.norm$genes$gene_id)
# Sorting by the coefficient of variance
means <- rowMeans(voomCounts)
Pvars <- rowVars(voomCounts)
cv2 <- Pvars / means ^ 2
select <-
  order(cv2, decreasing = TRUE)[seq_len(min(ntop, length(cv2)))]
head(select)

highly_variable_exp <- ((voomCounts)[select,])
dim(highly_variable_exp)
# Running PCA
pca_exp <- prcomp(t(highly_variable_exp), scale = F, center = T)
# scale a logical value indicating whether the variables should be scaled
# to have unit variance before the analysis takes place.
# a logical value indicating whether the variables should be shifted to be zero centered.
head(pca_exp$x)[, 1:3]
summary(pca_exp)
# Dataframe with the first 10 PCs
dim1_10 <- data.frame(pca_exp$x[, 1:10])
# Adding metadata
dim1_10$NPID <- rownames(dim1_10)
pcaWithMetadata <- merge(dim1_10, info_with_scale, by = "NPID", all = TRUE)
pcaWithMetadata$group <- pcaWithMetadata$TYPE

# Plotting
ggplot(data = pcaWithMetadata, aes(x = PC1, y = PC2, shape = group, color = group)) +
  geom_point(size = 2.5) +
  theme_bw() +
  scale_color_manual(values = colorbindColors) 

ggplot(data = pcaWithMetadata, aes(x = PC2, y = PC3, shape = group, color = group)) +
  geom_point(size = 2.5) +
  theme_bw() +
  scale_color_manual(values = colorbindColors) 

ggplot(data = pcaWithMetadata, aes(x = PC3, y = PC4, shape = group, color = group)) +
  geom_point(size = 2.5) +
  theme_bw()

ggplot(data = pcaWithMetadata, aes(x = PC5, y = PC6, shape = sex_inferred, color = sex_inferred)) +
  geom_point(size = 2.5) 
```
# Info with PCA output 
```{r}
write.table(sumTable, 
            paste0("../../rObjects/", condition, ".metadata.PCA.txt"), 
            quote = FALSE, sep = "\t")

```
# CCA PC1-10 & variables in model 
```{r CCA_PCA}
form_PCA <- ~ TYPE_sex_inferred + 
  flowcell_and_lane +
  PCT_CODING_BASES +
  PCT_INTERGENIC_BASES + 
  PCT_INTRONIC_BASES +
  RIN +
  Age +
  PC1 +
  PC2 +
  PC3 +
  PC4 +
  PC5 +
  PC6 +
  PC7 +
  PC8 +
  PC9 +
  PC10 

C = canCorPairs(form_PCA, pcaWithMetadata)
# Plot correlation matrix
cor.mtest <- function(C, ...) {
    C <- as.matrix(C)
    n <- ncol(C)
    p.mat<- matrix(NA, n, n)
    diag(p.mat) <- 0
    for (i in 1:(n - 1)) {
        for (j in (i + 1):n) {
            tmp <- cor.test(C[, i], C[, j], ...)
            p.mat[i, j] <- p.mat[j, i] <- tmp$p.value
        }
    }
  colnames(p.mat) <- rownames(p.mat) <- colnames(C)
  p.mat
}
# matrix of the p-value of the correlation
p.mat <- cor.mtest(C)
col <- colorRampPalette(c("#4477AA", "#77AADD", "#FFFFFF", "#EE9988", "#BB4444"))
  corrplot(C, method="color", col=col(200),  
         type="upper", order="hclust", 
         addCoef.col = "black", # Add coefficient of correlation
         tl.col="black", tl.srt=45, #Text label color and rotation
         diag=FALSE, col.lim = c(0, 1)
         )
path <- paste0("../../results/", tool ,"/varpart/", condition, ".CCA_PC1_10")
saveToPDF(paste0(path, ".pdf"), width = 20, height = 20)
```
# Glimma
```{r}
Sys.setenv(RSTUDIO_PANDOC="/usr/local/biotools/pandoc/3.1.2/bin")
glimmaMDS(dge.filtered.norm)
genes <- genes[,c(1,10,12)]
glimmaVolcano(veBayesFit, dge = dge.filtered.norm, coef = "male_XY_LBDvsContro", counts = dge.filtered.norm$counts, anno = genes)
```

```{r}
sessionInfo()
```