---
title: "WGCNA"
author: "Kimberly Olney, Ph.D."
date: "03/31/2023"
output:
  html_document:
    df_print: paged
  pdf_document: default
params:
  args: myarg
---
Tutorials: https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/Tutorials/index.html 
Background and glossary information https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/Tutorials/Simulated-00-Background.pdf 

Overview of steps:
1) data input
2) network construction
3) module detection
4) relating modules and genes to external information 

# Setup
```{r setup}
knitr::opts_knit$set(root.dir = ".")
```
# Install
GO enrichment calculations in WGCNA are deprecated recommended to use the R package anRichment which provides replacement for WGCNA functions GOenrichmentAnalysis() and userListEnrichment(). We will also use gprofiler. 
```{r install eval=FALSE}
BiocManager::install("WGCNA")
source("https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/GeneAnnotation/installAnRichment.R")
installAnRichment()
install.packages(c("AnnotationDBI", "GO.db", "org.Hs.eg.db", "org.Mm.eg.db", "TxDb.Hsapiens.UCSC.hg19.knownGene", "TxDb.Mmusculus.UCSC.mm10.knownGene"))
remotes::install_github("kevinblighe/CorLevelPlot")
```
# Libaries
```{r library}
library(WGCNA)
library(anRichmentMethods)
library(tidyverse)
library(CorLevelPlot)
library(gridExtra)
```
# User defined variables
```{r set_variables, warning=FALSE}
source(here::here("/LBD_CWOW/scripts/R", "file_paths_and_colours.R"))
condition <- c("TYPE") 
tool <- c("star")
```
# Read in DGE object & metadata
```{r dge}
dge.filtered.norm <- readRDS(paste0("../../rObjects/dge.filtered.norm.rds"))
dge.filtered.norm$samples$RIN <- impute(dge.filtered.norm$samples$RIN, median)
dge.filtered.norm$samples$flowcell_and_lane <- factor(dge.filtered.norm$samples$flowcell_and_lane)
dge.filtered.norm$samples$APOE <- factor(dge.filtered.norm$samples$APOE)

info <- as.data.frame(dge.filtered.norm$samples)
genes <- dge.filtered.norm$genes
```
# lcpm data 
```{r genes_lcpm}
lcpm <- edgeR::cpm(dge.filtered.norm$counts, log = TRUE) # obtain log CPM counts
```
# Add biomarker expression to dge metadata 
```{r biomarker_expression}
biomarkers <- c("ENO2", "GFAP", "OLIG2", "CD34", "P2RY12")

for (i in biomarkers) {
  biomarker <- subset(genes, gene_name == i) # gene of interest 
  biomarker_counts <- subset(lcpm, rownames(lcpm) %in% biomarker)
  biomarker_melt <- reshape2::melt(biomarker_counts) # reshape data 
  # rename columns to merge with metadata 
  names(biomarker_melt)[names(biomarker_melt) == "value"] <- i 
  # rename columns to merge with metadata 
  names(biomarker_melt)[names(biomarker_melt) == "Var2"] <- "NPID" 
  names(biomarker_melt)[names(biomarker_melt) == "Var1"] <- "gene_id"
  biomarker_melt$gene_id <- NULL
  assign(paste0(i),biomarker_melt)
}

# add gene expression values into one data frame 
# put all data frames into a list
df_list <- list(ENO2, OLIG2, CD34, P2RY12, GFAP)

# merge all data frames in list
cell_biomarker_lcpm <- df_list %>% reduce(full_join, by='NPID')
```
# Scale data
rescaling a predictor in a regression has absolutely no effect on the magnitude of the relation being studied—the slope itself will not change its steepness, nor will the p-values or variance explained be changed. Re-scaling is merely a means of communicating the nature of the regression line in different, hopefully more intuitive language.
```{r scale}
# New df with cell type biomarker expression and sample metadata 
df <- merge(cell_biomarker_lcpm, info, by = "NPID")
# scale data 
scaled.info <-
  df[c(
    "Race_numeric",
    "RIN",
    "Age",
    "PCT_CODING_BASES",
    "PCT_INTERGENIC_BASES",
    "PCT_INTRONIC_BASES",
    "APOE_E4_allele_count", 
    "ENO2", 
    "GFAP", 
    "OLIG2", 
    "CD34", 
    "P2RY12",
    "Brain.wt"
  )] %>% scale()
scaled.info.df <- as.data.frame(scaled.info)
# remove columns with unscaled data 
df <- (subset(df, select = -c(Race_numeric, RIN, Age, PCT_CODING_BASES, PCT_INTRONIC_BASES, APOE_E4_allele_count, ENO2, GFAP, OLIG2, CD34, P2RY12, Brain.wt)))
# Add scaled information to the metadata called "info"
info_with_scale <- cbind(df, scaled.info.df)
```
# replace sample information with scaled sample information
```{r replace_info}
all.equal(dge.filtered.norm$samples$NPID, as.character(info_with_scale$NPID))
# replace sample information with the updated info that includes biomakrer expression
dge.filtered.norm$samples <- info_with_scale
```
# Voom transform counts 
```{r voom, warning=FALSE}
formula <- (~ 0 + TYPE)
voom_with_weights <-
  variancePartition::voomWithDreamWeights(
    counts = dge.filtered.norm$counts,
    formula = formula,
    data = dge.filtered.norm$samples,
    BPPARAM = BiocParallel::SnowParam(cores),
    plot = FALSE
  )
voomCounts <- voom_with_weights$E
```
# Counts
The voom method (Law et al., 2014) stabilizes the variance of RNA-seq data. It does so by identifying the mean–variance relationship of the data at the gene-level.
WGCNA reads as input counts data that is stabilized. 
```{r counts}
# transpose data to be in the correct format for WGCNA 
counts <- t(voomCounts)
```
# Data traits
```{r data_traits}
dim(counts)
dim(info)
# keep only the columns that hold information we need. 
dataTraits <- info[, c(6,11:19,21,22,24,29,37,108,109,114,117:124)] # Keep
dim(dataTraits)
# disease type isn't numeric, make numeric as this is required down stream
numeric_traits <- sapply(dataTraits, as.numeric)
numeric_traits <- as.data.frame(numeric_traits) # save as data frame
rownames(numeric_traits) <- rownames(dataTraits) # add NPID as row names
```
# check data
```{r check_dfs}
all.equal(info$NPID, rownames(counts))
all.equal(info$NPID, colnames(lcpm))
```
# QC - outlier detection
goodSamplesGenes - This function checks data for missing entries, entries with weights below a threshold, and zero-variance genes, and returns a list of samples and genes that pass criteria on maximum number of missing or low weight values. If necessary, the filtering is iterated.
```{r QC_outliers}
gsg <- goodSamplesGenes(counts)
summary(gsg)

table(gsg$goodGenes) # outlier genes
table(gsg$goodSamples) # outlier samples
# no outliers
```
# hclust & pca
Hierarchical clustering analysis on a set of dissimilarities and methods for analyzing it.
default method is "complete" - The complete linkage method finds similar clusters. 

prcomp - Performs a principal components analysis on the given data matrix and returns the results as an object of class prcomp.
```{r hclust_pca}
# Hierarchical clustering 
htree <- hclust(dist(counts))
htree_plot <- plot(htree, xlab = "", ylab = "Height", main = "Sample clustering to detect outliers")
htree_plot

# PCA 
pca <- prcomp(counts)
pca.dat <- pca$x
pca.var <- pca$sdev^2
pca.var.percent <- round(pca.var/sum(pca.var)*100, digits = 2)
pca.dat <- as.data.frame(pca.dat)
ggplot(pca.dat, aes(PC1, PC2)) +
  geom_point() +
  geom_text(label = rownames(pca.dat)) +
  labs(x = paste0('PC1: ', pca.var.percent[1], '%'))
```

# soft threshold - power
which power gives a scale-free topology 
Analysis of scale free topology for multiple soft thresholding powers. The aim is to help the user pick an appropriate soft-thresholding power for network construction.

networkType - signed, the direction of the gene expression 
```{r soft_thres}
# choose a set of soft thresholding powers
power <- c(c(1:10), seq(from=12, to=20, by=2)) 
allowWGCNAThreads() # multithreads
# call network topology analysis function 
sft <- pickSoftThreshold(counts, powerVector = power, 
                         networkType = "signed", verbose = 5)
save(sft, file = "../../rObjects/sft.Rdata")
```
# Pick a power
```{r pick_power}
sft.data <- sft$fitIndices

# plot
scale_ind <- ggplot(sft.data, aes(Power, SFT.R.sq, label= Power)) +
  geom_text(label = sft.data$Power, color = "red") +
  geom_hline(yintercept = 0.8, color = "red") +
  labs(x = 'Soft Threshold (power)', y = 'Scale free topology model fit, signed R^2', title = "Scale independence") +
  theme_classic()

mean_con <- ggplot(sft.data, aes(Power, mean.k., label = Power)) +
  geom_text(color = "red") +
  labs(x = 'Soft Threshold (power)', y = 'Mean Connectivity', title = 'Mean Connectivity') +
  theme_classic()

ggarrange(
  scale_ind,
  mean_con,
  ncol = 2
  )
path <- paste0("../../results/star/WGCNA/scale_independence_mean_connectivity")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 4)
```
# Bloackwise 
This function performs automatic network construction and module detection on large expression datasets in a block-wise manner.

For each block of genes, the network is constructed and (if requested) topological overlap is calculated. If requested, the topological overlaps are returned as part of the return value list. Genes are then clustered using average linkage hierarchical clustering and modules are identified in the resulting dendrogram by the Dynamic Hybrid tree cut. Found modules are trimmed of genes whose correlation with module eigengene (KME) is less than minKMEtoStay. Modules in which fewer than minCoreKMESize genes have KME higher than minCoreKME are disbanded, i.e., their constituent genes are pronounced unassigned.

correlation type is pearson 
Will define which cor to use. There is cor in stats package that will throw an error if you don't specify that cor should be from the WGCNA package 
```{r blockwise}
counts[] <- sapply(counts, as.numeric)
soft_power <- 12
temp_cor <- cor
cor <- WGCNA::cor 
bwnet <- blockwiseModules(counts, 
                 maxBlockSize = 15000, # all genes
                 TOMType = "signed", # direction of the gene 
                 power = soft_power, # power decided from step above 
                 mergeCutHeight = 0.25, # height at which to merge modules 
                 numericLabels = FALSE, 
                 randomSeed = 28, # to get the same results again 
                 verbose = 3
                 )
# save 
save(bwnet, file = "../../rObjects/bwnet.Rdata")
save(counts, info, file = "../../WGCNA_allSampleInput.RData")
```
# load in data
```{r}
load("../../rObjects/bwnet.Rdata")
load("../../rObjects/WGCNA_allSampleInput.RData")
```

# Modules 
```{r modules}
# dataframe of sample IDs as rows and modeules as columns. The value is the correlation. 
module_eigengenes <- bwnet$MEs
head(module_eigengenes) # inspect 
ngenes_per_module <- table(bwnet$colors) # number of genes in each module 
write.table(ngenes_per_module, 
            "../../results/star/WGCNA/ngenes_per_module.txt", 
            sep = "\t", quote = FALSE, row.names = FALSE)

```
# Plot dendrogram
This function plots a hierarchical clustering dendrogram and color annotation(s) of objects in the dendrogram underneath.

guideHang - fraction of the dendrogram height to leave between the top end of the guide line and the dendrogram merge height. 
```{r plot_dendrogram}
plotDendroAndColors(bwnet$dendrograms[[1]], cbind(bwnet$unmergedColors, bwnet$colors), 
                    c("unmerged", "merged"), 
                    dendroLabels = FALSE, 
                    addGuide = TRUE, 
                    hang = 0.03,
                    guideHang = 0.05)
path <- paste0("../../results/star/WGCNA/module_dendrogram")
saveToPDF(paste0(path, ".pdf"), width = 8, height = 4)
```
# Association between modules and traits
Which modules are associated with our traits

binarizeCategoricalColumns - Given a data frame with (some) categorical columns, this function creates a set of indicator variables for the various possible sets of levels.

includePairwise - logical should pairwise binary indicators be included? For each pair of levels, the indicator is val1 for the lower level (earlier in levelOrder), val2 for the higher level and NA otherwise.
includeLevelVsAll - Logical: should binary indicators for each level be included? The indicator is val2 where x equals the level and val1 otherwise.
minCount - 	Levels of x for which there are fewer than minCount elements will be ignored.
```{r binary_traits}
controlvsdisease <- binarizeCategoricalColumns(dataTraits$TYPE, 
                           includePairwise = TRUE, 
                           includeLevelVsAll = TRUE, 
                           minCount = 1)
# remove data. from column names
names(controlvsdisease) <- gsub(x = names(controlvsdisease), 
                                pattern = "data.", replacement = "")  
traits <- cbind(numeric_traits, controlvsdisease) # column combine with dataTraits
# disease type isn't numeric, make numeric as this is required down stream
numeric_traits <- sapply(traits, as.numeric) # make sure all are numeric 
numeric_traits <- as.data.frame(numeric_traits) # save as data frame
rownames(numeric_traits) <- rownames(dataTraits) # add NPID as row names
```
# Correlation
```{r correlation}
nSamples <- nrow(counts) # number of samples
nGenes <- ncol(counts) # number of genes

# correlation between modules and traits 
module_trait_corr <- cor(module_eigengenes, numeric_traits, use = 'p')
# get the pval
# Calculates Student asymptotic p-value for given correlations.
module_trait_corr_pval <- corPvalueStudent(module_trait_corr, nSamples)
```
# Heatmap
```{r heatmap}
heatmap.data <- merge(module_eigengenes, numeric_traits, by = "row.names")
heatmap.data <- heatmap.data %>%
  column_to_rownames(var = 'Row.names')

# some data has missing values with na, such as lewy body count. 
# Replace na with zero 
heatmap.data["Cing.LB"][is.na(heatmap.data["Cing.LB"])] <- 0
heatmap.data["Cing.Synuclein"][is.na(heatmap.data["Cing.Synuclein"])] <- 0
plot_colors <- names(heatmap.data)[1:15]
plot_colors <- gsub("ME", "", plot_colors)

#par(mar= c(5, 5, 3, 3)) # bottom, left, top and right 
cor_disease_plot <- CorLevelPlot(heatmap.data, 
             x = c(names(heatmap.data)[43:44], names(heatmap.data)[47]), # which traits
             y = names(heatmap.data)[1:15], # gene modules 
             col = c("blue1", "skyblue", "white", "pink", "red"),
             rotLabX = 45, 
             colLabY = c(plot_colors), 
            # main = "Module-trait relationships", 
            # cexMain = 1, 
             fontLabX = 1,
             fontLabY = 1) # color scale
cor_disease_plot
path <- paste0("../../results/star/WGCNA/disease_vs_control_cor_module_heatmap")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 7)
```

```{r}
numeric_traits_selective <- numeric_traits[, c(1,2,13,14,16,17,27:29,32)] # Keep
module_trait_corr_selective <- module_trait_corr[, c(1,2,13,14,16,17,27:29,32)] # Keep
module_trait_corr_pval_selective <- module_trait_corr_pval[, c(1,2,13,14,16,17,27:29,32)] # Keep
#Print correlation heatmap between modules and traits
textMatrix= paste(signif(module_trait_corr_selective, 2), "\n(",
                        signif(module_trait_corr_pval_selective, 1), ")", sep= "")
dim(textMatrix)= dim(module_trait_corr_selective)
par(mar= c(8, 8.5, 3, 3)) # bottom, left, top and right 

labeledHeatmap(Matrix= module_trait_corr_selective,
            xLabels= names(numeric_traits_selective),
            yLabels= names(module_eigengenes),
            ySymbols= names(module_eigengenes),
            colorLabels= FALSE,
            colors= blueWhiteRed(50),
            textMatrix= textMatrix,
            setStdMargins= FALSE,
            cex.text= 1,
            zlim= c(-.4,.4),
            main= paste("Module-trait relationships"))
path <- paste0("../../results/star/WGCNA/module_trait_heatmap")
saveToPDF(paste0(path, ".pdf"), width = 10, height = 8)
```
# Genes in modules 
```{r genes_modules}
# data frame of module genes
module.gene.mapping <- as.data.frame(bwnet$colors)
module_colors <- c(unique(bwnet$colors))

for (i in module_colors){
  color <- module.gene.mapping %>%
    filter(bwnet$colors == i) %>%
    rownames()
  color <- as.data.frame(color)
  colnames(color)[1] ="gene_id"
  color_genes <- merge(color, genes, by = "gene_id")
  color_genes
  color_genes <- color_genes[, c(1:5,12)] # Keep
#  write.table(color_genes, paste0("../../results/star/WGCNA/", i, ".txt"), 
 #             row.names = FALSE, quote = FALSE, sep = "\t")
}
# save into one excel table
# first read in the files
for (i in module_colors) {
  filepath <- paste0("../../results/star/WGCNA/", i,".txt"
  )
  assign(paste0(i),
         read.delim(
           filepath,
           header = TRUE,
           sep = "\t",
           stringsAsFactors = FALSE
         ))
}
module_colors
list_of_datasets <- list("grey" = grey,
"blue" = blue,
"yellow" = yellow,
"turquoise" = turquoise,
"brown" = brown,
"red" = red,
"greenyellow" = greenyellow,
"magenta" = magenta,
"tan" = tan,
"green" = green,
"black" = black,
"pink" = pink,
"purple" = purple,
"salmon" = salmon,
"cyan" = cyan)
write.xlsx(list_of_datasets, file = paste0(
    "../../results/star/WGCNA/allSamples_genes_in_color_modules.xlsx"))

# plot
# Genes in modules
color_counts <- as.data.frame(bwnet$colors) # color modules 
color_counts$gene_id <- row.names(color_counts) # gene_ids as row names
colnames(color_counts)[1] ="hcolors"
color_counts$hcolors <- factor(color_counts$hcolors, levels = rev(plot_colors)) # arrange to match previous plot 
rev_plot_colors <- rev(plot_colors)
ggplot(color_counts, aes(x=hcolors)) + 
    geom_bar(fill = rev_plot_colors) +
  theme_classic() +
    ggtitle("Gene count in modules") +
    xlab("Module") +
    ylab("Gene count") +
  theme(axis.text.x=element_text(color=rev_plot_colors, angle = 65, vjust = .65)) +
  geom_text(stat='count', aes(label=..count..), vjust=-.25) 

path <- paste0("../../results/star/WGCNA/gene_counts_in_modules")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 5)
```

# Hubs
```{r}
library("limmaDE2")
graph<-wgcna2igraph(net = bwnet, datExpr = counts,
                    modules2plot = c("greenyellow"),
                    colors2plot = c("black"),
                    kME.threshold = 0.5, adjacency.threshold = 0.1,
                    adj.power = pow, verbose = T,
                    node.size = 0, frame.color = NA, node.color = NA,
                    edge.alpha = .5, edge.width =1)
plot(graph)
colorh = labels2colors(bwnet$colors)
hubs = chooseTopHubInEachModule(counts, colorh, omitColors = NA)
hubs
df_hubs <- as.data.frame(hubs)
colnames(df_hubs)[1] ="gene_id"
df_hubs$module <- row.names(df_hubs)
genes_selective_columns <- genes[,c(1,10,12)]
de <- merge(df_hubs, genes_selective_columns, by="gene_id")
de <- de[,c(2,3,4,1)]
write.table(de, "../../results/star/WGCNA/hub_gene_module.txt", sep = "\t", quote = FALSE, row.names = FALSE)
```

# DEGs in modules
```{r}
# read in DEGs 
LBDvsControl <- read.delim("../../results/star/DEGs/TYPE_LBDvsControl_gene_DEGs_FDRq0.05_logFC_0.25.txt")
ADvsControl <- read.delim("../../results/star/DEGs/TYPE_ADvsControl_gene_DEGs_FDRq0.05_logFC_0.25.txt")
# PAvsControl <- read.delim("../../results/star/DEGs/TYPE_PAvsControl_gene_DEGs_FDRq0.05_logFC_0.25.txt")
LBDvsAD <- read.delim("../../results/star/DEGs/TYPE_LBDvsAD_gene_DEGs_FDRq0.05_logFC_0.25.txt")

### AD vs Control
df_ADvsControl <- merge(color_counts, ADvsControl, by = "gene_id", all = TRUE)
df_ADvsControl$FDRq <- ifelse(df_ADvsControl$adj.P.Val < 0.05, "< 0.05", "No difference")
df_ADvsControl$FDRq <- df_ADvsControl$FDRq %>% replace_na('Not DE')
colnames(df_ADvsControl)[2] ="hcolors"

data <- df_ADvsControl %>% 
  group_by(FDRq,hcolors) %>%
  dplyr::count() %>% 
  ungroup(FDRq) %>% 
  mutate(perc = `n` / sum(`n`)) %>% 
  arrange(perc) %>%
  mutate(labels = scales::percent(perc))
data$hcolors <- factor(data$hcolors, levels = plot_colors) # arrange to match pervious plot 

ADvsControl_per_DEGs <- ggplot(data, aes(fill=FDRq, y=hcolors, x=perc)) + 
    geom_bar(position="fill", stat="identity") +
  theme_classic() +
  theme(axis.text.y=element_text(color=plot_colors)) +
  ggtitle("AD vs Control") +
    xlab("Percentage") +
    ylab("Gene modules") +
  scale_fill_manual("legend", values = c("< 0.05" = "salmon", "Not DE" = "grey45")) +
  theme(legend.position="none")
ADvsControl_per_DEGs
path <- paste0("../../results/star/WGCNA/percentage_ADvsControl_DEGs_gene_modules")
saveToPDF(paste0(path, ".pdf"), width = 4.5, height = 4)

#### LBD vs Control
df_LBDvsControl <- merge(color_counts, LBDvsControl, by = "gene_id", all = TRUE)
df_LBDvsControl$FDRq <- ifelse(df_LBDvsControl$adj.P.Val < 0.05, "< 0.05", "No difference")
df_LBDvsControl$FDRq <- df_LBDvsControl$FDRq %>% replace_na('Not DE')
colnames(df_LBDvsControl)[2] ="hcolors"

data <- df_LBDvsControl %>% 
  group_by(FDRq,hcolors) %>%
  dplyr::count() %>% 
  ungroup(FDRq) %>% 
  mutate(perc = `n` / sum(`n`)) %>% 
  arrange(perc) %>%
  mutate(labels = scales::percent(perc))
data$hcolors <- factor(data$hcolors, levels = plot_colors) # arrange to match pervious plot 

LBDvsControl_per_DEGs <- ggplot(data, aes(fill=FDRq, y=hcolors, x=perc)) + 
    geom_bar(position="fill", stat="identity") +
  theme_classic() +
  theme(axis.text.y=element_text(color=plot_colors)) +
  ggtitle("LBD vs Control ") +
    xlab("Percentage") +
    ylab("Gene modules") +
  scale_fill_manual("legend", values = c("< 0.05" = "salmon", "Not DE" = "grey45"))+
  theme(legend.position="none")
LBDvsControl_per_DEGs
path <- paste0("../../results/star/WGCNA/percentage_LBDvsControl_DEGs_gene_modules")
saveToPDF(paste0(path, ".pdf"), width = 4.5, height = 4)


#### LBD vs AD
df_LBDvsAD <- merge(color_counts, LBDvsAD, by = "gene_id", all = TRUE)
df_LBDvsAD$FDRq <- ifelse(df_LBDvsAD$adj.P.Val < 0.05, "< 0.05", "No difference")
df_LBDvsAD$FDRq <- df_LBDvsAD$FDRq %>% replace_na('Not DE')
colnames(df_LBDvsAD)[2] ="hcolors"

data <- df_LBDvsAD %>% 
  group_by(FDRq,hcolors) %>%
  dplyr::count() %>% 
  ungroup(FDRq) %>% 
  mutate(perc = `n` / sum(`n`)) %>% 
  arrange(perc) %>%
  mutate(labels = scales::percent(perc))
data$hcolors <- factor(data$hcolors, levels = plot_colors) # arrange to match pervious plot 

LBDvsAD_per_DEGs <- ggplot(data, aes(fill=FDRq, y=hcolors, x=perc)) + 
    geom_bar(position="fill", stat="identity") +
  theme_classic() +
  theme(axis.text.y=element_text(color=plot_colors)) +
  ggtitle("LBD vs AD ") +
    xlab("Percentage") +
    ylab("Gene modules") +
  scale_fill_manual("legend", values = c("< 0.05" = "salmon", "Not DE" = "grey45"))+
  theme(legend.position="none")
LBDvsAD_per_DEGs
path <- paste0("../../results/star/WGCNA/percentage_LBDvsAD_DEGs_gene_modules")
saveToPDF(paste0(path, ".pdf"), width = 4.5, height = 4)

ggarrange(
  ADvsControl_per_DEGs,
  LBDvsControl_per_DEGs,
  LBDvsAD_per_DEGs,
  ncol = 3
)
path <- paste0("../../results/star/WGCNA/ggarrange_percentage_DEGs_in_modules")
saveToPDF(paste0(path, ".pdf"), width = 12, height = 4)
```
# Hubs + DEGs
```{r}
hello <- as.data.frame(bwnet$colors)
atest <- subset(hello, rownames(hello) %in% LBDvsAD$gene_id)
colorh = labels2colors(atest)
hubs = chooseTopHubInEachModule(counts, colorh, omitColors = NA)
hubs
df_hubs <- as.data.frame(hubs)
colnames(df_hubs)[1] ="gene_id"
df_hubs$module <- row.names(df_hubs)
genes_selective_columns <- genes[,c(1,10,12)]
de <- merge(df_hubs, genes_selective_columns, by="gene_id")
de <- de[,c(2,3,4,1)]
write.table(de, "../../results/star/WGCNA/hub_LBDvsAD_DEGs_gene_module.txt", sep = "\t", quote = FALSE, row.names = FALSE)
```
# Membership intramodule connectivity
```{r}
module_membership_measure <- cor(module_eigengenes, counts, use = 'p')
module_membership_measure_pval <- corPvalueStudent(module_membership_measure, nSamples)

# high membership genes 
module_membership_measure_pval[1:10,1:10]

# calculate the gene significance and associated p-values
gene_sig_cor <- cor(counts, numeric_traits$LBD.vs.CONTROL, use = 'p')
nLBD_control <- subset(numeric_traits, TYPE == "4" | TYPE == "1")
nLBD_control <- nrow(nLBD_control) 
gene_sig_cor_pval <- corPvalueStudent(gene_sig_cor, nSamples = nLBD_control)

gene_sig_cor_pval %>%
  as.data.frame() %>%
  head(24)
```
