---
title: "Differential expression in LBD male XY samples"
author: "Kimberly Olney"
date: "03/17/2022"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
params:
  args: myarg
---

# Setup
```{r setup}
# Also do Session > Set Working Directory > Choose Directory
knitr::opts_knit$set(root.dir = ".")
```

```{r libraries, message=FALSE, warning=FALSE}
library(BiocParallel)  # SnowParam()
library(dplyr)  # left_join()
library(edgeR)  # DGEList()
library(limma)  # plotMDS()
library(ggrepel) # geom_text_repel()
library(ggplot2)  # ggplot()
library(gplots)  # heatmap.2()
library(grDevices)  # colorRampPalette()
library(rtracklayer)  # import()
library(stringr)  # str_match()
library(variancePartition)  # fitExtractVarPartModel()
library(reshape)  # melt()
library(Glimma)
library(plyr)
library(corrplot)
library(ggpubr)
```

# User defined variables
```{r set_variables}
sex_karyotype <- "XY"
control <- "CONTROL"
condition <- "LBD"
control_color <- "gray29"
condition_color <- "red"
myContrasts <- c("LBD - CONTROL")
tool = c("star")
pathToRef = c("/research/labs/neurology/fryer/projects/references/human/")
pathToRawData = c("/research/labs/neurology/fryer/projects/LBD_CWOW/")
```

# Read data
```{r read_data}
# read in metadata
metadata <- read.delim(paste0(pathToRawData, "metadata.tsv"), 
                       header = TRUE,
                       sep = "\t")

RIN <- read.delim(paste0(pathToRawData, "RIN.tsv"), 
                       header = TRUE,
                       sep = "\t")
RIN$NPID <- RIN$Sample
metadata <- merge(metadata, RIN, by = "NPID")
# read in counts data
counts <- read.delim(
  paste0("../../featureCounts/LBD_", sex_karyotype,".counts"),
  header = TRUE,
  sep = "\t", check.names=FALSE)
# get the sampleID from the counts file to match with metadata
sampleIDs <- colnames(counts)
samples <- as.data.frame(sampleIDs)
# split the sampleID by _ 
sample_count_id <- as.data.frame(str_split_fixed(samples$sampleID, "_", 2))
sample_count_id$counts_id <- samples$sampleIDs
# rename columns
names(sample_count_id)[names(sample_count_id) == 'V1'] <- 'NPID'
names(sample_count_id)[names(sample_count_id) == 'V2'] <- 'flow_lane'
# merge sample_count_id and metadata files
# this way metadata contains the sampleID to match the counts table 
counts_metadata <- merge(metadata, sample_count_id, by = "NPID")

# gene information
counts.geneid <- read.delim(
  "../../featureCounts/gene_info.txt",
  header = TRUE,
  sep = "\t"
)
counts.geneid <- as.data.frame(counts.geneid$Geneid)
colnames(counts.geneid) <- "gene_id"
# add gene_name as row names to counts file
rownames(counts) <- counts.geneid$gene_id

# read in annotation file
gtf.file <- paste0(pathToRef, "gencode.v38.annotation.gtf")
genes.gtf <- rtracklayer::import(gtf.file)
genes.gtf <- as.data.frame(genes.gtf)
genes.gtf <- genes.gtf[genes.gtf$type == "gene",]
table(genes.gtf$gene_type)

#counts.geneid
#names(counts.geneid)[1] <- 'gene_name'
joined.df <- join(counts.geneid, genes.gtf, type = "inner")

# check columns and rows match up between files
all.equal(rownames(counts), counts.geneid$gene_id)
all.equal(rownames(counts), genes.gtf$gene_id)

# reorder counts to be in the same order as metadata table
counts <- counts[, counts_metadata$counts_id]
all.equal(colnames(counts), (counts_metadata$counts_id))
```

# Create DGE object
```{r DGE_object}
# create object
dge <- DGEList(counts = counts,
               samples = counts_metadata,
               genes = genes.gtf)

table(dge$samples$TYPE)
# remove AD and PA controls
remove <- vector()
for (i in 1:nrow(dge$samples)) {
  if (dge$samples$TYPE[i] == "CONTROL - AD" | dge$samples$TYPE[i] == "CONTROL - PA") {
    remove <- c(remove, i)
  }
}
dge <- dge[,-remove, keep.lib.sizes = FALSE]
table(dge$samples$TYPE)
```
column 1 is pid id
column 15 is group information (LPS or control)
column 18 is date of birth
# Raw MDS with technical replicates
```{r MDS_techreps, warning=FALSE}
# set colors and get data
table(dge$samples$TYPE)
dge$samples$TYPE <- as.factor(dge$samples$TYPE)
group_colors <- c("black","red")[dge$samples$TYPE]
lcpm <- edgeR::cpm(dge$counts, log = TRUE)
par(bg = 'white')

# plot MDS
plotMDS(
  lcpm,
  top = 100, 
  labels = dge$samples$TYPE,
  cex = 1.5, 
  dim.plot = c(1,2), 
  plot = TRUE, 
  col = group_colors, 
  gene.selection = "common"
)
title(expression('Top 100 Genes - Raw (Log'[2]~'CPM)'))
#legend(
#  "bottomleft",
#  legend = c(control, condition),
#  pch = 16,
#  col = c("black","red"),
#  cex = 0.8
#)
path <- paste0("../../results/", tool, "/MDS/", condition,"_gene_MDS_dim1&2_techreps_", sex_karyotype)
saveToPDF(paste0(path, ".pdf"), width = 5, height = 5)
saveToPNG(paste0(path, ".png"), width = 4, height = 4, unit = "in", res = 300)

# plot MDS
plotMDS(
  lcpm,
  top = 100, 
  labels = dge$samples$NPID,
  cex = 1.5, 
  dim.plot = c(1,2), 
  plot = TRUE, 
  col = group_colors, 
  gene.selection = "common"
)
title(expression('Top 100 Genes - Raw (Log'[2]~'CPM)'))

path <- paste0("../../results/", tool, "/MDS/", condition,"_gene_MDS_dim2&3_techreps_", sex_karyotype)
saveToPDF(paste0(path, ".pdf"), width = 5, height = 5)
saveToPNG(paste0(path, ".png"), width = 4, height = 4, unit = "in", res = 300)
```

