---
title: "GOI"
author: "Kimberly Olney"
date: "3/21/2023"
output: html_document
---

# Setup
```{r setup}
knitr::opts_knit$set(root.dir = ".")
```

```{r libraries, message=FALSE, warning=FALSE}
```

# User defined variables
```{r set_variables}
source(here::here("scripts/R", "file_paths_and_colours.R"))
condition <- c("")
tool = c("star")
```
# read in voom counts matrix & dge object 
```{r}
dge.filtered.norm <- readRDS(paste0("../../rObjects/dge.filtered.norm.rds"))
voomCountsMatrix <- readRDS("/research/labs/neurology/fryer/m239830/LBD_CWOW/rObjects/TYPE.voomCountsMatrix.rds")
```

```{r}
# get the gene id that corresponds to the gene name 
genes <- dge.filtered.norm$genes # gene information 
goi <- subset(genes, gene_name == "ENO2" | gene_name == "MBP") # gene of interest 
# subset counts 
goi_counts <- subset(voomCountsMatrix, rownames(voomCountsMatrix) %in% rownames(goi))
goi_melt <- reshape2::melt(goi_counts) # reshape data 
names(goi_melt)[names(goi_melt) == "Var2"] <- "NPID" # rename columns to merge with metadata 
names(goi_melt)[names(goi_melt) == "Var1"] <- "gene_id"
df <- merge(goi_melt, metadata, by = "NPID")
```
# ENO2 & MBP expression
```{r}
ENO2 <- subset(df, gene_id == "ENSG00000111674.9")
MBP <- subset(df, gene_id == "ENSG00000197971.16")
df <- merge(ENO2, MBP, by = "NPID")
ggplot(df, aes(value.x, value.y, col = TYPE.x)) +
    geom_point() +
    theme_bw() +
    ggtitle("ENO2 and Brain weight expression") +
    xlab("Brain weight (kg)") +
    ylab("lcpm ENO2") +
    scale_fill_manual(values=TypeColors) 

cor(df$value.x, df$value.y, method = c("spearman"))
cor.test(df$value.x, df$value.y, method=c("spearman"))

# is the data normally distributed 
ggqqplot(df$value.x, ylab = "ENO2 expression")
ggqqplot(df$value.y, ylab = "MBP expression")
res <- cor.test(df$value.x, df$value.y, 
                method = "pearson")
res

ggscatter(df, x = "value.x", y = "value.y", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "spearman",
          ylab = "lcpm MBP", xlab = "lcpm ENO2") + 
  facet_grid(TYPE.x ~ sex_inferred.x) 
path <- paste0("../../results/clinical/TYPE.sex.ENO2_MBP")
saveToPDF(paste0(path, ".pdf"), width = 12, height = 10)
```


# Brain weight and ENO2 expression
```{r eval = FALSE}
ggplot(df, aes(Brain.wt, value, col = TYPE)) +
    geom_point() +
    theme_bw() +
    ggtitle("ENO2 and Brain weight expression") +
    xlab("Brain weight (kg)") +
    ylab("lcpm ENO2") +
    scale_fill_manual(values=TypeColors) 

cor(df$Brain.wt, df$value, method = c("spearman"))
cor.test(df$Brain.wt, df$value, method=c("spearman"))

# is the data normally distributed 
ggqqplot(df$Brain.wt, ylab = "Brain weight")
ggqqplot(df$value, ylab = "ENO2 expression")
res <- cor.test(df$Brain.wt, df$value, 
                method = "pearson")
res
ggscatter(df, x = "Brain.wt", y = "value", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "spearman",
          ylab = "lcpm ENO2", xlab = "Brain weight") 

path <- paste0("../../results/clinical/ENO2_brainweight")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 6)


ggscatter(df, x = "Brain.wt", y = "value", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "spearman",
          ylab = "lcpm ENO2", xlab = "Brain weight") + 
  facet_grid(. ~ sex_inferred)
path <- paste0("../../results/clinical/ENO2_brainweight_bySex")
saveToPDF(paste0(path, ".pdf"), width = 10, height = 6)
```
# Gene_id for gene of interest 
```{r}
column_variables <- c("ENSG00000111674.9") # gene_id to plot expression data for
colors <- TypeColors
```
# Function violin jitter of gene expression 
```{r}
# Disease type 
violin_plot_fun <- function(j) {
  df <- subset(df, gene_id == j)
  ggplot(df, aes(TYPE, value, fill = TYPE)) +
    geom_violin() +
    geom_boxplot(width = 0.1, outlier.shape=NA) + 
    geom_jitter(shape=16, position=position_jitter(0.2), alpha = 0.1) +
    theme_bw() +
    ggtitle("ENO2 expression") +
    xlab("TYPE") +
    ylab(j) +
    stat_compare_means() +
    scale_fill_manual(values=colors) +
    theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
}
violin_plots <-
  Map(violin_plot_fun, j = column_variables)
violin_plots

plotnames = imap(violin_plots, ~paste0("../../results/", tool, "/goi/ENO2.TYPE.pdf")) %>% #
  flatten()
plotnames
walk2(plotnames, violin_plots, ~ggsave(filename = .x, plot = .y, 
                                    height = 5.5, width = 8))


factor(df$LBD.type)
df$LBD.type <- factor(df$LBD.type, levels =c("no LBs", "TLBD", "DLBD", "BLBD"))

violin_plot_fun <- function(j) {
  df <- subset(df, gene_id == j)
  ggplot(df, aes(LBD.type, value, fill = LBD.type)) +
    geom_violin() +
    geom_boxplot(width = 0.1, outlier.shape=NA) + 
    geom_jitter(shape=16, position=position_jitter(0.2), alpha = 0.1) +
    theme_bw() +
    ggtitle("ENO2 expression") +
    xlab("LBD.type") +
    ylab(j) +
    stat_compare_means() +
    scale_fill_manual(values=colorbindColors) +
    theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
}
violin_plots <-
  Map(violin_plot_fun, j = column_variables)
violin_plots

plotnames = imap(violin_plots, ~paste0("../../results/", tool, "/goi/ENO2.LBD.type.pdf")) %>% #
  flatten()
plotnames
walk2(plotnames, violin_plots, ~ggsave(filename = .x, plot = .y, 
                                    height = 5.5, width = 8))


df <- subset(df, LBD.type == "TLBD" | LBD.type == "DLBD")
violin_plot_fun <- function(j) {
  df <- subset(df, gene_id == j)
  ggplot(df, aes(LBD.type, value, fill = LBD.type)) +
    geom_violin() +
    geom_boxplot(width = 0.1, outlier.shape=NA) + 
    geom_jitter(shape=16, position=position_jitter(0.2), alpha = 0.1) +
    theme_bw() +
    ggtitle("ENO2 expression") +
    xlab("LBD.type") +
    ylab(j) +
    stat_compare_means() +
    scale_fill_manual(values=colorbindColors) +
    theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
}
violin_plots <-
  Map(violin_plot_fun, j = column_variables)
violin_plots

plotnames = imap(violin_plots, ~paste0("../../results/", tool, "/goi/ENO2.TLBD_DLBD.pdf")) %>% #
  flatten()
plotnames
walk2(plotnames, violin_plots, ~ggsave(filename = .x, plot = .y, 
                                    height = 5.5, width = 6))


```
# GOI gene of interest histogram
```{r goi_hist}
voomCounts <- readRDS(paste0("../../rObjects/", condition, ".voomCountsMatrix.rds"))
voomCounts <- as.data.frame(voomCounts)
voomCounts$gene_id  <- rownames(voomCounts)
voom_genes <- merge(voomCounts, dge.filtered.norm$genes, by = "gene_id")
goi <- subset(voom_genes, gene_name == "TTR")
goi <- as.vector(goi[2:620])
goi_melt <- melt(goi)
summary(goi_melt$value)

# histogram
ggplot(goi_melt, aes(x = value)) + 
  geom_histogram(bins = 100, fill = "gray", color = "black") + 
  labs(title = "TTR", x=NULL, y=NULL) +
  xlab("TTR log2(CPM)") + ylab("# of samples") +
  theme_bw() 

path <- paste0("../../results/", tool, "/goi/TTR.voomcounts.histogram")
saveToPDF(paste0(path, ".pdf"), width = 8, height = 5)
```

# By TYPE, GOI gene of interest histogram 
```{r type_goi_hist}
lcpm <- edgeR::cpm(dge.filtered.norm$counts, log = TRUE)
cpm <- edgeR::cpm(dge.filtered.norm$counts, log = FALSE)

# using CPM data
cpm <- as.data.frame(cpm)
cpm$gene_id  <- rownames(cpm)
cpm_genes <- merge(cpm, dge.filtered.norm$genes, by = "gene_id")
goi <- subset(cpm_genes, gene_name == "TTR")
goi <- as.vector(goi[1:620]) # get only sample counts data. Columns 2-620 (i.e. 619 samples)
goi_melt <- melt(goi) # transform 
summary(goi_melt$value) 
colnames(goi_melt)[2] <- "NPID"
goi_melt_sample_info <-
  merge(goi_melt, dge.filtered.norm$samples, by = "NPID")
df <-
  with(goi_melt_sample_info, goi_melt_sample_info[order(TYPE, NPID), ])
df$NPID <- factor(df$NPID, levels = df$NPID)
# column plot
ggplot(df, aes(x = NPID, y = value, fill = TYPE)) +
  geom_col() +
  labs(title = "TTR", x = NULL, y = NULL) +
  xlab("TYPE") + ylab("CPM") +
  theme_bw() +
  scale_fill_manual(values = TypeColors)
path <- paste0("../../results/", tool, "/goi/TTR.CPM.bar")
saveToPDF(paste0(path, ".pdf"), width = 8, height = 5)

# using log2CPM data
lcpm <- as.data.frame(lcpm)
lcpm$gene_id  <- rownames(lcpm)
lcpm_genes <- merge(lcpm, dge.filtered.norm$genes, by = "gene_id")
goi <- subset(lcpm_genes, gene_name == "TTR")
goi <- as.vector(goi[1:620])
goi_melt <- melt(goi)
summary(goi_melt$value)
colnames(goi_melt)[2] <- "NPID"
goi_melt_sample_info <-
  merge(goi_melt, dge.filtered.norm$samples, by = "NPID")
df <-
  with(goi_melt_sample_info, goi_melt_sample_info[order(TYPE, NPID), ])
df$NPID <- factor(df$NPID, levels = df$NPID)
ggplot(df, aes(x = NPID, y = value, fill = TYPE)) +
  geom_col() +
  labs(title = "TTR", x = NULL, y = NULL) +
  xlab("TYPE") + ylab("log2(CPM)") +
  theme_bw() +
  scale_fill_manual(values = TypeColors)
path <- paste0("../../results/", tool, "/goi/TTR.log2CPM.bar")
saveToPDF(paste0(path, ".pdf"), width = 8, height = 5)
```



#-----------------


# read in genotype information
```{r}
# from Xu and Wolf at Mayo Clinic
GWAS <- read.delim("../../GWAS_phospho-ubiquitin/APOE_ZMIZ1_genotype.txt")
# rename npid to NPID
colnames(GWAS)[1] ="NPID"
# merge with metadata by NPID
info_GWAS <- merge(info, GWAS, by = "NPID")
```

```{r}
# get the gene id that corresponds to the gene name 
genes <- dge.filtered.norm$genes

bloodmarker_gene_info <- subset(genes, 
                                gene_name == "APOE" | 
                                gene_name == "ZMIZ1" |
                                gene_name == "GFAP" | 
                                gene_name == "MAP2" )
rownames(bloodmarker_gene_info)
bloodmarker_gene_id <- rownames(bloodmarker_gene_info)
# subset counts 
HBB_counts <- subset(voomCountsMatrix, rownames(voomCountsMatrix) %in% bloodmarker_gene_id)
HBB_melt <- reshape2::melt(HBB_counts)
names(HBB_melt)[names(HBB_melt) == "Var2"] <- "NPID"
names(HBB_melt)[names(HBB_melt) == "Var1"] <- "gene_id"
gene_id_gene_name <- data.frame(rownames(bloodmarker_gene_info), bloodmarker_gene_info$gene_name)
names(gene_id_gene_name)[names(gene_id_gene_name) == "rownames.bloodmarker_gene_info."] <- "gene_id"
names(gene_id_gene_name)[names(gene_id_gene_name) == "bloodmarker_gene_info.gene_name"] <- "gene_name"
df <- merge(gene_id_gene_name, HBB_melt, by = "gene_id")
df <- merge(df, info, by = "NPID")
```
plot APOE 
```{r}
column_variables <- c("APOE")
df$APOE <- as.factor(df$APOE)
my_comparisons <-
  list(
    c("E2E3", "E2E4"),
    c("E2E3", "E3E3"),
    c("E2E3", "E3E4"),
    c("E2E3", "E4E4"),
    c("E2E4", "E3E3"),
    c("E2E4", "E3E4"),
    c("E2E4", "E4E4"),
    c("E3E3", "E3E4"),
    c("E3E3", "E4E4"),
    c("E3E4", "E4E4")
  )
colors <- brewer.pal(n = 5, name = "RdBu")
violin_plot_fun <- function(j) {
  df <- subset(df, gene_name == j)
  ggplot(df, aes(APOE, value, fill = APOE)) +
    geom_violin() +
    geom_boxplot(width = 0.1, outlier.shape=NA) + 
    geom_jitter(shape=16, position=position_jitter(0.2)) +
    theme_bw() + 
    ggtitle(j) +
    xlab("APOE genotype") +
    ylab(j) +
    stat_compare_means(method = "anova") +
    scale_fill_manual(values=colors) +
    theme(legend.position = "none")
}
violin_plots <-
  Map(violin_plot_fun, j = column_variables)
violin_plots

plotnames = imap(violin_plots, ~paste0("../../results/", tool, "/goi/genotype.", .y, ".pdf")) %>%
  flatten()
plotnames
walk2(plotnames, violin_plots, ~ggsave(filename = .x, plot = .y, 
                                    height = 5.5, width = 6.5))
```


plot APOE 
```{r}
column_variables <- c("APOE")
df$APOE <- as.factor(df$APOE)
my_comparisons <-
  list(
    c("E2E3", "E2E4"),
    c("E2E3", "E3E3"),
    c("E2E3", "E3E4"),
    c("E2E3", "E4E4"),
    c("E2E4", "E3E3"),
    c("E2E4", "E3E4"),
    c("E2E4", "E4E4"),
    c("E3E3", "E3E4"),
    c("E3E3", "E4E4"),
    c("E3E4", "E4E4")
  )
colors <- brewer.pal(n = 5, name = "RdBu")
violin_plot_fun <- function(j) {
  df <- subset(df, gene_name == j)
  ggplot(df, aes(APOE, value, fill = APOE)) +
    geom_violin() +
    geom_boxplot(width = 0.1, outlier.shape=NA) + 
    geom_jitter(shape=16, position=position_jitter(0.2)) +
    theme_bw() + 
    ggtitle(j) +
    xlab("APOE genotype") +
    ylab(j) +
    stat_compare_means(method = "anova") +
    scale_fill_manual(values=colors) +
    theme(legend.position = "none")
}
violin_plots <-
  Map(violin_plot_fun, j = column_variables)
violin_plots

plotnames = imap(violin_plots, ~paste0("../../results/", tool, "/goi/genotype.", .y, ".pdf")) %>%
  flatten()
plotnames
walk2(plotnames, violin_plots, ~ggsave(filename = .x, plot = .y, 
                                    height = 5.5, width = 6.5))
```

```{r}
remove(df)
df <- merge(gene_id_gene_name, HBB_melt, by = "gene_id")
# merge by info_GWAS
df <- merge(df, info_GWAS, by = "NPID")
```
plot
```{r}
column_variables <- c("ZMIZ1")
df$rs6480922_cnt <- as.factor(df$rs6480922_cnt)
my_comparisons <-
  list(
    c("0", "1"),
    c("0", "2"),
    c("1", "2")
  )
colors <- brewer.pal(n = 5, name = "RdBu")
violin_plot_fun <- function(j) {
  df <- subset(df, gene_name == j)
  ggplot(df, aes(rs6480922_cnt, value, fill = rs6480922_cnt)) +
    geom_violin() +
    geom_boxplot(width = 0.1, outlier.shape=NA) + 
    geom_jitter(shape=16, position=position_jitter(0.2)) +
    theme_bw() + 
    ggtitle(j) +
    xlab("number of T alleles") +
    ylab(j) +
    stat_compare_means(comparisons = my_comparisons, method="wilcox.test") +
    scale_fill_manual(values=colors) +
    theme(legend.position = "none")
}
violin_plots <-
  Map(violin_plot_fun, j = column_variables)
violin_plots

plotnames = imap(violin_plots, ~paste0("../../results/", tool, "/goi/genotype.", .y, ".pdf")) %>%
  flatten()
plotnames
walk2(plotnames, violin_plots, ~ggsave(filename = .x, plot = .y, 
                                    height = 5.5, width = 6.5))
```

```{r}
df_compare_means <- subset(df, df$gene_name == "ZMIZ1")

shapiro.test(df_compare_means$value)

# Compute the analysis of variance
res.aov <- aov(value ~ rs6480922_cnt, data = df_compare_means)
# Summary of the analysis
summary(res.aov)

kruskal.test(value ~ rs6480922_cnt, data = df_compare_means)

pairwise.wilcox.test(df_compare_means$value, df_compare_means$rs6480922_cnt,
                 p.adjust.method = "BH")


write.table(df_compare_means, "ZMIZ1_genotype_and_log2CPM_expression.txt", row.names = F, quote = F, sep = "\t")
```
