---
title: "Create rObject of RNAseq counts data - LBD sub types"
author: "Kimberly Olney, Ph.D."
date: "08/23/2023"
output:
  html_document:
    df_print: paged
  pdf_document: default
params:
  args: myarg
---
# Setup
```{r setup}
knitr::opts_knit$set(root.dir = ".")
```
# User defined variables
```{r set_variables}
source(here::here("scripts/R", "file_paths_and_colours.R"))
#source(here::here("scripts/R", "gtf_path.R"))
condition <- c("LBD.type.") # change condition for file naming 
tool = c("star_LBD_sub_type") # change condition for folder naming 
typeOfCount <-  c(".bamReadsPerGene.out.tab")
```
# Read in DGE object
```{r}
dge <- readRDS(paste0("../../rObjects/dge.filtered.norm.rds"))
```
# MDS - normalized counts
multidimensional scaling
```{r MDS_norm}
lcpm <- edgeR::cpm(dge$counts, log = TRUE)
cpm <- edgeR::cpm(dge$counts, log = FALSE)

# LBD type colors 
dge$samples$LBD.type <- as.factor(dge$samples$LBD.type)
colors <- c(colorbindColors)[dge$samples$LBD.type]

par(bg = 'white')
plotMDS(
  lcpm,
  top = 100, 
  labels = dge$samples$LBD.type,
  cex = .8, 
  dim.plot = c(1,2), 
  plot = TRUE, 
  col = colors, 
  gene.selection = "common"
)
title(expression('Top 100 Genes (Log'[2]~'CPM)'))

path <- paste0("../../results/", tool, "/MDS/", 
               condition,"_dim1&2")
saveToPDF(paste0(path, ".pdf"), width = 5.2, height = 5.2)
```
# DGE object & metadata
```{r dge}
# impute missing values with median
dge$samples$RIN <- impute(dge$samples$RIN, median)
# one sample is missing VaD information
dge$samples$VaD <- impute(dge$samples$VaD, median)
dge$samples$flowcell_and_lane <- factor(dge$samples$flowcell_and_lane)
dge$samples$APOE <- factor(dge$samples$APOE)

info <- as.data.frame(dge$samples) # sample metadata
genes <- dge$genes # gene information 
```

# Add biomarker expression to dge metadata 
```{r biomarker_expression}
biomarkers <- c("ENO2", "GFAP", "OLIG2", "CD34", "P2RY12", "MBP")

for (i in biomarkers) {
  biomarker <- subset(genes, gene_name == i) # gene of interest 
  biomarker_counts <- subset(lcpm, rownames(lcpm) %in% biomarker)
  biomarker_melt <- reshape2::melt(biomarker_counts) # reshape data 
  # rename columns to merge with metadata 
  names(biomarker_melt)[names(biomarker_melt) == "value"] <- i 
  # rename columns to merge with metadata 
  names(biomarker_melt)[names(biomarker_melt) == "Var2"] <- "NPID" 
  names(biomarker_melt)[names(biomarker_melt) == "Var1"] <- "gene_id"
  biomarker_melt$gene_id <- NULL
  assign(paste0(i),biomarker_melt)
}

# add gene expression values into one dataframe 
# put all data frames into list
df_list <- list(ENO2, OLIG2, CD34, P2RY12, GFAP, MBP)

# merge all data frames in list
cell_biomarker_lcpm <- df_list %>% reduce(full_join, by='NPID')
remove(biomarker, biomarker_counts, biomarker_melt, biomarkers,
       ENO2, OLIG2, CD34, P2RY12, GFAP, MBP)
```

```{r}
plot(cell_biomarker_lcpm$ENO2, cell_biomarker_lcpm$MBP)

sp <- ggscatter(cell_biomarker_lcpm, x = "ENO2", y = "MBP",
                add = "reg.line",  # Add regression line
                add.params = list(color = "blue", fill = "lightgray"), 
                conf.int = TRUE)
# Add correlation coefficient
sp + stat_cor(method = "pearson", label.x = 7, label.y = 10)
```
# Scale data
Rescaling a predictor in a regression has absolutely no effect on the magnitude of the relation being studied. The slope itself will not change its steepness, nor will the p-values or variance explained be changed. Rescaling is merely a means of communicating the nature of the regression line in different, hopefully more intuitive language.
```{r scale}
df <- merge(cell_biomarker_lcpm, info, by = "NPID")
scaled.info <-
  df[c(
    "Race_numeric",
    "RIN",
    "Age",
    "PCT_CODING_BASES",
    "PCT_INTERGENIC_BASES",
    "PCT_INTRONIC_BASES",
    "APOE_E4_allele_count", 
    "ENO2", 
    "GFAP", 
    "OLIG2", 
    "CD34", 
    "P2RY12",
  #  "MBP",
    "Brain.wt"
  )] %>% scale()
scaled.info.df <- as.data.frame(scaled.info)
# remove columns with unscaled data 
df <- (subset(df, select = -c(Race_numeric, RIN, Age, PCT_CODING_BASES, PCT_INTRONIC_BASES, APOE_E4_allele_count, ENO2, GFAP, OLIG2, CD34, P2RY12, Brain.wt)))
# Add scaled information to the metadata called "info"
info_with_scale <- cbind(df, scaled.info.df)

remove(cell_biomarker_lcpm, cpm, df, df_list, info, scaled.info.df)
```

```{r}
# check that NPIDs match between dataframes
all.equal(dge$samples$NPID, as.character(info_with_scale$NPID))
# replace sample information with the updated info that includes biomakrer expression
dge$samples <- info_with_scale
```
# LBD type refined
```{r}
condition <- c("LBD.type.Braak") # change condition for file naming 

dge$samples$Braak.NFT.binary <-  ifelse(dge$samples$Braak.NFT > 3, 1, 0) 
dge$samples$LBD.type_Braak <- paste0(dge$samples$LBD.type, "_", dge$samples$Braak.NFT.binary)
dge$samples$LBD.type_Braak <- factor(dge$samples$LBD.type_Braak)

# BLBD_0 DLBD_0 DLBD_1 no LBs_0 no LBs_1 TLBD_0 TLBD_1
dge$samples$group <- ifelse(
  dge$samples$LBD.type_Braak == "BLBD_0",
  "BLBD",
  ifelse(
    dge$samples$LBD.type_Braak == "no LBs_0",
    "no_LBs",
    ifelse(
      dge$samples$LBD.type_Braak == "no LBs_1",
      "no_LBs",
      ifelse(
        dge$samples$LBD.type_Braak == "TLBD_0",
        "TLBD",
        ifelse(
          dge$samples$LBD.type_Braak == "TLBD_1",
          "TLBD",
          ifelse(dge$samples$LBD.type_Braak == "DLBD_0",
                 "DLBD_0", "DLBD_1")
        )
      )
    )
  )
)

dge$samples$group <- factor(dge$samples$group)
df <- as.data.frame(dge$samples$group)
ggplot(df, aes(x = df$`dge$samples$group`)) +
  geom_bar()
```

# Voom transform counts
Transform count data to log2-counts per million (logCPM), estimate the mean-variance relationship and use this to compute appropriate observation-level weights. The data are then ready for linear mixed modelling with dream(). This method is the same as limma::voom(), except that it allows random effects in the formula. 
```{r voom, warning=FALSE}
formula <- (~ 0 + LBD.type)
voom_with_weights <-
  variancePartition::voomWithDreamWeights(
    counts = dge$counts,
    formula = formula,
    data = dge$samples,
    BPPARAM = BiocParallel::SnowParam(cores),
    plot = TRUE
  )
path <- paste0("../../results/", tool, "/voom/", condition, ".raw.voom")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 4)
voomCounts <- voom_with_weights$E
```
# Design matrix
```{r design}
design <-
  model.matrix(~ 0 + 
      group + 
      sex_inferred + 
      flowcell_and_lane + 
      RIN +
      PCT_CODING_BASES +
      PCT_INTERGENIC_BASES + 
      PCT_INTRONIC_BASES +
      ENO2,
    dge$samples
  )
colnames(design) <-
  c("BLBD",
    "DLBD",
    "DLBD_Braak",
    "no_LBs",
    "TLBD", 
    "sex",
    "Batch1",
    "Batch2",
    "Batch3",
    "Batch4",
    "Batch5",
    "Batch6",
    "Batch7",
    "RIN",
    "PCT_CODING_BASES", 
    "PCT_INTERGENIC_BASES", 
    "PCT_INTRONIC_BASES",
    "ENO2"
  )
```
# Voom
When the library sizes are quite variable between samples, then the voom approach is theoretically more powerful than limma-trend. 
The voom method estimates the mean-variance relationship of the log-counts.Generates a precision weight for each observation and enters these into the limma empirical Bayes analysis pipeline.
```{r voom_BIC}
form <- (
  ~ 0 +
      group + 
      sex_inferred + 
      flowcell_and_lane + 
      RIN +
      PCT_CODING_BASES +
      PCT_INTERGENIC_BASES + 
      PCT_INTRONIC_BASES +
      ENO2 
)

voom_cov <-
  variancePartition::voomWithDreamWeights(
    counts = dge$counts,
    formula = form,
    data = dge$samples,
    BPPARAM = BiocParallel::SnowParam(cores),
    plot = TRUE
  )
path <-
  paste0("../../results/",
         tool,
         "/voom/",
         condition,
         ".voom.finalmodel")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 4)
voomCounts <- voom_cov$E
```
# Contrast 
Overview of steps:/
1.	lmFit computes coefficients, residual variances and standard errors./
2.	contrasts.fit converts the coefficients and standard errors to reflect the contrasts rather than the original design matrix, but does not compute t-statistics or p-values./
3.	eBayes computes t-statistics and p-values from the coefficients and standard errors./
https://support.bioconductor.org/p/117840/ 

```{r contrasts}
# fits linear model for each gene given a series of arrays
fit <- lmFit(voom_cov, design)
coef.fit <- fit$coefficients

contrasts <- makeContrasts(
  DLBD_BraakvsDLBD = DLBD_Braak - DLBD,
  levels = colnames(design))
head(contrasts)

# save contrast names
allComparisons <- colnames(contrasts)
allComparisons # check

# run contrast analysis
vfit <- contrasts.fit(fit, contrasts = contrasts)
```
# Empirical Bayes
Compute differential expression based on the empirical Bayes moderation of the standard errors towards a common value./
trend - 	logical, should an intensity-dependent trend be allowed for the prior variance? If FALSE then the prior variance is constant. Alternatively, trend can be a row-wise numeric vector, which will be used as the covariate for the prior variance./
robsut - logical, should the estimation of df.prior and var.prior be robustified against outlier sample variances./
```{r ebays}
veBayesFit <- eBayes(vfit, trend = TRUE, robust = TRUE)

plotSA(veBayesFit, main = "Final Model: Mean-variance Trend")
path <-
  paste0("../../results/",
         tool,
         "/voom/",
         condition,
         ".voom.eBayes.finalmodel")
saveToPDF(paste0(path, ".pdf"), width = 9, height = 5)
```

# DEGs summary
```{r DGE_summary}
pval <- 0.05
lfc.cutoff <- 0.25

sumTable <- 
  summary(decideTests(
    veBayesFit,  # object
    adjust.method = "BH", # by default the method = "separate"
    p.value = pval,
    lfc = lfc.cutoff  # numeric, minimum absolute log2-fold change required
  ))

print(paste0(" FDRq < ", pval,
             " & absolute log2-fold change > ", lfc.cutoff))
sumTable
write.table(sumTable, 
            paste0("../../results/", tool, "/DEGs/", condition, ".DEGs.summary.txt"), 
            quote = FALSE, sep = "\t")
```
# Add gene information to DEGs
reformat genes table to only include relevant information
```{r}
genes_relevant <- dplyr::select(genes, 1:4,10:12)
```
Check 
```{r DGE_check, eval=FALSE}
DEGs <- topTable(
  veBayesFit, 
  coef = "DLBD_BraakvsDLBD",  
  n = Inf, 
  p.value = 1,
  lfc = 0, 
  sort.by = "P", 
  genelist = genes_relevant, 
  confint = TRUE # column of confidence interval 
    )
```
# Save objects
```{r save_voom}
#saveRDS(veBayesFit, file = paste0("../../rObjects/", condition, ".veBayesFit.rds"))
#saveRDS(voomCounts, file = paste0("../../rObjects/", condition, ".voomCountsMatrix.rds"))
```
# Output DEG tables
```{r DGE_output}
coef <- 1

for (i in allComparisons) {
  vTopTableAll <- topTable(
    veBayesFit, 
    coef = coef,  
    n = Inf, 
    p.value = 1,
    lfc = 0, 
    sort.by = "P", 
    genelist = genes_relevant, 
    confint = TRUE # column of confidence interval 
    )
    saveRDS(vTopTableAll, file = 
            paste0("../../rObjects/gene_tables/", condition, "_", 
                   i,"_gene_table.rds"))
  path <- paste0("../../results/", tool, "/DEGs/", condition, "_", 
  i, "_gene_DEGs_FDRq1.00.txt", sep = "") 
  write.table(
    vTopTableAll,
    path,
    sep = "\t",
    row.names = FALSE,
    quote = FALSE
  )
  # p < 0.05, log2fc > 0
  vTopTable1 <-
    topTable( 
      veBayesFit,  
      coef = coef,  
      n = Inf, 
      p.value = pval,
      lfc = lfc.cutoff,
      genelist = genes_relevant, 
      confint = TRUE # column of confidence interval 
    )
  path <- paste0("../../results/", tool, "/DEGs/", condition, "_", 
  i, "_gene_DEGs_FDRq0.05_logFC_0.25.txt", sep = "") 
  write.table(
    vTopTable1,
    path,
    sep = "\t",
    row.names = FALSE,
    quote = FALSE
  )
  # increment 
  coef <- coef + 1
}
remove(coef, coef.fit, contrasts, design, fit, scaled.info, veBayesFit, vfit,
       voom_cov, voom_with_weights, voomCounts, vTopTable1, vTopTableAll)
```
# Volcano plots
```{r volcano, warning=FALSE}
for (i in allComparisons) {
  group1_vs_group2 <-
    read.delim(
      paste0(
        "../../results/",
        tool,
        "/DEGs/",
        condition,
        "_",
        i,
        "_gene_DEGs_FDRq1.00.txt"
      )
    )
  color_values <- vector()
  max <- nrow(group1_vs_group2)
  for (row in 1:max) {
    if (group1_vs_group2$adj.P.Val[row] < 0.05) {
      if (group1_vs_group2$logFC [row] > 0.25) {
        color_values <- c(color_values, 1)
      }
      else if (group1_vs_group2$logFC[row] < -.25) {
        color_values <- c(color_values, 2)
      }
      else {
        color_values <- c(color_values, 3)
      }
    }
    else{
      color_values <- c(color_values, 3)
    }
  }
  group1_vs_group2$color_adjpval_0.05 <- factor(color_values)
  data <- group1_vs_group2
  # plot only if there are DEGs with p_val_adj < 0.05
  num <- subset(data, (adj.P.Val < 0.05 & logFC < -.25)  | (adj.P.Val < 0.05 & logFC > .25 ))
  num <- nrow(num)
  if (num != 0) {
    up <- data[data$color_adjpval_0.05 == 1,]
    up10 <- up[1:10,]
    upFold <- subset(up, logFC > 0.5)
    upFold <- upFold[!(upFold$gene_name %in% up10$gene_name),]
    down <- data[data$color_adjpval_0.05 == 2,]
    down10 <- down[1:10,]
    downFold <- subset(down, logFC < -0.5)
    downFold <- downFold[!(downFold$gene_name %in% down10$gene_name),]
    if (!1 %in% unique(data$color_adjpval_0.05)) {
      my_colors <- c("blue", "gray")
    } else if (!2 %in% unique(data$color_adjpval_0.05)) {
      my_colors <- c("red", "gray")
    } else if (!1 %in% unique(data$color_adjpval_0.05) &&
               !2 %in% unique(data$color_adjpval_0.05)) {
      my_colors <- c("gray")
    } else {
      my_colors <- c("red", "blue", "gray")
    }
    hadjpval <- (-log10(max(data$P.Value[data$adj.P.Val < 0.05],
                            na.rm = TRUE)))
    negFC <- c(-.25)
    posFC <- c(.25) 
    i <- gsub("vs", " vs ", i)
    p <-
      ggplot(data = data,
             aes(
               x = logFC,
               y = -log10(P.Value),
               color = color_adjpval_0.05
             )) +
      geom_point(alpha = 0.8, size = 1) +
      theme_bw() +
      theme(legend.position = "none") +
      scale_color_manual(values = my_colors) +
      labs(
        title = "",
        x = expression(log[2](FC)),
        y = expression(-log[10] ~ "(" ~ italic("p") ~ "-value)")
      )  +
      geom_hline(yintercept = hadjpval,
                 #  horizontal line
                 colour = "#000000",
                 linetype = "dashed") +
      geom_vline(xintercept = negFC,
                 #  horizontal line
                 colour = "#000000",
                 linetype = "dashed") +
      geom_vline(xintercept = posFC,
                 #  horizontal line
                 colour = "#000000",
                 linetype = "dashed") +
      ggtitle(paste0(i)) +
      geom_text_repel(
        data = up10,
        aes(
          x = logFC,
          y = -log10(P.Value),
          label = gene_name
        ),
        color = "maroon",
        fontface = "italic",
        size = 3,
        max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
      ) +
      geom_text_repel(
        data = upFold,
        aes(
          x = logFC,
          y = -log10(P.Value),
          label = gene_name
        ),
        color = "maroon",
        fontface = "italic",
        size = 3,
        max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
      ) +
      geom_text_repel(
        data = down10,
        aes(
          x = logFC,
          y = -log10(P.Value),
          label = gene_name
        ),
        color = "navyblue",
        fontface = "italic",
        size = 3,
        max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
      ) +
      geom_text_repel(
        data = downFold,
        aes(
          x = logFC,
          y = -log10(P.Value),
          label = gene_name
        ),
        color = "navyblue",
        fontface = "italic",
        size = 3,
        max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
      )
    p
    i <- gsub(" vs ", "vs", i)
    # save
    path <-
      paste0(
        "../../results/",
        tool,
        "/volcano/",
        condition,
        "_",
        i,
        "_gene_volcano_FDRq0.05"
      )
    pdf(paste0(path, ".pdf"),
        height = 5,
        width = 5)
    print(p)
    dev.off()
    
    print(paste("i =", i))
  }
} 
```

```{r cleanup}
# clean up
remove(up, up10, upFold, group1_vs_group2, down, downFold, down10, data, p)
```
