<h1 id="setup">Setup</h1>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Also do Session &gt; Set Working Directory &gt; Choose Directory</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span>opts_knit<span class="sc">$</span><span class="fu">set</span>(<span class="at">root.dir =</span> <span class="st">&quot;.&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BiocParallel)  <span class="co"># SnowParam()</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)  <span class="co"># left_join()</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(edgeR)  <span class="co"># DGEList()</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(limma)  <span class="co"># plotMDS()</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel) <span class="co"># geom_text_repel()</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)  <span class="co"># ggplot()</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gplots)  <span class="co"># heatmap.2()</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(grDevices)  <span class="co"># colorRampPalette()</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(philentropy)  <span class="co"># JSD()</span></span></code></pre></div>
<pre><code>## Error in library(philentropy): there is no package called &#39;philentropy&#39;</code></pre>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rtracklayer)  <span class="co"># import()</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)  <span class="co"># str_match()</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(variancePartition)  <span class="co"># fitExtractVarPartModel()</span></span></code></pre></div>
<pre><code>## Error in library(variancePartition): there is no package called &#39;variancePartition&#39;</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reshape)  <span class="co"># melt()</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Glimma)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(plyr)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(corrplot)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggpubr)</span></code></pre></div>
<h1 id="user-defined-variables">User defined variables</h1>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>sex_karyotype <span class="ot">&lt;-</span> <span class="st">&quot;XX&quot;</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>control <span class="ot">&lt;-</span> <span class="st">&quot;CONTROL&quot;</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>condition <span class="ot">&lt;-</span> <span class="st">&quot;LBD&quot;</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>control_color <span class="ot">&lt;-</span> <span class="st">&quot;gray29&quot;</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>condition_color <span class="ot">&lt;-</span> <span class="st">&quot;red&quot;</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>myContrasts <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;LBD - CONTROL&quot;</span>)</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>tool <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;star&quot;</span>)</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>pathToRef <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;/research/labs/neurology/fryer/projects/references/human/&quot;</span>)</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>pathToRawData <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;/research/labs/neurology/fryer/projects/LBD_CWOW/&quot;</span>)</span></code></pre></div>
<h1 id="read-data">Read data</h1>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># read in metadata</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="fu">paste0</span>(pathToRawData, <span class="st">&quot;metadata.tsv&quot;</span>), </span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">header =</span> <span class="cn">TRUE</span>,</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>RIN <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="fu">paste0</span>(pathToRawData, <span class="st">&quot;RIN.tsv&quot;</span>), </span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>                       <span class="at">header =</span> <span class="cn">TRUE</span>,</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>                       <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>RIN<span class="sc">$</span>NPID <span class="ot">&lt;-</span> RIN<span class="sc">$</span>Sample</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> <span class="fu">merge</span>(metadata, RIN, <span class="at">by =</span> <span class="st">&quot;NPID&quot;</span>)</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a><span class="co"># read in counts data</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste0</span>(<span class="st">&quot;../../featureCounts/LBD_&quot;</span>, sex_karyotype,<span class="st">&quot;.counts&quot;</span>),</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">header =</span> <span class="cn">TRUE</span>,</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">check.names=</span><span class="cn">FALSE</span>)</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a><span class="co"># get the sampleID from the counts file to match with metadata</span></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>sampleIDs <span class="ot">&lt;-</span> <span class="fu">colnames</span>(counts)</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(sampleIDs)</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a><span class="co"># split the sampleID by _ </span></span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>sample_count_id <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">str_split_fixed</span>(samples<span class="sc">$</span>sampleID, <span class="st">&quot;_&quot;</span>, <span class="dv">2</span>))</span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>sample_count_id<span class="sc">$</span>counts_id <span class="ot">&lt;-</span> samples<span class="sc">$</span>sampleIDs</span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a><span class="co"># rename columns</span></span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sample_count_id)[<span class="fu">names</span>(sample_count_id) <span class="sc">==</span> <span class="st">&#39;V1&#39;</span>] <span class="ot">&lt;-</span> <span class="st">&#39;NPID&#39;</span></span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sample_count_id)[<span class="fu">names</span>(sample_count_id) <span class="sc">==</span> <span class="st">&#39;V2&#39;</span>] <span class="ot">&lt;-</span> <span class="st">&#39;flow_lane&#39;</span></span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a><span class="co"># merge sample_count_id and metadata files</span></span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a><span class="co"># this way metadata contains the sampleID to match the counts table </span></span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a>counts_metadata <span class="ot">&lt;-</span> <span class="fu">merge</span>(metadata, sample_count_id, <span class="at">by =</span> <span class="st">&quot;NPID&quot;</span>)</span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a><span class="co"># gene information</span></span>
<span id="cb8-30"><a href="#cb8-30" aria-hidden="true" tabindex="-1"></a>counts.geneid <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(</span>
<span id="cb8-31"><a href="#cb8-31" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;../../featureCounts/gene_info.txt&quot;</span>,</span>
<span id="cb8-32"><a href="#cb8-32" aria-hidden="true" tabindex="-1"></a>  <span class="at">header =</span> <span class="cn">TRUE</span>,</span>
<span id="cb8-33"><a href="#cb8-33" aria-hidden="true" tabindex="-1"></a>  <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span></span>
<span id="cb8-34"><a href="#cb8-34" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb8-35"><a href="#cb8-35" aria-hidden="true" tabindex="-1"></a>counts.geneid <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(counts.geneid<span class="sc">$</span>Geneid)</span>
<span id="cb8-36"><a href="#cb8-36" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(counts.geneid) <span class="ot">&lt;-</span> <span class="st">&quot;gene_id&quot;</span></span>
<span id="cb8-37"><a href="#cb8-37" aria-hidden="true" tabindex="-1"></a><span class="co"># add gene_name as row names to counts file</span></span>
<span id="cb8-38"><a href="#cb8-38" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(counts) <span class="ot">&lt;-</span> counts.geneid<span class="sc">$</span>gene_id</span>
<span id="cb8-39"><a href="#cb8-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-40"><a href="#cb8-40" aria-hidden="true" tabindex="-1"></a><span class="co"># read in annotation file</span></span>
<span id="cb8-41"><a href="#cb8-41" aria-hidden="true" tabindex="-1"></a>gtf.file <span class="ot">&lt;-</span> <span class="fu">paste0</span>(pathToRef, <span class="st">&quot;gencode.v38.annotation.gtf&quot;</span>)</span>
<span id="cb8-42"><a href="#cb8-42" aria-hidden="true" tabindex="-1"></a>genes.gtf <span class="ot">&lt;-</span> rtracklayer<span class="sc">::</span><span class="fu">import</span>(gtf.file)</span>
<span id="cb8-43"><a href="#cb8-43" aria-hidden="true" tabindex="-1"></a>genes.gtf <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(genes.gtf)</span>
<span id="cb8-44"><a href="#cb8-44" aria-hidden="true" tabindex="-1"></a>genes.gtf <span class="ot">&lt;-</span> genes.gtf[genes.gtf<span class="sc">$</span>type <span class="sc">==</span> <span class="st">&quot;gene&quot;</span>,]</span>
<span id="cb8-45"><a href="#cb8-45" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(genes.gtf<span class="sc">$</span>gene_type)</span></code></pre></div>
<pre><code>## 
##                          IG_C_gene                    IG_C_pseudogene 
##                                 14                                  9 
##                          IG_D_gene                          IG_J_gene 
##                                 37                                 18 
##                    IG_J_pseudogene                          IG_V_gene 
##                                  3                                145 
##                    IG_V_pseudogene                      IG_pseudogene 
##                                187                                  1 
##                            Mt_rRNA                            Mt_tRNA 
##                                  2                                 22 
##                                TEC                          TR_C_gene 
##                               1056                                  6 
##                          TR_D_gene                          TR_J_gene 
##                                  4                                 79 
##                    TR_J_pseudogene                          TR_V_gene 
##                                  4                                106 
##                    TR_V_pseudogene                             lncRNA 
##                                 33                              16888 
##                              miRNA                           misc_RNA 
##                               1879                               2212 
##             polymorphic_pseudogene               processed_pseudogene 
##                                 49                              10163 
##                     protein_coding                         pseudogene 
##                              19955                                 15 
##                               rRNA                    rRNA_pseudogene 
##                                 47                                497 
##                           ribozyme                               sRNA 
##                                  8                                  5 
##                              scRNA                             scaRNA 
##                                  1                                 49 
##                              snRNA                             snoRNA 
##                               1901                                943 
##   transcribed_processed_pseudogene     transcribed_unitary_pseudogene 
##                                502                                143 
## transcribed_unprocessed_pseudogene    translated_processed_pseudogene 
##                                950                                  2 
##  translated_unprocessed_pseudogene                 unitary_pseudogene 
##                                  1                                 98 
##             unprocessed_pseudogene                          vault_RNA 
##                               2614                                  1</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co">#counts.geneid</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co">#names(counts.geneid)[1] &lt;- &#39;gene_name&#39;</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>joined.df <span class="ot">&lt;-</span> <span class="fu">join</span>(counts.geneid, genes.gtf, <span class="at">type =</span> <span class="st">&quot;inner&quot;</span>)</span></code></pre></div>
<pre><code>## Joining by: gene_id</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># check columns and rows match up between files</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">all.equal</span>(<span class="fu">rownames</span>(counts), counts.geneid<span class="sc">$</span>gene_id)</span></code></pre></div>
<pre><code>## [1] &quot;Modes: character, numeric&quot;                      
## [2] &quot;Attributes: &lt; target is NULL, current is list &gt;&quot;
## [3] &quot;target is character, current is factor&quot;</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all.equal</span>(<span class="fu">rownames</span>(counts), genes.gtf<span class="sc">$</span>gene_id)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># reorder counts to be in the same order as metadata table</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> counts[, counts_metadata<span class="sc">$</span>counts_id]</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="fu">all.equal</span>(<span class="fu">colnames</span>(counts), (counts_metadata<span class="sc">$</span>counts_id))</span></code></pre></div>
<pre><code>## [1] &quot;Modes: character, numeric&quot;                      
## [2] &quot;Attributes: &lt; target is NULL, current is list &gt;&quot;
## [3] &quot;target is character, current is factor&quot;</code></pre>
<h1 id="create-dge-object">Create DGE object</h1>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create object</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">DGEList</span>(<span class="at">counts =</span> counts,</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">samples =</span> counts_metadata,</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">genes =</span> genes.gtf)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(dge<span class="sc">$</span>samples<span class="sc">$</span>TYPE)</span></code></pre></div>
<pre><code>## 
##      CONTROL CONTROL - AD CONTROL - PA          LBD 
##            8            8            2           38</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># remove AD and PA controls</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>remove <span class="ot">&lt;-</span> <span class="fu">vector</span>()</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(dge<span class="sc">$</span>samples)) {</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (dge<span class="sc">$</span>samples<span class="sc">$</span>TYPE[i] <span class="sc">==</span> <span class="st">&quot;CONTROL - AD&quot;</span> <span class="sc">|</span> dge<span class="sc">$</span>samples<span class="sc">$</span>TYPE[i] <span class="sc">==</span> <span class="st">&quot;CONTROL - PA&quot;</span>) {</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>    remove <span class="ot">&lt;-</span> <span class="fu">c</span>(remove, i)</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> dge[,<span class="sc">-</span>remove, keep.lib.sizes <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(dge<span class="sc">$</span>samples<span class="sc">$</span>TYPE)</span></code></pre></div>
<pre><code>## 
##      CONTROL CONTROL - AD CONTROL - PA          LBD 
##            8            0            0           38</code></pre>
<p>column 1 is pid id column 15 is group information (LPS or control)
column 18 is date of birth</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(dge, <span class="at">file =</span> <span class="fu">paste0</span>(<span class="st">&quot;../../rObjects/&quot;</span>, condition, <span class="st">&quot;_&quot;</span>, tool,<span class="st">&quot;_&quot;</span>, sex_karyotype, <span class="st">&quot;_gene_raw.rds&quot;</span>))</span></code></pre></div>
<h1 id="remove-mitochondrial-genes">Remove mitochondrial genes</h1>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dge)</span></code></pre></div>
<pre><code>## [1] 60649    46</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>removeMT <span class="ot">&lt;-</span> dge<span class="sc">$</span>genes<span class="sc">$</span>seqnames <span class="sc">!=</span> <span class="st">&quot;chrM&quot;</span>  <span class="co"># true when NOT MT</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> dge[removeMT,,keep.lib.sizes <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dge)</span></code></pre></div>
<pre><code>## [1] 60612    46</code></pre>
<h1 id="save-functions">Save functions</h1>
<p>These functions with help simultaneously save plots as a png, pdf,
and tiff file.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>saveToPDF <span class="ot">&lt;-</span> <span class="cf">function</span>(...) {</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>    d <span class="ot">=</span> <span class="fu">dev.copy</span>(pdf,...)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dev.off</span>(d)</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>saveToPNG <span class="ot">&lt;-</span> <span class="cf">function</span>(...) {</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>    d <span class="ot">=</span> <span class="fu">dev.copy</span>(png,...)</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dev.off</span>(d)</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<h1 id="raw-mds-with-technical-replicates">Raw MDS with technical
replicates</h1>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set colors and get data</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(dge<span class="sc">$</span>samples<span class="sc">$</span>TYPE)</span></code></pre></div>
<pre><code>## 
##      CONTROL CONTROL - AD CONTROL - PA          LBD 
##            8            0            0           38</code></pre>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>dge<span class="sc">$</span>samples<span class="sc">$</span>TYPE <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(dge<span class="sc">$</span>samples<span class="sc">$</span>TYPE)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>group_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;black&quot;</span>,<span class="st">&quot;red&quot;</span>)[dge<span class="sc">$</span>samples<span class="sc">$</span>TYPE]</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>lcpm <span class="ot">&lt;-</span> edgeR<span class="sc">::</span><span class="fu">cpm</span>(dge<span class="sc">$</span>counts, <span class="at">log =</span> <span class="cn">TRUE</span>)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">bg =</span> <span class="st">&#39;white&#39;</span>)</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a><span class="co"># plot MDS</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMDS</span>(</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>  lcpm,</span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">top =</span> <span class="dv">100</span>, </span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">labels =</span> dge<span class="sc">$</span>samples<span class="sc">$</span>NPID,</span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">cex =</span> <span class="fl">0.8</span>, </span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">dim.plot =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), </span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot =</span> <span class="cn">TRUE</span>, </span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> group_colors, </span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene.selection =</span> <span class="st">&quot;common&quot;</span></span>
<span id="cb30-16"><a href="#cb30-16" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb30-17"><a href="#cb30-17" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span>(<span class="fu">expression</span>(<span class="st">&#39;Top 100 Genes - Raw (Log&#39;</span>[<span class="dv">2</span>]<span class="sc">~</span><span class="st">&#39;CPM)&#39;</span>))</span></code></pre></div>
<figure>
<img src="figure/MDS_techreps-1.png" alt="plot of chunk MDS_techreps" />
<figcaption aria-hidden="true">plot of chunk MDS_techreps</figcaption>
</figure>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co">#legend(</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="co">#  &quot;bottomleft&quot;,</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="co">#  legend = c(control, condition),</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="co">#  pch = 16,</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="co">#  col = c(&quot;black&quot;,&quot;red&quot;),</span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a><span class="co">#  cex = 0.8</span></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a><span class="co">#)</span></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;../../results/&quot;</span>, tool, <span class="st">&quot;/MDS/&quot;</span>, condition,<span class="st">&quot;_gene_MDS_dim1&amp;2_techreps_&quot;</span>, sex_karyotype)</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPDF</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.pdf&quot;</span>), <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="dv">4</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPNG</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.png&quot;</span>), <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="dv">4</span>, <span class="at">unit =</span> <span class="st">&quot;in&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot MDS</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMDS</span>(</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  lcpm,</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">top =</span> <span class="dv">100</span>, </span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">labels =</span> dge<span class="sc">$</span>samples<span class="sc">$</span>NPID,</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">cex =</span> <span class="fl">0.8</span>, </span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">dim.plot =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">3</span>), </span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot =</span> <span class="cn">TRUE</span>, </span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> group_colors, </span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene.selection =</span> <span class="st">&quot;common&quot;</span></span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span>(<span class="fu">expression</span>(<span class="st">&#39;Top 100 Genes - Raw (Log&#39;</span>[<span class="dv">2</span>]<span class="sc">~</span><span class="st">&#39;CPM)&#39;</span>))</span></code></pre></div>
<figure>
<img src="figure/MDS_techreps-2.png" alt="plot of chunk MDS_techreps" />
<figcaption aria-hidden="true">plot of chunk MDS_techreps</figcaption>
</figure>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;../../results/&quot;</span>, tool, <span class="st">&quot;/MDS/&quot;</span>, condition,<span class="st">&quot;_gene_MDS_dim2&amp;3_techreps_&quot;</span>, sex_karyotype)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPDF</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.pdf&quot;</span>), <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="dv">4</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPNG</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.png&quot;</span>), <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="dv">4</span>, <span class="at">unit =</span> <span class="st">&quot;in&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<h1 id="rin-check-with-replicates">RIN check with replicates</h1>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first filter by expression and normalize the data</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>keep.expr <span class="ot">&lt;-</span> <span class="fu">filterByExpr</span>(dge, <span class="at">group =</span> dge<span class="sc">$</span>samples<span class="sc">$</span>TYPE)</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dge)</span></code></pre></div>
<pre><code>## [1] 60612    46</code></pre>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>dge.filtered.reps <span class="ot">&lt;-</span> dge[keep.expr, , keep.lib.sizes <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dge.filtered.reps)</span></code></pre></div>
<pre><code>## [1] 16319    46</code></pre>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(dge.filtered.reps<span class="sc">$</span>genes<span class="sc">$</span>gene_type)</span></code></pre></div>
<pre><code>## 
##                          IG_C_gene                                TEC 
##                                  1                                 91 
##                          TR_C_gene                             lncRNA 
##                                  3                               1942 
##                              miRNA                           misc_RNA 
##                                  3                                  5 
##             polymorphic_pseudogene               processed_pseudogene 
##                                  3                                121 
##                     protein_coding                              scRNA 
##                              13832                                  1 
##                             scaRNA                              snRNA 
##                                  1                                  5 
##                             snoRNA   transcribed_processed_pseudogene 
##                                  3                                 52 
##     transcribed_unitary_pseudogene transcribed_unprocessed_pseudogene 
##                                 30                                196 
##                 unitary_pseudogene             unprocessed_pseudogene 
##                                  1                                 29</code></pre>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Now, normalization by the method of trimmed mean of M-values (TMM)</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>dge.filtered.reps.norm <span class="ot">&lt;-</span> <span class="fu">calcNormFactors</span>(dge.filtered.reps, <span class="at">method =</span> <span class="st">&quot;TMM&quot;</span>)</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a><span class="co"># norm factor summary</span></span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(dge.filtered.reps.norm<span class="sc">$</span>samples<span class="sc">$</span>norm.factors)</span></code></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##  0.6909  0.9265  1.0322  1.0099  1.1152  1.1979</code></pre>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>log2cpm.norm.reps <span class="ot">&lt;-</span> edgeR<span class="sc">::</span><span class="fu">cpm</span>(dge.filtered.reps.norm, <span class="at">log =</span> T)</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>nsamples <span class="ot">&lt;-</span> <span class="fu">ncol</span>(dge.filtered.reps.norm)</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(log2cpm.norm.reps, </span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">main=</span><span class="st">&quot;Filtered normalized lcpm data with replicates&quot;</span>, </span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlab=</span><span class="st">&quot;RIN&quot;</span>, </span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab=</span><span class="fu">expression</span>(<span class="st">&#39;Counts per gene (Log&#39;</span>[<span class="dv">2</span>]<span class="sc">~</span><span class="st">&#39;CPM)&#39;</span>),</span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">axes=</span><span class="cn">FALSE</span>)</span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">2</span>)</span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">1</span>,<span class="at">at=</span><span class="dv">1</span><span class="sc">:</span>nsamples,<span class="at">labels=</span>(dge.filtered.reps.norm<span class="sc">$</span>samples<span class="sc">$</span>RIN),<span class="at">las=</span><span class="dv">2</span>,<span class="at">cex.axis=</span><span class="fl">0.8</span>)</span></code></pre></div>
<figure>
<img src="figure/unnamed-chunk-3-1.png"
alt="plot of chunk unnamed-chunk-3" />
<figcaption aria-hidden="true">plot of chunk
unnamed-chunk-3</figcaption>
</figure>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;../../results/&quot;</span>, tool, <span class="st">&quot;/boxplot/&quot;</span>, condition, <span class="st">&quot;_gene_boxplot_with_reps_with_RIN_&quot;</span>, sex_karyotype)</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPDF</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.pdf&quot;</span>), <span class="at">width =</span> <span class="dv">18</span>, <span class="at">height =</span> <span class="dv">6</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPNG</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.png&quot;</span>), <span class="at">width =</span> <span class="dv">18</span>, <span class="at">height =</span> <span class="dv">6</span>, <span class="at">unit =</span> <span class="st">&quot;in&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<h1 id="sum-technical-replicates">Sum technical replicates</h1>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co"># sum technical replicates</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dge)</span></code></pre></div>
<pre><code>## [1] 60612    46</code></pre>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>dge.tech <span class="ot">&lt;-</span> <span class="fu">sumTechReps</span>(dge, dge<span class="sc">$</span>samples<span class="sc">$</span>NPID)</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dge.tech<span class="sc">$</span>counts)</span></code></pre></div>
<pre><code>## [1] 60612    23</code></pre>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(dge.tech<span class="sc">$</span>counts) <span class="ot">&lt;-</span> dge.tech<span class="sc">$</span>samples<span class="sc">$</span>NPID</span></code></pre></div>
<h1 id="raw-mds">Raw MDS</h1>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set colors and get data</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>group_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>)[dge.tech<span class="sc">$</span>samples<span class="sc">$</span>TYPE]</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>lcpm <span class="ot">&lt;-</span> edgeR<span class="sc">::</span><span class="fu">cpm</span>(dge.tech<span class="sc">$</span>counts, <span class="at">log =</span> <span class="cn">TRUE</span>)</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">bg =</span> <span class="st">&#39;white&#39;</span>)</span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a><span class="co"># plot MDS</span></span>
<span id="cb58-8"><a href="#cb58-8" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMDS</span>(</span>
<span id="cb58-9"><a href="#cb58-9" aria-hidden="true" tabindex="-1"></a>  lcpm, </span>
<span id="cb58-10"><a href="#cb58-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">top =</span> <span class="dv">100</span>, </span>
<span id="cb58-11"><a href="#cb58-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">labels =</span> dge.tech<span class="sc">$</span>samples<span class="sc">$</span>NPID,</span>
<span id="cb58-12"><a href="#cb58-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">cex =</span> .<span class="dv">8</span>, </span>
<span id="cb58-13"><a href="#cb58-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">dim.plot =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), </span>
<span id="cb58-14"><a href="#cb58-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot =</span> <span class="cn">TRUE</span>, </span>
<span id="cb58-15"><a href="#cb58-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> group_colors,</span>
<span id="cb58-16"><a href="#cb58-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene.selection =</span> <span class="st">&quot;common&quot;</span></span>
<span id="cb58-17"><a href="#cb58-17" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb58-18"><a href="#cb58-18" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span>(<span class="fu">expression</span>(<span class="st">&#39;Top 100 Genes - Raw (Log&#39;</span>[<span class="dv">2</span>]<span class="sc">~</span><span class="st">&#39;CPM)&#39;</span>))</span></code></pre></div>
<figure>
<img src="figure/unnamed-chunk-4-1.png"
alt="plot of chunk unnamed-chunk-4" />
<figcaption aria-hidden="true">plot of chunk
unnamed-chunk-4</figcaption>
</figure>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co">#legend(</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="co">#  &quot;top&quot;,</span></span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a><span class="co">#  legend = c(control, condition),</span></span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a><span class="co">#  pch = 16,</span></span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a><span class="co">#  col = c(control_color, condition_color),</span></span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a><span class="co">#  cex = 0.8</span></span>
<span id="cb59-7"><a href="#cb59-7" aria-hidden="true" tabindex="-1"></a><span class="co">#)</span></span>
<span id="cb59-8"><a href="#cb59-8" aria-hidden="true" tabindex="-1"></a><span class="co"># save</span></span>
<span id="cb59-9"><a href="#cb59-9" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;../../results/&quot;</span>, tool, <span class="st">&quot;/MDS/&quot;</span>, condition, <span class="st">&quot;_gene_MDS_dim1&amp;2_raw_&quot;</span>, sex_karyotype)</span>
<span id="cb59-10"><a href="#cb59-10" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPDF</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.pdf&quot;</span>), <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="dv">4</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPNG</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.png&quot;</span>), <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="dv">4</span>, <span class="at">unit =</span> <span class="st">&quot;in&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot MDS</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMDS</span>(</span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>  lcpm, </span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">top =</span> <span class="dv">100</span>, </span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">labels =</span> dge.tech<span class="sc">$</span>samples<span class="sc">$</span>RIN,</span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">cex =</span> .<span class="dv">8</span>, </span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">dim.plot =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), </span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot =</span> <span class="cn">TRUE</span>, </span>
<span id="cb63-9"><a href="#cb63-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> group_colors,</span>
<span id="cb63-10"><a href="#cb63-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene.selection =</span> <span class="st">&quot;common&quot;</span></span>
<span id="cb63-11"><a href="#cb63-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb63-12"><a href="#cb63-12" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span>(<span class="fu">expression</span>(<span class="st">&#39;Top 100 Genes - Raw (Log&#39;</span>[<span class="dv">2</span>]<span class="sc">~</span><span class="st">&#39;CPM)&#39;</span>))</span></code></pre></div>
<figure>
<img src="figure/unnamed-chunk-4-2.png"
alt="plot of chunk unnamed-chunk-4" />
<figcaption aria-hidden="true">plot of chunk
unnamed-chunk-4</figcaption>
</figure>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="co">#legend(</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="co">#  &quot;top&quot;,</span></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a><span class="co">#  legend = c(control, condition),</span></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a><span class="co">#  pch = 16,</span></span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a><span class="co">#  col = c(control_color, condition_color),</span></span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a><span class="co">#  cex = 0.8</span></span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a><span class="co">#)</span></span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a><span class="co"># save</span></span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;../../results/&quot;</span>, tool, <span class="st">&quot;/MDS/&quot;</span>, condition, <span class="st">&quot;_gene_MDS_dim1&amp;2_raw_&quot;</span>, sex_karyotype, <span class="st">&quot;_label_RIN&quot;</span>)</span>
<span id="cb64-10"><a href="#cb64-10" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPDF</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.pdf&quot;</span>), <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="dv">4</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPNG</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.png&quot;</span>), <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="dv">4</span>, <span class="at">unit =</span> <span class="st">&quot;in&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot MDS</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMDS</span>(</span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>  lcpm, </span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">top =</span> <span class="dv">100</span>, </span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">labels =</span> dge.tech<span class="sc">$</span>samples<span class="sc">$</span>RIN,</span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">cex =</span> .<span class="dv">8</span>, </span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">dim.plot =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">3</span>), </span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot =</span> <span class="cn">TRUE</span>, </span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> group_colors,</span>
<span id="cb68-10"><a href="#cb68-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene.selection =</span> <span class="st">&quot;common&quot;</span></span>
<span id="cb68-11"><a href="#cb68-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb68-12"><a href="#cb68-12" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span>(<span class="fu">expression</span>(<span class="st">&#39;Top 100 Genes - Raw (Log&#39;</span>[<span class="dv">2</span>]<span class="sc">~</span><span class="st">&#39;CPM)&#39;</span>))</span></code></pre></div>
<figure>
<img src="figure/unnamed-chunk-4-3.png"
alt="plot of chunk unnamed-chunk-4" />
<figcaption aria-hidden="true">plot of chunk
unnamed-chunk-4</figcaption>
</figure>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;../../results/&quot;</span>, tool, <span class="st">&quot;/MDS/&quot;</span>, condition, <span class="st">&quot;_gene_MDS_dim2&amp;3_raw_&quot;</span>, sex_karyotype, <span class="st">&quot;_label_RIN&quot;</span>)</span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPDF</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.pdf&quot;</span>), <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="dv">4</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPNG</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.png&quot;</span>), <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="dv">4</span>, <span class="at">unit =</span> <span class="st">&quot;in&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<h1 id="filter-lowly-expressed-genes">Filter lowly expressed genes</h1>
<p>The filterByExpr() function in the edgeR package determines which
genes have a great enough count value to keep. We will filter by group.
This means at least 6 samples (6 is the smallest group sample size) must
express a minimum count of 10 (in cpm, default value).</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>keep.expr <span class="ot">&lt;-</span> <span class="fu">filterByExpr</span>(dge.tech, <span class="at">group =</span> dge.tech<span class="sc">$</span>samples<span class="sc">$</span>TYPE)</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dge.tech)</span></code></pre></div>
<pre><code>## [1] 60612    23</code></pre>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>dge.filtered <span class="ot">&lt;-</span> dge.tech[keep.expr, , keep.lib.sizes <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dge.filtered)</span></code></pre></div>
<pre><code>## [1] 18389    23</code></pre>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(dge.filtered<span class="sc">$</span>genes<span class="sc">$</span>gene_type)</span></code></pre></div>
<pre><code>## 
##                          IG_C_gene                                TEC 
##                                  2                                156 
##                          TR_C_gene                             lncRNA 
##                                  4                               2918 
##                              miRNA                           misc_RNA 
##                                  6                                 15 
##             polymorphic_pseudogene               processed_pseudogene 
##                                  5                                231 
##                     protein_coding                              scRNA 
##                              14584                                  1 
##                             scaRNA                              snRNA 
##                                  1                                 14 
##                             snoRNA   transcribed_processed_pseudogene 
##                                 14                                 79 
##     transcribed_unitary_pseudogene transcribed_unprocessed_pseudogene 
##                                 42                                257 
##                 unitary_pseudogene             unprocessed_pseudogene 
##                                  1                                 59</code></pre>
<h1 id="tmm-normalization">TMM normalization</h1>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Now, normalization by the method of trimmed mean of M-values (TMM)</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>dge.filtered.norm <span class="ot">&lt;-</span> <span class="fu">calcNormFactors</span>(dge.filtered, <span class="at">method =</span> <span class="st">&quot;TMM&quot;</span>)</span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a><span class="co"># norm factor summary</span></span>
<span id="cb79-5"><a href="#cb79-5" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(dge.filtered.norm<span class="sc">$</span>samples<span class="sc">$</span>norm.factors)</span></code></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##  0.6855  0.9293  1.0383  1.0100  1.1155  1.1967</code></pre>
<h1 id="density-plot">Density plot</h1>
<p>Density plots of log-intensity distribution of each library can be
superposed on a single graph for a better comparison between libraries
and for identification of libraries with weird distribution.</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set graphical parameter</span></span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>))</span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize data for library size and expression intesntiy</span></span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a>log2cpm.tech <span class="ot">&lt;-</span> edgeR<span class="sc">::</span><span class="fu">cpm</span>(dge.tech, <span class="at">log =</span> <span class="cn">TRUE</span>)</span>
<span id="cb81-6"><a href="#cb81-6" aria-hidden="true" tabindex="-1"></a>log2cpm.filtered <span class="ot">&lt;-</span> edgeR<span class="sc">::</span><span class="fu">cpm</span>(dge.filtered, <span class="at">log =</span> <span class="cn">TRUE</span>)</span>
<span id="cb81-7"><a href="#cb81-7" aria-hidden="true" tabindex="-1"></a>log2cpm.norm <span class="ot">&lt;-</span> edgeR<span class="sc">::</span><span class="fu">cpm</span>(dge.filtered.norm, <span class="at">log =</span> <span class="cn">TRUE</span>)</span>
<span id="cb81-8"><a href="#cb81-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-9"><a href="#cb81-9" aria-hidden="true" tabindex="-1"></a><span class="co"># set colors</span></span>
<span id="cb81-10"><a href="#cb81-10" aria-hidden="true" tabindex="-1"></a>colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;red&quot;</span>,<span class="st">&quot;orange&quot;</span>,<span class="st">&quot;green&quot;</span>,<span class="st">&quot;yellow&quot;</span>,<span class="st">&quot;blue&quot;</span>,<span class="st">&quot;purple&quot;</span>, </span>
<span id="cb81-11"><a href="#cb81-11" aria-hidden="true" tabindex="-1"></a>            <span class="st">&quot;lightgray&quot;</span>,<span class="st">&quot;brown&quot;</span>,<span class="st">&quot;pink&quot;</span>,<span class="st">&quot;cyan&quot;</span>)</span>
<span id="cb81-12"><a href="#cb81-12" aria-hidden="true" tabindex="-1"></a>nsamples <span class="ot">&lt;-</span> <span class="fu">ncol</span>(dge.tech)</span>
<span id="cb81-13"><a href="#cb81-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-14"><a href="#cb81-14" aria-hidden="true" tabindex="-1"></a><span class="co"># First, plot the first column of the log2cpm.tech density</span></span>
<span id="cb81-15"><a href="#cb81-15" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(log2cpm.tech[,<span class="dv">1</span>]), <span class="at">col =</span> colors[<span class="dv">1</span>], <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="fl">0.25</span>), </span>
<span id="cb81-16"><a href="#cb81-16" aria-hidden="true" tabindex="-1"></a>     <span class="at">las =</span> <span class="dv">2</span>, <span class="at">main =</span> <span class="st">&quot;A. Raw&quot;</span>, <span class="at">xlab =</span> <span class="fu">expression</span>(<span class="st">&#39;Log&#39;</span>[<span class="dv">2</span>]<span class="sc">~</span>CPM))</span>
<span id="cb81-17"><a href="#cb81-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-18"><a href="#cb81-18" aria-hidden="true" tabindex="-1"></a><span class="co"># For each sample plot the lcpm density</span></span>
<span id="cb81-19"><a href="#cb81-19" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>nsamples){</span>
<span id="cb81-20"><a href="#cb81-20" aria-hidden="true" tabindex="-1"></a>  den <span class="ot">&lt;-</span> <span class="fu">density</span>(log2cpm.tech[,i]) <span class="co">#subset each column</span></span>
<span id="cb81-21"><a href="#cb81-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lines</span>(den<span class="sc">$</span>x, den<span class="sc">$</span>y, <span class="at">col =</span> colors[i], <span class="at">lwd =</span> <span class="dv">2</span>) </span>
<span id="cb81-22"><a href="#cb81-22" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb81-23"><a href="#cb81-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-24"><a href="#cb81-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Second, plot log2cpm.filtered</span></span>
<span id="cb81-25"><a href="#cb81-25" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(log2cpm.filtered[,<span class="dv">1</span>]), <span class="at">col =</span> colors[<span class="dv">1</span>], <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="fl">0.25</span>), </span>
<span id="cb81-26"><a href="#cb81-26" aria-hidden="true" tabindex="-1"></a>     <span class="at">las =</span> <span class="dv">2</span>, <span class="at">main =</span> <span class="st">&quot;B. Filtered&quot;</span>, <span class="at">xlab =</span> <span class="fu">expression</span>(<span class="st">&#39;Log&#39;</span>[<span class="dv">2</span>]<span class="sc">~</span>CPM))</span>
<span id="cb81-27"><a href="#cb81-27" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> edgeR<span class="sc">::</span><span class="fu">cpm</span>(<span class="dv">3</span>, <span class="at">log =</span> <span class="cn">TRUE</span>), <span class="at">lty =</span> <span class="dv">3</span>)</span>
<span id="cb81-28"><a href="#cb81-28" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>nsamples) {</span>
<span id="cb81-29"><a href="#cb81-29" aria-hidden="true" tabindex="-1"></a>  den <span class="ot">&lt;-</span> <span class="fu">density</span>(log2cpm.filtered[,i])</span>
<span id="cb81-30"><a href="#cb81-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lines</span>(den<span class="sc">$</span>x, den<span class="sc">$</span>y, <span class="at">col =</span> colors[i], <span class="at">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb81-31"><a href="#cb81-31" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb81-32"><a href="#cb81-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-33"><a href="#cb81-33" aria-hidden="true" tabindex="-1"></a><span class="co"># Third, plot log2cpm.norm</span></span>
<span id="cb81-34"><a href="#cb81-34" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(log2cpm.norm[,<span class="dv">1</span>]), <span class="at">col =</span> colors[<span class="dv">1</span>], <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="fl">0.25</span>), </span>
<span id="cb81-35"><a href="#cb81-35" aria-hidden="true" tabindex="-1"></a>     <span class="at">las =</span> <span class="dv">2</span>, <span class="at">main =</span> <span class="st">&quot;C. Normalized&quot;</span>, <span class="at">xlab =</span> <span class="fu">expression</span>(<span class="st">&#39;Log&#39;</span>[<span class="dv">2</span>]<span class="sc">~</span>CPM))</span>
<span id="cb81-36"><a href="#cb81-36" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> edgeR<span class="sc">::</span><span class="fu">cpm</span>(<span class="dv">3</span>, <span class="at">log =</span> <span class="cn">TRUE</span>), <span class="at">lty =</span> <span class="dv">3</span>)</span>
<span id="cb81-37"><a href="#cb81-37" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>nsamples) {</span>
<span id="cb81-38"><a href="#cb81-38" aria-hidden="true" tabindex="-1"></a>  den <span class="ot">&lt;-</span> <span class="fu">density</span>(log2cpm.norm[,i])</span>
<span id="cb81-39"><a href="#cb81-39" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lines</span>(den<span class="sc">$</span>x, den<span class="sc">$</span>y, <span class="at">col =</span> colors[i], <span class="at">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb81-40"><a href="#cb81-40" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<figure>
<img src="figure/density_plots-1.png"
alt="plot of chunk density_plots" />
<figcaption aria-hidden="true">plot of chunk density_plots</figcaption>
</figure>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save</span></span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;../../results/&quot;</span>, tool, <span class="st">&quot;/density/&quot;</span>, condition, <span class="st">&quot;_gene_density_&quot;</span>, sex_karyotype)</span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPDF</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.pdf&quot;</span>), <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">4</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPNG</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.png&quot;</span>), <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">4</span>, <span class="at">unit =</span> <span class="st">&quot;in&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<h1 id="boxplots">Boxplots</h1>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set parameters</span></span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>))</span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a><span class="co"># First look at dge.tech</span></span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(log2cpm.tech, </span>
<span id="cb86-6"><a href="#cb86-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">main=</span><span class="st">&quot;A. Raw&quot;</span>, </span>
<span id="cb86-7"><a href="#cb86-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlab=</span><span class="st">&quot;&quot;</span>, </span>
<span id="cb86-8"><a href="#cb86-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab=</span><span class="fu">expression</span>(<span class="st">&#39;Counts per gene (Log&#39;</span>[<span class="dv">2</span>]<span class="sc">~</span><span class="st">&#39;CPM)&#39;</span>),</span>
<span id="cb86-9"><a href="#cb86-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">axes=</span><span class="cn">FALSE</span>,</span>
<span id="cb86-10"><a href="#cb86-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> colors</span>
<span id="cb86-11"><a href="#cb86-11" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb86-12"><a href="#cb86-12" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">2</span>) <span class="co"># 2 = left </span></span>
<span id="cb86-13"><a href="#cb86-13" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">1</span>, <span class="co"># 1 = below </span></span>
<span id="cb86-14"><a href="#cb86-14" aria-hidden="true" tabindex="-1"></a>     <span class="at">at =</span> <span class="dv">1</span><span class="sc">:</span>nsamples, <span class="co"># points at which tick-marks should be drawn</span></span>
<span id="cb86-15"><a href="#cb86-15" aria-hidden="true" tabindex="-1"></a>     <span class="at">labels =</span> <span class="fu">colnames</span>(log2cpm.tech),</span>
<span id="cb86-16"><a href="#cb86-16" aria-hidden="true" tabindex="-1"></a>     <span class="at">las =</span> <span class="dv">2</span>,</span>
<span id="cb86-17"><a href="#cb86-17" aria-hidden="true" tabindex="-1"></a>     <span class="at">cex.axis =</span> <span class="fl">0.8</span> <span class="co"># size of axis</span></span>
<span id="cb86-18"><a href="#cb86-18" aria-hidden="true" tabindex="-1"></a>     )</span>
<span id="cb86-19"><a href="#cb86-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-20"><a href="#cb86-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Second, look at dge.filtered</span></span>
<span id="cb86-21"><a href="#cb86-21" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(log2cpm.filtered, </span>
<span id="cb86-22"><a href="#cb86-22" aria-hidden="true" tabindex="-1"></a>        <span class="at">main=</span><span class="st">&quot;B. Filtered&quot;</span>, </span>
<span id="cb86-23"><a href="#cb86-23" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlab=</span><span class="st">&quot;&quot;</span>, </span>
<span id="cb86-24"><a href="#cb86-24" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab=</span><span class="fu">expression</span>(<span class="st">&#39;Counts per gene (Log&#39;</span>[<span class="dv">2</span>]<span class="sc">~</span><span class="st">&#39;CPM)&#39;</span>),</span>
<span id="cb86-25"><a href="#cb86-25" aria-hidden="true" tabindex="-1"></a>        <span class="at">axes=</span><span class="cn">FALSE</span>,</span>
<span id="cb86-26"><a href="#cb86-26" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> colors</span>
<span id="cb86-27"><a href="#cb86-27" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb86-28"><a href="#cb86-28" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">2</span>)</span>
<span id="cb86-29"><a href="#cb86-29" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">1</span>, <span class="at">at=</span><span class="dv">1</span><span class="sc">:</span>nsamples,<span class="at">labels=</span><span class="fu">colnames</span>(log2cpm.filtered),<span class="at">las=</span><span class="dv">2</span>,<span class="at">cex.axis=</span><span class="fl">0.8</span>)</span>
<span id="cb86-30"><a href="#cb86-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-31"><a href="#cb86-31" aria-hidden="true" tabindex="-1"></a><span class="co"># Third, look at dge.norm</span></span>
<span id="cb86-32"><a href="#cb86-32" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(log2cpm.norm, </span>
<span id="cb86-33"><a href="#cb86-33" aria-hidden="true" tabindex="-1"></a>        <span class="at">main=</span><span class="st">&quot;C. Normalized&quot;</span>, </span>
<span id="cb86-34"><a href="#cb86-34" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlab=</span><span class="st">&quot;&quot;</span>, </span>
<span id="cb86-35"><a href="#cb86-35" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab=</span><span class="fu">expression</span>(<span class="st">&#39;Counts per gene (Log&#39;</span>[<span class="dv">2</span>]<span class="sc">~</span><span class="st">&#39;CPM)&#39;</span>),</span>
<span id="cb86-36"><a href="#cb86-36" aria-hidden="true" tabindex="-1"></a>        <span class="at">axes=</span><span class="cn">FALSE</span>,</span>
<span id="cb86-37"><a href="#cb86-37" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> colors)</span>
<span id="cb86-38"><a href="#cb86-38" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">2</span>)</span>
<span id="cb86-39"><a href="#cb86-39" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">1</span>,<span class="at">at=</span><span class="dv">1</span><span class="sc">:</span>nsamples,<span class="at">labels=</span><span class="fu">colnames</span>(log2cpm.norm),<span class="at">las=</span><span class="dv">2</span>,<span class="at">cex.axis=</span><span class="fl">0.8</span>)</span></code></pre></div>
<figure>
<img src="figure/boxplots-1.png" alt="plot of chunk boxplots" />
<figcaption aria-hidden="true">plot of chunk boxplots</figcaption>
</figure>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save</span></span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;../../results/&quot;</span>, tool, <span class="st">&quot;/boxplot/&quot;</span>, condition, <span class="st">&quot;_gene_boxplot_&quot;</span>, sex_karyotype)</span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPDF</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.pdf&quot;</span>), <span class="at">width =</span> <span class="dv">10</span>, <span class="at">height =</span> <span class="dv">6</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPNG</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.png&quot;</span>), <span class="at">width =</span> <span class="dv">10</span>, <span class="at">height =</span> <span class="dv">6</span>, <span class="at">unit =</span> <span class="st">&quot;in&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<h1 id="rin-check-with-replicates-summed">RIN check with replicates
summed</h1>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(log2cpm.norm, </span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">main=</span><span class="st">&quot;Filtered normalized lcpm data&quot;</span>, </span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlab=</span><span class="st">&quot;RIN&quot;</span>, </span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab=</span><span class="fu">expression</span>(<span class="st">&#39;Counts per gene (Log&#39;</span>[<span class="dv">2</span>]<span class="sc">~</span><span class="st">&#39;CPM)&#39;</span>),</span>
<span id="cb91-5"><a href="#cb91-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">axes=</span><span class="cn">FALSE</span>,</span>
<span id="cb91-6"><a href="#cb91-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> colors)</span>
<span id="cb91-7"><a href="#cb91-7" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">2</span>)</span>
<span id="cb91-8"><a href="#cb91-8" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">1</span>,<span class="at">at=</span><span class="dv">1</span><span class="sc">:</span>nsamples,<span class="at">labels=</span>(dge.filtered.norm<span class="sc">$</span>samples<span class="sc">$</span>RIN),<span class="at">las=</span><span class="dv">2</span>,<span class="at">cex.axis=</span><span class="fl">0.8</span>)</span></code></pre></div>
<figure>
<img src="figure/unnamed-chunk-5-1.png"
alt="plot of chunk unnamed-chunk-5" />
<figcaption aria-hidden="true">plot of chunk
unnamed-chunk-5</figcaption>
</figure>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;../../results/&quot;</span>, tool, <span class="st">&quot;/boxplot/&quot;</span>, condition, <span class="st">&quot;_gene_boxplot_sum_reps_&quot;</span>, sex_karyotype)</span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPDF</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.pdf&quot;</span>), <span class="at">width =</span> <span class="dv">12</span>, <span class="at">height =</span> <span class="dv">6</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPNG</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.png&quot;</span>), <span class="at">width =</span> <span class="dv">12</span>, <span class="at">height =</span> <span class="dv">6</span>, <span class="at">unit =</span> <span class="st">&quot;in&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="co"># check if there is correlation between RIN and library size</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>box <span class="ot">&lt;-</span> dge.filtered.norm<span class="sc">$</span>samples</span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(box<span class="sc">$</span>RIN, box<span class="sc">$</span>lib.size)</span></code></pre></div>
<figure>
<img src="figure/unnamed-chunk-5-2.png"
alt="plot of chunk unnamed-chunk-5" />
<figcaption aria-hidden="true">plot of chunk
unnamed-chunk-5</figcaption>
</figure>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(box<span class="sc">$</span>RIN, box<span class="sc">$</span>lib.size, <span class="at">method =</span> <span class="fu">c</span>(<span class="st">&quot;pearson&quot;</span>, <span class="st">&quot;kendall&quot;</span>, <span class="st">&quot;spearman&quot;</span>))</span></code></pre></div>
<pre><code>## [1] 0.2219294</code></pre>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cor.test</span>(box<span class="sc">$</span>RIN, box<span class="sc">$</span>lib.size, <span class="at">method=</span><span class="fu">c</span>(<span class="st">&quot;pearson&quot;</span>, <span class="st">&quot;kendall&quot;</span>, <span class="st">&quot;spearman&quot;</span>))</span></code></pre></div>
<pre><code>## 
##  Pearson&#39;s product-moment correlation
## 
## data:  box$RIN and box$lib.size
## t = 1.043, df = 21, p-value = 0.3088
## alternative hypothesis: true correlation is not equal to 0
## 95 percent confidence interval:
##  -0.2094316  0.5809833
## sample estimates:
##       cor 
## 0.2219294</code></pre>
<div class="sourceCode" id="cb101"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="co"># is the data normally distrubited </span></span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggqqplot</span>(box<span class="sc">$</span>lib.size, <span class="at">ylab =</span> <span class="st">&quot;library size&quot;</span>)</span></code></pre></div>
<figure>
<img src="figure/unnamed-chunk-5-3.png"
alt="plot of chunk unnamed-chunk-5" />
<figcaption aria-hidden="true">plot of chunk
unnamed-chunk-5</figcaption>
</figure>
<div class="sourceCode" id="cb102"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="co"># wt</span></span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggqqplot</span>(box<span class="sc">$</span>RIN, <span class="at">ylab =</span> <span class="st">&quot;RIN&quot;</span>)</span></code></pre></div>
<figure>
<img src="figure/unnamed-chunk-5-4.png"
alt="plot of chunk unnamed-chunk-5" />
<figcaption aria-hidden="true">plot of chunk
unnamed-chunk-5</figcaption>
</figure>
<div class="sourceCode" id="cb103"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">cor.test</span>(box<span class="sc">$</span>lib.size, box<span class="sc">$</span>RIN, </span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a>                <span class="at">method =</span> <span class="st">&quot;pearson&quot;</span>)</span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a>res</span></code></pre></div>
<pre><code>## 
##  Pearson&#39;s product-moment correlation
## 
## data:  box$lib.size and box$RIN
## t = 1.043, df = 21, p-value = 0.3088
## alternative hypothesis: true correlation is not equal to 0
## 95 percent confidence interval:
##  -0.2094316  0.5809833
## sample estimates:
##       cor 
## 0.2219294</code></pre>
<div class="sourceCode" id="cb105"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggscatter</span>(box, <span class="at">x =</span> <span class="st">&quot;RIN&quot;</span>, <span class="at">y =</span> <span class="st">&quot;lib.size&quot;</span>, </span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a>          <span class="at">add =</span> <span class="st">&quot;reg.line&quot;</span>, <span class="at">conf.int =</span> <span class="cn">TRUE</span>, </span>
<span id="cb105-3"><a href="#cb105-3" aria-hidden="true" tabindex="-1"></a>          <span class="at">cor.coef =</span> <span class="cn">TRUE</span>, <span class="at">cor.method =</span> <span class="st">&quot;pearson&quot;</span>,</span>
<span id="cb105-4"><a href="#cb105-4" aria-hidden="true" tabindex="-1"></a>          <span class="at">ylab =</span> <span class="st">&quot;library size&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;RIN value&quot;</span>) </span></code></pre></div>
<pre><code>## `geom_smooth()` using formula &#39;y ~ x&#39;</code></pre>
<figure>
<img src="figure/unnamed-chunk-5-5.png"
alt="plot of chunk unnamed-chunk-5" />
<figcaption aria-hidden="true">plot of chunk
unnamed-chunk-5</figcaption>
</figure>
<div class="sourceCode" id="cb107"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;../../results/&quot;</span>, tool, <span class="st">&quot;/library/&quot;</span>, condition, <span class="st">&quot;_corr_RIN_lib_size_&quot;</span>, sex_karyotype)</span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPDF</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.pdf&quot;</span>), <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">6</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb109"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPNG</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.png&quot;</span>), <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">6</span>, <span class="at">unit =</span> <span class="st">&quot;in&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<h1 id="cca-heatmap">CCA Heatmap</h1>
<div class="sourceCode" id="cb111"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="co">#form &lt;- ~ TYPE + Age + Sex + ClinicalDx +lib.size + LBD.type +CDLB</span></span>
<span id="cb111-2"><a href="#cb111-2" aria-hidden="true" tabindex="-1"></a>form <span class="ot">&lt;-</span> <span class="er">~</span> TYPE <span class="sc">+</span> PathDx <span class="sc">+</span> AD.subtype <span class="sc">+</span> LBD.type <span class="sc">+</span> CDLB <span class="sc">+</span> Braak.NFT <span class="sc">+</span> Thal.amyloid <span class="sc">+</span> MF.SP <span class="sc">+</span> MF.NFT <span class="sc">+</span> MF.LB <span class="sc">+</span> Cing.LB <span class="sc">+</span> MF.Amyloid <span class="sc">+</span> MF.Tau <span class="sc">+</span> Cing.Synuclein <span class="sc">+</span> CWOW.Category <span class="sc">+</span> VaD  <span class="sc">+</span> TDP.type <span class="sc">+</span> Brain.wt <span class="sc">+</span> ClinicalDx <span class="sc">+</span> FHx <span class="sc">+</span> Duration <span class="sc">+</span> Age  <span class="sc">+</span> PMI <span class="sc">+</span> APOE <span class="sc">+</span> MAPT <span class="sc">+</span> GRN <span class="sc">+</span> TMEM106b <span class="sc">+</span> RIN <span class="sc">+</span> Total.RNA.ng </span>
<span id="cb111-3"><a href="#cb111-3" aria-hidden="true" tabindex="-1"></a><span class="co"># removed Race since there is only white samples in this set. </span></span>
<span id="cb111-4"><a href="#cb111-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb111-5"><a href="#cb111-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute Canonical Correlation Analysis (CCA)</span></span>
<span id="cb111-6"><a href="#cb111-6" aria-hidden="true" tabindex="-1"></a><span class="co"># between all pairs of variables</span></span>
<span id="cb111-7"><a href="#cb111-7" aria-hidden="true" tabindex="-1"></a><span class="co"># returns absolute correlation value</span></span>
<span id="cb111-8"><a href="#cb111-8" aria-hidden="true" tabindex="-1"></a>info <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(dge.filtered.norm<span class="sc">$</span>samples)</span>
<span id="cb111-9"><a href="#cb111-9" aria-hidden="true" tabindex="-1"></a>C <span class="ot">=</span> <span class="fu">canCorPairs</span>( form, info)</span></code></pre></div>
<pre><code>## Error in canCorPairs(form, info): could not find function &quot;canCorPairs&quot;</code></pre>
<div class="sourceCode" id="cb113"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot correlation matrix</span></span>
<span id="cb113-2"><a href="#cb113-2" aria-hidden="true" tabindex="-1"></a><span class="co">#plotCorrMatrix( C , key.xlab = &quot;correlation&quot;)</span></span>
<span id="cb113-3"><a href="#cb113-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb113-4"><a href="#cb113-4" aria-hidden="true" tabindex="-1"></a><span class="co"># replace NA with zero</span></span>
<span id="cb113-5"><a href="#cb113-5" aria-hidden="true" tabindex="-1"></a>C[<span class="fu">is.na</span>(C)] <span class="ot">=</span> <span class="dv">0</span></span></code></pre></div>
<pre><code>## Warning in is.na(C): is.na() applied to non-(list or vector) of type &#39;closure&#39;</code></pre>
<pre><code>## Error in C[is.na(C)] = 0: object of type &#39;closure&#39; is not subsettable</code></pre>
<div class="sourceCode" id="cb116"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a>cor.mtest <span class="ot">&lt;-</span> <span class="cf">function</span>(C, ...) {</span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a>    C <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(C)</span>
<span id="cb116-3"><a href="#cb116-3" aria-hidden="true" tabindex="-1"></a>    n <span class="ot">&lt;-</span> <span class="fu">ncol</span>(C)</span>
<span id="cb116-4"><a href="#cb116-4" aria-hidden="true" tabindex="-1"></a>    p.mat<span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>, n, n)</span>
<span id="cb116-5"><a href="#cb116-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">diag</span>(p.mat) <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb116-6"><a href="#cb116-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>(n <span class="sc">-</span> <span class="dv">1</span>)) {</span>
<span id="cb116-7"><a href="#cb116-7" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> (j <span class="cf">in</span> (i <span class="sc">+</span> <span class="dv">1</span>)<span class="sc">:</span>n) {</span>
<span id="cb116-8"><a href="#cb116-8" aria-hidden="true" tabindex="-1"></a>            tmp <span class="ot">&lt;-</span> <span class="fu">cor.test</span>(C[, i], C[, j], ...)</span>
<span id="cb116-9"><a href="#cb116-9" aria-hidden="true" tabindex="-1"></a>            p.mat[i, j] <span class="ot">&lt;-</span> p.mat[j, i] <span class="ot">&lt;-</span> tmp<span class="sc">$</span>p.value</span>
<span id="cb116-10"><a href="#cb116-10" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb116-11"><a href="#cb116-11" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb116-12"><a href="#cb116-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(p.mat) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(p.mat) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(C)</span>
<span id="cb116-13"><a href="#cb116-13" aria-hidden="true" tabindex="-1"></a>  p.mat</span>
<span id="cb116-14"><a href="#cb116-14" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb116-15"><a href="#cb116-15" aria-hidden="true" tabindex="-1"></a><span class="co"># matrix of the p-value of the correlation</span></span>
<span id="cb116-16"><a href="#cb116-16" aria-hidden="true" tabindex="-1"></a>p.mat <span class="ot">&lt;-</span> <span class="fu">cor.mtest</span>(C)</span></code></pre></div>
<pre><code>## Error in as.vector(x, mode): cannot coerce type &#39;closure&#39; to vector of type &#39;any&#39;</code></pre>
<div class="sourceCode" id="cb118"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a>col <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">&quot;#BB4444&quot;</span>, <span class="st">&quot;#EE9988&quot;</span>, <span class="st">&quot;#FFFFFF&quot;</span>, <span class="st">&quot;#77AADD&quot;</span>, <span class="st">&quot;#4477AA&quot;</span>))</span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a><span class="fu">corrplot</span>(C, <span class="at">method=</span><span class="st">&quot;color&quot;</span>, <span class="at">col=</span><span class="fu">col</span>(<span class="dv">200</span>),  </span>
<span id="cb118-3"><a href="#cb118-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">type=</span><span class="st">&quot;upper&quot;</span>, <span class="at">order=</span><span class="st">&quot;hclust&quot;</span>, </span>
<span id="cb118-4"><a href="#cb118-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">addCoef.col =</span> <span class="st">&quot;black&quot;</span>, <span class="co"># Add coefficient of correlation</span></span>
<span id="cb118-5"><a href="#cb118-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">tl.col=</span><span class="st">&quot;black&quot;</span>, <span class="at">tl.srt=</span><span class="dv">45</span>, <span class="co">#Text label color and rotation</span></span>
<span id="cb118-6"><a href="#cb118-6" aria-hidden="true" tabindex="-1"></a>         <span class="co"># Combine with significance</span></span>
<span id="cb118-7"><a href="#cb118-7" aria-hidden="true" tabindex="-1"></a>        <span class="co"># p.mat = p.mat, sig.level = 0.1, insig = &quot;blank&quot;, </span></span>
<span id="cb118-8"><a href="#cb118-8" aria-hidden="true" tabindex="-1"></a>         <span class="co"># hide correlation coefficient on the principal diagonal</span></span>
<span id="cb118-9"><a href="#cb118-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">diag=</span><span class="cn">FALSE</span> </span>
<span id="cb118-10"><a href="#cb118-10" aria-hidden="true" tabindex="-1"></a>         )</span></code></pre></div>
<pre><code>## Error in corrplot(C, method = &quot;color&quot;, col = col(200), type = &quot;upper&quot;, : Need a matrix or data frame!</code></pre>
<div class="sourceCode" id="cb120"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;../../results/&quot;</span>, tool ,<span class="st">&quot;/varpart/LBD_gene_CCA&quot;</span>, sex_karyotype)</span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPDF</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.pdf&quot;</span>), <span class="at">width =</span> <span class="dv">25</span>, <span class="at">height =</span> <span class="dv">25</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb122"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPNG</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.png&quot;</span>), <span class="at">width =</span> <span class="dv">23</span>, <span class="at">height =</span> <span class="dv">25</span>, <span class="at">unit =</span> <span class="st">&quot;in&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb124"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a><span class="co"># CCA for only a handful of variables </span></span>
<span id="cb124-2"><a href="#cb124-2" aria-hidden="true" tabindex="-1"></a>form <span class="ot">&lt;-</span> <span class="er">~</span> TYPE <span class="sc">+</span>  Brain.wt <span class="sc">+</span> Duration <span class="sc">+</span> Age <span class="sc">+</span> PMI <span class="sc">+</span> RIN</span>
<span id="cb124-3"><a href="#cb124-3" aria-hidden="true" tabindex="-1"></a><span class="co"># removed Race since there is only white samples in this set. </span></span>
<span id="cb124-4"><a href="#cb124-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb124-5"><a href="#cb124-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute Canonical Correlation Analysis (CCA)</span></span>
<span id="cb124-6"><a href="#cb124-6" aria-hidden="true" tabindex="-1"></a><span class="co"># between all pairs of variables</span></span>
<span id="cb124-7"><a href="#cb124-7" aria-hidden="true" tabindex="-1"></a><span class="co"># returns absolute correlation value</span></span>
<span id="cb124-8"><a href="#cb124-8" aria-hidden="true" tabindex="-1"></a>info <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(dge.filtered.norm<span class="sc">$</span>samples)</span>
<span id="cb124-9"><a href="#cb124-9" aria-hidden="true" tabindex="-1"></a>C <span class="ot">=</span> <span class="fu">canCorPairs</span>( form, info)</span></code></pre></div>
<pre><code>## Error in canCorPairs(form, info): could not find function &quot;canCorPairs&quot;</code></pre>
<div class="sourceCode" id="cb126"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot correlation matrix</span></span>
<span id="cb126-2"><a href="#cb126-2" aria-hidden="true" tabindex="-1"></a><span class="co">#plotCorrMatrix( C , key.xlab = &quot;correlation&quot;)</span></span>
<span id="cb126-3"><a href="#cb126-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb126-4"><a href="#cb126-4" aria-hidden="true" tabindex="-1"></a><span class="co"># replace NA with zero</span></span>
<span id="cb126-5"><a href="#cb126-5" aria-hidden="true" tabindex="-1"></a>C[<span class="fu">is.na</span>(C)] <span class="ot">=</span> <span class="dv">0</span></span></code></pre></div>
<pre><code>## Warning in is.na(C): is.na() applied to non-(list or vector) of type &#39;closure&#39;</code></pre>
<pre><code>## Error in C[is.na(C)] = 0: object of type &#39;closure&#39; is not subsettable</code></pre>
<div class="sourceCode" id="cb129"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a>cor.mtest <span class="ot">&lt;-</span> <span class="cf">function</span>(C, ...) {</span>
<span id="cb129-2"><a href="#cb129-2" aria-hidden="true" tabindex="-1"></a>    C <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(C)</span>
<span id="cb129-3"><a href="#cb129-3" aria-hidden="true" tabindex="-1"></a>    n <span class="ot">&lt;-</span> <span class="fu">ncol</span>(C)</span>
<span id="cb129-4"><a href="#cb129-4" aria-hidden="true" tabindex="-1"></a>    p.mat<span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>, n, n)</span>
<span id="cb129-5"><a href="#cb129-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">diag</span>(p.mat) <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb129-6"><a href="#cb129-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>(n <span class="sc">-</span> <span class="dv">1</span>)) {</span>
<span id="cb129-7"><a href="#cb129-7" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> (j <span class="cf">in</span> (i <span class="sc">+</span> <span class="dv">1</span>)<span class="sc">:</span>n) {</span>
<span id="cb129-8"><a href="#cb129-8" aria-hidden="true" tabindex="-1"></a>            tmp <span class="ot">&lt;-</span> <span class="fu">cor.test</span>(C[, i], C[, j], ...)</span>
<span id="cb129-9"><a href="#cb129-9" aria-hidden="true" tabindex="-1"></a>            p.mat[i, j] <span class="ot">&lt;-</span> p.mat[j, i] <span class="ot">&lt;-</span> tmp<span class="sc">$</span>p.value</span>
<span id="cb129-10"><a href="#cb129-10" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb129-11"><a href="#cb129-11" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb129-12"><a href="#cb129-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(p.mat) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(p.mat) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(C)</span>
<span id="cb129-13"><a href="#cb129-13" aria-hidden="true" tabindex="-1"></a>  p.mat</span>
<span id="cb129-14"><a href="#cb129-14" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb129-15"><a href="#cb129-15" aria-hidden="true" tabindex="-1"></a><span class="co"># matrix of the p-value of the correlation</span></span>
<span id="cb129-16"><a href="#cb129-16" aria-hidden="true" tabindex="-1"></a>p.mat <span class="ot">&lt;-</span> <span class="fu">cor.mtest</span>(C)</span></code></pre></div>
<pre><code>## Error in as.vector(x, mode): cannot coerce type &#39;closure&#39; to vector of type &#39;any&#39;</code></pre>
<div class="sourceCode" id="cb131"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a>col <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">&quot;#BB4444&quot;</span>, <span class="st">&quot;#EE9988&quot;</span>, <span class="st">&quot;#FFFFFF&quot;</span>, <span class="st">&quot;#77AADD&quot;</span>, <span class="st">&quot;#4477AA&quot;</span>))</span>
<span id="cb131-2"><a href="#cb131-2" aria-hidden="true" tabindex="-1"></a><span class="fu">corrplot</span>(C, <span class="at">method=</span><span class="st">&quot;color&quot;</span>, <span class="at">col=</span><span class="fu">col</span>(<span class="dv">200</span>),  </span>
<span id="cb131-3"><a href="#cb131-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">type=</span><span class="st">&quot;upper&quot;</span>, <span class="at">order=</span><span class="st">&quot;hclust&quot;</span>, </span>
<span id="cb131-4"><a href="#cb131-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">addCoef.col =</span> <span class="st">&quot;black&quot;</span>, <span class="co"># Add coefficient of correlation</span></span>
<span id="cb131-5"><a href="#cb131-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">tl.col=</span><span class="st">&quot;black&quot;</span>, <span class="at">tl.srt=</span><span class="dv">45</span>, <span class="co">#Text label color and rotation</span></span>
<span id="cb131-6"><a href="#cb131-6" aria-hidden="true" tabindex="-1"></a>         <span class="co"># Combine with significance</span></span>
<span id="cb131-7"><a href="#cb131-7" aria-hidden="true" tabindex="-1"></a>        <span class="co"># p.mat = p.mat, sig.level = 0.1, insig = &quot;blank&quot;, </span></span>
<span id="cb131-8"><a href="#cb131-8" aria-hidden="true" tabindex="-1"></a>         <span class="co"># hide correlation coefficient on the principal diagonal</span></span>
<span id="cb131-9"><a href="#cb131-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">diag=</span><span class="cn">FALSE</span> </span>
<span id="cb131-10"><a href="#cb131-10" aria-hidden="true" tabindex="-1"></a>         )</span></code></pre></div>
<pre><code>## Error in corrplot(C, method = &quot;color&quot;, col = col(200), type = &quot;upper&quot;, : Need a matrix or data frame!</code></pre>
<h1 id="save-r-object">Save R object</h1>
<div class="sourceCode" id="cb133"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(dge.filtered.norm, <span class="at">file =</span> <span class="fu">paste0</span>(<span class="st">&quot;../../rObjects/&quot;</span>, condition, <span class="st">&quot;_dge.filtered.norm&quot;</span>, sex_karyotype, <span class="st">&quot;.rds&quot;</span>))</span>
<span id="cb133-2"><a href="#cb133-2" aria-hidden="true" tabindex="-1"></a><span class="co">#dge.filtered.norm &lt;- readRDS( file = paste0(&quot;../../rObjects/&quot;, condition, &quot;_dge.filtered.norm&quot;, sex_karyotype, &quot;.rds&quot;))</span></span></code></pre></div>
<h1 id="variance-partition">Variance partition</h1>
<div class="sourceCode" id="cb134"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a><span class="fu">register</span>(<span class="fu">SnowParam</span>(<span class="fu">detectCores</span>())) <span class="co"># work in parallel (takes a while to run)</span></span>
<span id="cb134-2"><a href="#cb134-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb134-3"><a href="#cb134-3" aria-hidden="true" tabindex="-1"></a><span class="co"># geneExpr: matrix of gene expression values</span></span>
<span id="cb134-4"><a href="#cb134-4" aria-hidden="true" tabindex="-1"></a><span class="co"># info: information/metadata about each sample</span></span>
<span id="cb134-5"><a href="#cb134-5" aria-hidden="true" tabindex="-1"></a>geneExpr <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(dge.filtered.norm<span class="sc">$</span>counts)</span>
<span id="cb134-6"><a href="#cb134-6" aria-hidden="true" tabindex="-1"></a>info <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(dge.filtered.norm<span class="sc">$</span>samples)</span>
<span id="cb134-7"><a href="#cb134-7" aria-hidden="true" tabindex="-1"></a>info<span class="sc">$</span>RIN <span class="ot">&lt;-</span> <span class="fu">as.integer</span>(info<span class="sc">$</span>RIN)</span>
<span id="cb134-8"><a href="#cb134-8" aria-hidden="true" tabindex="-1"></a>info<span class="sc">$</span>TYPE <span class="ot">&lt;-</span> <span class="fu">as.character</span>(info<span class="sc">$</span>TYPE)</span>
<span id="cb134-9"><a href="#cb134-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb134-10"><a href="#cb134-10" aria-hidden="true" tabindex="-1"></a><span class="co"># age is usually a continuous so model it as a fixed effect &quot;age&quot;</span></span>
<span id="cb134-11"><a href="#cb134-11" aria-hidden="true" tabindex="-1"></a><span class="co"># group is categorical, so model them as random effects &quot;(1|group)&quot;</span></span>
<span id="cb134-12"><a href="#cb134-12" aria-hidden="true" tabindex="-1"></a><span class="co"># note the syntax</span></span>
<span id="cb134-13"><a href="#cb134-13" aria-hidden="true" tabindex="-1"></a>form <span class="ot">&lt;-</span> <span class="er">~</span>Age <span class="sc">+</span> PMI <span class="sc">+</span> RIN <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>TYPE)<span class="co"># + (1 | CWOW.Category)</span></span>
<span id="cb134-14"><a href="#cb134-14" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb134-15"><a href="#cb134-15" aria-hidden="true" tabindex="-1"></a><span class="co">#(1 | LBD.type) + </span></span>
<span id="cb134-16"><a href="#cb134-16" aria-hidden="true" tabindex="-1"></a><span class="co">#(1 | TDP.type)  + </span></span>
<span id="cb134-17"><a href="#cb134-17" aria-hidden="true" tabindex="-1"></a><span class="co">#(1 | ClinicalDx) + </span></span>
<span id="cb134-18"><a href="#cb134-18" aria-hidden="true" tabindex="-1"></a><span class="co">#(1 | FHx) + </span></span>
<span id="cb134-19"><a href="#cb134-19" aria-hidden="true" tabindex="-1"></a><span class="co">#(1 | Sex) + </span></span>
<span id="cb134-20"><a href="#cb134-20" aria-hidden="true" tabindex="-1"></a><span class="co">#(1 | Race) + </span></span>
<span id="cb134-21"><a href="#cb134-21" aria-hidden="true" tabindex="-1"></a><span class="co">#(1 | TMEM106b) +</span></span>
<span id="cb134-22"><a href="#cb134-22" aria-hidden="true" tabindex="-1"></a><span class="co">#Braak.NFT + </span></span>
<span id="cb134-23"><a href="#cb134-23" aria-hidden="true" tabindex="-1"></a><span class="co">#Thal.amyloid + </span></span>
<span id="cb134-24"><a href="#cb134-24" aria-hidden="true" tabindex="-1"></a><span class="co">#MF.SP  + </span></span>
<span id="cb134-25"><a href="#cb134-25" aria-hidden="true" tabindex="-1"></a><span class="co">#MF.LB + </span></span>
<span id="cb134-26"><a href="#cb134-26" aria-hidden="true" tabindex="-1"></a><span class="co">#Cing.LB + </span></span>
<span id="cb134-27"><a href="#cb134-27" aria-hidden="true" tabindex="-1"></a><span class="co">#MF.Amyloid + </span></span>
<span id="cb134-28"><a href="#cb134-28" aria-hidden="true" tabindex="-1"></a><span class="co">#MF.Tau + </span></span>
<span id="cb134-29"><a href="#cb134-29" aria-hidden="true" tabindex="-1"></a><span class="co">#Cing.Synuclein + </span></span>
<span id="cb134-30"><a href="#cb134-30" aria-hidden="true" tabindex="-1"></a><span class="co">#VaD + </span></span>
<span id="cb134-31"><a href="#cb134-31" aria-hidden="true" tabindex="-1"></a><span class="co">#TDP.43 + </span></span>
<span id="cb134-32"><a href="#cb134-32" aria-hidden="true" tabindex="-1"></a><span class="co">#Brain.wt + </span></span>
<span id="cb134-33"><a href="#cb134-33" aria-hidden="true" tabindex="-1"></a><span class="co">#Duration + </span></span>
<span id="cb134-34"><a href="#cb134-34" aria-hidden="true" tabindex="-1"></a><span class="co">#Age + </span></span>
<span id="cb134-35"><a href="#cb134-35" aria-hidden="true" tabindex="-1"></a><span class="co">#PMI + </span></span>
<span id="cb134-36"><a href="#cb134-36" aria-hidden="true" tabindex="-1"></a><span class="co">#RIN</span></span>
<span id="cb134-37"><a href="#cb134-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb134-38"><a href="#cb134-38" aria-hidden="true" tabindex="-1"></a>varPart <span class="ot">&lt;-</span> <span class="fu">fitExtractVarPartModel</span>(geneExpr, form, info)</span></code></pre></div>
<pre><code>## Error in fitExtractVarPartModel(geneExpr, form, info): could not find function &quot;fitExtractVarPartModel&quot;</code></pre>
<div class="sourceCode" id="cb136"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a>vp <span class="ot">&lt;-</span> <span class="fu">sortCols</span>(varPart)</span></code></pre></div>
<pre><code>## Error in sortCols(varPart): could not find function &quot;sortCols&quot;</code></pre>
<div class="sourceCode" id="cb138"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot</span></span>
<span id="cb138-2"><a href="#cb138-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotVarPart</span>(vp)</span></code></pre></div>
<pre><code>## Error in plotVarPart(vp): could not find function &quot;plotVarPart&quot;</code></pre>
<div class="sourceCode" id="cb140"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb140-1"><a href="#cb140-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save</span></span>
<span id="cb140-2"><a href="#cb140-2" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;../../results/&quot;</span>, tool, <span class="st">&quot;/varpart/&quot;</span>, condition, <span class="st">&quot;_gene_varpart_violins_&quot;</span>, sex_karyotype)</span>
<span id="cb140-3"><a href="#cb140-3" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPDF</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.pdf&quot;</span>), <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">4</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb142"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPNG</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.png&quot;</span>), <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">4</span>, <span class="at">unit =</span> <span class="st">&quot;in&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<h1 id="add-gene-name-to-gene_id">Add gene name to gene_id</h1>
<div class="sourceCode" id="cb144"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb144-1"><a href="#cb144-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot</span></span>
<span id="cb144-2"><a href="#cb144-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPercentBars</span>(vp[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,])</span></code></pre></div>
<pre><code>## Error in plotPercentBars(vp[1:10, ]): could not find function &quot;plotPercentBars&quot;</code></pre>
<div class="sourceCode" id="cb146"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make a column called gene_id to later merge with genes.gtf</span></span>
<span id="cb146-2"><a href="#cb146-2" aria-hidden="true" tabindex="-1"></a>vp<span class="sc">$</span>gene_id <span class="ot">&lt;-</span> <span class="fu">row.names</span>(vp)</span></code></pre></div>
<pre><code>## Error in row.names(vp): object &#39;vp&#39; not found</code></pre>
<div class="sourceCode" id="cb148"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a><span class="co"># dataframe of gene_id and gene_name </span></span>
<span id="cb148-2"><a href="#cb148-2" aria-hidden="true" tabindex="-1"></a>gene_id_name <span class="ot">&lt;-</span> <span class="fu">cbind</span>(genes.gtf<span class="sc">$</span>gene_id, genes.gtf<span class="sc">$</span>gene_name)</span>
<span id="cb148-3"><a href="#cb148-3" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(gene_id_name)</span>
<span id="cb148-4"><a href="#cb148-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(df)[<span class="fu">names</span>(df) <span class="sc">==</span> <span class="st">&#39;V1&#39;</span>] <span class="ot">&lt;-</span> <span class="st">&#39;gene_id&#39;</span></span>
<span id="cb148-5"><a href="#cb148-5" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(df)[<span class="fu">names</span>(df) <span class="sc">==</span> <span class="st">&#39;V2&#39;</span>] <span class="ot">&lt;-</span> <span class="st">&#39;gene_name&#39;</span></span>
<span id="cb148-6"><a href="#cb148-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb148-7"><a href="#cb148-7" aria-hidden="true" tabindex="-1"></a>vp_df <span class="ot">&lt;-</span> <span class="fu">merge</span>(vp, df)</span></code></pre></div>
<pre><code>## Error in merge(vp, df): object &#39;vp&#39; not found</code></pre>
<div class="sourceCode" id="cb150"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb150-1"><a href="#cb150-1" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(vp_df) <span class="ot">&lt;-</span> <span class="fu">make.names</span>(vp_df<span class="sc">$</span>gene_name, <span class="at">unique =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## Error in make.names(vp_df$gene_name, unique = TRUE): object &#39;vp_df&#39; not found</code></pre>
<div class="sourceCode" id="cb152"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb152-1"><a href="#cb152-1" aria-hidden="true" tabindex="-1"></a>vp_df<span class="sc">$</span>gene_id <span class="ot">&lt;-</span> <span class="cn">NULL</span></span></code></pre></div>
<pre><code>## Error in vp_df$gene_id &lt;- NULL: object &#39;vp_df&#39; not found</code></pre>
<div class="sourceCode" id="cb154"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb154-1"><a href="#cb154-1" aria-hidden="true" tabindex="-1"></a>vp_df<span class="sc">$</span>gene_name <span class="ot">&lt;-</span> <span class="cn">NULL</span></span></code></pre></div>
<pre><code>## Error in vp_df$gene_name &lt;- NULL: object &#39;vp_df&#39; not found</code></pre>
<div class="sourceCode" id="cb156"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb156-1"><a href="#cb156-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPercentBars</span>(vp_df[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,])</span></code></pre></div>
<pre><code>## Error in plotPercentBars(vp_df[1:10, ]): could not find function &quot;plotPercentBars&quot;</code></pre>
<div class="sourceCode" id="cb158"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb158-1"><a href="#cb158-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save</span></span>
<span id="cb158-2"><a href="#cb158-2" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;../../results/&quot;</span>, tool, <span class="st">&quot;/varpart/&quot;</span>, condition, <span class="st">&quot;_gene_percent_bars&quot;</span>, sex_karyotype)</span>
<span id="cb158-3"><a href="#cb158-3" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPDF</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.pdf&quot;</span>), <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">6</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb160"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb160-1"><a href="#cb160-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPNG</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.png&quot;</span>), <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">6</span>, <span class="at">unit =</span> <span class="st">&quot;in&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<h1 id="top-variable-genes">Top variable genes</h1>
<div class="sourceCode" id="cb162"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb162-1"><a href="#cb162-1" aria-hidden="true" tabindex="-1"></a>varPart.df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(vp_df)</span>
<span id="cb162-2"><a href="#cb162-2" aria-hidden="true" tabindex="-1"></a><span class="co"># sort genes based on variance explained by group</span></span>
<span id="cb162-3"><a href="#cb162-3" aria-hidden="true" tabindex="-1"></a>order.varPart.df <span class="ot">&lt;-</span> varPart.df[<span class="fu">order</span>(varPart.df<span class="sc">$</span>RIN, <span class="at">decreasing =</span> <span class="cn">TRUE</span>),]</span>
<span id="cb162-4"><a href="#cb162-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(order.varPart.df[<span class="st">&quot;RIN&quot;</span>], <span class="dv">10</span>)</span>
<span id="cb162-5"><a href="#cb162-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb162-6"><a href="#cb162-6" aria-hidden="true" tabindex="-1"></a><span class="co"># sort genes based on variance explained by Race</span></span>
<span id="cb162-7"><a href="#cb162-7" aria-hidden="true" tabindex="-1"></a>order.varPart.df <span class="ot">&lt;-</span> varPart.df[<span class="fu">order</span>(varPart.df<span class="sc">$</span>Race, <span class="at">decreasing =</span> <span class="cn">TRUE</span>),]</span>
<span id="cb162-8"><a href="#cb162-8" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(order.varPart.df[<span class="st">&quot;Race&quot;</span>], <span class="dv">10</span>)</span></code></pre></div>
<h1 id="design-matrix">Design matrix</h1>
<div class="sourceCode" id="cb163"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb163-1"><a href="#cb163-1" aria-hidden="true" tabindex="-1"></a>age <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(dge.filtered.norm<span class="sc">$</span>samples<span class="sc">$</span>Age)</span>
<span id="cb163-2"><a href="#cb163-2" aria-hidden="true" tabindex="-1"></a>RIN <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(dge.filtered.norm<span class="sc">$</span>samples<span class="sc">$</span>RIN)</span>
<span id="cb163-3"><a href="#cb163-3" aria-hidden="true" tabindex="-1"></a>race <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(dge.filtered.norm<span class="sc">$</span>samples<span class="sc">$</span>Race)</span>
<span id="cb163-4"><a href="#cb163-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb163-5"><a href="#cb163-5" aria-hidden="true" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">interaction</span>(dge.filtered.norm<span class="sc">$</span>samples<span class="sc">$</span>TYPE)</span>
<span id="cb163-6"><a href="#cb163-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb163-7"><a href="#cb163-7" aria-hidden="true" tabindex="-1"></a>design <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> <span class="dv">0</span> <span class="sc">+</span> group)</span>
<span id="cb163-8"><a href="#cb163-8" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(design) <span class="ot">&lt;-</span> <span class="fu">c</span>(control,condition)</span></code></pre></div>
<pre><code>## Error in dimnames(x) &lt;- dn: length of &#39;dimnames&#39; [2] not equal to array extent</code></pre>
<div class="sourceCode" id="cb165"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb165-1"><a href="#cb165-1" aria-hidden="true" tabindex="-1"></a>design</span></code></pre></div>
<pre><code>##    groupCONTROL groupCONTROL - AD groupCONTROL - PA groupLBD
## 1             0                 0                 0        1
## 2             1                 0                 0        0
## 3             1                 0                 0        0
## 4             0                 0                 0        1
## 5             0                 0                 0        1
## 6             0                 0                 0        1
## 7             1                 0                 0        0
## 8             0                 0                 0        1
## 9             0                 0                 0        1
## 10            0                 0                 0        1
## 11            0                 0                 0        1
## 12            0                 0                 0        1
## 13            0                 0                 0        1
## 14            0                 0                 0        1
## 15            0                 0                 0        1
## 16            0                 0                 0        1
## 17            0                 0                 0        1
## 18            0                 0                 0        1
## 19            0                 0                 0        1
## 20            0                 0                 0        1
## 21            1                 0                 0        0
## 22            0                 0                 0        1
## 23            0                 0                 0        1
## attr(,&quot;assign&quot;)
## [1] 1 1 1 1
## attr(,&quot;contrasts&quot;)
## attr(,&quot;contrasts&quot;)$group
## [1] &quot;contr.treatment&quot;</code></pre>
<h1 id="voom">Voom</h1>
<div class="sourceCode" id="cb167"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb167-1"><a href="#cb167-1" aria-hidden="true" tabindex="-1"></a><span class="co"># voom transform counts</span></span>
<span id="cb167-2"><a href="#cb167-2" aria-hidden="true" tabindex="-1"></a>v <span class="ot">&lt;-</span> <span class="fu">voomWithQualityWeights</span>(dge.filtered.norm,</span>
<span id="cb167-3"><a href="#cb167-3" aria-hidden="true" tabindex="-1"></a>                            design,</span>
<span id="cb167-4"><a href="#cb167-4" aria-hidden="true" tabindex="-1"></a>                            <span class="at">plot =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## Coefficients not estimable: groupCONTROL - AD groupCONTROL - PA</code></pre>
<pre><code>## Warning: Partial NA coefficients for 18389 probe(s)</code></pre>
<pre><code>## Coefficients not estimable: groupCONTROL - AD groupCONTROL - PA</code></pre>
<pre><code>## Warning: Partial NA coefficients for 18389 probe(s)</code></pre>
<figure>
<img src="figure/voom-1.png" alt="plot of chunk voom" />
<figcaption aria-hidden="true">plot of chunk voom</figcaption>
</figure>
<div class="sourceCode" id="cb172"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb172-1"><a href="#cb172-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save</span></span>
<span id="cb172-2"><a href="#cb172-2" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;../../results/&quot;</span>, tool, <span class="st">&quot;/voom/&quot;</span>, condition, <span class="st">&quot;_gene_mean_var_weights_&quot;</span>, sex_karyotype)</span>
<span id="cb172-3"><a href="#cb172-3" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPDF</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.pdf&quot;</span>), <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">4</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb174"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb174-1"><a href="#cb174-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPNG</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.png&quot;</span>), <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">4</span>, <span class="at">unit =</span> <span class="st">&quot;in&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb176"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb176-1"><a href="#cb176-1" aria-hidden="true" tabindex="-1"></a><span class="co"># fits linear model for each gene given a series of arrays</span></span>
<span id="cb176-2"><a href="#cb176-2" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">lmFit</span>(v, design)</span></code></pre></div>
<pre><code>## Coefficients not estimable: groupCONTROL - AD groupCONTROL - PA</code></pre>
<pre><code>## Warning: Partial NA coefficients for 18389 probe(s)</code></pre>
<div class="sourceCode" id="cb179"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb179-1"><a href="#cb179-1" aria-hidden="true" tabindex="-1"></a><span class="co"># contrast design for differential expression</span></span>
<span id="cb179-2"><a href="#cb179-2" aria-hidden="true" tabindex="-1"></a>contrasts <span class="ot">&lt;-</span> <span class="fu">makeContrasts</span>(</span>
<span id="cb179-3"><a href="#cb179-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> myContrasts,  <span class="co"># myContrasts was user input from beginning</span></span>
<span id="cb179-4"><a href="#cb179-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> <span class="fu">colnames</span>(design))</span></code></pre></div>
<pre><code>## Error in makeContrasts(title = myContrasts, levels = colnames(design)): The levels must by syntactically valid names in R, see help(make.names).  Non-valid names: groupCONTROL - AD,groupCONTROL - PA</code></pre>
<div class="sourceCode" id="cb181"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb181-1"><a href="#cb181-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(contrasts)</span></code></pre></div>
<pre><code>##                                                  
## 1 function (x, contrasts = TRUE, sparse = FALSE) 
## 2 {                                              
## 3     if (is.logical(x))                         
## 4         x &lt;- factor(x, levels = c(FALSE, TRUE))
## 5     if (!is.factor(x))                         
## 6         stop(&quot;contrasts apply only to factors&quot;)</code></pre>
<div class="sourceCode" id="cb183"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb183-1"><a href="#cb183-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save contrast names</span></span>
<span id="cb183-2"><a href="#cb183-2" aria-hidden="true" tabindex="-1"></a>allComparisons <span class="ot">&lt;-</span> <span class="fu">colnames</span>(contrasts)</span>
<span id="cb183-3"><a href="#cb183-3" aria-hidden="true" tabindex="-1"></a>allComparisons <span class="co"># check</span></span></code></pre></div>
<pre><code>## NULL</code></pre>
<div class="sourceCode" id="cb185"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb185-1"><a href="#cb185-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run contrast analysis</span></span>
<span id="cb185-2"><a href="#cb185-2" aria-hidden="true" tabindex="-1"></a>vfit <span class="ot">&lt;-</span> <span class="fu">contrasts.fit</span>(fit, <span class="at">contrasts =</span> contrasts)</span></code></pre></div>
<pre><code>## Error in as.vector(x, mode): cannot coerce type &#39;closure&#39; to vector of type &#39;any&#39;</code></pre>
<div class="sourceCode" id="cb187"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb187-1"><a href="#cb187-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute differential expression based on the empirical Bayes moderation of the</span></span>
<span id="cb187-2"><a href="#cb187-2" aria-hidden="true" tabindex="-1"></a><span class="co"># standard errors towards a common value.</span></span>
<span id="cb187-3"><a href="#cb187-3" aria-hidden="true" tabindex="-1"></a>veBayesFit <span class="ot">&lt;-</span> <span class="fu">eBayes</span>(vfit)</span></code></pre></div>
<pre><code>## Error in .ebayes(fit = fit, proportion = proportion, stdev.coef.lim = stdev.coef.lim, : object &#39;vfit&#39; not found</code></pre>
<div class="sourceCode" id="cb189"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb189-1"><a href="#cb189-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotSA</span>(veBayesFit, <span class="at">main =</span> <span class="st">&quot;Final Model: Mean-variance Trend&quot;</span>)</span></code></pre></div>
<pre><code>## Error in is(fit, &quot;MArrayLM&quot;): object &#39;veBayesFit&#39; not found</code></pre>
<div class="sourceCode" id="cb191"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb191-1"><a href="#cb191-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save</span></span>
<span id="cb191-2"><a href="#cb191-2" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;../../results/&quot;</span>, tool, <span class="st">&quot;/voom/&quot;</span>, condition, <span class="st">&quot;_gene_final_mean_var_&quot;</span>, sex_karyotype)</span>
<span id="cb191-3"><a href="#cb191-3" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPDF</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.pdf&quot;</span>), <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">4</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb193"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb193-1"><a href="#cb193-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPNG</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.png&quot;</span>), <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">4</span>, <span class="at">unit =</span> <span class="st">&quot;in&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<h1 id="voom-mds-plot">Voom MDS Plot</h1>
<div class="sourceCode" id="cb195"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb195-1"><a href="#cb195-1" aria-hidden="true" tabindex="-1"></a>group_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>)[v<span class="sc">$</span>targets<span class="sc">$</span>TYPE]</span>
<span id="cb195-2"><a href="#cb195-2" aria-hidden="true" tabindex="-1"></a>names <span class="ot">&lt;-</span> v<span class="sc">$</span>targets<span class="sc">$</span>NPID</span>
<span id="cb195-3"><a href="#cb195-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb195-4"><a href="#cb195-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMDS</span>(</span>
<span id="cb195-5"><a href="#cb195-5" aria-hidden="true" tabindex="-1"></a>  v, </span>
<span id="cb195-6"><a href="#cb195-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">top =</span> <span class="dv">100</span>, </span>
<span id="cb195-7"><a href="#cb195-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">labels =</span> names,</span>
<span id="cb195-8"><a href="#cb195-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">cex =</span> <span class="dv">1</span>, </span>
<span id="cb195-9"><a href="#cb195-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">dim.plot =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), </span>
<span id="cb195-10"><a href="#cb195-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot =</span> <span class="cn">TRUE</span>, </span>
<span id="cb195-11"><a href="#cb195-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> group_colors</span>
<span id="cb195-12"><a href="#cb195-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb195-13"><a href="#cb195-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb195-14"><a href="#cb195-14" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span>(<span class="fu">expression</span>(<span class="st">&#39;Top 100 Genes - Voom (Log&#39;</span>[<span class="dv">2</span>]<span class="sc">~</span><span class="st">&#39;CPM)&#39;</span>))</span></code></pre></div>
<figure>
<img src="figure/MDS_voom-1.png" alt="plot of chunk MDS_voom" />
<figcaption aria-hidden="true">plot of chunk MDS_voom</figcaption>
</figure>
<div class="sourceCode" id="cb196"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb196-1"><a href="#cb196-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save</span></span>
<span id="cb196-2"><a href="#cb196-2" aria-hidden="true" tabindex="-1"></a><span class="co"># save</span></span>
<span id="cb196-3"><a href="#cb196-3" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;../../results/&quot;</span>, tool, <span class="st">&quot;/MDS/&quot;</span>, condition, <span class="st">&quot;_gene_MDS_normalized__dim1&amp;2_&quot;</span>, sex_karyotype)</span>
<span id="cb196-4"><a href="#cb196-4" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPDF</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.pdf&quot;</span>), <span class="at">width =</span> <span class="dv">5</span>, <span class="at">height =</span> <span class="dv">5</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb198"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb198-1"><a href="#cb198-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPNG</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.png&quot;</span>), <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="dv">4</span>, <span class="at">unit =</span> <span class="st">&quot;in&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb200"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb200-1"><a href="#cb200-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMDS</span>(</span>
<span id="cb200-2"><a href="#cb200-2" aria-hidden="true" tabindex="-1"></a>  v, </span>
<span id="cb200-3"><a href="#cb200-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">top =</span> <span class="dv">100</span>, </span>
<span id="cb200-4"><a href="#cb200-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">labels =</span> names,</span>
<span id="cb200-5"><a href="#cb200-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">cex =</span> <span class="dv">1</span>, </span>
<span id="cb200-6"><a href="#cb200-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">dim.plot =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">3</span>), </span>
<span id="cb200-7"><a href="#cb200-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot =</span> <span class="cn">TRUE</span>, </span>
<span id="cb200-8"><a href="#cb200-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> group_colors</span>
<span id="cb200-9"><a href="#cb200-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb200-10"><a href="#cb200-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb200-11"><a href="#cb200-11" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span>(<span class="fu">expression</span>(<span class="st">&#39;Top 100 Genes - Voom (Log&#39;</span>[<span class="dv">2</span>]<span class="sc">~</span><span class="st">&#39;CPM)&#39;</span>))</span></code></pre></div>
<figure>
<img src="figure/MDS_voom-2.png" alt="plot of chunk MDS_voom" />
<figcaption aria-hidden="true">plot of chunk MDS_voom</figcaption>
</figure>
<div class="sourceCode" id="cb201"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb201-1"><a href="#cb201-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save</span></span>
<span id="cb201-2"><a href="#cb201-2" aria-hidden="true" tabindex="-1"></a><span class="co"># save</span></span>
<span id="cb201-3"><a href="#cb201-3" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;../../results/&quot;</span>, tool, <span class="st">&quot;/MDS/&quot;</span>, condition, <span class="st">&quot;_gene_MDS_normalized_dim2&amp;3_&quot;</span>, sex_karyotype)</span>
<span id="cb201-4"><a href="#cb201-4" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPDF</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.pdf&quot;</span>), <span class="at">width =</span> <span class="dv">5</span>, <span class="at">height =</span> <span class="dv">5</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb203"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb203-1"><a href="#cb203-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPNG</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.png&quot;</span>), <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="dv">4</span>, <span class="at">unit =</span> <span class="st">&quot;in&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb205"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb205-1"><a href="#cb205-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(v, <span class="at">file =</span> <span class="fu">paste0</span>(<span class="st">&quot;../../rObjects/&quot;</span>, condition, <span class="st">&quot;_gene_voom&quot;</span>, sex_karyotype, <span class="st">&quot;.rds&quot;</span>))</span></code></pre></div>
<h1 id="number-of-degs">Number of DEGs</h1>
<p>Identify number of differential expressed genes.</p>
<div class="sourceCode" id="cb206"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb206-1"><a href="#cb206-1" aria-hidden="true" tabindex="-1"></a>pval <span class="ot">&lt;-</span> <span class="fl">0.05</span></span>
<span id="cb206-2"><a href="#cb206-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-3"><a href="#cb206-3" aria-hidden="true" tabindex="-1"></a>sumTable <span class="ot">&lt;-</span> </span>
<span id="cb206-4"><a href="#cb206-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summary</span>(<span class="fu">decideTests</span>(</span>
<span id="cb206-5"><a href="#cb206-5" aria-hidden="true" tabindex="-1"></a>    vfit,  <span class="co"># object</span></span>
<span id="cb206-6"><a href="#cb206-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">adjust.method =</span> <span class="st">&quot;BH&quot;</span>, <span class="co"># by default the method = &quot;separate&quot;</span></span>
<span id="cb206-7"><a href="#cb206-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">p.value =</span> pval,</span>
<span id="cb206-8"><a href="#cb206-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">lfc =</span> <span class="dv">0</span>  <span class="co"># numeric, minimum absolute log2-fold change required</span></span>
<span id="cb206-9"><a href="#cb206-9" aria-hidden="true" tabindex="-1"></a>  ))</span></code></pre></div>
<pre><code>## Error in decideTests(vfit, adjust.method = &quot;BH&quot;, p.value = pval, lfc = 0): object &#39;vfit&#39; not found</code></pre>
<div class="sourceCode" id="cb208"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb208-1"><a href="#cb208-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">&quot; FDRq &lt; &quot;</span>, pval))</span></code></pre></div>
<pre><code>## [1] &quot; FDRq &lt; 0.05&quot;</code></pre>
<div class="sourceCode" id="cb210"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb210-1"><a href="#cb210-1" aria-hidden="true" tabindex="-1"></a>sumTable</span></code></pre></div>
<pre><code>## Error in eval(expr, envir, enclos): object &#39;sumTable&#39; not found</code></pre>
<h1 id="output-deg-tables">Output DEG tables</h1>
<div class="sourceCode" id="cb212"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb212-1"><a href="#cb212-1" aria-hidden="true" tabindex="-1"></a>coef <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb212-2"><a href="#cb212-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb212-3"><a href="#cb212-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> allComparisons) {</span>
<span id="cb212-4"><a href="#cb212-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># p &lt; 1, log2fc &gt; 0 </span></span>
<span id="cb212-5"><a href="#cb212-5" aria-hidden="true" tabindex="-1"></a>  vTopTableAll <span class="ot">&lt;-</span></span>
<span id="cb212-6"><a href="#cb212-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">topTable</span>(</span>
<span id="cb212-7"><a href="#cb212-7" aria-hidden="true" tabindex="-1"></a>      veBayesFit, </span>
<span id="cb212-8"><a href="#cb212-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">coef =</span> coef,  </span>
<span id="cb212-9"><a href="#cb212-9" aria-hidden="true" tabindex="-1"></a>      <span class="at">n =</span> <span class="cn">Inf</span>, </span>
<span id="cb212-10"><a href="#cb212-10" aria-hidden="true" tabindex="-1"></a>      <span class="at">p.value =</span> <span class="dv">1</span>,</span>
<span id="cb212-11"><a href="#cb212-11" aria-hidden="true" tabindex="-1"></a>      <span class="at">lfc =</span> <span class="dv">0</span> </span>
<span id="cb212-12"><a href="#cb212-12" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb212-13"><a href="#cb212-13" aria-hidden="true" tabindex="-1"></a>  path <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&quot;../../results/&quot;</span>, tool, <span class="st">&quot;/DEGs/&quot;</span>, condition, <span class="st">&quot;_gene_DEGs_FDRq1.00_&quot;</span>, sex_karyotype, <span class="st">&quot;.txt&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>) </span>
<span id="cb212-14"><a href="#cb212-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write.table</span>(</span>
<span id="cb212-15"><a href="#cb212-15" aria-hidden="true" tabindex="-1"></a>    vTopTableAll,</span>
<span id="cb212-16"><a href="#cb212-16" aria-hidden="true" tabindex="-1"></a>    path,</span>
<span id="cb212-17"><a href="#cb212-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>,</span>
<span id="cb212-18"><a href="#cb212-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">row.names =</span> <span class="cn">FALSE</span>,</span>
<span id="cb212-19"><a href="#cb212-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">quote =</span> <span class="cn">FALSE</span></span>
<span id="cb212-20"><a href="#cb212-20" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb212-21"><a href="#cb212-21" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb212-22"><a href="#cb212-22" aria-hidden="true" tabindex="-1"></a>  <span class="co"># p &lt; 0.05, log2fc &gt; 0</span></span>
<span id="cb212-23"><a href="#cb212-23" aria-hidden="true" tabindex="-1"></a>  vTopTable1 <span class="ot">&lt;-</span></span>
<span id="cb212-24"><a href="#cb212-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">topTable</span>( </span>
<span id="cb212-25"><a href="#cb212-25" aria-hidden="true" tabindex="-1"></a>      veBayesFit,  </span>
<span id="cb212-26"><a href="#cb212-26" aria-hidden="true" tabindex="-1"></a>      <span class="at">coef =</span> coef,  </span>
<span id="cb212-27"><a href="#cb212-27" aria-hidden="true" tabindex="-1"></a>      <span class="at">n =</span> <span class="cn">Inf</span>, </span>
<span id="cb212-28"><a href="#cb212-28" aria-hidden="true" tabindex="-1"></a>      <span class="at">p.value =</span> <span class="fl">0.05</span>,</span>
<span id="cb212-29"><a href="#cb212-29" aria-hidden="true" tabindex="-1"></a>      <span class="at">lfc =</span> <span class="dv">0</span></span>
<span id="cb212-30"><a href="#cb212-30" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb212-31"><a href="#cb212-31" aria-hidden="true" tabindex="-1"></a>  path <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&quot;../../results/&quot;</span>, tool, <span class="st">&quot;/DEGs/&quot;</span>, condition, <span class="st">&quot;_gene_DEGs_FDRq0.05_&quot;</span>, sex_karyotype, <span class="st">&quot;.txt&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>) </span>
<span id="cb212-32"><a href="#cb212-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write.table</span>(</span>
<span id="cb212-33"><a href="#cb212-33" aria-hidden="true" tabindex="-1"></a>    vTopTable1,</span>
<span id="cb212-34"><a href="#cb212-34" aria-hidden="true" tabindex="-1"></a>    path,</span>
<span id="cb212-35"><a href="#cb212-35" aria-hidden="true" tabindex="-1"></a>    <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>,</span>
<span id="cb212-36"><a href="#cb212-36" aria-hidden="true" tabindex="-1"></a>    <span class="at">row.names =</span> <span class="cn">FALSE</span>,</span>
<span id="cb212-37"><a href="#cb212-37" aria-hidden="true" tabindex="-1"></a>    <span class="at">quote =</span> <span class="cn">FALSE</span></span>
<span id="cb212-38"><a href="#cb212-38" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb212-39"><a href="#cb212-39" aria-hidden="true" tabindex="-1"></a>  <span class="co"># increment -----------------------------------------------------------------</span></span>
<span id="cb212-40"><a href="#cb212-40" aria-hidden="true" tabindex="-1"></a>  coef <span class="ot">&lt;-</span> coef <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb212-41"><a href="#cb212-41" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>Read and save table with all genes (FDRq = 1).</p>
<div class="sourceCode" id="cb213"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb213-1"><a href="#cb213-1" aria-hidden="true" tabindex="-1"></a>condition_vs_control <span class="ot">&lt;-</span> <span class="fu">read.table</span>(</span>
<span id="cb213-2"><a href="#cb213-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste0</span>(<span class="st">&quot;../../results/&quot;</span>, tool, <span class="st">&quot;/DEGs/&quot;</span>, condition, <span class="st">&quot;_gene_DEGs_FDRq1.00_&quot;</span>, sex_karyotype, <span class="st">&quot;.txt&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>),</span>
<span id="cb213-3"><a href="#cb213-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">header =</span> <span class="cn">TRUE</span>,</span>
<span id="cb213-4"><a href="#cb213-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>,</span>
<span id="cb213-5"><a href="#cb213-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb213-6"><a href="#cb213-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb213-7"><a href="#cb213-7" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(condition_vs_control, <span class="at">file =</span> <span class="fu">paste0</span>(<span class="st">&quot;../../rObjects/&quot;</span>, condition, <span class="st">&quot;_gene_table_&quot;</span>, sex_karyotype, <span class="st">&quot;.rds&quot;</span>))</span></code></pre></div>
<h1 id="assign-colors">Assign colors</h1>
<p>Assign colors values based on FDRq cutoff of 0.2.</p>
<div class="sourceCode" id="cb214"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb214-1"><a href="#cb214-1" aria-hidden="true" tabindex="-1"></a>color_values <span class="ot">&lt;-</span> <span class="fu">vector</span>()</span>
<span id="cb214-2"><a href="#cb214-2" aria-hidden="true" tabindex="-1"></a>max <span class="ot">&lt;-</span> <span class="fu">nrow</span>(condition_vs_control)</span>
<span id="cb214-3"><a href="#cb214-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb214-4"><a href="#cb214-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>max){</span>
<span id="cb214-5"><a href="#cb214-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (condition_vs_control<span class="sc">$</span>adj.P.Val[i] <span class="sc">&lt;</span> <span class="fl">0.2</span>){</span>
<span id="cb214-6"><a href="#cb214-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (condition_vs_control<span class="sc">$</span>logFC[i] <span class="sc">&gt;</span> <span class="dv">0</span>){</span>
<span id="cb214-7"><a href="#cb214-7" aria-hidden="true" tabindex="-1"></a>      color_values <span class="ot">&lt;-</span> <span class="fu">c</span>(color_values, <span class="dv">1</span>) <span class="co"># 1 when logFC &gt; 0 and FDRq &lt; 0.05</span></span>
<span id="cb214-8"><a href="#cb214-8" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb214-9"><a href="#cb214-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span> <span class="cf">if</span> (condition_vs_control<span class="sc">$</span>logFC[i] <span class="sc">&lt;</span> <span class="dv">0</span>){</span>
<span id="cb214-10"><a href="#cb214-10" aria-hidden="true" tabindex="-1"></a>      color_values <span class="ot">&lt;-</span> <span class="fu">c</span>(color_values, <span class="dv">2</span>) <span class="co"># 2 when logFC &lt; 0 and FDRq &lt; 0.05</span></span>
<span id="cb214-11"><a href="#cb214-11" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb214-12"><a href="#cb214-12" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb214-13"><a href="#cb214-13" aria-hidden="true" tabindex="-1"></a>  <span class="cf">else</span>{</span>
<span id="cb214-14"><a href="#cb214-14" aria-hidden="true" tabindex="-1"></a>    color_values <span class="ot">&lt;-</span> <span class="fu">c</span>(color_values, <span class="dv">3</span>) <span class="co"># 3 when FDRq &gt;= 0.05</span></span>
<span id="cb214-15"><a href="#cb214-15" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb214-16"><a href="#cb214-16" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb214-17"><a href="#cb214-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb214-18"><a href="#cb214-18" aria-hidden="true" tabindex="-1"></a>condition_vs_control<span class="sc">$</span>color_p0<span class="fl">.05</span> <span class="ot">&lt;-</span> <span class="fu">factor</span>(color_values)</span></code></pre></div>
<h1 id="subset-genes-to-label">Subset genes to label</h1>
<p>Subset the top 10 up and down-regulated genes</p>
<div class="sourceCode" id="cb215"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb215-1"><a href="#cb215-1" aria-hidden="true" tabindex="-1"></a>up <span class="ot">&lt;-</span> condition_vs_control[condition_vs_control<span class="sc">$</span>color_p0<span class="fl">.05</span> <span class="sc">==</span> <span class="dv">1</span>,]</span>
<span id="cb215-2"><a href="#cb215-2" aria-hidden="true" tabindex="-1"></a>up10 <span class="ot">&lt;-</span> up[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,]</span>
<span id="cb215-3"><a href="#cb215-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb215-4"><a href="#cb215-4" aria-hidden="true" tabindex="-1"></a>down <span class="ot">&lt;-</span> condition_vs_control[condition_vs_control<span class="sc">$</span>color_p0<span class="fl">.05</span> <span class="sc">==</span> <span class="dv">2</span>,]</span>
<span id="cb215-5"><a href="#cb215-5" aria-hidden="true" tabindex="-1"></a>down <span class="ot">&lt;-</span> <span class="fu">subset</span>(down, down<span class="sc">$</span>logFC <span class="sc">&lt;</span> <span class="sc">-</span><span class="fl">1.5</span>)</span>
<span id="cb215-6"><a href="#cb215-6" aria-hidden="true" tabindex="-1"></a>down10 <span class="ot">&lt;-</span> down[<span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>,]</span></code></pre></div>
<h1 id="volcano-plot">Volcano plot</h1>
<div class="sourceCode" id="cb216"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb216-1"><a href="#cb216-1" aria-hidden="true" tabindex="-1"></a>hadjpval <span class="ot">&lt;-</span> (<span class="sc">-</span><span class="fu">log10</span>(<span class="fu">max</span>(</span>
<span id="cb216-2"><a href="#cb216-2" aria-hidden="true" tabindex="-1"></a>  condition_vs_control<span class="sc">$</span>P.Value[condition_vs_control<span class="sc">$</span>adj.P.Val <span class="sc">&lt;</span> <span class="fl">0.2</span>], </span>
<span id="cb216-3"><a href="#cb216-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">na.rm=</span><span class="cn">TRUE</span>)))</span>
<span id="cb216-4"><a href="#cb216-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb216-5"><a href="#cb216-5" aria-hidden="true" tabindex="-1"></a>p_vol <span class="ot">&lt;-</span></span>
<span id="cb216-6"><a href="#cb216-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="at">data =</span> condition_vs_control, </span>
<span id="cb216-7"><a href="#cb216-7" aria-hidden="true" tabindex="-1"></a>         <span class="fu">aes</span>(<span class="at">x =</span> logFC,  <span class="co"># x-axis is logFC</span></span>
<span id="cb216-8"><a href="#cb216-8" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(P.Value),  <span class="co"># y-axis will be -log10 of P.Value</span></span>
<span id="cb216-9"><a href="#cb216-9" aria-hidden="true" tabindex="-1"></a>             <span class="at">color =</span> color_p0<span class="fl">.05</span>)) <span class="sc">+</span>  <span class="co"># color is based on factored color column</span></span>
<span id="cb216-10"><a href="#cb216-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.8</span>, <span class="at">size =</span> <span class="fl">1.7</span>) <span class="sc">+</span>  <span class="co"># create scatterplot, alpha makes points transparent</span></span>
<span id="cb216-11"><a href="#cb216-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span>  <span class="co"># set color theme</span></span>
<span id="cb216-12"><a href="#cb216-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>) <span class="sc">+</span>  <span class="co"># no legend</span></span>
<span id="cb216-13"><a href="#cb216-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;red&quot;</span>, <span class="st">&quot;blue&quot;</span>,<span class="st">&quot;grey&quot;</span>)) <span class="sc">+</span>  <span class="co"># set factor colors</span></span>
<span id="cb216-14"><a href="#cb216-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb216-15"><a href="#cb216-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">&quot;&quot;</span>, <span class="co"># no main title</span></span>
<span id="cb216-16"><a href="#cb216-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="fu">expression</span>(log[<span class="dv">2</span>](FC)), <span class="co"># x-axis title</span></span>
<span id="cb216-17"><a href="#cb216-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="fu">expression</span>(<span class="sc">-</span>log[<span class="dv">10</span>] <span class="sc">~</span> <span class="st">&quot;(&quot;</span> <span class="sc">~</span> <span class="fu">italic</span>(<span class="st">&quot;p&quot;</span>) <span class="sc">~</span> <span class="st">&quot;-value)&quot;</span>) <span class="co"># y-axis title</span></span>
<span id="cb216-18"><a href="#cb216-18" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb216-19"><a href="#cb216-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">10</span>),</span>
<span id="cb216-20"><a href="#cb216-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">10</span>)) <span class="sc">+</span></span>
<span id="cb216-21"><a href="#cb216-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">10</span>),</span>
<span id="cb216-22"><a href="#cb216-22" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">10</span>)) <span class="sc">+</span></span>
<span id="cb216-23"><a href="#cb216-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> hadjpval,  <span class="co">#  horizontal line</span></span>
<span id="cb216-24"><a href="#cb216-24" aria-hidden="true" tabindex="-1"></a>                     <span class="at">colour =</span> <span class="st">&quot;#000000&quot;</span>,</span>
<span id="cb216-25"><a href="#cb216-25" aria-hidden="true" tabindex="-1"></a>                     <span class="at">linetype =</span> <span class="st">&quot;dashed&quot;</span>) <span class="sc">+</span></span>
<span id="cb216-26"><a href="#cb216-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;LBD vs Control</span><span class="sc">\n</span><span class="st">FDRq &lt; 0.05&quot;</span>) <span class="sc">+</span></span>
<span id="cb216-27"><a href="#cb216-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">10</span>)) <span class="sc">+</span></span>
<span id="cb216-28"><a href="#cb216-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>(<span class="at">data =</span> up10,</span>
<span id="cb216-29"><a href="#cb216-29" aria-hidden="true" tabindex="-1"></a>                  <span class="fu">aes</span>(<span class="at">x =</span> logFC, <span class="at">y=</span> <span class="sc">-</span><span class="fu">log10</span>(P.Value), <span class="at">label =</span> gene_name), </span>
<span id="cb216-30"><a href="#cb216-30" aria-hidden="true" tabindex="-1"></a>                  <span class="at">color =</span> <span class="st">&quot;maroon&quot;</span>, </span>
<span id="cb216-31"><a href="#cb216-31" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fontface=</span><span class="st">&quot;italic&quot;</span>,</span>
<span id="cb216-32"><a href="#cb216-32" aria-hidden="true" tabindex="-1"></a>                  <span class="at">size =</span> <span class="dv">3</span>, </span>
<span id="cb216-33"><a href="#cb216-33" aria-hidden="true" tabindex="-1"></a>                  <span class="at">max.overlaps =</span> <span class="fu">getOption</span>(<span class="st">&quot;ggrepel.max.overlaps&quot;</span>, <span class="at">default =</span> <span class="dv">30</span>)</span>
<span id="cb216-34"><a href="#cb216-34" aria-hidden="true" tabindex="-1"></a>                  ) <span class="sc">+</span></span>
<span id="cb216-35"><a href="#cb216-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>(<span class="at">data =</span> down10,</span>
<span id="cb216-36"><a href="#cb216-36" aria-hidden="true" tabindex="-1"></a>                  <span class="fu">aes</span>(<span class="at">x =</span> logFC, <span class="at">y=</span> <span class="sc">-</span><span class="fu">log10</span>(P.Value), <span class="at">label =</span> gene_name), </span>
<span id="cb216-37"><a href="#cb216-37" aria-hidden="true" tabindex="-1"></a>                  <span class="at">color =</span> <span class="st">&quot;navyblue&quot;</span>, </span>
<span id="cb216-38"><a href="#cb216-38" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fontface=</span><span class="st">&quot;italic&quot;</span>,</span>
<span id="cb216-39"><a href="#cb216-39" aria-hidden="true" tabindex="-1"></a>                  <span class="at">size =</span> <span class="dv">3</span>, </span>
<span id="cb216-40"><a href="#cb216-40" aria-hidden="true" tabindex="-1"></a>                  <span class="at">max.overlaps =</span> <span class="fu">getOption</span>(<span class="st">&quot;ggrepel.max.overlaps&quot;</span>, <span class="at">default =</span> <span class="dv">15</span>)</span>
<span id="cb216-41"><a href="#cb216-41" aria-hidden="true" tabindex="-1"></a>                  ) <span class="co"># +</span></span>
<span id="cb216-42"><a href="#cb216-42" aria-hidden="true" tabindex="-1"></a> <span class="co"># scale_y_continuous(breaks = seq(0,12,by=1), limits = c(0,12)) +</span></span>
<span id="cb216-43"><a href="#cb216-43" aria-hidden="true" tabindex="-1"></a> <span class="co"># scale_x_continuous(breaks = seq(-8,8,by=2), limits = c(-8,8))</span></span>
<span id="cb216-44"><a href="#cb216-44" aria-hidden="true" tabindex="-1"></a>p_vol</span></code></pre></div>
<pre><code>## Warning: Removed 9 rows containing missing values (geom_text_repel).</code></pre>
<figure>
<img src="figure/volcano-1.png" alt="plot of chunk volcano" />
<figcaption aria-hidden="true">plot of chunk volcano</figcaption>
</figure>
<div class="sourceCode" id="cb218"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb218-1"><a href="#cb218-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save</span></span>
<span id="cb218-2"><a href="#cb218-2" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;../../results/&quot;</span>, tool, <span class="st">&quot;/volcano/&quot;</span>, condition, <span class="st">&quot;_gene_volcano_FDRq0.05_&quot;</span>, sex_karyotype)</span>
<span id="cb218-3"><a href="#cb218-3" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPDF</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.pdf&quot;</span>), <span class="at">width =</span> <span class="dv">8</span>, <span class="at">height =</span> <span class="dv">6</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb220"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb220-1"><a href="#cb220-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPNG</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.png&quot;</span>), <span class="at">width =</span> <span class="dv">8</span>, <span class="at">height =</span> <span class="dv">6</span>, <span class="at">unit =</span> <span class="st">&quot;in&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<h1 id="glimma">Glimma</h1>
<div class="sourceCode" id="cb222"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb222-1"><a href="#cb222-1" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(dge.filtered.norm<span class="sc">$</span>genes) <span class="ot">&lt;-</span> <span class="fu">make.names</span>(dge.filtered.norm<span class="sc">$</span>genes<span class="sc">$</span>gene_name, <span class="at">unique =</span> T)</span>
<span id="cb222-2"><a href="#cb222-2" aria-hidden="true" tabindex="-1"></a><span class="fu">glimmaVolcano</span>(veBayesFit, <span class="at">dge =</span> dge.filtered.norm, <span class="at">groups =</span> dge.filtered.norm<span class="sc">$</span>samples<span class="sc">$</span>TYPE)</span></code></pre></div>
<pre><code>## Error in glimmaVolcano(veBayesFit, dge = dge.filtered.norm, groups = dge.filtered.norm$samples$TYPE): could not find function &quot;glimmaVolcano&quot;</code></pre>
<div class="sourceCode" id="cb224"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb224-1"><a href="#cb224-1" aria-hidden="true" tabindex="-1"></a><span class="fu">help</span>(glimmaVolcano)</span></code></pre></div>
<pre><code>## No documentation for &#39;glimmaVolcano&#39; in specified packages and libraries:
## you could try &#39;??glimmaVolcano&#39;</code></pre>
<div class="sourceCode" id="cb226"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb226-1"><a href="#cb226-1" aria-hidden="true" tabindex="-1"></a>dt <span class="ot">&lt;-</span> <span class="fu">decideTests</span>(veBayesFit)</span></code></pre></div>
<pre><code>## Error in decideTests(veBayesFit): object &#39;veBayesFit&#39; not found</code></pre>
<div class="sourceCode" id="cb228"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb228-1"><a href="#cb228-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(veBayesFit)</span></code></pre></div>
<pre><code>## Error in summary(veBayesFit): object &#39;veBayesFit&#39; not found</code></pre>
<div class="sourceCode" id="cb230"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb230-1"><a href="#cb230-1" aria-hidden="true" tabindex="-1"></a><span class="co">#glMDPlot(veBayesFit, coef=1, status=dt, main=colnames(efit)[1],</span></span>
<span id="cb230-2"><a href="#cb230-2" aria-hidden="true" tabindex="-1"></a><span class="co">#         side.main=&quot;ENTREZID&quot;, counts=v$E, groups=TYPE, launch=T)</span></span>
<span id="cb230-3"><a href="#cb230-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb230-4"><a href="#cb230-4" aria-hidden="true" tabindex="-1"></a>TYPE <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(dge.filtered.norm<span class="sc">$</span>samples<span class="sc">$</span>TYPE)</span>
<span id="cb230-5"><a href="#cb230-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gplots)</span>
<span id="cb230-6"><a href="#cb230-6" aria-hidden="true" tabindex="-1"></a>basal.vs.lp <span class="ot">&lt;-</span> <span class="fu">topTreat</span>(veBayesFit, <span class="at">coef=</span><span class="dv">1</span>, <span class="at">n=</span><span class="cn">Inf</span>)</span></code></pre></div>
<pre><code>## Error in is(fit, &quot;MArrayLM&quot;): object &#39;veBayesFit&#39; not found</code></pre>
<div class="sourceCode" id="cb232"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb232-1"><a href="#cb232-1" aria-hidden="true" tabindex="-1"></a>basal.vs.lp.topgenes <span class="ot">&lt;-</span> basal.vs.lp<span class="sc">$</span>gene_name[<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>]</span></code></pre></div>
<pre><code>## Error in eval(expr, envir, enclos): object &#39;basal.vs.lp&#39; not found</code></pre>
<div class="sourceCode" id="cb234"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb234-1"><a href="#cb234-1" aria-hidden="true" tabindex="-1"></a>i <span class="ot">&lt;-</span> <span class="fu">which</span>(v<span class="sc">$</span>genes<span class="sc">$</span>gene_name <span class="sc">%in%</span> basal.vs.lp.topgenes)</span></code></pre></div>
<pre><code>## Error in v$genes$gene_name %in% basal.vs.lp.topgenes: object &#39;basal.vs.lp.topgenes&#39; not found</code></pre>
<div class="sourceCode" id="cb236"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb236-1"><a href="#cb236-1" aria-hidden="true" tabindex="-1"></a>mycol <span class="ot">&lt;-</span> <span class="fu">colorpanel</span>(<span class="dv">1000</span>,<span class="st">&quot;blue&quot;</span>,<span class="st">&quot;white&quot;</span>,<span class="st">&quot;red&quot;</span>)</span>
<span id="cb236-2"><a href="#cb236-2" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap.2</span>(v<span class="sc">$</span>E[i,], <span class="at">scale=</span><span class="st">&quot;row&quot;</span>,</span>
<span id="cb236-3"><a href="#cb236-3" aria-hidden="true" tabindex="-1"></a>   <span class="at">labRow=</span>v<span class="sc">$</span>genes<span class="sc">$</span>gene_name[i], <span class="at">labCol=</span>TYPE, </span>
<span id="cb236-4"><a href="#cb236-4" aria-hidden="true" tabindex="-1"></a>   <span class="at">col=</span>mycol, <span class="at">trace=</span><span class="st">&quot;none&quot;</span>, <span class="at">density.info=</span><span class="st">&quot;none&quot;</span>, </span>
<span id="cb236-5"><a href="#cb236-5" aria-hidden="true" tabindex="-1"></a>   <span class="at">margin=</span><span class="fu">c</span>(<span class="dv">8</span>,<span class="dv">6</span>), <span class="at">lhei=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">10</span>), <span class="at">dendrogram=</span><span class="st">&quot;column&quot;</span>)</span></code></pre></div>
<pre><code>## Error in heatmap.2(v$E[i, ], scale = &quot;row&quot;, labRow = v$genes$gene_name[i], : `x&#39; must be a numeric matrix</code></pre>
<div class="sourceCode" id="cb238"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb238-1"><a href="#cb238-1" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;../../results/&quot;</span>, tool, <span class="st">&quot;/volcano/LBD_gene_volcano_FDRq0.05_heatmap&quot;</span>, sex_karyotype)</span>
<span id="cb238-2"><a href="#cb238-2" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPDF</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.pdf&quot;</span>), <span class="at">width =</span> <span class="dv">12</span>, <span class="at">height =</span> <span class="dv">8</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb240"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb240-1"><a href="#cb240-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveToPNG</span>(<span class="fu">paste0</span>(path, <span class="st">&quot;.png&quot;</span>), <span class="at">width =</span> <span class="dv">8</span>, <span class="at">height =</span> <span class="dv">6</span>, <span class="at">unit =</span> <span class="st">&quot;in&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>)</span></code></pre></div>
<pre><code>## png 
##   2</code></pre>
<h1 id="bar-plot-gene-of-interest">Bar plot gene of interest</h1>
<div class="sourceCode" id="cb242"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb242-1"><a href="#cb242-1" aria-hidden="true" tabindex="-1"></a>GOI <span class="ot">&lt;-</span> v[v<span class="sc">$</span>genes<span class="sc">$</span>gene_name <span class="sc">%in%</span> <span class="st">&quot;TTR&quot;</span>,]</span>
<span id="cb242-2"><a href="#cb242-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Barplot</span></span>
<span id="cb242-3"><a href="#cb242-3" aria-hidden="true" tabindex="-1"></a>out_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(GOI<span class="sc">$</span>E)</span>
<span id="cb242-4"><a href="#cb242-4" aria-hidden="true" tabindex="-1"></a>df_melt <span class="ot">&lt;-</span> <span class="fu">melt</span>(out_df)</span></code></pre></div>
<pre><code>## Using  as id variables</code></pre>
<div class="sourceCode" id="cb244"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb244-1"><a href="#cb244-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(df_melt<span class="sc">$</span>value, <span class="at">breaks =</span> <span class="dv">40</span>)</span></code></pre></div>
<figure>
<img src="figure/unnamed-chunk-9-1.png"
alt="plot of chunk unnamed-chunk-9" />
<figcaption aria-hidden="true">plot of chunk
unnamed-chunk-9</figcaption>
</figure>
<div class="sourceCode" id="cb245"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb245-1"><a href="#cb245-1" aria-hidden="true" tabindex="-1"></a>df_melt<span class="sc">$</span>NPID <span class="ot">&lt;-</span> df_melt<span class="sc">$</span>variable</span>
<span id="cb245-2"><a href="#cb245-2" aria-hidden="true" tabindex="-1"></a>df_melt <span class="ot">&lt;-</span> <span class="fu">merge</span>(df_melt, metadata, <span class="at">by =</span> <span class="st">&quot;NPID&quot;</span>)</span>
<span id="cb245-3"><a href="#cb245-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_melt, <span class="fu">aes</span>(<span class="fu">reorder</span>(<span class="at">x=</span>variable, <span class="sc">-</span>value), <span class="at">y=</span>value, <span class="at">color =</span> TYPE, <span class="at">fill =</span> TYPE)) <span class="sc">+</span> <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">&quot;identity&quot;</span>)</span></code></pre></div>
<figure>
<img src="figure/unnamed-chunk-9-2.png"
alt="plot of chunk unnamed-chunk-9" />
<figcaption aria-hidden="true">plot of chunk
unnamed-chunk-9</figcaption>
</figure>
<div class="sourceCode" id="cb246"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb246-1"><a href="#cb246-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 3.6.2 (2019-12-12)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Red Hat Enterprise Linux Server 7.9 (Maipo)
## 
## Matrix products: default
## BLAS:   /usr/lib64/libblas.so.3.4.2
## LAPACK: /usr/lib64/liblapack.so.3.4.2
## 
## locale:
## [1] C
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] ggpubr_0.4.0         corrplot_0.84        plyr_1.8.6          
##  [4] Glimma_1.14.0        reshape_0.8.8        stringr_1.4.0       
##  [7] rtracklayer_1.46.0   GenomicRanges_1.38.0 GenomeInfoDb_1.22.0 
## [10] IRanges_2.20.2       S4Vectors_0.24.4     BiocGenerics_0.32.0 
## [13] gplots_3.1.1         ggrepel_0.9.1        ggplot2_3.3.3       
## [16] edgeR_3.28.1         limma_3.42.2         dplyr_1.0.4         
## [19] BiocParallel_1.20.1  knitr_1.31          
## 
## loaded via a namespace (and not attached):
##  [1] nlme_3.1-152                bitops_1.0-6               
##  [3] matrixStats_0.58.0          tools_3.6.2                
##  [5] backports_1.2.1             utf8_1.1.4                 
##  [7] R6_2.5.0                    KernSmooth_2.23-18         
##  [9] mgcv_1.8-34                 DBI_1.1.1                  
## [11] colorspace_2.0-0            withr_2.4.1                
## [13] tidyselect_1.1.0            curl_4.3                   
## [15] compiler_3.6.2              Biobase_2.46.0             
## [17] DelayedArray_0.12.3         labeling_0.4.2             
## [19] caTools_1.18.1              scales_1.1.1               
## [21] digest_0.6.27               Rsamtools_2.2.3            
## [23] foreign_0.8-76              rio_0.5.16                 
## [25] XVector_0.26.0              pkgconfig_2.0.3            
## [27] highr_0.8                   rlang_0.4.10               
## [29] readxl_1.3.1                farver_2.0.3               
## [31] generics_0.1.0              jsonlite_1.7.2             
## [33] gtools_3.8.2                zip_2.1.1                  
## [35] car_3.0-10                  RCurl_1.98-1.2             
## [37] magrittr_2.0.1              GenomeInfoDbData_1.2.2     
## [39] Matrix_1.3-2                Rcpp_1.0.6                 
## [41] munsell_0.5.0               fansi_0.4.2                
## [43] abind_1.4-5                 lifecycle_1.0.0            
## [45] stringi_1.5.3               yaml_2.2.1                 
## [47] carData_3.0-4               debugme_1.1.0              
## [49] SummarizedExperiment_1.16.1 zlibbioc_1.32.0            
## [51] grid_3.6.2                  forcats_0.5.1              
## [53] crayon_1.4.1                lattice_0.20-41            
## [55] splines_3.6.2               Biostrings_2.54.0          
## [57] haven_2.3.1                 hms_1.0.0                  
## [59] locfit_1.5-9.4              pillar_1.5.0               
## [61] ggsignif_0.6.0              XML_3.99-0.3               
## [63] glue_1.4.2                  evaluate_0.14              
## [65] data.table_1.14.0           vctrs_0.3.6                
## [67] cellranger_1.1.0            gtable_0.3.0               
## [69] purrr_0.3.4                 tidyr_1.1.2                
## [71] assertthat_0.2.1            xfun_0.21                  
## [73] openxlsx_4.2.3              broom_0.7.5                
## [75] rstatix_0.7.0               tibble_3.0.6               
## [77] GenomicAlignments_1.22.1    ellipsis_0.3.1</code></pre>
