---
title: "LBD CWOW <br> January 26th, 2023"
#author: "Kimberly Olney, Ph.D"
#date: "01/26/2023"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

# Data
```{r setup, include=FALSE}
# Also do Session > Set Working Directory > Choose Directory
knitr::opts_knit$set(root.dir = ".")
```

```{r library, include=FALSE}
#library(gtools)
library(tidyr)
library(wesanderson) # colors
library(gridExtra)
library(grid)
library(knitr)
```

# Explore Metadata
```{r metadata, include=FALSE}
source(here::here("scripts/R", "file_paths_and_colours.R"))

metadata <- read_delim(
  paste0(pathToRawData, "RNA_metadata.tsv"))
```

Alzheimer's disease (AD), path aging (PA), Lewy Body Dementia (LBD)
```{r echo=FALSE, results='asis'}
metadata$TYPE <- factor(metadata$TYPE, levels = c("CONTROL", "AD", "PA", "LBD"))
disease_count <- table(metadata[3:3])
kableExtra::kable(disease_count[1:4], "simple", col.names = c("diagnosis", "count"))
```

Visualize the distribution of data. 
```{r echo=FALSE, warning = FALSE, message = FALSE, results='asis'}
bar <- ggplot(metadata, aes(TYPE, after_stat(count), fill = TYPE)) +
  geom_bar() +
  theme_bw() +
  xlab("Disease group") +
  ggtitle("Sample count") +
  geom_text(
    stat = 'count',
    aes(label = after_stat(count)),
    vjust = 1.6,
    color = "white",
    size = 3.5
  ) +
    theme_bw() + 
      theme(legend.position = "none", 
          axis.title.x = element_blank(),
          axis.title.y = element_blank()) +
  scale_fill_manual(values = TypeColors)

metadata_continuous <-
  data.frame(metadata$Age, metadata$PF_READS, metadata$PMI, metadata$RIN, metadata$PCT_CODING_BASES, metadata$PCT_INTERGENIC_BASES, metadata$PCT_INTRONIC_BASES) 

column_variables <- c("age of death", "PF_READS", "PMI", "RIN", "PCT_CODING_BASES", "PCT_INTERGENIC_BASES", "PCT_INTRONIC_BASES")

my_comparisons <- list(c("CONTROL", "AD"), c("CONTROL", "PA"), c("CONTROL", "LBD"), c("AD", "PA"), c("AD", "LBD"), c("PA", "LBD"))

violin_plot_fun <- function(i, j) {
  ggplot(metadata, aes(TYPE, i, fill = TYPE)) +
    geom_violin() +
    geom_boxplot(width = 0.1, outlier.shape = NA) + 
    geom_jitter(shape=16, position=position_jitter(0.2), size = 0.5) +
    theme_bw() + 
    ggtitle(j) +
  #  xlab("Disease group") +
  #  ylab(j) +
  #  stat_compare_means(comparisons = my_comparisons) +
    scale_fill_manual(values=TypeColors) +
    theme(legend.position = "none", 
          axis.title.x = element_blank(),
          axis.title.y = element_blank())
}
violin_plots <-
  Map(violin_plot_fun, i = metadata_continuous, j = column_variables)


row1 <- ggarrange(
  bar,
  violin_plots$metadata.Age,
  ncol = 2
  )

row2 <- ggarrange(
  violin_plots$metadata.RIN,
  violin_plots$metadata.PMI,
  ncol = 2
  )

row3 <- ggarrange(
  violin_plots$metadata.PF_READS,
  violin_plots$metadata.PCT_CODING_BASES,
  ncol = 2
  )

row4 <- ggarrange(
  violin_plots$metadata.PCT_INTERGENIC_BASES,
  violin_plots$metadata.PCT_INTRONIC_BASES,
  ncol = 2
  )

combind1 <-
  ggarrange(
    row1,
    row2,
    nrow = 2
  )
combind1

combind2 <-
  ggarrange(
    row3,
    row4,
    nrow = 2
  )
combind2
```

Visualize gene expression of sex chromosomes linked genes when aligned to the default reference GRCh38.
```{r  echo=FALSE, warning = FALSE, message = FALSE, results='asis'}
df <- read.delim(paste0("../../results/star/sex_check/sex_check_data.txt"), sep = "\t")
wrong_sex <- read.delim(paste0("../../results/star/sex_check/wrong_sex_samples.txt"), sep = "\t")
leg_lab <- "reported sex"
geneticSEXgenes_plot <- ggplot(df, aes(x = Geneid, y = value)) +
  geom_jitter(aes(color = Sex, shape = Sex),
              width = 0.25,
              size = 1.2) +
  scale_color_manual(leg_lab, values = c("mediumorchid4", "orange")) + 
  scale_shape_manual(leg_lab, values = c(19, 15)) +
  labs(x = "", y = "lcpm", title = "") +
  facet_grid(
    . ~ group + geneComb,
    switch = "x",
    scales = "free_x",
    space = "free_x"
  ) +
  theme_bw() +
  theme(
    strip.background = element_blank(),
    strip.text.x = element_blank(),
    axis.ticks = element_blank(),
    axis.text.x = element_text(face = "italic", angle = 90, vjust = 0.5, hjust=1),
    legend.position="top",
    panel.border = element_rect(
      colour = "black",
      fill = NA,
      size = 1
    )
  ) +
  scale_y_continuous(breaks = seq(-4, 9, by = 1), limits = c(-4, 9)) 
geneticSEXgenes_plot
```
```{r echo=FALSE, warning = FALSE, message = FALSE, results='asis'}
UTY <- subset(df, Geneid == "UTY")
XIST <- subset(df, Geneid == "XIST")
df2 <- merge(XIST, UTY, by = "NPID")
df2_wrong_sex <- subset(df2, NPID %in% wrong_sex$NPID) # reported male
df2_wrong_sex_M <- subset(df2_wrong_sex, Sex.x == "M") # reported male
df2_wrong_sex_F <- subset(df2_wrong_sex, Sex.x == "F") # reported female


leg_lab <- "reported sex"
XIST_UTY_plot <- ggplot(df2, aes(x = value.x, y = value.y)) +
  geom_jitter(aes(color = Sex.x, shape = Sex.x),
              width = 0.25,
              size = 1.35) +
  scale_color_manual(leg_lab, values = c("mediumorchid4", "orange")) + 
  scale_shape_manual(leg_lab, values = c(19, 15)) +
  labs(title = "", # no main title
    x = expression(log[2] ~ "(" ~ italic("XIST") ~ ")"), # x-axis title
    y = expression(log[2] ~ "(" ~ italic("UTY") ~ ")") # y-axis title
  ) +
  theme_bw() +
    geom_text_repel(
    data = df2_wrong_sex_M,
    aes(x = value.x,
        y = value.y,
        label = NPID),
    color = "orange",
    size = 3.5,
    nudge_y = .5,
    min.segment.length = unit(0, 'lines'),
    max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
  ) +
    geom_text_repel(
    data = df2_wrong_sex_F,
    aes(x = value.x,
        y = value.y,
        label = NPID),
    color = "mediumorchid4",
    size = 3.5,
    nudge_x = .5,
    min.segment.length = unit(0, 'lines'),
    max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
  )
XIST_UTY_plot
```

Sample swaps identified by discordant sex label: NA04-053, NA05-083, NA05-388\n

Samples are realigned to the sex chromosome complement (Olney et al. 2020) https://doi.org/10.1186/s13293-020-00312-9 

```{r  echo=FALSE, warning = FALSE, message = FALSE, results='asis', out.width = "70%"}
bar_sex <- ggplot(metadata, aes(x = TYPE, fill = sex_inferred)) + 
  geom_bar(aes(y = after_stat(count)), position = "dodge") + 
  theme_bw() + 
  theme(axis.title.x = element_blank()) +
  ggtitle("Inferred sex sample count") +
  geom_text(stat='count', aes(x = TYPE, label=after_stat(count)), 
            position = position_dodge(width = 1), 
              vjust=-.25, 
            color="black", size=3.5) + 
  scale_fill_manual(values = c("mediumorchid4", "orange")) +
  guides(fill = guide_legend(title = "Sex inferred"))+
  scale_y_continuous(breaks = seq(0, 350, by = 100), limits = c(0, 350))
bar_sex
```

Visualize the relationships between covariates.

```{r echo=FALSE, warning = FALSE, message = FALSE, results='asis'}
metadata$APOE_E4_allele_count <-
  ifelse(
    metadata$APOE == "E2E3",
    0,
    ifelse(
      metadata$APOE == "E2E4",
      1,
      ifelse(
        metadata$APOE == "E3E3",
        0,
        ifelse(
          metadata$APOE == "E3E4",
          0,
          ifelse(
            metadata$APOE == "E4E4", 2, "NA")))))
metadata$APOE_E4_allele_count <- as.numeric(metadata$APOE_E4_allele_count)
metadata$sex_numeric <- ifelse(metadata$sex_inferred == "male", 0, 1)
metadata$Race_numeric <- ifelse(metadata$Race == "Caucasian", 0, 1)

metadata2 <-
  metadata[, c("Box", "Age", "sex_numeric", "Race_numeric", "RIN", "APOE_E4_allele_count", "PF_READS", "PCT_CODING_BASES", "PCT_INTERGENIC_BASES", "PCT_INTRONIC_BASES")] %>% scale() # PMI
colnames(metadata2)[1] ="Batch"
colnames(metadata2)[3] ="Sex.inferred"
colnames(metadata2)[4] ="Race.reported"
colnames(metadata2)[6] ="APOE4"

cor_mat <- rcorr(as.matrix(metadata2))

corrplot(
  cor_mat$r,   
  tl.col = "black",     
  col = correlationColors(200))
```
# Filter Genes
Remove genes that have less than 1 counts per million (CPM) in at least 0.5 of samples per specified condition.

```{r echo=FALSE, warning = FALSE, message = FALSE, results='asis'}
condition <- c("All_samples")
dge <- readRDS(paste0("../../rObjects/", condition, ".dge.raw.rds"))
cpm <- edgeR::cpm(dge$counts, log = FALSE)

dge$samples$sex_inferred <- as.factor(dge$samples$sex_inferred)

keep.expr <- filterByExpr(cpm, dge$samples$TYPE, min.count = 1, min.prop=0.5)
dge.filtered <- dge[keep.expr, , keep.lib.sizes = FALSE]
# Now, normalization by the method of trimmed mean of M-values (TMM)
dge.filtered.norm <- calcNormFactors(dge.filtered, method = "TMM")

genes <- dge.filtered$genes
#write.table(genes, "genes.txt", quote = FALSE, row.names = FALSE, sep = "\t")

expressed_genes <- table(dge.filtered.norm$genes$gene_type)
exp_genes_df <- as.data.frame(expressed_genes)
exp_genes_df$Fraction <- exp_genes_df$Freq/16627
exp_genes_df$Freq <- NULL
kableExtra::kable(exp_genes_df, "simple", col.names = c("gene_biotype", "fraction"))
```





16,627 genes are considered expressed and will be kept for downstream analysis. 

# Identify Outliers
```{r  echo=FALSE, warning = FALSE, message = FALSE, results='asis'}
# Setting the N of genes to use
ntop = length(dge.filtered.norm$counts)
cpm <- edgeR::cpm(dge.filtered.norm$counts, log = TRUE)
means <- rowMeans(cpm)
Pvars <- rowVars(cpm)
cv2 <- Pvars / means ^ 2
select <- order(cv2, decreasing = TRUE)[seq_len(min(ntop, length(cv2)))]
highly_variable_exp <- ((cpm)[select,])
# Running PCA
pca_exp <- prcomp(t(highly_variable_exp), scale = T, center = T)
# scale a logical value indicating whether the variables should be scaled
# to have unit variance before the analysis takes place.
# a logical value indicating whether the variables should be shifted to be zero centered.

# Dataframe with the first 10 PCs
dim1_10 <- data.frame(pca_exp$x[, 1:10])
# Adding metadata
pcaWithMetadata <- merge(dim1_10, dge.filtered$samples, by = 0, all = TRUE)
pcaWithMetadata$diagnosis <- factor(pcaWithMetadata$TYPE,
                                 levels = c(CONTROL, AD, PA, LBD, NA))

pcaWithMetadata$batch <- factor(pcaWithMetadata$Box)
pcaWithMetadata$batch <- as.character(pcaWithMetadata$batch)
pcaWithMetadata$RIN2 <- as.factor(pcaWithMetadata$RIN)
# Plotting
ggplot(data = pcaWithMetadata, aes(
  x = PC1,
  y = PC2,
  shape = diagnosis,
  color = RIN2
)) +
  geom_point(size = 2.5) +
  theme_bw() +
  theme(legend.position = "none")


ggplot(data = pcaWithMetadata, aes(
  x = PC2,
  y = PC3,
  shape = diagnosis,
  color = diagnosis
)) +
  geom_point(size = 2.5) +
  theme_bw() +
  scale_color_manual(values = c(control_color, AD_color,
                                PA_color, LBD_color,
                                "azure3")) 

```
```{r}
```
# Significant Covariates
```{r echo=FALSE, warning = FALSE, message = FALSE, results='asis'}
PCA1_10 <-
  pcaWithMetadata[, c("PC1", "PC2", "PC3")] 
metadata2_df <- as.data.frame(metadata2)
hello <- cbind(metadata2, PCA1_10)
cor_mat <- rcorr(as.matrix(hello))

corrplot(
  cor_mat$r,   
  tl.col = "black",     
  col = correlationColors(200))
```
```{r}
```
# Model Identification
To be determined 
```{r}
```
# Differential Expression
To-do
```{r}
```
