---
title: "Create rObject of RNAseq counts data for repersentative_samples"
author: "Kimberly Olney, Ph.D."
date: "05/11/2023"
output:
  html_document:
    df_print: paged
  pdf_document: default
params:
  args: myarg
---
Samples have already been sex checked and realigned to the SCC reference. 
# Setup
```{r setup}
knitr::opts_knit$set(root.dir = ".")
```
# User defined variables
```{r set_variables}
source(here::here("scripts/R", "file_paths_and_colours.R"))
source(here::here("scripts/R", "gtf_path.R"))
condition <- c("representative_samples") # change condition to rename files for comparing between different approaches. 
tool = c("star_representative_samples")
typeOfCount <-  c(".bamReadsPerGene.out.tab")

# remove metadata since we only want the subset data
remove(metadata)

# read in metadata
metadata <-
  read.delim(paste0("repersentative_samples.tsv"),
             header = TRUE,
             sep = "\t")
metadata$TYPE <-
  factor(metadata$TYPE, levels = c("CONTROL", "PA", "AD", "LBD"))

# remove library size and normalization factor
metadata$lib.size <- NULL
metadata$norm.factors <- NULL
```

# further subset data to only include 4 per sex per disease type 
```{r subset_metadata}
set.seed(28) # to get the same random next time
disease_types <- c("CONTROL", "PA", "AD", "LBD")
for (i in disease_types){
  female <- subset(metadata, TYPE == i & sex_inferred == "female")
  female_4 <- sample_n(female, 4)
  assign(paste0(i, "_female"), female_4)
  male <- subset(metadata, TYPE == i & sex_inferred == "male")
  male_4 <- sample_n(male, 4)
  assign(paste0(i, "_male"), male_4)
  remove(female, female_4, male, male_4)
}

metadata <-
  rbind(CONTROL_female,
        CONTROL_male,
        PA_female,
        PA_male,
        AD_female,
        AD_male,
        LBD_female,
        LBD_male)
```
# Read data
### Metadata
```{r read_data}
# example of file naming:
# NA12-254_FCHCLTJDMXY_L1_STAR_XY.bamReadsPerGene.out.tab
# path to counts files
count_files <-
  file.path(paste0(
    "../../starAligned_SCC/",
    metadata$NPID,
    "_STAR_",
    metadata$sex_chr,
    typeOfCount
  ))
# add sample name to counts files
names(count_files) <- paste0(metadata$NPID)
# add sample count path information to metadata
metadata$path <- count_files
# make sure there is only unique paths
```
### Create a counts matrix
```{r counts_matrix}
counts <- data.frame(fread(count_files[1]))[c(1, 2)]
# Loop and read the 2nd column remaining files
for (i in 2:length(count_files)) {
  counts <- cbind(counts, data.frame(fread(count_files[i]))[2])
}
```
# Star mapping metrics
```{r mapping_metrics}
star_metrics <- data.frame(counts[counts$V1 %like% "N_", ])
star_metrics$V1 <- as.vector(star_metrics$V1)
melt_star_metrics <- reshape::melt(star_metrics, id = c("V1"))
ggplot(melt_star_metrics, aes(x = value, fill = V1)) + 
  geom_histogram(position = "identity", bins = 100) + 
  facet_grid(V1 ~ .)  + theme_bw()
```
clean up
```{r}
# clean up
remove(melt_star_metrics, 
       params, 
       star_metrics)
```
# Format counts table
```{r counts_table}
# remove star metric information
counts <- counts[!grepl("N_", counts$V1),]
# set gene_id (i.e. V1) as the row name
row.names(counts) <- counts$V1
counts$V1 <- NULL
# set column names to NPID sample ID
colnames(counts) <- metadata$NPID
```
# Check inputs
```{r check_inputs}
all.equal(rownames(counts), genes.gtf$gene_id)
all.equal(colnames(counts), (metadata$NPID))
```
# Create DGE object
```{r DGE_object}
# create object
dge <- DGEList(counts = counts,
               samples = metadata,
               genes = genes.gtf)

table(dge$samples$TYPE)
table(dge$samples$ATS_names)
```

# Save R object
```{r rOjects}
saveRDS(dge, file = paste0("../../rObjects/", condition, "dge.raw.rds"))
dge <- readRDS(paste0("../../rObjects/", condition, "dge.raw.rds"))
```
# MDS
convert counts to cpm and lcpm
set colors and get data
```{r cpm, warning=FALSE}
lcpm <- edgeR::cpm(dge$counts, log = TRUE)
cpm <- edgeR::cpm(dge$counts, log = FALSE)

# sex colors 
dge$samples$sex_inferred <- as.factor(dge$samples$sex_inferred)
sex_colors <- c(SexColors)[dge$samples$sex_inferred]

par(bg = 'white')
plotMDS(
  lcpm,
  top = 100, 
  labels = dge$samples$sex_inferred,
  cex = .8, 
  dim.plot = c(1,2), 
  plot = TRUE, 
  col = sex_colors,
  gene.selection = "common"
)
title(expression('Top 100 Genes (Log'[2]~'CPM)'))

path <- paste0("../../results/", tool, "/MDS/", 
               condition,"label_sex_dim1&2")
saveToPDF(paste0(path, ".pdf"), width = 5.2, height = 5.2)

plotMDS(
  lcpm,
  top = 100, 
  labels = dge$samples$sex_inferred,
  cex = .8, 
  dim.plot = c(2,3), 
  plot = TRUE, 
  col = sex_colors,
  gene.selection = "common"
)
title(expression('Top 100 Genes (Log'[2]~'CPM)'))
path <- paste0("../../results/", tool, "/MDS/", 
               condition,"label_sex_dim2&3")
saveToPDF(paste0(path, ".pdf"), width = 5.2, height = 5.2)
```

# Filtering and RIN check
The filterByExpr() function in the edgeR package determines which genes have a 
great enough count value to keep.  
We will filter by group.  
This means at least 6 samples (6 is the smallest group sample size) must express a minimum count of 
1 (in cpm, default value is 10).
```{r filter_counts}
# first filter by expression
dim(dge)
keep.expr <-
  filterByExpr(
    edgeR::cpm(dge$counts),
    dge$samples$TYPE, # by ATS groups
    min.count = 1, # min count of 1 CPM 
    min.prop = 0.3 # at least 30% of the smallest group 
  )
dge.filtered <- dge[keep.expr, , keep.lib.sizes = FALSE]
dim(dge.filtered)
table(dge.filtered$genes$gene_type)
```
# Remove mitochondrial genes & keep only protein coding genes
```{r MT_genes, eval = TRUE}
dim(dge.filtered)
removeMT <- dge.filtered$genes$seqnames != "chrM"  # true when NOT MT
dge.filtered <- dge.filtered[removeMT,,keep.lib.sizes = FALSE]

keepProtein <- dge.filtered$genes$gene_type == "protein_coding"
dge.filtered <- dge.filtered[keepProtein,,keep.lib.sizes = FALSE]
dim(dge.filtered)
```
# Library sizes
```{r library}
# before filtering
L <- mean(dge$samples$lib.size) * 1e-6
M <- median(dge$samples$lib.size) * 1e-6
c(L, M)

# after filtering 
L <- mean(dge.filtered$samples$lib.size) * 1e-6
M <- median(dge.filtered$samples$lib.size) * 1e-6
c(L, M)
```

# TMM
For estimating relative RNA production levels from RNA-seq data. 
The TMM method estimates scale factors between samples that can be incorporated 
into currently used statistical methods for DE analysis.
```{r TMM}
# Now, method of trimmed mean of M-values (TMM)
# Calculate scaling factors to convert raw library sizes into effective library sizes.
dge.filtered.norm <- calcNormFactors(dge.filtered, method = "TMM")

# norm factor summary
summary(dge.filtered.norm$samples$norm.factors)
normfactors <- (dge.filtered.norm$samples$norm.factors)
metadata$normfactors <- normfactors

# examine normalization factors 
plot(metadata$TYPE, metadata$normfactors)
```

```{r plot_RIN}
log2cpm.norm <- edgeR::cpm(dge.filtered.norm, log = T)
nsamples <- ncol(dge.filtered.norm)
boxplot(log2cpm.norm, 
        main="Filtered normalized lcpm data", 
        xlab="RIN", 
        ylab=expression('Counts per gene (Log'[2]~'CPM)'),
        axes=FALSE)
axis(2)
axis(1,at=1:nsamples,labels=(dge.filtered.norm$samples$RIN),las=2,cex.axis=0.8)

path <- paste0("../../results/", tool, "/library/", condition,
               "boxplot_RIN")
saveToPDF(paste0(path, ".pdf"), width = 12, height = 6)
```

# RIN check and library size
```{r RIN_check, warning=FALSE}
# check if there is correlation between RIN and library size
box <- dge.filtered.norm$samples

cor(box$RIN, box$lib.size, method = c("pearson", "kendall", "spearman"))
cor.test(box$RIN, box$lib.size, method=c("pearson", "kendall", "spearman"))

# is the data normally distributed 
ggqqplot(box$lib.size, ylab = "library size")
ggqqplot(box$RIN, ylab = "RIN")
res <- cor.test(box$lib.size, box$RIN, 
                method = "pearson")
res
ggscatter(box, x = "RIN", y = "lib.size", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          ylab = "library size", xlab = "RIN value") 

path <- paste0("../../results/", tool, "/library/", condition, 
               "corr_RIN_library_size")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 6)
```

# MDS - normalized counts
multidimensional scaling
```{r MDS_norm}
par(bg = 'white')
plotMDS(
  log2cpm.norm,
  top = 100, 
  labels = dge.filtered.norm$samples$sex_inferred,
  cex = .8, 
  dim.plot = c(1,2), 
  plot = TRUE, 
  col = sex_colors,
  gene.selection = "common"
)
title(expression('Top 100 Genes (Log'[2]~'CPM)'))

path <- paste0("../../results/", tool, "/MDS/", 
               condition, "label_sex_dim1&2")
saveToPDF(paste0(path, ".pdf"), width = 5.2, height = 5.2)
```
# Density plot
Density plots of log-intensity distribution of each library can be superposed on a single graph for a better comparison between libraries and for identification of libraries with weird distribution. 
```{r density_plots}
# set graphical parameter
par(mfrow = c(1,3))

# Normalize data for library size and expression intensity
log2cpm.tech <- edgeR::cpm(dge, log = TRUE)
log2cpm.filtered <- edgeR::cpm(dge.filtered, log = TRUE)
log2cpm.norm <- edgeR::cpm(dge.filtered.norm, log = TRUE)

# set colors
colors <- TypeColors[dge$samples$TYPE]
nsamples <- ncol(dge)

# First, plot the first column of the log2cpm.tech density
plot(density(log2cpm.tech[,1]), col = colors[1], lwd = 2, ylim = c(0,0.5), 
     las = 2, main = "A. Raw", xlab = expression('Log'[2]~CPM))
# For each sample plot the lcpm density
for (i in 1:nsamples){
  den <- density(log2cpm.tech[,i]) #subset each column
  lines(den$x, den$y, col = colors[i], lwd = 2) 
}

# Second, plot log2cpm.filtered
plot(density(log2cpm.filtered[,1]), col = colors[1], lwd = 2, ylim = c(0,0.5), 
     las = 2, main = "B. Filtered", xlab = expression('Log'[2]~CPM))
for (i in 2:nsamples) {
  den <- density(log2cpm.filtered[,i])
  lines(den$x, den$y, col = colors[i], lwd = 2)
}

# Third, plot log2cpm.norm
plot(density(log2cpm.norm[,1]), col = colors[1], lwd = 2, ylim = c(0,0.5), 
     las = 2, main = "C. TMM", xlab = expression('Log'[2]~CPM))
for (i in 2:nsamples) {
  den <- density(log2cpm.norm[,i])
  lines(den$x, den$y, col = colors[i], lwd = 2)
}

# save
path <- paste0("../../results/", tool, "/library/TYPE.gene_density")
saveToPDF(paste0(path, ".pdf"), width = 7, height = 5)
```
clean up 
```{r}
remove(box,
       res)
```
clean up 
```{r}
remove(
       dge, 
       dge.filtered,
       lcpm,
       log2cpm.filtered,
       log2cpm.norm,
       log2cpm.tech)
```

# Save R object
```{r save_dge_final}
saveRDS(dge.filtered.norm, file = paste0("../../rObjects/", condition, "dge.filtered.norm.rds"))
check_info <- dge.filtered.norm$samples
```

```{r}
sessionInfo()
```
