---
title: "LBD CWOW Brain Single Nucleus Find Markers"
author: "Kimberly Olney"
date: "06/08/2023"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

# Set working directory
```{r working_directory}
knitr::opts_knit$set(root.dir = ".")
```
# Libraris, paths, colors
```{r library, message=FALSE, warning=FALSE}
source(here::here("snRNA/scripts/R", "file_paths_and_colours.R"))
tissue <- "Brain"
treatment <- "pilot"
```
# Read in object
```{r read_object, message=FALSE, warning=FALSE}
# read object
dataObject.unannotated <- readRDS("../../rObjects/brain_unannotated.rds")
```
## Define resolution and groups
```{r define_res, message=FALSE, warning=FALSE}
# set levels and idents
Idents(dataObject.unannotated) <- "seurat_clusters"
# normalize if SCTtransform was used
DefaultAssay(dataObject.unannotated) <- "RNA"
```

# Clustering QC before annotation
## Unannotated UMAP
```{r color_blind_umap, message=FALSE, warning=FALSE}
dittoDimPlot(object = dataObject.unannotated,
             var = "seurat_clusters",
             reduction.use = "umap",
             do.label = TRUE,
             labels.highlight = TRUE)
```


## Cells per cluster
```{r cells_per_cluster}
dataObject.unannotated$sample <- factor(dataObject.unannotated$sample, levels = c("beads", "debris"))

n_cells <- FetchData(dataObject.unannotated, 
  vars = c("ident", "sample")) %>%
  dplyr::count(ident,sample) %>%
  tidyr::spread(ident, n)
n_cells
```

## Cluster tree
```{r cluster_tree, warning=FALSE, message=FALSE}
dataObject.unannotated <- BuildClusterTree(object = dataObject.unannotated,
                                     dims = 1:20,
                                     reorder = FALSE,
                                     reorder.numeric = FALSE)

tree <- dataObject.unannotated@tools$BuildClusterTree
tree$tip.label <- paste0("Cluster ", tree$tip.label)

tree_1 <- ggtree::ggtree(tree, aes(x, y)) +
  scale_y_reverse() +
  ggtree::geom_tree() +
  ggtree::theme_tree() +
  ggtree::geom_tiplab(offset = 1) +
#  ggtree::geom_tippoint(color = colors[1:length(tree$tip.label)], 
  #                      shape = 16, size = 5) +
  coord_cartesian(clip = 'off') +
  theme(plot.margin = unit(c(0,2.5,0,0), 'cm'))
tree_1
```

```{r, echo=FALSE, message=FALSE, warning=FALSE, eval=FALSE}
# save
path <- paste0("../../results/seurat/tree/",treatment,"_",tolower(tissue),
               "_tree_cluster_numbers")
tree_1
saveToPDF(paste0(path, ".pdf"), width = 4, height = 4)
remove(tree)
```

# Marker exploration
FindMarkers will find markers between two different identity groups - you have to specify both identity groups. This is useful for comparing the differences between two specific groups.

FindAllMarkers will find markers differentially expressed in each identity group by comparing it to all of the others - you don't have to manually define anything. Note that markers may bleed over between closely-related groups - they are not forced to be specific to only one group. This is what most people use (and likely what you want).

FindConservedMarkers will find markers that are conserved between two groups - this can be useful if you want to find markers that are conserved between a treated and untreated condition for a specific cell type or group of cells. It means they are differentially expressed compared to other groups, but have similar expression between the two groups you're actually comparing.
https://www.biostars.org/p/409790/
### Find all markers 
```{r read_all_markers}
# read object
all.markers <- readRDS("../../rObjects/unannotated_all_markers.rds")

# more stringent filtering
all.markers <- all.markers[all.markers$p_val_adj < 0.05,]
markers.strict <- all.markers[
  all.markers$delta_pct > summary(all.markers$delta_pct)[5],]

# compare 
table(all.markers$cluster)
table(markers.strict$cluster)

# subset
cluster0 <- markers.strict[markers.strict$cluster == 0,]
cluster1 <- markers.strict[markers.strict$cluster == 1,]
cluster2 <- markers.strict[markers.strict$cluster == 2,]
cluster3 <- markers.strict[markers.strict$cluster == 3,]
cluster4 <- markers.strict[markers.strict$cluster == 4,]
cluster5 <- markers.strict[markers.strict$cluster == 5,]
cluster6 <- markers.strict[markers.strict$cluster == 6,]
cluster7 <- markers.strict[markers.strict$cluster == 7,]
cluster8 <- markers.strict[markers.strict$cluster == 8,]
cluster9 <- markers.strict[markers.strict$cluster == 9,]
cluster10 <- markers.strict[markers.strict$cluster == 10,]
cluster11 <- markers.strict[markers.strict$cluster == 11,]
cluster12 <- markers.strict[markers.strict$cluster == 12,]
cluster13 <- markers.strict[markers.strict$cluster == 13,]
cluster14 <- markers.strict[markers.strict$cluster == 14,]
cluster15 <- markers.strict[markers.strict$cluster == 15,]
cluster16 <- markers.strict[markers.strict$cluster == 16,]
cluster17 <- markers.strict[markers.strict$cluster == 17,]
cluster18 <- markers.strict[markers.strict$cluster == 18,]
cluster19 <- markers.strict[markers.strict$cluster == 19,]
cluster20 <- markers.strict[markers.strict$cluster == 20,]
cluster21 <- markers.strict[markers.strict$cluster == 21,]
```
## Conserved markers 
Identify conserved cell type markers
To identify canonical cell type marker genes that are conserved across conditions, we provide the FindConservedMarkers() function. This function performs differential gene expression testing for each dataset/group and combines the p-values using meta-analysis methods from the MetaDE R package. For example, we can calculated the genes that are conserved markers irrespective of stimulation condition in cluster 6 (NK cells).
```{r conserved_markers}
conserved.markers <- readRDS("../../rObjects/brain_conserved_markers.rds")
conserved.markers.strict <- readRDS("../../rObjects/brain_conserved_markers_strict.rds")

# compare 
table(conserved.markers$cluster)
table(conserved.markers.strict$cluster)


conserved.cluster0 <- conserved.markers.strict[conserved.markers.strict$cluster == 0,]
conserved.cluster1 <- conserved.markers.strict[conserved.markers.strict$cluster == 1,]
conserved.cluster2 <- conserved.markers.strict[conserved.markers.strict$cluster == 2,]
conserved.cluster3 <- conserved.markers.strict[conserved.markers.strict$cluster == 3,]
conserved.cluster4 <- conserved.markers.strict[conserved.markers.strict$cluster == 4,]
conserved.cluster5 <- conserved.markers.strict[conserved.markers.strict$cluster == 5,]
conserved.cluster6 <- conserved.markers.strict[conserved.markers.strict$cluster == 6,]
conserved.cluster7 <- conserved.markers.strict[conserved.markers.strict$cluster == 7,]
conserved.cluster8 <- conserved.markers.strict[conserved.markers.strict$cluster == 8,]
conserved.cluster9 <- conserved.markers.strict[conserved.markers.strict$cluster == 9,]
conserved.cluster10 <- conserved.markers.strict[conserved.markers.strict$cluster == 10,]
conserved.cluster11 <- conserved.markers.strict[conserved.markers.strict$cluster == 11,]
conserved.cluster12 <- conserved.markers.strict[conserved.markers.strict$cluster == 12,]
conserved.cluster13 <- conserved.markers.strict[conserved.markers.strict$cluster == 13,]
conserved.cluster14 <- conserved.markers.strict[conserved.markers.strict$cluster == 14,]
conserved.cluster15 <- conserved.markers.strict[conserved.markers.strict$cluster == 15,]
conserved.cluster16 <- conserved.markers.strict[conserved.markers.strict$cluster == 16,]
conserved.cluster17 <- conserved.markers.strict[conserved.markers.strict$cluster == 17,]
conserved.cluster18 <- conserved.markers.strict[conserved.markers.strict$cluster == 18,]
conserved.cluster19 <- conserved.markers.strict[conserved.markers.strict$cluster == 19,]
conserved.cluster20 <- conserved.markers.strict[conserved.markers.strict$cluster == 20,]
conserved.cluster21 <- conserved.markers.strict[conserved.markers.strict$cluster == 21,]
```

## PCs
```{r print_pcs, message=FALSE}
# Printing out the most variable genes driving PCs
print(x = dataObject.unannotated[["pca"]], 
      dims = 1:5, 
      nfeatures = 10)
```

## Top variable genes
```{r top_variable, message=FALSE}
# print top variable genes
top100 <- dataObject.unannotated@assays$SCT@var.features[1:100]
top100
```


# Cluster Annotation
## Cluster 0 - oligodendrocyte / polydendrocyte
OPALIN - oligodendrocyte\
MAG  - polydendrocyte\
CLDN11 - polydendrocyte / oligodendrocyte\
ANLN - oligodendrocyte\
APOD - fibroblast-like / oligodendrocyte\
```{r cluster0,warning=FALSE,message=FALSE}
# Number of cells per condition
n_cells[,c(1,2)]

# UMAP with only cluster 0
DimPlot(object = subset(dataObject.unannotated, seurat_clusters == "0"),
        reduction = "umap", 
        label = TRUE,
        label.box = TRUE,
        label.size = 3,
        repel = TRUE,
        cols = "#E69F00")

VlnPlot(dataObject.unannotated,
        features = cluster0$gene[1:10],
        cols = colors,
        stack = TRUE,
        flip = TRUE,
        split.by = "seurat_clusters")

VlnPlot(dataObject.unannotated,
        features = conserved.cluster0$gene[1:10],
        cols = colors,
        stack = TRUE,
        flip = TRUE,
        split.by = "seurat_clusters")

cluster0$gene[1:10]
conserved.cluster0$gene[1:10]
```